// @fluentui/accessibility/dist/commonjs/behaviors/List/selectableListItemBehavior.js
"use strict";

exports.__esModule = true;
exports.selectableListItemBehavior = void 0;

var _keyboardKey = require("../../keyboard-key");

var _attributes = require("../../attributes");

/**
 * @specification
 * Adds role='option'. This role is used for a selectable item in a list.
 * Adds attribute 'data-is-focusable=true' to 'root' slot.
 * Adds attribute 'aria-selected=true' based on the property 'selected'. Based on this screen readers will recognize the selected state of the item.
 * Triggers 'performClick' action with 'Enter' or 'Spacebar' on 'root'.
 */
var selectableListItemBehavior = function selectableListItemBehavior(props) {
  var _root;

  return {
    attributes: {
      root: (_root = {
        role: 'option',
        'aria-selected': !!props.selected
      }, _root[_attributes.IS_FOCUSABLE_ATTRIBUTE] = true, _root)
    },
    keyActions: {
      root: {
        performClick: {
          keyCombinations: [{
            keyCode: _keyboardKey.keyboardKey.Enter
          }, {
            keyCode: _keyboardKey.SpacebarKey
          }]
        }
      }
    }
  };
};

exports.selectableListItemBehavior = selectableListItemBehavior;
