// @fluentui/accessibility/dist/commonjs/behaviors/Datepicker/datepickerCalendarBehavior.js
"use strict";

exports.__esModule = true;
exports.datepickerCalendarBehavior = void 0;

var _datepickerCalendarGridBehavior = require("./datepickerCalendarGridBehavior");

var _datepickerCalendarGridRowBehavior = require("./datepickerCalendarGridRowBehavior");

var _keyboardKey = require("../../keyboard-key");

/**
 * @description
 * Behavior for a datepicked calendar component
 * @specification
 * Applies 'datepickerCalendarGridBehavior' for 'calendarGrid' child component.
 */
var datepickerCalendarBehavior = function datepickerCalendarBehavior() {
  return {
    childBehaviors: {
      calendarGrid: _datepickerCalendarGridBehavior.datepickerCalendarGridBehavior,
      calendarGridRow: _datepickerCalendarGridRowBehavior.datepickerCalendarGridRowBehavior
    },
    keyActions: {
      calendarCell: {
        addWeek: {
          keyCombinations: [{
            keyCode: _keyboardKey.keyboardKey.ArrowDown
          }]
        },
        subtractWeek: {
          keyCombinations: [{
            keyCode: _keyboardKey.keyboardKey.ArrowUp
          }]
        },
        addDay: {
          keyCombinations: [{
            keyCode: _keyboardKey.keyboardKey.ArrowRight
          }]
        },
        subtractDay: {
          keyCombinations: [{
            keyCode: _keyboardKey.keyboardKey.ArrowLeft
          }]
        },
        moveToStartOfWeek: {
          keyCombinations: [{
            keyCode: _keyboardKey.keyboardKey.Home,
            ctrlKey: false
          }]
        },
        moveToEndOfWeek: {
          keyCombinations: [{
            keyCode: _keyboardKey.keyboardKey.End,
            ctrlKey: false
          }]
        },
        moveToStartOfColumn: {
          keyCombinations: [{
            keyCode: _keyboardKey.keyboardKey.PageUp
          }]
        },
        moveToEndOfColumn: {
          keyCombinations: [{
            keyCode: _keyboardKey.keyboardKey.PageDown
          }]
        }
      }
    }
  };
};

exports.datepickerCalendarBehavior = datepickerCalendarBehavior;
