// @fluentui/react-northstar/dist/commonjs/themes/teams/components/TextArea/textAreaStyles.js
"use strict";

exports.__esModule = true;
exports.textAreaStyles = void 0;

var _utils = require("../../../../utils");

var textAreaStyles = {
  root: function root(_ref) {
    var p = _ref.props,
        v = _ref.variables;
    return Object.assign({
      margin: v.margin,
      height: v.height,
      backgroundColor: v.backgroundColor
    }, p.inverted && {
      backgroundColor: v.invertedBackgroundColor
    }, {
      color: v.fontColor,
      borderColor: v.borderColor,
      borderRadius: v.borderRadius,
      borderStyle: 'solid',
      borderWidth: v.borderWidth,
      outline: 0,
      padding: v.padding,
      resize: p.resize || 'none'
    }, p.fluid && {
      width: '100%'
    }, p.disabled && {
      pointerEvents: 'none',
      color: v.disabledColor,
      boxShadow: 'none'
    }, p.error && {
      border: (0, _utils.pxToRem)(1) + " solid " + v.borderColorError
    }, {
      '::placeholder': Object.assign({
        color: v.placeholderColor,
        opacity: 1
      }, p.disabled && {
        color: v.disabledColor
      }),
      ':focus': {
        borderColor: v.borderColorFocus
      }
    });
  }
};
exports.textAreaStyles = textAreaStyles;
