// @fluentui/styles/dist/dts/src/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.withDebugId = exports.objectKeyToValues = exports.mergeSiteVariables = exports.mergeThemeVariables = exports.mergeThemeStyles = exports.mergeStyles = exports.mergeStaticStyles = exports.mergeFontFaces = exports.mergeComponentStyles = exports.mergeComponentVariables = exports.emptyTheme = exports.mergeThemes = exports.deepmerge = exports.isDebugEnabled = exports.createTheme = exports.callable = void 0;
var callable_1 = require("./callable");
Object.defineProperty(exports, "callable", { enumerable: true, get: function () { return callable_1.callable; } });
var createTheme_1 = require("./createTheme");
Object.defineProperty(exports, "createTheme", { enumerable: true, get: function () { return createTheme_1.createTheme; } });
var debugEnabled_1 = require("./debugEnabled");
Object.defineProperty(exports, "isDebugEnabled", { enumerable: true, get: function () { return debugEnabled_1.isEnabled; } });
var deepmerge_1 = require("./deepmerge");
Object.defineProperty(exports, "deepmerge", { enumerable: true, get: function () { return deepmerge_1.deepmerge; } });
var mergeThemes_1 = require("./mergeThemes");
Object.defineProperty(exports, "mergeThemes", { enumerable: true, get: function () { return mergeThemes_1.mergeThemes; } });
Object.defineProperty(exports, "emptyTheme", { enumerable: true, get: function () { return mergeThemes_1.emptyTheme; } });
Object.defineProperty(exports, "mergeComponentVariables", { enumerable: true, get: function () { return mergeThemes_1.mergeComponentVariables; } });
Object.defineProperty(exports, "mergeComponentStyles", { enumerable: true, get: function () { return mergeThemes_1.mergeComponentStyles; } });
Object.defineProperty(exports, "mergeFontFaces", { enumerable: true, get: function () { return mergeThemes_1.mergeFontFaces; } });
Object.defineProperty(exports, "mergeStaticStyles", { enumerable: true, get: function () { return mergeThemes_1.mergeStaticStyles; } });
Object.defineProperty(exports, "mergeStyles", { enumerable: true, get: function () { return mergeThemes_1.mergeStyles; } });
Object.defineProperty(exports, "mergeThemeStyles", { enumerable: true, get: function () { return mergeThemes_1.mergeThemeStyles; } });
Object.defineProperty(exports, "mergeThemeVariables", { enumerable: true, get: function () { return mergeThemes_1.mergeThemeVariables; } });
Object.defineProperty(exports, "mergeSiteVariables", { enumerable: true, get: function () { return mergeThemes_1.mergeSiteVariables; } });
var objectKeysToValues_1 = require("./objectKeysToValues");
Object.defineProperty(exports, "objectKeyToValues", { enumerable: true, get: function () { return objectKeysToValues_1.objectKeyToValues; } });
var withDebugId_1 = require("./withDebugId");
Object.defineProperty(exports, "withDebugId", { enumerable: true, get: function () { return withDebugId_1.withDebugId; } });
__exportStar(require("./types"), exports);
