// @fluentui/react-northstar/dist/dts/src/utils/date-time-utilities/dateFormatting/dateFormatting.defaults.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_CALENDAR_STRINGS = exports.DEFAULT_DATE_FORMATTING = exports.DEFAULT_LOCALIZED_STRINGS = void 0;
var formatDay_1 = require("./formatDay");
var formatYear_1 = require("./formatYear");
var formatMonthDayYear_1 = require("./formatMonthDayYear");
var formatMonthYear_1 = require("./formatMonthYear");
exports.DEFAULT_LOCALIZED_STRINGS = {
    months: [
        'January',
        'February',
        'March',
        'April',
        'May',
        'June',
        'July',
        'August',
        'September',
        'October',
        'November',
        'December',
    ],
    shortMonths: ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'],
    days: ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'],
    shortDays: ['S', 'M', 'T', 'W', 'T', 'F', 'S'],
};
exports.DEFAULT_DATE_FORMATTING = __assign({ formatDay: formatDay_1.formatDay, formatYear: formatYear_1.formatYear, formatMonthDayYear: formatMonthDayYear_1.formatMonthDayYear, formatMonthYear: formatMonthYear_1.formatMonthYear, parseDate: function (dateStr) {
        var date = Date.parse(dateStr);
        if (date) {
            return new Date(date);
        }
        return null;
    } }, exports.DEFAULT_LOCALIZED_STRINGS);
exports.DEFAULT_CALENDAR_STRINGS = __assign(__assign({}, exports.DEFAULT_DATE_FORMATTING), { goToToday: 'Go to today', openCalendarTitle: 'Open calendar', inputPlaceholder: 'Select a date...', weekNumberFormatString: 'Week number {0}', prevMonthAriaLabel: 'Previous month', nextMonthAriaLabel: 'Next month', prevYearAriaLabel: 'Previous year', nextYearAriaLabel: 'Next year', prevYearRangeAriaLabel: 'Previous year range', nextYearRangeAriaLabel: 'Next year range', closeButtonAriaLabel: 'Close', selectedDateFormatString: 'Selected date {0}', todayDateFormatString: "Today's date {0}", monthPickerHeaderAriaLabel: '{0}, select to change the year', yearPickerHeaderAriaLabel: '{0}, select to change the month', isRequiredErrorMessage: 'A date selection is required', invalidInputErrorMessage: 'Manually entered date is not in correct format.', isOutOfBoundsErrorMessage: 'The selected date is from the restricted range.', inputAriaLabel: 'Select a date.', inputBoundedFormatString: 'Input a date between {0} and {1}.', inputMinBoundedFormatString: 'Input a date starting from {0}.', inputMaxBoundedFormatString: 'Input a date ending at {0}.' });
