// @fluentui/react-northstar/dist/dts/src/themes/teams/components/Skeleton/skeletonVariables.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.skeletonVariables = void 0;
var utils_1 = require("../../../../utils");
var skeletonVariables = function (siteVariables) {
    return {
        lineBackground: siteVariables.colorScheme.default.foreground6,
        shapeBackground: siteVariables.colorScheme.default.foreground6,
        animationBackground: siteVariables.colorScheme.default.foreground6,
        animationBackgroundSecondary: siteVariables.colorScheme.default.foreground6,
        lineMargin: "0 0 " + utils_1.pxToRem(2) + " 0",
        shapeMargin: "0 0 " + utils_1.pxToRem(2) + " 0",
        // Skeleton Button
        buttonHeight: utils_1.pxToRem(32),
        buttonWidth: utils_1.pxToRem(96),
        buttonSmallHeight: utils_1.pxToRem(24),
        buttonSmallWidth: utils_1.pxToRem(72),
        buttonBackground: siteVariables.colorScheme.default.foreground6,
        buttonCircularBorderRadius: utils_1.pxToRem(999),
        // Skeleton Text
        textBackground: siteVariables.colorScheme.default.foreground6,
        textWidth: '100%',
        textSmallerHeight: utils_1.pxToRem(14),
        textSmallHeight: utils_1.pxToRem(16),
        textMediumHeight: utils_1.pxToRem(19),
        textLargeHeight: utils_1.pxToRem(24),
        textLargerHeight: utils_1.pxToRem(32),
        // Skeleton Input
        inputHeight: utils_1.pxToRem(32),
        inputWidth: utils_1.pxToRem(154),
        inputBackground: siteVariables.colorScheme.default.foreground6,
        // Skeleton Avatar
        avatarBackground: siteVariables.colorScheme.default.foreground6,
        avatarSmallest: utils_1.pxToRem(20),
        avatarSmaller: utils_1.pxToRem(24),
        avatarSmall: utils_1.pxToRem(28),
        avatarMedium: utils_1.pxToRem(32),
        avatarLarge: utils_1.pxToRem(44),
        avatarLarger: utils_1.pxToRem(64),
        avatarLargest: utils_1.pxToRem(96),
    };
};
exports.skeletonVariables = skeletonVariables;
