// @fluentui/accessibility/dist/dts/src/behaviors/Checkbox/checkboxBehavior.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.checkboxBehavior = void 0;
var keyboard_key_1 = require("../../keyboard-key");
var attributes_1 = require("../../attributes");
/**
 * @specification
 * Adds role='checkbox'. This allows screen readers to handle the component as a checkbox button.
 * Adds attribute 'aria-checked=true' based on the property 'checked'.
 * Adds attribute 'aria-disabled=true' based on the property 'disabled'.
 * Adds attribute 'tabIndex=0' to 'root' slot.
 * Adds attribute 'data-is-focusable=true' to 'root' slot.
 */
var checkboxBehavior = function (props) {
    var _a;
    return ({
        attributes: {
            root: (_a = {
                    'aria-checked': props.checked === 'mixed' ? 'mixed' : !!props.checked,
                    'aria-disabled': props.disabled,
                    role: 'checkbox',
                    tabIndex: 0
                },
                _a[attributes_1.IS_FOCUSABLE_ATTRIBUTE] = true,
                _a),
        },
        keyActions: {
            root: {
                performClick: {
                    keyCombinations: [{ keyCode: keyboard_key_1.SpacebarKey }],
                },
            },
        },
    });
};
exports.checkboxBehavior = checkboxBehavior;
