// @fluentui/react-northstar/dist/dts/src/components/Attachment/AttachmentAction.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AttachmentAction = exports.attachmentActionClassName = void 0;
var accessibility_1 = require("@fluentui/accessibility");
var customPropTypes = require("@fluentui/react-proptypes");
var PropTypes = require("prop-types");
var react_bindings_1 = require("@fluentui/react-bindings");
var utils_1 = require("../../utils");
var Button_1 = require("../Button/Button");
exports.attachmentActionClassName = 'ui-attachment__action';
/**
 * An AttachmentAction provides a slot for actions in the Attachment.
 */
exports.AttachmentAction = react_bindings_1.compose(Button_1.Button, {
    className: exports.attachmentActionClassName,
    displayName: 'AttachmentAction',
    overrideStyles: true,
    shorthandConfig: {
        mappedProp: 'content',
    },
});
exports.AttachmentAction.defaultProps = {
    accessibility: accessibility_1.buttonBehavior,
    as: 'button',
};
exports.AttachmentAction.propTypes = __assign(__assign({}, utils_1.commonPropTypes.createCommon({
    content: 'shorthand',
})), { disabled: PropTypes.bool, icon: customPropTypes.shorthandAllowingChildren, iconPosition: PropTypes.oneOf(['before', 'after']), loader: customPropTypes.itemShorthandWithoutJSX, loading: PropTypes.bool, onClick: PropTypes.func, onFocus: PropTypes.func, primary: customPropTypes.every([customPropTypes.disallow(['secondary']), PropTypes.bool]), secondary: customPropTypes.every([customPropTypes.disallow(['primary']), PropTypes.bool]) });
