// @fluentui/react-northstar/dist/dts/src/themes/teams/components/Toolbar/toolbarItemStyles.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.toolbarItemStyles = void 0;
var getIconFillOrOutlineStyles_1 = require("../../getIconFillOrOutlineStyles");
var colors_1 = require("../../colors");
var getBorderFocusStyles_1 = require("../../getBorderFocusStyles");
exports.toolbarItemStyles = {
    root: function (_a) {
        var p = _a.props, v = _a.variables, siteVariables = _a.theme.siteVariables;
        var colors = colors_1.getColorScheme(v.colorScheme);
        var borderWidth = siteVariables.borderWidth;
        var borderFocusStyles = getBorderFocusStyles_1.getBorderFocusStyles({
            variables: siteVariables,
        });
        return __assign(__assign(__assign({ position: 'relative', backgroundColor: v.background, borderWidth: borderWidth, borderStyle: 'solid', borderColor: 'transparent', borderRadius: v.borderRadius, height: v.itemHeight, minWidth: v.itemHeight, padding: v.itemPadding, color: v.foreground || colors.foreground1, cursor: 'pointer', userSelect: 'none', ':focus': borderFocusStyles[':focus'] }, (p.active && __assign({ color: v.foregroundActive || colors.foregroundActive, backgroundColor: v.backgroundActive }, getIconFillOrOutlineStyles_1.getIconFillOrOutlineStyles({ outline: false })))), { ':hover': __assign({ color: v.foregroundHover || colors.foregroundHover, backgroundColor: v.backgroundHover || colors.backgroundHover }, getIconFillOrOutlineStyles_1.getIconFillOrOutlineStyles({ outline: false })), ':focus-visible': borderFocusStyles[':focus-visible'] }), (p.disabled && {
            color: v.foregroundDisabled || colors.foregroundDisabled1,
            backgroundColor: v.backgroundDisabled,
            cursor: 'default',
            pointerEvents: 'none',
            ':hover': {
            // empty to overwrite all existing hover styles
            },
        }));
    },
};
