// @fluentui/react-northstar/dist/dts/src/themes/teams/components/Toolbar/toolbarCustomItemStyles.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.toolbarCustomItemStyles = void 0;
var colors_1 = require("../../colors");
var getBorderFocusStyles_1 = require("../../getBorderFocusStyles");
exports.toolbarCustomItemStyles = {
    root: function (_a) {
        var p = _a.props, v = _a.variables, siteVariables = _a.theme.siteVariables;
        var colors = colors_1.getColorScheme(v.colorScheme);
        var borderWidth = siteVariables.borderWidth;
        var borderFocusStyles = getBorderFocusStyles_1.getBorderFocusStyles({
            variables: siteVariables,
        });
        return __assign(__assign(__assign({ position: 'relative', backgroundColor: v.background, borderColor: 'transparent', borderWidth: borderWidth, borderStyle: 'solid', height: v.itemHeight, color: v.foreground || colors.foreground1, display: 'flex', alignItems: 'center', justifyContent: 'center', userSelect: 'none' }, (p.fitted !== true &&
            p.fitted !== 'horizontally' && {
            paddingLeft: v.customItemHorizontalPadding,
            paddingRight: v.customItemHorizontalPadding,
        })), (p.fitted !== true &&
            p.fitted !== 'vertically' && {
            paddingTop: v.customItemVerticalPadding,
            paddingBottom: v.customItemVerticalPadding,
        })), { ':focus': borderFocusStyles[':focus'], ':focus-visible': borderFocusStyles[':focus-visible'] });
    },
};
