// @fluentui/react-northstar/dist/dts/src/themes/teams/components/Toolbar/toolbarMenuItemStyles.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.toolbarMenuItemStyles = void 0;
var colors_1 = require("../../colors");
var getBorderFocusStyles_1 = require("../../getBorderFocusStyles");
var activeIndicatorUrl_1 = require("./activeIndicatorUrl");
var ToolbarMenuItemSubmenuIndicator_1 = require("../../../../components/Toolbar/ToolbarMenuItemSubmenuIndicator");
var ToolbarMenuItemActiveIndicator_1 = require("../../../../components/Toolbar/ToolbarMenuItemActiveIndicator");
exports.toolbarMenuItemStyles = {
    root: function (_a) {
        var _b, _c;
        var p = _a.props, v = _a.variables, siteVariables = _a.theme.siteVariables;
        var colors = colors_1.getColorScheme(v.colorScheme);
        var borderFocusStyles = getBorderFocusStyles_1.getBorderFocusStyles({
            variables: siteVariables,
            borderRadius: 0,
        });
        return __assign({ position: 'relative', color: v.menuItemForeground || colors.foreground1, borderWidth: v.menuBorderWidth, backgroundColor: 'transparent', borderColor: 'transparent', display: 'flex', alignItems: 'center', textAlign: 'left', width: '100%', maxWidth: '100%', padding: v.menuItemPadding, cursor: 'pointer', minHeight: v.itemHeight, lineHeight: v.lineHeightBase, userSelect: 'none', ':focus': {
                outline: 0,
            }, ':hover': (_b = {
                    color: v.menuItemForegroundHover || colors.menuItemForegroundHover,
                    backgroundColor: v.menuItemBackgroundHover || colors.menuItemBackgroundHover
                },
                _b["& ." + ToolbarMenuItemSubmenuIndicator_1.toolbarMenuItemSubmenuIndicatorClassName] = {
                    color: v.menuItemForegroundHover || colors.menuItemForegroundHover,
                },
                _b["& ." + ToolbarMenuItemActiveIndicator_1.toolbarMenuItemActiveIndicatorClassName] = {
                    backgroundImage: activeIndicatorUrl_1.activeIndicatorUrl(v.menuItemForegroundHover || colors.menuItemForegroundHover),
                },
                _b), ':focus-visible': borderFocusStyles[':focus-visible'] }, (p.disabled && (_c = {
                cursor: 'default',
                color: v.menuItemForegroundDisabled || colors.foregroundDisabled1,
                backgroundColor: v.menuItemBackgroundDisabled
            },
            _c["& ." + ToolbarMenuItemSubmenuIndicator_1.toolbarMenuItemSubmenuIndicatorClassName] = {
                color: v.menuItemForegroundDisabled || colors.foregroundDisabled1,
            },
            _c["& ." + ToolbarMenuItemActiveIndicator_1.toolbarMenuItemActiveIndicatorClassName] = {
                backgroundImage: activeIndicatorUrl_1.activeIndicatorUrl(v.menuItemForegroundDisabled || colors.foregroundDisabled1),
            },
            _c[':hover'] = {
            // empty to overwrite all existing hover styles
            },
            _c)));
    },
    wrapper: function () { return ({
        display: 'block',
    }); },
};
