// @fluentui/react-northstar/dist/dts/src/themes/teams/components/Toolbar/toolbarVariables.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.toolbarVariables = exports.toolbarColorAreas = void 0;
var utils_1 = require("../../../../utils");
var colorUtils_1 = require("../../../colorUtils");
exports.toolbarColorAreas = utils_1.stringLiteralsArray('foreground1', 'background', 'border', 'foregroundActive', 'foregroundHover', 'backgroundHover', 'foregroundDisabled1', 
// custom
'menuItemForegroundHover', 'menuItemBackgroundHover');
var toolbarVariables = function (siteVars) { return ({
    colorScheme: colorUtils_1.pickValuesFromColorScheme(colorUtils_1.extendColorScheme(siteVars.colorScheme, {
        default: {
            foregroundHover: siteVars.colorScheme.brand.foregroundHover,
            backgroundHover: 'transparent',
            foregroundActive: siteVars.colorScheme.brand.foregroundActive,
            menuItemForegroundHover: siteVars.colorScheme.default.foregroundHover,
            menuItemBackgroundHover: siteVars.colorScheme.default.backgroundHover,
        },
    }), exports.toolbarColorAreas),
    foreground: undefined,
    background: 'transparent',
    dividerBorder: undefined,
    foregroundHover: undefined,
    backgroundHover: undefined,
    foregroundActive: undefined,
    backgroundActive: 'transparent',
    foregroundDisabled: undefined,
    backgroundDisabled: 'transparent',
    itemHeight: utils_1.pxToRem(32),
    itemPadding: '0',
    borderWidth: '2px',
    borderRadius: siteVars.borderRadiusMedium,
    dividerMargin: utils_1.pxToRem(10) + " " + utils_1.pxToRem(4),
    menuMaxWidth: utils_1.pxToRem(312),
    menuPadding: utils_1.pxToRem(8) + " 0",
    menuBackground: undefined,
    menuBorder: undefined,
    menuBorderWidth: '1px',
    menuBorderRadius: siteVars.borderRadiusMedium,
    menuBoxShadow: siteVars.shadowLevel3,
    menuItemForeground: undefined,
    menuItemForegroundHover: undefined,
    menuItemBackgroundHover: undefined,
    menuItemForegroundDisabled: undefined,
    menuItemBackgroundDisabled: 'transparent',
    menuItemPadding: utils_1.pxToRem(5) + " " + utils_1.pxToRem(15),
    menuDividerBorder: undefined,
    menuDividerMargin: utils_1.pxToRem(8) + " 0",
    customItemHorizontalPadding: utils_1.pxToRem(16),
    customItemVerticalPadding: utils_1.pxToRem(4),
    overlayZIndex: siteVars.zIndexes.overlay,
    lineHeightBase: siteVars.lineHeightMedium,
}); };
exports.toolbarVariables = toolbarVariables;
