// @fluentui/react-northstar/dist/dts/src/themes/teams/components/Slider/sliderVariables.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.sliderVariables = void 0;
var utils_1 = require("../../../../utils");
var sliderVariables = function (siteVars) {
    var colorScheme = siteVars.colorScheme;
    return {
        height: utils_1.pxToRem(16),
        length: utils_1.pxToRem(320),
        railColor: colorScheme.default.border,
        disabledRailColor: colorScheme.default.backgroundDisabled1,
        railHeight: utils_1.pxToRem(2),
        thumbColor: colorScheme.default.foreground2,
        activeThumbColor: colorScheme.default.foreground1,
        disabledThumbColor: colorScheme.default.foregroundDisabled1,
        thumbHeight: utils_1.pxToRem(10),
        activeThumbHeight: utils_1.pxToRem(14),
        thumbBorderPadding: utils_1.pxToRem(4),
        thumbWidth: utils_1.pxToRem(10),
        activeThumbWidth: utils_1.pxToRem(14),
        trackColor: colorScheme.brand.foregroundActive,
        disabledTrackColor: colorScheme.default.foregroundDisabled1,
    };
};
exports.sliderVariables = sliderVariables;
