// @fluentui/react-northstar/dist/es/themes/teams-dark/siteVariables.js
import { colors } from '../teams/siteVariables';
export { colorScheme, categoryColorScheme } from './colors'; //
// BORDER STYLES
//

export var focusInnerBorderColor = colors.black;
export var focusOuterBorderColor = colors.white; //
// SHADOW LEVELS
//

export var shadowLevel1 = '0 .2rem .4rem -.075rem rgba(0, 0, 0, .25)';
export var shadowLevel2 = '0 .4rem 0.7rem -0.1rem rgba(0, 0, 0, .25)';
export var shadowLevel3 = '0 .8rem 1rem -0.2rem rgba(0, 0, 0, .25)';
export var shadowLevel4 = '0 .6rem 1.8rem -0.4rem rgba(0, 0, 0, .25)';
export var shadow2 = '0 0 2px rgba(0, 0, 0, .24), 0 1px 2px rgba(0, 0, 0, .28)';
export var shadow4 = '0 0 2px rgba(0, 0, 0, .24), 0 2px 4px rgba(0, 0, 0, .28)';
export var shadow8 = '0 0 2px rgba(0, 0, 0, .24), 0 4px 8px rgba(0, 0, 0, .28)';
export var shadow16 = '0 0 2px rgba(0, 0, 0, .24), 0 8px 16px rgba(0, 0, 0, .28)';
export var shadow28 = '0 0 8px rgba(0, 0, 0, .40), 0 14px 28px rgba(0, 0, 0, .48)';
export var shadow64 = '0 0 8px rgba(0, 0, 0, .40), 0 32px 64px rgba(0, 0, 0, .48)'; //
// SEMANTIC ASSIGNMENTS
//

export var bodyBackground = colors.grey[650];
export var bodyColor = colors.white;
