// @fluentui/accessibility/dist/dts/src/behaviors/Toolbar/toolbarRadioGroupItemBehavior.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.toolbarRadioGroupItemBehavior = void 0;
var buttonBehavior_1 = require("../Button/buttonBehavior");
/**
 * @specification
 *  Adds role='radio'. This allows screen readers to handle the component as a radio button.
 *  Adds attribute 'aria-checked=true' based on the property 'active'.
 *  Triggers 'performClick' action with 'Enter' or 'Spacebar' on 'root'.
 *  Adds attribute 'disabled=true' based on the property 'disabled'.
 *  Adds attribute 'aria-disabled=true' based on the property 'disabledFocusable'.
 */
var toolbarRadioGroupItemBehavior = function (props) {
    var definition = {
        attributes: {
            root: {
                role: 'radio',
                'aria-checked': props.active,
                disabled: props.disabled,
                'aria-disabled': props.disabledFocusable,
            },
        },
        keyActions: buttonBehavior_1.buttonBehavior(props).keyActions,
    };
    if (process.env.NODE_ENV !== 'production') {
        // Override the default trigger's accessibility schema class.
        definition.attributes.root['data-aa-class'] = 'RadioGroupItem';
    }
    return definition;
};
exports.toolbarRadioGroupItemBehavior = toolbarRadioGroupItemBehavior;
