// @fluentui/accessibility/dist/dts/src/behaviors/Toolbar/toolbarMenuBehavior.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.toolbarMenuBehavior = void 0;
var types_1 = require("../../focusZone/types");
var toolbarMenuItemBehavior_1 = require("./toolbarMenuItemBehavior");
var keyboard_key_1 = require("../../keyboard-key");
/**
 * @description
 * The 'menu' role is used to identify an element that creates a list of common actions or functions that a user can invoke.
 *
 * @specification
 * Adds role='menu'.
 * Provides arrow key navigation in vertical direction.
 * Keyboard navigation is circular.
 * Component will get focus when mounted.
 */
var toolbarMenuBehavior = function () { return ({
    attributes: {
        root: {
            role: 'menu',
        },
    },
    keyActions: {
        root: {
            performClick: {
                keyCombinations: [{ keyCode: keyboard_key_1.keyboardKey.Enter }, { keyCode: keyboard_key_1.SpacebarKey }],
            },
        },
    },
    focusZone: {
        props: {
            isCircularNavigation: true,
            shouldFocusOnMount: true,
            direction: types_1.FocusZoneDirection.vertical,
        },
    },
    childBehaviors: { item: toolbarMenuItemBehavior_1.toolbarMenuItemBehavior },
}); };
exports.toolbarMenuBehavior = toolbarMenuBehavior;
