// @fluentui/accessibility/dist/dts/src/behaviors/Toolbar/toolbarMenuItemRadioBehavior.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.toolbarMenuItemRadioBehavior = void 0;
var keyboard_key_1 = require("../../keyboard-key");
var attributes_1 = require("../../attributes");
/**
 * @specification
 *  Adds attribute 'aria-checked=true' based on the property 'active'.
 *  Adds attribute 'aria-disabled=true' based on the property 'disabled'.
 *  Adds role='menuitemradio'.
 */
var toolbarMenuItemRadioBehavior = function (props) {
    var _a;
    return ({
        attributes: {
            root: (_a = {},
                _a[attributes_1.IS_FOCUSABLE_ATTRIBUTE] = true,
                _a['aria-checked'] = props.active,
                _a['aria-disabled'] = props.disabled,
                _a.role = 'menuitemradio',
                _a),
        },
        keyActions: {
            root: {
                performClick: {
                    keyCombinations: [{ keyCode: keyboard_key_1.keyboardKey.Enter }, { keyCode: keyboard_key_1.SpacebarKey }],
                },
            },
        },
    });
};
exports.toolbarMenuItemRadioBehavior = toolbarMenuItemRadioBehavior;
