// @fluentui/react-northstar/dist/es/themes/teams/components/Menu/menuItemIndicatorStyles.js
import { pxToRem } from '../../../../utils';
import { submenuIndicatorDirection } from './submenuIndicatorDirection';
export var menuItemIndicatorStyles = {
  root: function root(_ref) {
    var p = _ref.props,
        v = _ref.variables;
    return Object.assign({
      position: 'relative',
      float: 'right',
      left: pxToRem(12),
      userSelect: 'none',
      marginRight: pxToRem(4)
    }, p.inSubmenu && {
      position: 'absolute',
      top: pxToRem(6),
      right: pxToRem(2),
      left: 'unset'
    }, {
      content: '" "',
      display: 'block',
      overflow: 'hidden',
      height: pxToRem(16),
      width: pxToRem(16),
      backgroundSize: pxToRem(16),
      color: v.indicatorColor
    }, p.active && Object.assign({
      color: v.activeIndicatorColor
    }, p.primary && Object.assign({
      color: v.activePrimaryIndicatorColor
    }, p.vertical && {
      color: v.activePrimaryVerticalIndicatorColor
    })), p.underlined && {
      color: v.indicatorColor
    }, p.iconOnly && {
      color: v.indicatorColor
    }, submenuIndicatorDirection(p.vertical));
  }
};
