// @fluentui/react-northstar/dist/commonjs/themes/teams/components/Breadcrumb/breadcrumbVariables.js
"use strict";

exports.__esModule = true;
exports.breadcrumbVariables = void 0;

var _utils = require("../../../../utils");

var breadcrumbVariables = function breadcrumbVariables(siteVars) {
  return {
    dividerMargin: "0 " + (0, _utils.pxToRem)(8),
    fontSizeSmaller: (0, _utils.pxToRem)(12),
    fontSizeSmall: (0, _utils.pxToRem)(12),
    fontSizeMedium: (0, _utils.pxToRem)(14),
    fontSizeLarge: (0, _utils.pxToRem)(18),
    // Breadcrumb Item
    itemCurrentFontWeight: siteVars.fontWeightBold,
    // Breadcrumb Link
    linkPaddingTop: (0, _utils.pxToRem)(6),
    linkPaddingBottom: (0, _utils.pxToRem)(6),
    linkPaddingLeftSmaller: (0, _utils.pxToRem)(4),
    linkPaddingRightSmaller: (0, _utils.pxToRem)(4),
    linkPaddingLeftSmall: (0, _utils.pxToRem)(8),
    linkPaddingRightSmall: (0, _utils.pxToRem)(8),
    linkPaddingLeftMedium: (0, _utils.pxToRem)(12),
    linkPaddingRightMedium: (0, _utils.pxToRem)(12),
    linkPaddingLeftLarge: (0, _utils.pxToRem)(12),
    linkPaddingRightLarge: (0, _utils.pxToRem)(12),
    linkSmallerGap: (0, _utils.pxToRem)(2),
    linkSmallGap: (0, _utils.pxToRem)(8),
    linkMediumGap: (0, _utils.pxToRem)(8),
    linkLargeGap: (0, _utils.pxToRem)(8),
    disabledColor: siteVars.colorScheme.default.foregroundDisabled
  };
};

exports.breadcrumbVariables = breadcrumbVariables;
