// @fluentui/accessibility/dist/dts/src/behaviors/Table/gridRowNestedBehavior.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.gridRowNestedBehavior = void 0;
var attributes_1 = require("../../attributes");
var types_1 = require("../../focusZone/types");
var keyboard_key_1 = require("../../keyboard-key");
var gridCellBehavior_1 = require("./gridCellBehavior");
/**
 * @specification
 * Adds role='row'.
 * Adds attribute 'data-is-focusable=true' to 'root' slot.
 * Adds attribute 'aria-selected=true' based on the property 'selected'. Based on this screen readers will recognize the selected state of the item.
 * Focus can be moved inside a child component with embeded inner FocusZone by pressing a specified key.
 * Provides arrow key navigation in horizontal direction.
 * Triggers 'performClick' action with 'Enter' or 'Spacebar' on 'root'.
 * Triggers 'unsetRowTabbable' action using 'shiftKey' + 'Tab' key on 'root'.
 * Applies 'gridCellBehavior' for 'cell' child component.
 */
var gridRowNestedBehavior = function (props) {
    var _a;
    return ({
        attributes: {
            root: (_a = {},
                _a[attributes_1.IS_FOCUSABLE_ATTRIBUTE] = true,
                _a.role = 'row',
                _a['aria-selected'] = props.selected,
                _a),
        },
        focusZone: {
            props: {
                shouldEnterInnerZone: function (event) { return keyboard_key_1.getCode(event) === keyboard_key_1.keyboardKey.Enter; },
                direction: types_1.FocusZoneDirection.horizontal,
            },
        },
        keyActions: {
            root: {
                unsetRowTabbable: {
                    keyCombinations: [{ keyCode: keyboard_key_1.keyboardKey.Tab, shiftKey: true }],
                },
                performClick: {
                    keyCombinations: [{ keyCode: keyboard_key_1.keyboardKey.Enter }, { keyCode: keyboard_key_1.SpacebarKey }],
                },
            },
        },
        childBehaviors: {
            cell: gridCellBehavior_1.gridCellBehavior,
        },
    });
};
exports.gridRowNestedBehavior = gridRowNestedBehavior;
