// @fluentui/react-northstar/dist/dts/src/themes/teams/components/Tooltip/tooltipContentVariables.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.tooltipContentVariables = void 0;
var utils_1 = require("../../../../utils");
var tooltipContentVariables = function (siteVars) { return ({
    boxShadow: siteVars.shadowLevel2,
    borderRadius: siteVars.borderRadiusMedium,
    borderSize: '1px',
    padding: utils_1.pxToRem(5) + " " + utils_1.pxToRem(12) + " " + utils_1.pxToRem(7) + " " + utils_1.pxToRem(12),
    maxWidth: utils_1.pxToRem(246),
    pointerMargin: utils_1.pxToRem(6),
    pointerGap: utils_1.pxToRem(5),
    pointerWidth: utils_1.pxToRem(16),
    pointerHeight: utils_1.pxToRem(6),
    color: siteVars.colorScheme.default.foreground3,
    backgroundColor: siteVars.colors.grey[500],
    subtleBackgroundColor: siteVars.colorScheme.default.background,
    subtleForegroundColor: siteVars.colorScheme.default.foreground,
    subtleBorderColor: siteVars.colorScheme.onyx.border1,
    zIndex: siteVars.zIndexes.overlayPriority,
}); };
exports.tooltipContentVariables = tooltipContentVariables;
