// @fluentui/react-northstar/dist/dts/src/themes/teams/components/List/listItemVariables.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.listItemVariables = void 0;
var utils_1 = require("../../../../utils");
var listItemVariables = function (siteVariables) {
    return {
        minHeight: utils_1.pxToRem(48),
        rootPadding: "0 " + utils_1.pxToRem(18) + " 0 " + utils_1.pxToRem(20),
        gap: utils_1.pxToRem(8),
        zIndex: siteVariables.zIndexes.foreground,
        // Header
        // TODO: prod app uses 17.5px here, it should be 16px per the design guide!
        headerLineHeight: siteVariables.lineHeightSmall,
        headerFontSize: siteVariables.fontSizes.medium,
        // Header Media
        headerMediaFontSize: siteVariables.fontSizes.small,
        // TODO: prod app uses 17.5px here, it should be 16px per the design guide!
        headerMediaLineHeight: siteVariables.lineHeightSmall,
        importantFontWeight: siteVariables.fontWeightBold,
        // Content
        contentFontSize: siteVariables.fontSizes.small,
        contentLineHeight: siteVariables.lineHeightSmall,
        // Content Media
        contentMediaFontSize: siteVariables.fontSizes.small,
        contentMediaLineHeight: siteVariables.lineHeightSmall,
        // Selectable
        selectableFocusHoverColor: siteVariables.colorScheme.default.foregroundHover,
        selectableFocusHoverBackgroundColor: siteVariables.colorScheme.default.backgroundHover,
        // Selected
        selectedColor: siteVariables.colorScheme.default.foregroundPressed,
        selectedBackgroundColor: siteVariables.colorScheme.default.backgroundActive1,
    };
};
exports.listItemVariables = listItemVariables;
