// @fluentui/react-northstar/dist/dts/src/themes/teams/components/List/listItemMediaStyles.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.listItemMediaStyles = void 0;
var utils_1 = require("../../../../utils");
exports.listItemMediaStyles = {
    root: function (_a) {
        var p = _a.props, v = _a.variables;
        return (__assign(__assign({}, (p.important && {
            '::before': {
                content: '""',
                position: 'absolute',
                left: utils_1.pxToRem(8),
                width: utils_1.pxToRem(2),
                height: utils_1.pxToRem(2),
                background: '#000',
            },
        })), ((p.hasHeader || p.hasContent) && {
            marginRight: v.gap,
        })));
    },
};
