// @fluentui/react-northstar/dist/commonjs/themes/teams/components/Avatar/avatarLabelStyles.js
"use strict";

exports.__esModule = true;
exports.avatarLabelStyles = void 0;

var _utils = require("../../../../utils");

var sizeToPxValue = {
  smallest: 20,
  smaller: 24,
  small: 28,
  medium: 32,
  large: 44,
  larger: 64,
  largest: 96
};
var avatarLabelStyles = {
  root: function root(_ref) {
    var p = _ref.props,
        v = _ref.variables;
    var sizeInRem = (0, _utils.pxToRem)(sizeToPxValue[p.size]);
    return Object.assign({
      alignItems: 'center',
      overflow: 'hidden',
      color: v.labelColor,
      backgroundColor: v.labelBackground,
      borderRadius: '50%',
      display: 'inline-block',
      width: sizeInRem,
      height: sizeInRem,
      lineHeight: sizeInRem,
      fontSize: (0, _utils.pxToRem)(sizeToPxValue[p.size] / 2.333),
      verticalAlign: 'top',
      textAlign: 'center',
      padding: '0'
    }, p.square && {
      borderRadius: v.squareAvatarBorderRadius
    }, p.circular && {
      borderRadius: v.labelCircularRadius
    });
  }
};
exports.avatarLabelStyles = avatarLabelStyles;
