// @fluentui/react-northstar/dist/commonjs/themes/teams/components/Avatar/avatarVariables.js
"use strict";

exports.__esModule = true;
exports.avatarVariables = exports.labelColorAreas = void 0;

var _utils = require("../../../../utils");

var labelColorAreas = (0, _utils.stringLiteralsArray)('foreground', 'background');
exports.labelColorAreas = labelColorAreas;

var avatarVariables = function avatarVariables(siteVariables) {
  return {
    avatarBorderColor: '',
    avatarBorderWidth: '0',
    squareAvatarBorderRadius: siteVariables.borderRadiusMedium,
    iconColor: siteVariables.colors.white,
    iconBackgroundColor: siteVariables.colors.brand[600],
    statusBorderWidth: '2px',
    statusIconSize: (0, _utils.pxToRem)(7),
    statusBorderColor: siteVariables.bodyBackground,
    statusSuccessBackgroundColor: siteVariables.colorScheme.green.background,
    statusSuccessColor: siteVariables.colorScheme.green.foreground1,
    statusInfoBackgroundColor: siteVariables.colorScheme.brand.background,
    statusInfoColor: siteVariables.colorScheme.default.foreground2,
    statusWarningBackgroundColor: siteVariables.colorScheme.yellow.background,
    statusWarningColor: siteVariables.colorScheme.yellow.foreground2,
    statusErrorBackgroundColor: siteVariables.colorScheme.red.background,
    statusErrorColor: siteVariables.colorScheme.red.foreground2,
    statusBackgroundColor: siteVariables.colorScheme.default.background5,
    statusColor: siteVariables.colorScheme.default.foreground4,
    imageWidth: undefined,
    imageHeight: undefined,
    imageAvatarRadius: (0, _utils.pxToRem)(9999),
    imageAvatarSize: (0, _utils.pxToRem)(32),
    imageCircularRadius: (0, _utils.pxToRem)(9999),
    labelCircularRadius: (0, _utils.pxToRem)(9999),
    labelColor: 'rgba(0, 0, 0, 0.6)',
    labelBackground: 'rgb(232, 232, 232)'
  };
};

exports.avatarVariables = avatarVariables;
