// @fluentui/react-northstar/dist/dts/src/themes/teams/components/Input/inputLabelStyles.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.inputLabelStyles = void 0;
exports.inputLabelStyles = {
    root: function (_a) {
        var p = _a.props, v = _a.variables;
        return (__assign(__assign(__assign({ display: 'block', transition: 'all .2s', lineHeight: v.lineHeight, marginBottom: v.marginBottom }, (p.labelPosition === 'inside' && __assign({ bottom: v.insideLabelBottom, top: 0, left: 0, margin: 0, position: 'absolute', display: 'flex', alignItems: 'center', zIndex: 100, paddingLeft: v.insideLabelPaddingLeft }, (p.hasValue && {
            transform: 'translateY(-16px)',
            fontSize: v.insideLabelActiveFontSize,
        })))), (p.labelPosition === 'inline' && {
            paddingRight: v.inlineLabelPaddingRight,
        })), (p.required && {
            '::after': {
                content: '"*"',
            },
        })));
    },
};
