// @uifabric/utilities/lib-amd/index.js
define(["require", "exports", "tslib", "./Async", "./AutoScroll", "./BaseComponent", "./DelayedRender", "./EventGroup", "./FabricPerformance", "./GlobalSettings", "./KeyCodes", "./Rectangle", "./appendFunction", "./aria", "./array", "./asAsync", "./assertNever", "./classNamesFunction", "./componentAs/composeComponentAs", "./controlled", "./css", "./customizations/Customizations", "./customizations/Customizer", "./customizations/CustomizerContext", "./customizations/customizable", "./customizations/useCustomizationSettings", "./customizations/mergeCustomizations", "./customizations/mergeSettings", "./dom", "./extendComponent", "./focus", "./getId", "./getNativeElementProps", "./hoist", "./hoistStatics", "./initializeComponentRef", "./initializeFocusRects", "./useFocusRects", "./initials", "./keyboard", "./language", "./math", "./memoize", "./merge", "./mobileDetector", "./modalize", "./object", "./osDetector", "./overflow", "./properties", "./renderFunction/composeRenderFunction", "./resources", "./rtl", "./safeRequestAnimationFrame", "./safeSetTimeout", "./scroll", "./selection/index", "./string", "./styled", "./warn", "./ie11Detector", "./getPropsWithDefaults", "./setFocusVisibility", "./dom/setSSR", "./createMergedRef", "./version"], function (require, exports, tslib_1, Async_1, AutoScroll_1, BaseComponent_1, DelayedRender_1, EventGroup_1, FabricPerformance_1, GlobalSettings_1, KeyCodes_1, Rectangle_1, appendFunction_1, aria_1, array_1, asAsync_1, assertNever_1, classNamesFunction_1, composeComponentAs_1, controlled_1, css_1, Customizations_1, Customizer_1, CustomizerContext_1, customizable_1, useCustomizationSettings_1, mergeCustomizations_1, mergeSettings_1, dom_1, extendComponent_1, focus_1, getId_1, getNativeElementProps_1, hoist_1, hoistStatics_1, initializeComponentRef_1, initializeFocusRects_1, useFocusRects_1, initials_1, keyboard_1, language_1, math_1, memoize_1, merge_1, mobileDetector_1, modalize_1, object_1, osDetector_1, overflow_1, properties_1, composeRenderFunction_1, resources_1, rtl_1, safeRequestAnimationFrame_1, safeSetTimeout_1, scroll_1, index_1, string_1, styled_1, warn_1, ie11Detector_1, getPropsWithDefaults_1, setFocusVisibility_1, setSSR_1, createMergedRef_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    tslib_1.__exportStar(Async_1, exports);
    tslib_1.__exportStar(AutoScroll_1, exports);
    tslib_1.__exportStar(BaseComponent_1, exports);
    tslib_1.__exportStar(DelayedRender_1, exports);
    tslib_1.__exportStar(EventGroup_1, exports);
    tslib_1.__exportStar(FabricPerformance_1, exports);
    tslib_1.__exportStar(GlobalSettings_1, exports);
    tslib_1.__exportStar(KeyCodes_1, exports);
    tslib_1.__exportStar(Rectangle_1, exports);
    tslib_1.__exportStar(appendFunction_1, exports);
    tslib_1.__exportStar(aria_1, exports);
    tslib_1.__exportStar(array_1, exports);
    tslib_1.__exportStar(asAsync_1, exports);
    tslib_1.__exportStar(assertNever_1, exports);
    tslib_1.__exportStar(classNamesFunction_1, exports);
    tslib_1.__exportStar(composeComponentAs_1, exports);
    tslib_1.__exportStar(controlled_1, exports);
    tslib_1.__exportStar(css_1, exports);
    tslib_1.__exportStar(Customizations_1, exports);
    tslib_1.__exportStar(Customizer_1, exports);
    tslib_1.__exportStar(CustomizerContext_1, exports);
    tslib_1.__exportStar(customizable_1, exports);
    tslib_1.__exportStar(useCustomizationSettings_1, exports);
    tslib_1.__exportStar(mergeCustomizations_1, exports);
    tslib_1.__exportStar(mergeSettings_1, exports);
    tslib_1.__exportStar(dom_1, exports);
    tslib_1.__exportStar(extendComponent_1, exports);
    tslib_1.__exportStar(focus_1, exports);
    tslib_1.__exportStar(getId_1, exports);
    tslib_1.__exportStar(getNativeElementProps_1, exports);
    tslib_1.__exportStar(hoist_1, exports);
    tslib_1.__exportStar(hoistStatics_1, exports);
    tslib_1.__exportStar(initializeComponentRef_1, exports);
    tslib_1.__exportStar(initializeFocusRects_1, exports);
    tslib_1.__exportStar(useFocusRects_1, exports);
    tslib_1.__exportStar(initials_1, exports);
    tslib_1.__exportStar(keyboard_1, exports);
    tslib_1.__exportStar(language_1, exports);
    tslib_1.__exportStar(math_1, exports);
    tslib_1.__exportStar(memoize_1, exports);
    tslib_1.__exportStar(merge_1, exports);
    tslib_1.__exportStar(mobileDetector_1, exports);
    tslib_1.__exportStar(modalize_1, exports);
    exports.assign = object_1.assign;
    exports.filteredAssign = object_1.filteredAssign;
    exports.mapEnumByName = object_1.mapEnumByName;
    exports.shallowCompare = object_1.shallowCompare;
    exports.values = object_1.values;
    exports.omit = object_1.omit;
    tslib_1.__exportStar(osDetector_1, exports);
    tslib_1.__exportStar(overflow_1, exports);
    tslib_1.__exportStar(properties_1, exports);
    tslib_1.__exportStar(composeRenderFunction_1, exports);
    tslib_1.__exportStar(resources_1, exports);
    tslib_1.__exportStar(rtl_1, exports);
    tslib_1.__exportStar(safeRequestAnimationFrame_1, exports);
    tslib_1.__exportStar(safeSetTimeout_1, exports);
    tslib_1.__exportStar(scroll_1, exports);
    tslib_1.__exportStar(index_1, exports);
    tslib_1.__exportStar(string_1, exports);
    tslib_1.__exportStar(styled_1, exports);
    tslib_1.__exportStar(warn_1, exports);
    tslib_1.__exportStar(ie11Detector_1, exports);
    tslib_1.__exportStar(getPropsWithDefaults_1, exports);
    exports.setFocusVisibility = setFocusVisibility_1.setFocusVisibility;
    exports.IsFocusVisibleClassName = setFocusVisibility_1.IsFocusVisibleClassName;
    exports.setSSR = setSSR_1.setSSR;
    exports.createMergedRef = createMergedRef_1.createMergedRef;
});
