// @fluentui/react-northstar/dist/commonjs/themes/teams-dark/componentVariables.js
"use strict";

exports.__esModule = true;
exports.SplitButtonDivider = exports.DatepickerCalendarCellButton = exports.DatepickerCalendarCell = exports.Label = exports.Provider = exports.Reaction = exports.SvgIcon = exports.MenuItemWrapper = exports.MenuItemIndicator = exports.MenuItemContent = exports.MenuItemIcon = exports.MenuItem = exports.MenuDivider = exports.Menu = exports.Text = exports.Segment = exports.RadioGroupItem = exports.HeaderDescription = exports.Header = exports.Divider = exports.Dialog = exports.ChatMessage = exports.Avatar = void 0;

var _avatarVariables = require("./components/Avatar/avatarVariables");

exports.Avatar = _avatarVariables.avatarVariables;

var _chatMessageVariables = require("./components/Chat/chatMessageVariables");

exports.ChatMessage = _chatMessageVariables.chatMessageVariables;

var _dialogVariables = require("./components/Dialog/dialogVariables");

exports.Dialog = _dialogVariables.dialogVariables;

var _dividerVariables = require("./components/Divider/dividerVariables");

exports.Divider = _dividerVariables.dividerVariables;

var _headerVariables = require("./components/Header/headerVariables");

exports.Header = _headerVariables.headerVariables;

var _headerDescriptionVariables = require("./components/Header/headerDescriptionVariables");

exports.HeaderDescription = _headerDescriptionVariables.headerDescriptionVariables;

var _radioGroupItemVariables = require("./components/RadioGroup/radioGroupItemVariables");

exports.RadioGroupItem = _radioGroupItemVariables.radioGroupItemVariables;

var _segmentVariables = require("./components/Segment/segmentVariables");

exports.Segment = _segmentVariables.segmentVariables;

var _textVariables = require("./components/Text/textVariables");

exports.Text = _textVariables.textVariables;

var _menuVariables = require("./components/Menu/menuVariables");

exports.Menu = _menuVariables.menuVariables;

var _menuDividerVariables = require("./components/Menu/menuDividerVariables");

exports.MenuDivider = _menuDividerVariables.menuDividerVariables;

var _menuItemVariables = require("./components/Menu/menuItemVariables");

exports.MenuItem = _menuItemVariables.menuItemVariables;

var _menuItemIconVariables = require("./components/Menu/menuItemIconVariables");

exports.MenuItemIcon = _menuItemIconVariables.menuItemIconVariables;

var _menuItemContentVariables = require("./components/Menu/menuItemContentVariables");

exports.MenuItemContent = _menuItemContentVariables.menuItemContentVariables;

var _menuItemIndicatorVariables = require("./components/Menu/menuItemIndicatorVariables");

exports.MenuItemIndicator = _menuItemIndicatorVariables.menuItemIndicatorVariables;

var _menuItemWrapperVariables = require("./components/Menu/menuItemWrapperVariables");

exports.MenuItemWrapper = _menuItemWrapperVariables.menuItemWrapperVariables;

var _svgIconVariables = require("./components/SvgIcon/svgIconVariables");

exports.SvgIcon = _svgIconVariables.svgIconVariables;

var _reactionVariables = require("./components/Reaction/reactionVariables");

exports.Reaction = _reactionVariables.reactionVariables;

var _providerVariables = require("./components/Provider/providerVariables");

exports.Provider = _providerVariables.providerVariables;

var _labelVariables = require("./components/Label/labelVariables");

exports.Label = _labelVariables.labelVariables;

var _datepickerCalendarCellVariables = require("./components/Datepicker/datepickerCalendarCellVariables");

exports.DatepickerCalendarCell = _datepickerCalendarCellVariables.datepickerCalendarCellVariables;

var _datepickerCalendarCellButtonVariables = require("./components/Datepicker/datepickerCalendarCellButtonVariables");

exports.DatepickerCalendarCellButton = _datepickerCalendarCellButtonVariables.datepickerCalendarCellButtonVariables;

var _splitButtonVariables = require("./components/SplitButton/splitButtonVariables");

exports.SplitButtonDivider = _splitButtonVariables.splitButtonVariables;
