// @fluentui/react-northstar/dist/commonjs/themes/teams-dark/siteVariables.js
"use strict";

exports.__esModule = true;
exports.bodyColor = exports.bodyBackground = exports.shadow64 = exports.shadow28 = exports.shadow16 = exports.shadow8 = exports.shadow4 = exports.shadow2 = exports.shadowLevel4 = exports.shadowLevel3 = exports.shadowLevel2 = exports.shadowLevel1 = exports.focusOuterBorderColor = exports.focusInnerBorderColor = exports.categoryColorScheme = exports.colorScheme = void 0;

var _siteVariables = require("../teams/siteVariables");

var _colors = require("./colors");

exports.colorScheme = _colors.colorScheme;
exports.categoryColorScheme = _colors.categoryColorScheme;
//
// BORDER STYLES
//
var focusInnerBorderColor = _siteVariables.colors.black;
exports.focusInnerBorderColor = focusInnerBorderColor;
var focusOuterBorderColor = _siteVariables.colors.white; //
// SHADOW LEVELS
//

exports.focusOuterBorderColor = focusOuterBorderColor;
var shadowLevel1 = '0 .2rem .4rem -.075rem rgba(0, 0, 0, .25)';
exports.shadowLevel1 = shadowLevel1;
var shadowLevel2 = '0 .4rem 0.7rem -0.1rem rgba(0, 0, 0, .25)';
exports.shadowLevel2 = shadowLevel2;
var shadowLevel3 = '0 .8rem 1rem -0.2rem rgba(0, 0, 0, .25)';
exports.shadowLevel3 = shadowLevel3;
var shadowLevel4 = '0 .6rem 1.8rem -0.4rem rgba(0, 0, 0, .25)';
exports.shadowLevel4 = shadowLevel4;
var shadow2 = '0 0 2px rgba(0, 0, 0, .24), 0 1px 2px rgba(0, 0, 0, .28)';
exports.shadow2 = shadow2;
var shadow4 = '0 0 2px rgba(0, 0, 0, .24), 0 2px 4px rgba(0, 0, 0, .28)';
exports.shadow4 = shadow4;
var shadow8 = '0 0 2px rgba(0, 0, 0, .24), 0 4px 8px rgba(0, 0, 0, .28)';
exports.shadow8 = shadow8;
var shadow16 = '0 0 2px rgba(0, 0, 0, .24), 0 8px 16px rgba(0, 0, 0, .28)';
exports.shadow16 = shadow16;
var shadow28 = '0 0 8px rgba(0, 0, 0, .40), 0 14px 28px rgba(0, 0, 0, .48)';
exports.shadow28 = shadow28;
var shadow64 = '0 0 8px rgba(0, 0, 0, .40), 0 32px 64px rgba(0, 0, 0, .48)'; //
// SEMANTIC ASSIGNMENTS
//

exports.shadow64 = shadow64;
var bodyBackground = _siteVariables.colors.grey[650];
exports.bodyBackground = bodyBackground;
var bodyColor = _siteVariables.colors.white;
exports.bodyColor = bodyColor;
