// @fluentui/react-northstar/dist/dts/src/components/SplitButton/SplitButtonDivider.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SplitButtonDivider = exports.SplitButtonDividerClassName = void 0;
var React = require("react");
var PropTypes = require("prop-types");
var react_bindings_1 = require("@fluentui/react-bindings");
var utils_1 = require("../../utils");
exports.SplitButtonDividerClassName = 'ui-splitButtonDivider';
/**
 * A SplitButtonDivider visually segments content.
 */
exports.SplitButtonDivider = React.forwardRef(function (props, ref) {
    var context = react_bindings_1.useFluentContext();
    var _a = react_bindings_1.useTelemetry(exports.SplitButtonDivider.displayName, context.telemetry), setStart = _a.setStart, setEnd = _a.setEnd;
    setStart();
    var className = props.className, design = props.design, styles = props.styles, variables = props.variables, accessibility = props.accessibility;
    var ElementType = react_bindings_1.getElementType(props);
    var primary = props.primary;
    var unhandledProps = react_bindings_1.useUnhandledProps(exports.SplitButtonDivider.handledProps, props);
    var getA11yProps = react_bindings_1.useAccessibility(accessibility, {
        debugName: exports.SplitButtonDivider.displayName,
        rtl: context.rtl,
    });
    var classes = react_bindings_1.useStyles(exports.SplitButtonDivider.displayName, {
        className: exports.SplitButtonDividerClassName,
        mapPropsToStyles: function () { return ({
            primary: primary,
        }); },
        mapPropsToInlineStyles: function () { return ({
            className: className,
            design: design,
            styles: styles,
            variables: variables,
        }); },
        rtl: context.rtl,
    }).classes;
    var element = (React.createElement(ElementType, __assign({}, getA11yProps('root', __assign({ className: classes.root, ref: ref }, unhandledProps)))));
    setEnd();
    return element;
});
exports.SplitButtonDivider.displayName = 'SplitButtonDivider';
exports.SplitButtonDivider.propTypes = __assign(__assign({}, utils_1.commonPropTypes.createCommon()), { primary: PropTypes.bool });
exports.SplitButtonDivider.handledProps = Object.keys(exports.SplitButtonDivider.propTypes);
