// @fluentui/react-northstar/dist/commonjs/utils/createComponent.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.createComponent = void 0;

var _pick2 = _interopRequireDefault(require("lodash/pick"));

var _createComponentInternal = require("./createComponentInternal");

var createComponent = function createComponent(_ref) {
  var displayName = _ref.displayName,
      className = _ref.className,
      _render = _ref.render,
      defaultProps = _ref.defaultProps,
      actionHandlers = _ref.actionHandlers;
  return (0, _createComponentInternal.createComponentInternal)({
    displayName: displayName,
    className: className,
    render: function render(config, props) {
      var filteredConfig = (0, _pick2.default)(config, ['accessibility', 'classes', 'rtl', 'styles']);
      return _render(Object.assign({
        config: filteredConfig
      }, props));
    },
    defaultProps: defaultProps,
    actionHandlers: actionHandlers
  });
};

exports.createComponent = createComponent;
