// @fluentui/react-northstar/dist/commonjs/utils/index.js
"use strict";

exports.__esModule = true;
var _exportNames = {
  childrenExist: true,
  rtlTextContainer: true,
  stringLiteralsArray: true,
  getOrGenerateIdFromShorthand: true,
  mergeProviderContexts: true,
  renderComponent: true,
  htmlImageProps: true,
  htmlInputAttrs: true,
  htmlInputEvents: true,
  htmlInputProps: true,
  partitionHTMLProps: true,
  isBrowser: true,
  doesNodeContainClick: true,
  pxToRem: true,
  createComponent: true,
  getKindProp: true,
  shouldPreventDefaultOnKeyDown: true,
  screenReaderContainerStyles: true
};
exports.screenReaderContainerStyles = exports.shouldPreventDefaultOnKeyDown = exports.getKindProp = exports.createComponent = exports.pxToRem = exports.doesNodeContainClick = exports.isBrowser = exports.partitionHTMLProps = exports.htmlInputProps = exports.htmlInputEvents = exports.htmlInputAttrs = exports.htmlImageProps = exports.renderComponent = exports.mergeProviderContexts = exports.getOrGenerateIdFromShorthand = exports.stringLiteralsArray = exports.rtlTextContainer = exports.childrenExist = void 0;

var _reactBindings = require("@fluentui/react-bindings");

exports.childrenExist = _reactBindings.childrenExist;

var _rtlTextContainer = require("./rtlTextContainer");

exports.rtlTextContainer = _rtlTextContainer.rtlTextContainer;

var _stringLiteralsArray = require("./stringLiteralsArray");

exports.stringLiteralsArray = _stringLiteralsArray.stringLiteralsArray;

var _getOrGenerateIdFromShorthand = require("./getOrGenerateIdFromShorthand");

exports.getOrGenerateIdFromShorthand = _getOrGenerateIdFromShorthand.getOrGenerateIdFromShorthand;

var _factories = require("./factories");

Object.keys(_factories).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _factories[key]) return;
  exports[key] = _factories[key];
});

var _mergeProviderContexts = require("./mergeProviderContexts");

exports.mergeProviderContexts = _mergeProviderContexts.mergeProviderContexts;

var _renderComponent = require("./renderComponent");

exports.renderComponent = _renderComponent.renderComponent;
Object.keys(_renderComponent).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _renderComponent[key]) return;
  exports[key] = _renderComponent[key];
});

var _htmlPropsUtils = require("./htmlPropsUtils");

exports.htmlImageProps = _htmlPropsUtils.htmlImageProps;
exports.htmlInputAttrs = _htmlPropsUtils.htmlInputAttrs;
exports.htmlInputEvents = _htmlPropsUtils.htmlInputEvents;
exports.htmlInputProps = _htmlPropsUtils.htmlInputProps;
exports.partitionHTMLProps = _htmlPropsUtils.partitionHTMLProps;

var _isBrowser = require("./isBrowser");

exports.isBrowser = _isBrowser.isBrowser;

var _doesNodeContainClick = require("./doesNodeContainClick");

exports.doesNodeContainClick = _doesNodeContainClick.doesNodeContainClick;

var _fontSizeUtility = require("./fontSizeUtility");

exports.pxToRem = _fontSizeUtility.pxToRem;

var _createComponent = require("./createComponent");

exports.createComponent = _createComponent.createComponent;

var _getKindProp = require("./getKindProp");

exports.getKindProp = _getKindProp.getKindProp;

var _shouldPreventDefaultOnKeyDown = require("./shouldPreventDefaultOnKeyDown");

exports.shouldPreventDefaultOnKeyDown = _shouldPreventDefaultOnKeyDown.shouldPreventDefaultOnKeyDown;

var _whatInput = require("./whatInput");

Object.keys(_whatInput).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _whatInput[key]) return;
  exports[key] = _whatInput[key];
});

var _accessibilityStyles = require("./accessibility/Styles/accessibilityStyles");

exports.screenReaderContainerStyles = _accessibilityStyles.screenReaderContainerStyles;

var _commonPropTypes = require("./commonPropTypes");

Object.keys(_commonPropTypes).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _commonPropTypes[key]) return;
  exports[key] = _commonPropTypes[key];
});
