// @fluentui/react-northstar/dist/dts/test/specs/utils/date-time-utils/dateGrid/getBoundedDateRange-test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var DateGrid = require("../../../../../src/utils/date-time-utilities/dateGrid/getBoundedDateRange");
var Months;
(function (Months) {
    Months[Months["Jan"] = 0] = "Jan";
    Months[Months["Feb"] = 1] = "Feb";
    Months[Months["Mar"] = 2] = "Mar";
    Months[Months["Apr"] = 3] = "Apr";
    Months[Months["May"] = 4] = "May";
    Months[Months["Jun"] = 5] = "Jun";
    Months[Months["Jul"] = 6] = "Jul";
    Months[Months["Aug"] = 7] = "Aug";
    Months[Months["Sep"] = 8] = "Sep";
    Months[Months["Oct"] = 9] = "Oct";
    Months[Months["Nov"] = 10] = "Nov";
    Months[Months["Dec"] = 11] = "Dec";
})(Months || (Months = {}));
describe('getBoundedDateRange', function () {
    var defaultRange = [
        new Date(2016, Months.Apr, 1),
        new Date(2016, Months.Apr, 2),
        new Date(2016, Months.Apr, 3),
        new Date(2016, Months.Apr, 4),
    ];
    it('returns same range if min and max dates are empty', function () {
        var result = DateGrid.getBoundedDateRange(defaultRange);
        var expected = defaultRange;
        expect(result).toEqual(expected);
    });
    it('returns adjusted range if min date is present', function () {
        var result = DateGrid.getBoundedDateRange(defaultRange, new Date(2016, Months.Apr, 3));
        var expected = [new Date(2016, Months.Apr, 3), new Date(2016, Months.Apr, 4)];
        expect(result).toEqual(expected);
    });
    it('returns adjusted range if max date is present', function () {
        var result = DateGrid.getBoundedDateRange(defaultRange, undefined, new Date(2016, Months.Apr, 2));
        var expected = [new Date(2016, Months.Apr, 1), new Date(2016, Months.Apr, 2)];
        expect(result).toEqual(expected);
    });
    it('returns adjusted range if min and max dates are present', function () {
        var result = DateGrid.getBoundedDateRange(defaultRange, new Date(2016, Months.Apr, 3), new Date(2016, Months.Apr, 3));
        var expected = [new Date(2016, Months.Apr, 3)];
        expect(result).toEqual(expected);
    });
});
