// @fluentui/react-northstar/dist/dts/src/themes/teams/components/Accordion/accordionTitleStyles.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.accordionTitleStyles = void 0;
var activeIndicatorUrl_1 = require("./activeIndicatorUrl");
var utils_1 = require("../../../../utils");
exports.accordionTitleStyles = {
    root: function (_a) {
        var p = _a.props;
        return ({
            display: 'inline-block',
            verticalAlign: 'middle',
            padding: '.5rem 0',
            cursor: p.disabled ? 'default' : 'pointer',
        });
    },
    contentWrapper: function (_a) {
        var p = _a.props;
        return (__assign({ display: 'grid', gridTemplateColumns: 'auto', msGridColumns: 'auto', userSelect: 'none' }, (p.content && {
            gridTemplateColumns: 'auto 1fr',
            msGridColumns: 'auto 1fr',
        })));
    },
    indicator: function (_a) {
        var p = _a.props, v = _a.variables, rtl = _a.rtl;
        return (__assign(__assign({ alignItems: 'center', display: 'inline-flex', backgroundImage: activeIndicatorUrl_1.activeIndicatorUrl(v.color, p.active), backgroundPosition: 'center' }, (rtl && {
            transform: "scaleX(-1)",
        })), { content: '" "', overflow: 'hidden', height: '100%', width: utils_1.pxToRem(16), msGridColumn: '1', marginRight: "" + utils_1.pxToRem(4) }));
    },
    content: function () { return ({
        alignItems: 'center',
        display: 'grid',
        msGridColumn: '2',
    }); },
};
