// @fluentui/react-northstar/dist/commonjs/components/Datepicker/DatepickerCalendarGridRow.js
"use strict";

exports.__esModule = true;
exports.DatepickerCalendarGridRow = exports.datepickerCalendarGridRowClassName = void 0;

var _reactBindings = require("@fluentui/react-bindings");

var _utils = require("../../utils");

var _Box = require("../Box/Box");

var datepickerCalendarGridRowClassName = 'ui-datepicker__calendargridrow';
/**
 * A DatepickerCalendarGridRow is used to display the calendar grid row.
 */

exports.datepickerCalendarGridRowClassName = datepickerCalendarGridRowClassName;
var DatepickerCalendarGridRow = (0, _reactBindings.compose)(_Box.Box, {
  className: datepickerCalendarGridRowClassName,
  displayName: 'DatepickerCalendarGridRow',
  handledProps: ['isRowSelectionActive'],
  overrideStyles: true,
  mapPropsToStylesProps: function mapPropsToStylesProps(_ref) {
    var isRowSelectionActive = _ref.isRowSelectionActive;
    return {
      isRowSelectionActive: isRowSelectionActive
    };
  },
  shorthandConfig: {
    mappedProp: 'content'
  }
});
exports.DatepickerCalendarGridRow = DatepickerCalendarGridRow;
DatepickerCalendarGridRow.defaultProps = {
  as: 'tr'
};
DatepickerCalendarGridRow.propTypes = _utils.commonPropTypes.createCommon();
