// @fluentui/react-northstar/dist/commonjs/components/Datepicker/DatepickerCalendarHeader.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.DatepickerCalendarHeader = exports.datepickerCalendarHeaderClassName = void 0;

var _invoke2 = _interopRequireDefault(require("lodash/invoke"));

var React = _interopRequireWildcard(require("react"));

var _accessibility = require("@fluentui/accessibility");

var _dateTimeUtilities = require("../../utils/date-time-utilities");

var _reactBindings = require("@fluentui/react-bindings");

var customPropTypes = _interopRequireWildcard(require("@fluentui/react-proptypes"));

var PropTypes = _interopRequireWildcard(require("prop-types"));

var _utils = require("../../utils");

var _DatepickerCalendarHeaderAction = require("./DatepickerCalendarHeaderAction");

var _Text = require("../Text/Text");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var datepickerCalendarHeaderClassName = 'ui-datepicker__calendarheader';
/**
 * A DatepickerCalendarHeader is used to display header block above calendar grid.
 */

exports.datepickerCalendarHeaderClassName = datepickerCalendarHeaderClassName;
var DatepickerCalendarHeader = /*#__PURE__*/React.forwardRef(function (props, ref) {
  var context = (0, _reactBindings.useFluentContext)();

  var _useTelemetry = (0, _reactBindings.useTelemetry)(DatepickerCalendarHeader.displayName, context.telemetry),
      setStart = _useTelemetry.setStart,
      setEnd = _useTelemetry.setEnd;

  setStart();
  var className = props.className,
      design = props.design,
      styles = props.styles,
      variables = props.variables,
      label = props.label,
      nextButton = props.nextButton,
      previousButton = props.previousButton,
      onPreviousClick = props.onPreviousClick,
      onNextClick = props.onNextClick;
  var ElementType = (0, _reactBindings.getElementType)(props);
  var unhandledProps = (0, _reactBindings.useUnhandledProps)(DatepickerCalendarHeader.handledProps, props);
  var getA11yProps = (0, _reactBindings.useAccessibility)(props.accessibility, {
    debugName: DatepickerCalendarHeader.displayName,
    actionHandlers: {},
    rtl: context.rtl
  });

  var _useStyles = (0, _reactBindings.useStyles)(DatepickerCalendarHeader.displayName, {
    className: datepickerCalendarHeaderClassName,
    mapPropsToInlineStyles: function mapPropsToInlineStyles() {
      return {
        className: className,
        design: design,
        styles: styles,
        variables: variables
      };
    },
    rtl: context.rtl
  }),
      classes = _useStyles.classes;

  var element = /*#__PURE__*/React.createElement(ElementType, getA11yProps('root', Object.assign({
    className: classes.root,
    ref: ref
  }, unhandledProps)), (0, _utils.createShorthand)(_Text.Text, label, {
    defaultProps: function defaultProps() {
      return getA11yProps('label', {
        className: classes.label
      });
    }
  }), (0, _utils.createShorthand)(_DatepickerCalendarHeaderAction.DatepickerCalendarHeaderAction, previousButton, {
    defaultProps: function defaultProps() {
      return getA11yProps('previousButton', {
        title: props.prevMonthAriaLabel,
        direction: 'previous',
        'aria-disabled': props.disabledPreviousButton,
        disabledNavigatableButton: props.disabledPreviousButton
      });
    },
    overrideProps: function overrideProps(predefinedProps) {
      return {
        onClick: function onClick(e, data) {
          if (!props.disabledPreviousButton) {
            onPreviousClick(e, data);
            (0, _invoke2.default)(predefinedProps, 'onClick', e, data);
          }
        }
      };
    }
  }), (0, _utils.createShorthand)(_DatepickerCalendarHeaderAction.DatepickerCalendarHeaderAction, nextButton, {
    defaultProps: function defaultProps() {
      return getA11yProps('nextButton', {
        title: props.nextMonthAriaLabel,
        direction: 'next',
        'aria-disabled': props.disabledNextButton,
        disabledNavigatableButton: props.disabledNextButton
      });
    },
    overrideProps: function overrideProps(predefinedProps) {
      return {
        onClick: function onClick(e, data) {
          if (!props.disabledNextButton) {
            onNextClick(e, data);
            (0, _invoke2.default)(predefinedProps, 'onClick', e, data);
          }
        }
      };
    }
  }));
  setEnd();
  return element;
});
exports.DatepickerCalendarHeader = DatepickerCalendarHeader;
DatepickerCalendarHeader.displayName = 'DatepickerCalendarHeader';
DatepickerCalendarHeader.propTypes = Object.assign({}, _utils.commonPropTypes.createCommon(), {
  label: customPropTypes.itemShorthand,
  nextButton: customPropTypes.itemShorthand,
  previousButton: customPropTypes.itemShorthand,
  onPreviousClick: PropTypes.func,
  onNextClick: PropTypes.func,
  disabledNextButton: PropTypes.bool,
  disabledPreviousButton: PropTypes.bool,
  prevMonthAriaLabel: PropTypes.string,
  nextMonthAriaLabel: PropTypes.string
});
DatepickerCalendarHeader.defaultProps = {
  accessibility: _accessibility.datepickerCalendarHeaderBehavior,
  nextButton: {},
  previousButton: {},
  label: {},
  prevMonthAriaLabel: _dateTimeUtilities.DEFAULT_CALENDAR_STRINGS.prevMonthAriaLabel,
  nextMonthAriaLabel: _dateTimeUtilities.DEFAULT_CALENDAR_STRINGS.nextMonthAriaLabel
};
DatepickerCalendarHeader.handledProps = Object.keys(DatepickerCalendarHeader.propTypes);
DatepickerCalendarHeader.create = (0, _utils.createShorthandFactory)({
  Component: DatepickerCalendarHeader
});
