// @fluentui/react-northstar/dist/commonjs/themes/teams/components/Tree/treeItemStyles.js
"use strict";

exports.__esModule = true;
exports.treeItemStyles = void 0;

var _getBorderFocusStyles = require("../../getBorderFocusStyles");

var _TreeTitle = require("../../../../components/Tree/TreeTitle");

var treeItemStyles = {
  root: function root(_ref) {
    var _ref2, _ref3, _Object$assign;

    var siteVariables = _ref.theme.siteVariables,
        p = _ref.props;
    var borderFocusStyles = (0, _getBorderFocusStyles.getBorderFocusStyles)({
      variables: siteVariables
    });
    return {
      listStyleType: 'none',
      ':focus': Object.assign({}, p.selectable && (_ref2 = {}, _ref2["& ." + _TreeTitle.treeTitleSlotClassNames.indicator] = {
        display: 'inline-block'
      }, _ref2), borderFocusStyles[':focus']),
      ':focus-visible': Object.assign({
        outline: 0
      }, p.selectable && (_ref3 = {}, _ref3["& ." + _TreeTitle.treeTitleSlotClassNames.indicator] = {
        display: 'inline-block'
      }, _ref3), (_Object$assign = {}, _Object$assign["> ." + _TreeTitle.treeTitleClassName] = Object.assign({
        position: 'relative'
      }, borderFocusStyles[':focus-visible']), _Object$assign))
    };
  }
};
exports.treeItemStyles = treeItemStyles;
