// @fluentui/react-northstar/dist/commonjs/themes/teams/components/Tree/treeTitleVariables.js
"use strict";

exports.__esModule = true;
exports.treeTitleVariables = void 0;

var _utils = require("../../../../utils");

var treeTitleVariables = function treeTitleVariables(siteVars) {
  return {
    color: siteVars.colorScheme.default.foreground,
    padding: (0, _utils.pxToRem)(1) + " 0",
    paddingRight: (0, _utils.pxToRem)(20),
    paddingLeft: (0, _utils.pxToRem)(20),
    borderColor: siteVars.colorScheme.default.foreground1,
    borderStyle: 'solid',
    borderRadius: siteVars.borderRadiusMedium,
    borderWidth: (0, _utils.pxToRem)(1),
    indicatorColor: 'transparent',
    selectionIndicatorMargin: "0 0 0 0",
    background: 'transparent',
    hoverBackground: siteVars.colorScheme.default.background1,
    checkedBackground: siteVars.colorScheme.brand.backgroundActive1,
    checkedBorderColor: siteVars.colorScheme.brand.backgroundActive1,
    checkedIndicatorColor: siteVars.colorScheme.default.background,
    disabledBackground: siteVars.colorScheme.default.background,
    disabledBackgroundChecked: siteVars.colorScheme.default.backgroundDisabled,
    disabledBorderColor: siteVars.colorScheme.default.foregroundDisabled1,
    disabledCheckedIndicatorColor: siteVars.colorScheme.default.foregroundDisabled
  };
};

exports.treeTitleVariables = treeTitleVariables;
