// @fluentui/react-northstar/dist/dts/src/themes/teams/components/Attachment/attachmentVariables.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.attachmentVariables = void 0;
var utils_1 = require("../../../../utils");
var attachmentVariables = function (siteVariables) { return ({
    padding: utils_1.pxToRem(7) + " " + utils_1.pxToRem(3) + " " + utils_1.pxToRem(7) + " " + utils_1.pxToRem(11),
    iconSpace: utils_1.pxToRem(12),
    iconSize: utils_1.pxToRem(32),
    borderColor: siteVariables.colorScheme.default.border3,
    borderRadius: siteVariables.borderRadiusMedium,
    backgroundColor: siteVariables.colorScheme.default.background4,
    backgroundColorHover: siteVariables.colorScheme.default.backgroundHover1,
    textColor: siteVariables.colorScheme.default.foreground,
    textColorHover: siteVariables.colorScheme.default.foregroundHover,
    boxShadow: siteVariables.shadowLevel1,
    focusBackgroundColor: undefined,
    focusColor: undefined,
    progressColor: siteVariables.colorScheme.green.background,
    progressHeight: utils_1.pxToRem(4),
    headerFontSize: siteVariables.fontSizes.medium,
    headerFontWeight: siteVariables.fontWeightSemibold,
    headerLineHeight: siteVariables.lineHeightMedium,
    descriptionFontSize: siteVariables.fontSizes.small,
    descriptionFontWeight: siteVariables.fontWeightRegular,
    descriptionLineHeight: siteVariables.lineHeightDefault,
    // action variables
    actionHeight: utils_1.pxToRem(32),
    actionMaxWidth: utils_1.pxToRem(280),
    actionColor: siteVariables.colorScheme.default.foreground,
    actionPrimaryColor: siteVariables.colorScheme.brand.foreground,
    actionColorDisabled: siteVariables.colorScheme.brand.foregroundDisabled1,
    actionIconSize: utils_1.pxToRem(16),
    actionLoaderBorderSize: utils_1.pxToRem(2),
    actionLoaderSize: utils_1.pxToRem(20),
    actionLoaderSvgHeight: utils_1.pxToRem(1220),
    actionLoaderSvgAnimationHeight: utils_1.pxToRem(-1200),
    actionFocusBorderRadius: siteVariables.borderRadiusMedium,
}); };
exports.attachmentVariables = attachmentVariables;
