// @fluentui/react-northstar/dist/dts/src/themes/teams/components/Attachment/attachmentStyles.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.attachmentStyles = void 0;
var react_icons_northstar_1 = require("@fluentui/react-icons-northstar");
var AttachmentAction_1 = require("../../../../components/Attachment/AttachmentAction");
var utils_1 = require("../../../../utils");
var getBorderFocusStyles_1 = require("../../getBorderFocusStyles");
exports.attachmentStyles = {
    root: function (_a) {
        var _b, _c;
        var p = _a.props, v = _a.variables, siteVariables = _a.theme.siteVariables;
        var borderFocusStyles = getBorderFocusStyles_1.getBorderFocusStyles({
            variables: siteVariables,
            borderRadius: v.borderRadius,
        });
        return __assign(__assign(__assign({ position: 'relative', display: 'inline-flex', alignItems: 'center', width: '100%', maxWidth: utils_1.pxToRem(440), minHeight: utils_1.pxToRem(48), padding: v.padding, marginBottom: utils_1.pxToRem(2), marginRight: utils_1.pxToRem(2), background: v.backgroundColor, color: v.textColor, boxShadow: v.boxShadow, border: siteVariables.borderWidth + " solid " + v.borderColor, borderRadius: v.borderRadius }, borderFocusStyles), { '& .ui-attachment__progress': {
                transition: 'width 0.2s',
                position: 'absolute',
                display: 'block',
                bottom: 0,
                left: 0,
                maxWidth: '100%',
                height: v.progressHeight,
                background: v.progressColor,
            } }), (p.actionable && {
            cursor: 'pointer',
            ':focus-visible': __assign(__assign({}, borderFocusStyles[':focus-visible']), (_b = { backgroundColor: v.focusBackgroundColor, color: v.focusColor }, _b["& ." + AttachmentAction_1.attachmentActionClassName] = {
                color: v.focusColor,
            }, _b["& ." + react_icons_northstar_1.svgIconClassName] = {
                color: v.focusColor,
            }, _b)),
            ':hover': (_c = {
                    backgroundColor: v.backgroundColorHover,
                    color: v.textColorHover
                },
                _c["& ." + AttachmentAction_1.attachmentActionClassName] = {
                    color: v.textColorHover,
                },
                _c["& ." + react_icons_northstar_1.svgIconClassName] = {
                    color: v.textColorHover,
                },
                _c),
        }));
    },
};
