// @fluentui/react-northstar/dist/dts/src/components/Alert/AlertDismissAction.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AlertDismissAction = exports.alertDismissActionSlotClassNames = exports.alertDismissActionClassName = void 0;
var accessibility_1 = require("@fluentui/accessibility");
var PropTypes = require("prop-types");
var React = require("react");
var _ = require("lodash");
var utils_1 = require("../../utils");
var react_bindings_1 = require("@fluentui/react-bindings");
var Box_1 = require("../Box/Box");
exports.alertDismissActionClassName = 'ui-alert__dismissaction';
exports.alertDismissActionSlotClassNames = {
    content: exports.alertDismissActionClassName + "__content",
};
/**
 * A AlertDismissAction allows users to customize the dismissAction slot  inside the Alert component.
 *
 * @accessibility
 * Implements [ARIA Button](https://www.w3.org/TR/wai-aria-practices-1.1/#button) design pattern.
 */
exports.AlertDismissAction = React.forwardRef(function (props, ref) {
    var context = react_bindings_1.useFluentContext();
    var _a = react_bindings_1.useTelemetry(exports.AlertDismissAction.displayName, context.telemetry), setStart = _a.setStart, setEnd = _a.setEnd;
    setStart();
    var accessibility = props.accessibility, as = props.as, children = props.children, className = props.className, content = props.content, disabled = props.disabled, design = props.design, styles = props.styles, variables = props.variables, danger = props.danger, warning = props.warning, info = props.info, success = props.success;
    var hasChildren = utils_1.childrenExist(children);
    var hasContent = !!content && !_.isEmpty(content);
    var getA11Props = react_bindings_1.useAccessibility(accessibility, {
        debugName: exports.AlertDismissAction.displayName,
        mapPropsToBehavior: function () { return ({
            as: String(as),
            disabled: disabled,
        }); },
        actionHandlers: {
            performClick: function (event) {
                event.preventDefault();
                handleClick(event);
            },
        },
        rtl: context.rtl,
    });
    var _b = react_bindings_1.useStyles(exports.AlertDismissAction.displayName, {
        className: exports.alertDismissActionClassName,
        mapPropsToStyles: function () { return ({
            disabled: disabled,
            danger: danger,
            warning: warning,
            info: info,
            success: success,
            hasContent: hasContent || hasChildren,
        }); },
        mapPropsToInlineStyles: function () { return ({
            className: className,
            design: design,
            styles: styles,
            variables: variables,
        }); },
        rtl: context.rtl,
    }), classes = _b.classes, resolvedStyles = _b.styles;
    var unhandledProps = react_bindings_1.useUnhandledProps(exports.AlertDismissAction.handledProps, props);
    var ElementType = react_bindings_1.getElementType(props);
    var handleClick = function (e) {
        if (disabled) {
            e.preventDefault();
            return;
        }
        _.invoke(props, 'onClick', e, props);
    };
    var result = (React.createElement(ElementType, __assign({}, utils_1.rtlTextContainer.getAttributes({ forElements: [children] }), getA11Props('root', __assign(__assign({ onClick: handleClick, disabled: disabled, className: classes.root }, unhandledProps), { ref: ref }))), hasChildren
        ? children
        : Box_1.Box.create(content, {
            defaultProps: function () {
                return getA11Props('content', {
                    as: 'span',
                    className: exports.alertDismissActionSlotClassNames.content,
                    styles: resolvedStyles.content,
                });
            },
        })));
    setEnd();
    return result;
});
exports.AlertDismissAction.defaultProps = {
    as: 'button',
    accessibility: accessibility_1.buttonBehavior,
    content: {},
};
exports.AlertDismissAction.displayName = 'AlertDismissAction';
exports.AlertDismissAction.propTypes = __assign(__assign({}, utils_1.commonPropTypes.createCommon({
    content: 'shorthand',
})), { disabled: PropTypes.bool, danger: PropTypes.bool, warning: PropTypes.bool, info: PropTypes.bool, onClick: PropTypes.func, success: PropTypes.bool });
exports.AlertDismissAction.handledProps = Object.keys(exports.AlertDismissAction.propTypes);
exports.AlertDismissAction.create = utils_1.createShorthandFactory({ Component: exports.AlertDismissAction, mappedProp: 'content' });
