// botframework-streaming/lib/webSocket/index.js
"use strict";
/**
 * @module botframework-streaming
 */
/**
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebSocketTransport = exports.WebSocketServer = exports.WebSocketClient = exports.NodeWebSocket = exports.NodeWebSocketFactoryBase = exports.NodeWebSocketFactory = void 0;
var factories_1 = require("./factories");
Object.defineProperty(exports, "NodeWebSocketFactory", { enumerable: true, get: function () { return factories_1.NodeWebSocketFactory; } });
Object.defineProperty(exports, "NodeWebSocketFactoryBase", { enumerable: true, get: function () { return factories_1.NodeWebSocketFactoryBase; } });
var nodeWebSocket_1 = require("./nodeWebSocket");
Object.defineProperty(exports, "NodeWebSocket", { enumerable: true, get: function () { return nodeWebSocket_1.NodeWebSocket; } });
var nodeWebSocketClient_1 = require("./nodeWebSocketClient");
Object.defineProperty(exports, "WebSocketClient", { enumerable: true, get: function () { return nodeWebSocketClient_1.WebSocketClient; } });
var webSocketServer_1 = require("./webSocketServer");
Object.defineProperty(exports, "WebSocketServer", { enumerable: true, get: function () { return webSocketServer_1.WebSocketServer; } });
/** @internal */
var webSocketTransport_1 = require("./webSocketTransport");
Object.defineProperty(exports, "WebSocketTransport", { enumerable: true, get: function () { return webSocketTransport_1.WebSocketTransport; } });
