// botframework-connector/lib/connectorApi/connectorClient.js
"use strict";
/**
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConnectorMappers = exports.ConnectorModels = exports.ConnectorClientContext = exports.ConnectorClient = void 0;
const Mappers = require("./models/mappers");
exports.ConnectorMappers = Mappers;
const Models = require("./models");
exports.ConnectorModels = Models;
const operations = require("./operations");
const connectorClientContext_1 = require("./connectorClientContext");
Object.defineProperty(exports, "ConnectorClientContext", { enumerable: true, get: function () { return connectorClientContext_1.ConnectorClientContext; } });
class ConnectorClient extends connectorClientContext_1.ConnectorClientContext {
    /**
     * Initializes a new instance of the ConnectorClient class.
     * @param credentials Subscription credentials which uniquely identify client subscription.
     * @param [options] The parameter options
     */
    constructor(credentials, options) {
        super(credentials, options);
        this.attachments = new operations.Attachments(this);
        this.conversations = new operations.Conversations(this);
    }
}
exports.ConnectorClient = ConnectorClient;
// Operation Specifications
__exportStar(require("./operations"), exports);
