// @fluentui/react-northstar/dist/dts/src/components/Skeleton/SkeletonShape.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SkeletonShape = exports.skeletonShapeClassName = void 0;
var PropTypes = require("prop-types");
var react_bindings_1 = require("@fluentui/react-bindings");
var utils_1 = require("../../utils");
var Box_1 = require("../Box/Box");
exports.skeletonShapeClassName = 'ui-skeleton__shape';
/**
 * A SkeletonShape represents a shape (Image/Button/etc...) that will be loaded
 */
exports.SkeletonShape = react_bindings_1.compose(Box_1.Box, {
    className: exports.skeletonShapeClassName,
    displayName: 'SkeletonShape',
    overrideStyles: true,
    shorthandConfig: {},
    handledProps: ['round', 'width', 'height'],
    mapPropsToStylesProps: function (_a) {
        var width = _a.width, height = _a.height, round = _a.round;
        return ({ width: width, height: height, round: round });
    },
});
exports.SkeletonShape.propTypes = __assign(__assign({}, utils_1.commonPropTypes.createCommon()), { width: PropTypes.string, height: PropTypes.string, round: PropTypes.bool });
exports.SkeletonShape.defaultProps = {
    as: 'span',
    width: '100px',
    height: '100px',
    round: false,
};
