// @fluentui/react-northstar/dist/dts/src/components/Skeleton/Skeleton.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Skeleton = exports.skeletonClassName = void 0;
var PropTypes = require("prop-types");
var react_bindings_1 = require("@fluentui/react-bindings");
var utils_1 = require("../../utils");
var Box_1 = require("../Box/Box");
var SkeletonLine_1 = require("./SkeletonLine");
var SkeletonShape_1 = require("./SkeletonShape");
var SkeletonButton_1 = require("./SkeletonButton");
var SkeletonText_1 = require("./SkeletonText");
var SkeletonInput_1 = require("./SkeletonInput");
var SkeletonAvatar_1 = require("./SkeletonAvatar");
exports.skeletonClassName = 'ui-skeleton';
/**
 * A Skeleton is a component to be placed while the content is loading.
 */
exports.Skeleton = react_bindings_1.compose(Box_1.Box, {
    className: exports.skeletonClassName,
    displayName: 'Skeleton',
    overrideStyles: true,
    shorthandConfig: {},
    handledProps: ['animation'],
    mapPropsToStylesProps: function (_a) {
        var animation = _a.animation;
        return ({
            animation: animation,
        });
    },
});
exports.Skeleton.Line = SkeletonLine_1.SkeletonLine;
exports.Skeleton.Shape = SkeletonShape_1.SkeletonShape;
exports.Skeleton.Button = SkeletonButton_1.SkeletonButton;
exports.Skeleton.Text = SkeletonText_1.SkeletonText;
exports.Skeleton.Input = SkeletonInput_1.SkeletonInput;
exports.Skeleton.Avatar = SkeletonAvatar_1.SkeletonAvatar;
exports.Skeleton.propTypes = __assign(__assign({}, utils_1.commonPropTypes.createCommon({
    accessibility: false,
})), { animation: PropTypes.oneOf(['wave', 'pulse']) });
