// @fluentui/react-northstar/dist/dts/src/components/Skeleton/SkeletonAvatar.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SkeletonAvatar = exports.skeletonAvatarClassName = void 0;
var customPropTypes = require("@fluentui/react-proptypes");
var react_bindings_1 = require("@fluentui/react-bindings");
var utils_1 = require("../../utils");
var SkeletonShape_1 = require("./SkeletonShape");
exports.skeletonAvatarClassName = 'ui-skeleton__avatar';
/**
 * An SkeletonAvatar represents an avatar component that will be loaded
 */
exports.SkeletonAvatar = react_bindings_1.compose(SkeletonShape_1.SkeletonShape, {
    className: exports.skeletonAvatarClassName,
    displayName: 'SkeletonAvatar',
    overrideStyles: true,
    shorthandConfig: {},
    handledProps: ['size'],
    mapPropsToStylesProps: function (_a) {
        var size = _a.size;
        return ({
            size: size,
        });
    },
});
exports.SkeletonAvatar.propTypes = __assign(__assign({}, utils_1.commonPropTypes.createCommon()), { size: customPropTypes.size });
exports.SkeletonAvatar.defaultProps = {
    as: 'span',
    size: 'medium',
};
