// @fluentui/react-northstar/dist/dts/src/components/Skeleton/SkeletonLine.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SkeletonLine = exports.skeletonLineClassName = void 0;
var react_bindings_1 = require("@fluentui/react-bindings");
var PropTypes = require("prop-types");
var utils_1 = require("../../utils");
var Box_1 = require("../Box/Box");
exports.skeletonLineClassName = 'ui-skeleton__line';
/**
 * A SkeletonLine represents a text line that will be loaded
 */
exports.SkeletonLine = react_bindings_1.compose(Box_1.Box, {
    className: exports.skeletonLineClassName,
    displayName: 'SkeletonLine',
    overrideStyles: true,
    shorthandConfig: {},
    handledProps: ['width', 'height'],
    mapPropsToStylesProps: function (_a) {
        var width = _a.width, height = _a.height;
        return ({ width: width, height: height });
    },
});
exports.SkeletonLine.propTypes = __assign(__assign({}, utils_1.commonPropTypes.createCommon()), { width: PropTypes.string, height: PropTypes.string });
exports.SkeletonLine.defaultProps = {
    as: 'span',
    width: '100%',
    height: '1rem',
};
