// @fluentui/react-northstar/dist/dts/src/components/Skeleton/SkeletonButton.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SkeletonButton = exports.skeletonButtonClassName = void 0;
var customPropTypes = require("@fluentui/react-proptypes");
var PropTypes = require("prop-types");
var react_bindings_1 = require("@fluentui/react-bindings");
var utils_1 = require("../../utils");
var SkeletonShape_1 = require("./SkeletonShape");
exports.skeletonButtonClassName = 'ui-skeleton__button';
/**
 * An SkeletonButton represents a buton component that will be loaded
 */
exports.SkeletonButton = react_bindings_1.compose(SkeletonShape_1.SkeletonShape, {
    className: exports.skeletonButtonClassName,
    displayName: 'SkeletonButton',
    overrideStyles: true,
    shorthandConfig: {},
    handledProps: ['size', 'circular', 'iconOnly', 'fluid'],
    mapPropsToStylesProps: function (_a) {
        var size = _a.size, fluid = _a.fluid, iconOnly = _a.iconOnly, circular = _a.circular;
        return ({
            size: size,
            fluid: fluid,
            iconOnly: iconOnly,
            circular: circular,
        });
    },
});
exports.SkeletonButton.propTypes = __assign(__assign({}, utils_1.commonPropTypes.createCommon()), { size: customPropTypes.size, circular: PropTypes.bool, iconOnly: PropTypes.bool, fluid: PropTypes.bool });
exports.SkeletonButton.defaultProps = {
    as: 'span',
    size: 'medium',
};
