// @fluentui/react-northstar/dist/dts/src/components/Skeleton/SkeletonText.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SkeletonText = exports.skeletonTextClassName = void 0;
var customPropTypes = require("@fluentui/react-proptypes");
var react_bindings_1 = require("@fluentui/react-bindings");
var utils_1 = require("../../utils");
var SkeletonLine_1 = require("./SkeletonLine");
exports.skeletonTextClassName = 'ui-skeleton__text';
/**
 * An SkeletonText represents a text component that will be loaded
 */
exports.SkeletonText = react_bindings_1.compose(SkeletonLine_1.SkeletonLine, {
    className: exports.skeletonTextClassName,
    displayName: 'SkeletonText',
    overrideStyles: true,
    shorthandConfig: {},
    handledProps: ['size'],
    mapPropsToStylesProps: function (_a) {
        var size = _a.size;
        return ({
            size: size,
        });
    },
});
exports.SkeletonText.propTypes = __assign(__assign({}, utils_1.commonPropTypes.createCommon()), { size: customPropTypes.size });
exports.SkeletonText.defaultProps = {
    as: 'span',
    size: 'medium',
};
