// @fluentui/react-northstar/dist/commonjs/components/Tooltip/TooltipContent.js
"use strict";

exports.__esModule = true;
exports.TooltipContent = exports.tooltipContentClassName = void 0;

var _reactBindings = require("@fluentui/react-bindings");

var customPropTypes = _interopRequireWildcard(require("@fluentui/react-proptypes"));

var PropTypes = _interopRequireWildcard(require("prop-types"));

var React = _interopRequireWildcard(require("react"));

var _utils = require("../../utils");

var _positioner = require("../../utils/positioner");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var tooltipContentClassName = 'ui-tooltip__content';
/**
 * A TooltipContent contains the content of a Tooltip component.
 */

exports.tooltipContentClassName = tooltipContentClassName;
var TooltipContent = /*#__PURE__*/React.forwardRef(function (props, ref) {
  var context = (0, _reactBindings.useFluentContext)();

  var _useTelemetry = (0, _reactBindings.useTelemetry)(TooltipContent.displayName, context.telemetry),
      setStart = _useTelemetry.setStart,
      setEnd = _useTelemetry.setEnd;

  setStart();
  var accessibility = props.accessibility,
      children = props.children,
      className = props.className,
      content = props.content,
      design = props.design,
      open = props.open,
      placement = props.placement,
      pointing = props.pointing,
      pointerRef = props.pointerRef,
      styles = props.styles,
      variables = props.variables,
      subtle = props.subtle;
  var getA11Props = (0, _reactBindings.useAccessibility)(accessibility, {
    debugName: TooltipContent.displayName,
    rtl: context.rtl
  });

  var _useStyles = (0, _reactBindings.useStyles)(TooltipContent.displayName, {
    className: tooltipContentClassName,
    mapPropsToStyles: function mapPropsToStyles() {
      return {
        basePlacement: (0, _positioner.getBasePlacement)(placement, context.rtl),
        open: open,
        pointing: pointing,
        subtle: subtle
      };
    },
    mapPropsToInlineStyles: function mapPropsToInlineStyles() {
      return {
        className: className,
        design: design,
        styles: styles,
        variables: variables
      };
    },
    rtl: context.rtl
  }),
      classes = _useStyles.classes;

  var ElementType = (0, _reactBindings.getElementType)(props);
  var unhandledProps = (0, _reactBindings.useUnhandledProps)(TooltipContent.handledProps, props);
  var element = /*#__PURE__*/React.createElement(ElementType, getA11Props('root', Object.assign({
    className: classes.root,
    ref: ref
  }, _utils.rtlTextContainer.getAttributes({
    forElements: [children, content]
  }), unhandledProps)), open && pointing && /*#__PURE__*/React.createElement("div", {
    className: classes.pointer,
    ref: pointerRef
  }), /*#__PURE__*/React.createElement("div", getA11Props('content', {
    className: classes.content
  }), (0, _utils.childrenExist)(children) ? children : content));
  setEnd();
  return element;
});
exports.TooltipContent = TooltipContent;
TooltipContent.displayName = 'TooltipContent';
TooltipContent.propTypes = Object.assign({}, _utils.commonPropTypes.createCommon(), {
  placement: PropTypes.oneOf(['auto-start', 'auto', 'auto-end', 'top-start', 'top', 'top-end', 'right-start', 'right', 'right-end', 'bottom-end', 'bottom', 'bottom-start', 'left-end', 'left', 'left-start']),
  pointing: PropTypes.bool,
  pointerRef: customPropTypes.ref,
  subtle: PropTypes.bool
});
TooltipContent.handledProps = Object.keys(TooltipContent.propTypes);
TooltipContent.create = (0, _utils.createShorthandFactory)({
  Component: TooltipContent,
  mappedProp: 'content'
});
