// @fluentui/react-northstar/dist/dts/src/components/Toolbar/ToolbarMenuRadioGroupWrapper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ToolbarMenuRadioGroupWrapper = exports.toolbarMenuRadioGroupWrapperClassName = void 0;
var react_bindings_1 = require("@fluentui/react-bindings");
var accessibility_1 = require("@fluentui/accessibility");
var utils_1 = require("../../utils");
var Box_1 = require("../Box/Box");
exports.toolbarMenuRadioGroupWrapperClassName = 'ui-toolbars'; // FIXME: required by getComponentInfo/isConformant. But this is group inside a toolbar not a group of toolbars
/**
 * An ToolbarMenuRadioGroupWrapper provides a wrapping slot for in  ToolbarMenuRadioGroup.
 */
exports.ToolbarMenuRadioGroupWrapper = react_bindings_1.compose(Box_1.Box, {
    className: exports.toolbarMenuRadioGroupWrapperClassName,
    displayName: 'ToolbarMenuRadioGroupWrapper',
    overrideStyles: true,
    shorthandConfig: { mappedProp: 'content' },
});
exports.ToolbarMenuRadioGroupWrapper.defaultProps = {
    as: 'li',
    accessibility: accessibility_1.toolbarMenuRadioGroupWrapperBehavior,
};
exports.ToolbarMenuRadioGroupWrapper.propTypes = utils_1.commonPropTypes.createCommon();
