// @fluentui/react-northstar/dist/dts/src/components/Toolbar/ToolbarMenuItem.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ToolbarMenuItem = exports.toolbarMenuItemSlotClassNames = exports.toolbarMenuItemClassName = void 0;
var accessibility_1 = require("@fluentui/accessibility");
var React = require("react");
var _ = require("lodash");
var classnames_1 = require("classnames");
var PropTypes = require("prop-types");
var react_component_event_listener_1 = require("@fluentui/react-component-event-listener");
var react_component_ref_1 = require("@fluentui/react-component-ref");
var customPropTypes = require("@fluentui/react-proptypes");
var react_bindings_1 = require("@fluentui/react-bindings");
var react_component_nesting_registry_1 = require("@fluentui/react-component-nesting-registry");
var utils_1 = require("../../utils");
var positioner_1 = require("../../utils/positioner");
var Box_1 = require("../Box/Box");
var Popup_1 = require("../Popup/Popup");
var ToolbarMenu_1 = require("./ToolbarMenu");
var ToolbarMenuItemIcon_1 = require("./ToolbarMenuItemIcon");
var toolbarVariablesContext_1 = require("./toolbarVariablesContext");
var ToolbarMenuItemSubmenuIndicator_1 = require("./ToolbarMenuItemSubmenuIndicator");
var ToolbarMenuItemActiveIndicator_1 = require("./ToolbarMenuItemActiveIndicator");
var toolbarMenuContext_1 = require("./toolbarMenuContext");
var ToolbarMenuItemContent_1 = require("./ToolbarMenuItemContent");
var react_icons_northstar_1 = require("@fluentui/react-icons-northstar");
exports.toolbarMenuItemClassName = 'ui-toolbar__menuitem';
exports.toolbarMenuItemSlotClassNames = {
    wrapper: exports.toolbarMenuItemClassName + "__wrapper",
    submenu: exports.toolbarMenuItemClassName + "__submenu",
};
/**
 * A ToolbarMenuItem renders ToolbarMenu item as button.
 */
exports.ToolbarMenuItem = react_bindings_1.compose(function (props, ref, composeOptions) {
    var context = react_bindings_1.useFluentContext();
    var _a = react_bindings_1.useTelemetry(composeOptions.displayName, context.telemetry), setStart = _a.setStart, setEnd = _a.setEnd;
    setStart();
    var active = props.active, activeIndicator = props.activeIndicator, children = props.children, content = props.content, disabled = props.disabled, disabledFocusable = props.disabledFocusable, submenuIndicator = props.submenuIndicator, icon = props.icon, popup = props.popup, wrapper = props.wrapper, inSubmenu = props.inSubmenu, className = props.className, design = props.design, styles = props.styles, variables = props.variables;
    var _b = positioner_1.partitionPopperPropsFromShorthand(props.menu), menu = _b[0], menuPositioningProps = _b[1];
    var _c = react_bindings_1.useAutoControlled({
        defaultValue: props.defaultMenuOpen,
        value: props.menuOpen,
        initialValue: false,
    }), menuOpen = _c[0], setMenuOpen = _c[1];
    var itemRef = React.useRef();
    var menuRef = React.useRef();
    var menuSlot = react_bindings_1.useContextSelectors(toolbarMenuContext_1.ToolbarMenuContext, {
        menuSlot: function (v) { return v.slots.menu; },
    }).menuSlot; // TODO: we should improve typings for the useContextSelectors
    var parentVariables = React.useContext(toolbarVariablesContext_1.ToolbarVariablesContext);
    var mergedVariables = react_bindings_1.mergeVariablesOverrides(parentVariables, variables);
    var ElementType = react_bindings_1.getElementType(props);
    var slotProps = composeOptions.resolveSlotProps(props);
    var unhandledProps = react_bindings_1.useUnhandledProps(composeOptions.handledProps, props);
    var getA11yProps = react_bindings_1.useAccessibility(props.accessibility, {
        debugName: composeOptions.displayName,
        mapPropsToBehavior: function () { return ({
            hasMenu: !!menu,
            active: active,
            menuOpen: menuOpen,
            disabled: disabled,
            disabledFocusable: disabledFocusable,
            'aria-label': props['aria-label'],
            'aria-labelledby': props['aria-labelledby'],
            'aria-describedby': props['aria-describedby'],
        }); },
        actionHandlers: {
            performClick: function (event) {
                event.preventDefault();
                handleClick(event);
            },
            openMenu: function (event) { return openMenu(event); },
            closeAllMenusAndFocusNextParentItem: function (event) { return closeAllMenus(event); },
            closeMenu: function (event) { return closeMenu(event); },
            closeMenuAndFocusTrigger: function (event) { return closeMenu(event); },
            doNotNavigateNextParentItem: function (event) {
                event.stopPropagation();
            },
            closeAllMenus: function (event) { return closeAllMenus(event); },
        },
        rtl: context.rtl,
    });
    var _d = react_bindings_1.useStyles(composeOptions.displayName, {
        className: composeOptions.className,
        composeOptions: composeOptions,
        mapPropsToStyles: function () { return ({
            disabled: disabled || disabledFocusable,
            hasContent: !!content,
        }); },
        mapPropsToInlineStyles: function () { return ({
            className: className,
            design: design,
            styles: styles,
            variables: mergedVariables,
        }); },
        rtl: context.rtl,
        unstable_props: props,
    }), classes = _d.classes, resolvedStyles = _d.styles;
    var openMenu = function (e) {
        if (menu && !menuOpen) {
            trySetMenuOpen(true, e);
            e.stopPropagation();
            e.preventDefault();
        }
    };
    var closeMenu = function (e) {
        if (!isSubmenuOpen()) {
            return;
        }
        trySetMenuOpen(false, e, function () {
            react_bindings_1.focusAsync(itemRef.current);
        });
        e.stopPropagation();
    };
    var closeAllMenus = function (e) {
        if (!isSubmenuOpen()) {
            return;
        }
        trySetMenuOpen(false, e, function () {
            if (!inSubmenu) {
                react_bindings_1.focusAsync(itemRef.current);
            }
        });
        // avoid spacebar scrolling the page
        if (!inSubmenu) {
            e.preventDefault();
        }
    };
    var isSubmenuOpen = function () {
        return !!(menu && menuOpen);
    };
    var trySetMenuOpen = function (newValue, e, onStateChanged) {
        setMenuOpen(newValue);
        // The reason why post-effect is not passed as callback to trySetState method
        // is that in 'controlled' mode the post-effect is applied before final re-rendering
        // which cause a broken behavior: for e.g. when it is needed to focus submenu trigger on ESC.
        // TODO: all DOM post-effects should be applied at componentDidMount & componentDidUpdated stages.
        onStateChanged && onStateChanged();
        _.invoke(props, 'onMenuOpenChange', e, __assign(__assign({}, props), { menuOpen: newValue }));
    };
    var outsideClickHandler = function (getRefs) { return function (e) {
        var isItemClick = utils_1.doesNodeContainClick(itemRef.current, e, context.target);
        var isNestedClick = _.some(getRefs(), function (childRef) {
            return utils_1.doesNodeContainClick(childRef.current, e, context.target);
        });
        var isInside = isItemClick || isNestedClick;
        if (!isInside) {
            trySetMenuOpen(false, e);
        }
    }; };
    var handleMenuOverrides = function (predefinedProps) { return ({
        onItemClick: function (e, itemProps) {
            var popup = itemProps.popup, menuOpen = itemProps.menuOpen;
            _.invoke(predefinedProps, 'onItemClick', e, itemProps);
            if (popup) {
                return;
            }
            trySetMenuOpen(menuOpen, e);
            if (!menuOpen) {
                _.invoke(itemRef.current, 'focus');
            }
        },
    }); };
    var handleClick = function (e) {
        if (disabled) {
            e.preventDefault();
            return;
        }
        if (menu) {
            // the menuItem element was clicked => toggle the open/close and stop propagation
            trySetMenuOpen(!menuOpen, e);
            e.stopPropagation();
            e.preventDefault();
        }
        if (popup) {
            e.stopPropagation();
            e.preventDefault();
            return;
        }
        _.invoke(props, 'onClick', e, props);
    };
    var element = (React.createElement(ElementType, __assign({}, getA11yProps('root', __assign({ className: classes.root, onClick: handleClick, disabled: disabled, ref: ref }, unhandledProps))), utils_1.childrenExist(children) ? (children) : (React.createElement(React.Fragment, null,
        utils_1.createShorthand(composeOptions.slots.icon, icon, { defaultProps: function () { return slotProps.icon; } }),
        utils_1.createShorthand(composeOptions.slots.content, content, {
            defaultProps: function () { return getA11yProps('content', slotProps.content); },
        }),
        active &&
            utils_1.createShorthand(composeOptions.slots.activeIndicator, activeIndicator, {
                defaultProps: function () { return slotProps.activeIndicator; },
            }),
        menu &&
            utils_1.createShorthand(composeOptions.slots.submenuIndicator, submenuIndicator, {
                defaultProps: function () { return slotProps.submenuIndicator; },
            })))));
    var hasChildren = utils_1.childrenExist(children);
    if (popup && !hasChildren) {
        var popupElement = utils_1.createShorthand(composeOptions.slots.popup, popup, {
            defaultProps: function () { return (__assign(__assign({}, slotProps.popup), { onOpenChange: function (e) {
                    e.stopPropagation();
                } })); },
            overrideProps: {
                trigger: element,
                children: undefined, // force-reset `children` defined for `Popup` as it collides with the `trigger`
            },
        });
        setEnd();
        return popupElement;
    }
    var menuItemInner = hasChildren ? children : React.createElement(react_component_ref_1.Ref, { innerRef: itemRef }, element);
    var maybeSubmenu = menu && menuOpen ? (React.createElement(react_component_nesting_registry_1.Unstable_NestingAuto, null, function (getRefs, nestingRef) {
        return (React.createElement(React.Fragment, null,
            React.createElement(react_component_ref_1.Ref, { innerRef: function (node) {
                    nestingRef.current = node;
                    menuRef.current = node;
                } },
                React.createElement(positioner_1.Popper, __assign({ align: "top", position: context.rtl ? 'before' : 'after', targetRef: itemRef }, menuPositioningProps),
                    React.createElement(toolbarVariablesContext_1.ToolbarVariablesProvider, { value: mergedVariables }, utils_1.createShorthand(composeOptions.slots.menu || menuSlot || ToolbarMenu_1.ToolbarMenu, menu, {
                        defaultProps: function () { return (__assign({ className: exports.toolbarMenuItemSlotClassNames.submenu, styles: resolvedStyles.menu, submenu: true, submenuIndicator: submenuIndicator }, slotProps.menu)); },
                        overrideProps: handleMenuOverrides,
                    })))),
            React.createElement(react_component_event_listener_1.EventListener, { listener: outsideClickHandler(getRefs), target: context.target, type: "click" })));
    })) : null;
    if (!wrapper) {
        setEnd();
        return menuItemInner;
    }
    var wrapperElement = Box_1.Box.create(wrapper, {
        defaultProps: function () {
            return getA11yProps('wrapper', {
                className: classnames_1.default(exports.toolbarMenuItemSlotClassNames.wrapper, classes.wrapper),
            });
        },
        overrideProps: function () { return ({
            children: (React.createElement(React.Fragment, null,
                menuItemInner,
                maybeSubmenu)),
        }); },
    });
    setEnd();
    return wrapperElement;
}, {
    className: exports.toolbarMenuItemClassName,
    displayName: 'ToolbarMenuItem',
    slots: {
        icon: ToolbarMenuItemIcon_1.ToolbarMenuItemIcon,
        submenuIndicator: ToolbarMenuItemSubmenuIndicator_1.ToolbarMenuItemSubmenuIndicator,
        activeIndicator: ToolbarMenuItemActiveIndicator_1.ToolbarMenuItemActiveIndicator,
        popup: Popup_1.Popup,
        content: ToolbarMenuItemContent_1.ToolbarMenuItemContent,
    },
    slotProps: function (props) { return ({
        icon: {
            hasContent: !!props.content,
        },
        submenuIndicator: {
            accessibility: accessibility_1.indicatorBehavior,
        },
        activeIndicator: {
            accessibility: accessibility_1.indicatorBehavior,
        },
        popup: {
            trapFocus: true,
        },
        content: {},
    }); },
    shorthandConfig: {
        mappedProp: 'content',
    },
    handledProps: [
        'accessibility',
        'as',
        'children',
        'className',
        'content',
        'design',
        'styles',
        'variables',
        'disabledFocusable',
        'active',
        'activeIndicator',
        'defaultMenuOpen',
        'disabled',
        'icon',
        'index',
        'submenuIndicator',
        'inSubmenu',
        'menu',
        'menuOpen',
        'onClick',
        'onMenuOpenChange',
        'popup',
        'wrapper',
    ],
});
exports.ToolbarMenuItem.propTypes = __assign(__assign({}, utils_1.commonPropTypes.createCommon()), { active: PropTypes.bool, activeIndicator: customPropTypes.shorthandAllowingChildren, defaultMenuOpen: PropTypes.bool, disabled: PropTypes.bool, disabledFocusable: PropTypes.bool, icon: customPropTypes.shorthandAllowingChildren, index: PropTypes.number, submenuIndicator: customPropTypes.shorthandAllowingChildren, inSubmenu: PropTypes.bool, menu: PropTypes.oneOfType([customPropTypes.itemShorthand, customPropTypes.collectionShorthand]), menuOpen: PropTypes.bool, onClick: PropTypes.func, onMenuOpenChange: PropTypes.func, popup: PropTypes.oneOfType([
        PropTypes.shape(__assign(__assign({}, Popup_1.Popup.propTypes), { trigger: customPropTypes.never, children: customPropTypes.never })),
        PropTypes.string,
    ]), wrapper: customPropTypes.itemShorthand });
exports.ToolbarMenuItem.defaultProps = {
    as: 'button',
    accessibility: accessibility_1.toolbarMenuItemBehavior,
    activeIndicator: {},
    submenuIndicator: React.createElement(react_icons_northstar_1.ChevronEndIcon, { outline: true }),
    wrapper: { as: 'li' },
};
