// @fluentui/react-northstar/dist/dts/src/components/Toolbar/ToolbarMenuItemSubmenuIndicator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ToolbarMenuItemSubmenuIndicator = exports.toolbarMenuItemSubmenuIndicatorClassName = void 0;
var react_bindings_1 = require("@fluentui/react-bindings");
var utils_1 = require("../../utils");
var Box_1 = require("../Box/Box");
exports.toolbarMenuItemSubmenuIndicatorClassName = 'ui-toolbar__menuitemsubmenuindicator';
/**
 * A ToolbarMenuItemSubmenuIndicator allows a user to have a dedicated component that can be targeted from the theme.
 */
exports.ToolbarMenuItemSubmenuIndicator = react_bindings_1.compose(Box_1.Box, {
    className: exports.toolbarMenuItemSubmenuIndicatorClassName,
    displayName: 'ToolbarMenuItemSubmenuIndicator',
    mapPropsToStylesProps: function (props) { return ({
        hasContent: props.hasContent,
    }); },
    shorthandConfig: {
        mappedProp: 'content',
    },
    overrideStyles: true,
});
exports.ToolbarMenuItemSubmenuIndicator.defaultProps = {
    as: 'span',
};
exports.ToolbarMenuItemSubmenuIndicator.propTypes = utils_1.commonPropTypes.createCommon();
