// @fluentui/react-northstar/dist/dts/src/components/Toolbar/ToolbarItem.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ToolbarItem = exports.toolbarItemClassName = void 0;
var accessibility_1 = require("@fluentui/accessibility");
var react_bindings_1 = require("@fluentui/react-bindings");
var react_component_ref_1 = require("@fluentui/react-component-ref");
var react_component_event_listener_1 = require("@fluentui/react-component-event-listener");
var react_component_nesting_registry_1 = require("@fluentui/react-component-nesting-registry");
var customPropTypes = require("@fluentui/react-proptypes");
var _ = require("lodash");
var PropTypes = require("prop-types");
var React = require("react");
var utils_1 = require("../../utils");
var positioner_1 = require("../../utils/positioner");
var ToolbarMenu_1 = require("./ToolbarMenu");
var Popup_1 = require("../Popup/Popup");
var toolbarVariablesContext_1 = require("./toolbarVariablesContext");
var ToolbarItemWrapper_1 = require("./ToolbarItemWrapper");
var ToolbarItemIcon_1 = require("./ToolbarItemIcon");
var toolbarMenuContext_1 = require("./toolbarMenuContext");
exports.toolbarItemClassName = 'ui-toolbar__item';
/**
 * A ToolbarItem renders Toolbar item as a button with an icon.
 */
exports.ToolbarItem = react_bindings_1.compose(function (props, ref, composeOptions) {
    var context = react_bindings_1.useFluentContext();
    var _a = react_bindings_1.useTelemetry(composeOptions.displayName, context.telemetry), setStart = _a.setStart, setEnd = _a.setEnd;
    setStart();
    var accessibility = props.accessibility, active = props.active, className = props.className, design = props.design, icon = props.icon, children = props.children, disabled = props.disabled, disabledFocusable = props.disabledFocusable, popup = props.popup, menuOpen = props.menuOpen, wrapper = props.wrapper, styles = props.styles, variables = props.variables;
    var _b = positioner_1.partitionPopperPropsFromShorthand(props.menu), menu = _b[0], positioningProps = _b[1];
    var itemRef = React.useRef();
    var menuRef = React.useRef();
    var parentVariables = React.useContext(toolbarVariablesContext_1.ToolbarVariablesContext);
    var mergedVariables = react_bindings_1.mergeVariablesOverrides(parentVariables, variables);
    var menuSlot = react_bindings_1.useContextSelectors(toolbarMenuContext_1.ToolbarMenuContext, {
        menuSlot: function (v) { return v.slots.menu; },
    }).menuSlot; // TODO: we should improve typings for the useContextSelectors
    var getA11yProps = react_bindings_1.useAccessibility(accessibility, {
        debugName: composeOptions.displayName,
        actionHandlers: {
            performClick: function (event) {
                event.preventDefault();
                handleClick(event);
            },
            performWrapperClick: function (event) {
                handleWrapperClick(event);
            },
            closeMenuAndFocusTrigger: function (event) {
                trySetMenuOpen(false, event);
                _.invoke(itemRef.current, 'focus');
            },
            doNotNavigateNextToolbarItem: function (event) {
                event.stopPropagation();
            },
        },
        mapPropsToBehavior: function () { return ({
            as: String(props.as),
            disabled: disabled,
            disabledFocusable: disabledFocusable,
            hasMenu: !!menu,
            hasPopup: !!popup,
            menuOpen: menuOpen,
            active: active,
        }); },
        rtl: context.rtl,
    });
    var classes = react_bindings_1.useStyles(composeOptions.displayName, {
        className: composeOptions.className,
        composeOptions: composeOptions,
        mapPropsToStyles: function () { return ({ active: active, disabled: disabled || disabledFocusable }); },
        mapPropsToInlineStyles: function () { return ({
            className: className,
            design: design,
            styles: styles,
            variables: mergedVariables,
        }); },
        rtl: context.rtl,
        unstable_props: props,
    }).classes;
    var handleBlur = function (e) {
        _.invoke(props, 'onBlur', e, props);
    };
    var handleFocus = function (e) {
        _.invoke(props, 'onFocus', e, props);
    };
    var handleClick = function (e) {
        if (disabled || disabledFocusable) {
            e.preventDefault();
            return;
        }
        if (menu) {
            trySetMenuOpen(!menuOpen, e);
        }
        _.invoke(props, 'onClick', e, props);
    };
    var handleWrapperClick = function (e) {
        if (menu) {
            if (utils_1.doesNodeContainClick(menuRef.current, e.nativeEvent, context.target, false)) {
                trySetMenuOpen(false, e);
            }
        }
    };
    var handleOutsideClick = function (getRefs) { return function (e) {
        var isItemClick = utils_1.doesNodeContainClick(itemRef.current, e, context.target);
        var isNestedClick = _.some(getRefs(), function (childRef) {
            return utils_1.doesNodeContainClick(childRef.current, e, context.target);
        });
        var isInside = isItemClick || isNestedClick;
        if (!isInside) {
            trySetMenuOpen(false, e);
        }
    }; };
    var trySetMenuOpen = function (newValue, e) {
        _.invoke(props, 'onMenuOpenChange', e, __assign(__assign({}, props), { menuOpen: newValue }));
    };
    var handleMenuOverrides = function (getRefs) { return function (predefinedProps) { return ({
        onBlur: function (e) {
            var isInsideOrMenuTrigger = _.some(getRefs(), function (childRef) {
                return (childRef.current.contains(e.relatedTarget) ||
                    itemRef.current.contains(e.relatedTarget));
            });
            if (!isInsideOrMenuTrigger) {
                trySetMenuOpen(false, e);
            }
        },
        onItemClick: function (e, itemProps) {
            var popup = itemProps.popup, menuOpen = itemProps.menuOpen;
            _.invoke(predefinedProps, 'onItemClick', e, itemProps);
            if (popup) {
                return;
            }
            // TODO: should we pass toolbarMenuItem to the user callback so he can decide if he wants to close the menu?
            trySetMenuOpen(menuOpen, e);
        },
    }); }; };
    var ElementType = react_bindings_1.getElementType(props);
    var slotProps = composeOptions.resolveSlotProps(props);
    var unhandledProps = react_bindings_1.useUnhandledProps(composeOptions.handledProps, props);
    var itemElement = (React.createElement(react_component_ref_1.Ref, { innerRef: function (node) {
            itemRef.current = node;
            react_component_ref_1.handleRef(ref, node);
        } },
        React.createElement(ElementType, __assign({}, getA11yProps('root', __assign(__assign({}, unhandledProps), { disabled: disabled, className: classes.root, onBlur: handleBlur, onFocus: handleFocus, onClick: handleClick }))), utils_1.childrenExist(children) ? children : utils_1.createShorthand(composeOptions.slots.icon, icon, slotProps.icon))));
    var submenuElement = menuOpen ? (React.createElement(react_component_nesting_registry_1.Unstable_NestingAuto, null, function (getRefs, nestingRef) {
        return (React.createElement(React.Fragment, null,
            React.createElement(react_component_ref_1.Ref, { innerRef: function (node) {
                    nestingRef.current = node;
                    menuRef.current = node;
                } },
                React.createElement(positioner_1.Popper, __assign({ align: "start", position: "above", targetRef: itemRef }, positioningProps),
                    React.createElement(toolbarVariablesContext_1.ToolbarVariablesProvider, { value: mergedVariables }, utils_1.createShorthand(composeOptions.slots.menu || menuSlot || ToolbarMenu_1.ToolbarMenu, menu, {
                        defaultProps: function () { return slotProps.menu; },
                        overrideProps: handleMenuOverrides(getRefs),
                    })))),
            React.createElement(react_component_event_listener_1.EventListener, { listener: handleOutsideClick(getRefs), target: context.target, type: "click", capture: true })));
    })) : null;
    if (popup) {
        var popupElement = utils_1.createShorthand(composeOptions.slots.popup, popup, {
            defaultProps: function () { return slotProps.popup; },
            overrideProps: {
                trigger: itemElement,
                children: undefined, // force-reset `children` defined for `Popup` as it collides with the `trigger`
            },
        });
        setEnd();
        return popupElement;
    }
    // wrap the item if it has menu (even if it is closed = not rendered)
    if (menu) {
        var contentElement_1 = (React.createElement(React.Fragment, null,
            itemElement,
            submenuElement));
        if (wrapper) {
            var wrapperElement = utils_1.createShorthand(composeOptions.slots.wrapper, wrapper, {
                defaultProps: function () { return getA11yProps('wrapper', slotProps.wrapper || {}); },
                overrideProps: function (predefinedProps) { return ({
                    children: contentElement_1,
                    onClick: function (e) {
                        handleWrapperClick(e);
                        _.invoke(predefinedProps, 'onClick', e);
                    },
                }); },
            });
            setEnd();
            return wrapperElement;
        }
        setEnd();
        return contentElement_1;
    }
    setEnd();
    return itemElement;
}, {
    className: exports.toolbarItemClassName,
    displayName: 'ToolbarItem',
    slots: {
        icon: ToolbarItemIcon_1.ToolbarItemIcon,
        wrapper: ToolbarItemWrapper_1.ToolbarItemWrapper,
        popup: Popup_1.Popup, // TODO: compose Popup to ToolbarItemPopup once it has compose functionality
    },
    slotProps: function () { return ({
        popup: { trapFocus: true },
    }); },
    shorthandConfig: { mappedProp: 'content' },
    handledProps: [
        'accessibility',
        'as',
        'children',
        'className',
        'content',
        'design',
        'styles',
        'disabledFocusable',
        'variables',
        'active',
        'disabled',
        'icon',
        'menu',
        'menuOpen',
        'onMenuOpenChange',
        'onClick',
        'onFocus',
        'onBlur',
        'popup',
        'wrapper',
    ],
});
exports.ToolbarItem.propTypes = __assign(__assign({}, utils_1.commonPropTypes.createCommon()), { active: PropTypes.bool, disabled: PropTypes.bool, disabledFocusable: PropTypes.bool, icon: customPropTypes.shorthandAllowingChildren, menu: PropTypes.oneOfType([
        customPropTypes.shorthandAllowingChildren,
        PropTypes.arrayOf(customPropTypes.shorthandAllowingChildren),
    ]), menuOpen: PropTypes.bool, onMenuOpenChange: PropTypes.func, onClick: PropTypes.func, onFocus: PropTypes.func, onBlur: PropTypes.func, popup: PropTypes.oneOfType([
        PropTypes.shape(__assign(__assign({}, Popup_1.Popup.propTypes), { trigger: customPropTypes.never, children: customPropTypes.never })),
        PropTypes.string,
    ]), wrapper: customPropTypes.shorthandAllowingChildren });
exports.ToolbarItem.defaultProps = {
    as: 'button',
    accessibility: accessibility_1.toolbarItemBehavior,
    wrapper: {},
};
