// @fluentui/react-northstar/dist/dts/src/themes/teams/animations/fade.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.fadeInOutAnimations = void 0;
var timingFunctions_1 = require("./timingFunctions");
var durations_1 = require("./durations");
exports.fadeInOutAnimations = {
    // Fade Ins
    // Basic Fade In Animation -- Ultra Fast
    fadeEnterUltraFast: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': { opacity: delta },
                '100%': { opacity: 1 },
            });
        },
        keyframeParams: { delta: 0 },
        duration: durations_1.ultraFast,
        timingFunction: timingFunctions_1.linear,
        fillMode: 'forwards',
    },
    // Basic Fade In Animation -- Faster
    fadeEnterFaster: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': { opacity: delta },
                '100%': { opacity: 1 },
            });
        },
        keyframeParams: { delta: 0 },
        duration: durations_1.faster,
        timingFunction: timingFunctions_1.linear,
        fillMode: 'forwards',
    },
    // Basic Fade In Animation -- Fast
    fadeEnterFast: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': { opacity: delta },
                '100%': { opacity: 1 },
            });
        },
        keyframeParams: { delta: 0 },
        duration: durations_1.fast,
        timingFunction: timingFunctions_1.linear,
        fillMode: 'forwards',
    },
    // Basic Fade In Animation --Normal
    fadeEnterNormal: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': { opacity: delta },
                '100%': { opacity: 1 },
            });
        },
        keyframeParams: { delta: 0 },
        duration: durations_1.normal,
        timingFunction: timingFunctions_1.linear,
        fillMode: 'forwards',
    },
    // Basic Fade In Animation --Slow
    fadeEnterSlow: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': { opacity: delta },
                '100%': { opacity: 1 },
            });
        },
        keyframeParams: { delta: 0 },
        duration: durations_1.slow,
        timingFunction: timingFunctions_1.linear,
        fillMode: 'forwards',
    },
    // Basic Fade In Animation --Slow
    fadeEnterSlower: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': { opacity: delta },
                '100%': { opacity: 1 },
            });
        },
        keyframeParams: { delta: 0 },
        duration: durations_1.slower,
        timingFunction: timingFunctions_1.linear,
        fillMode: 'forwards',
    },
    // Basic Fade In Animation --UltraSlow
    fadeEnterUltraSlow: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': { opacity: delta },
                '100%': { opacity: 1 },
            });
        },
        keyframeParams: { delta: 0 },
        duration: durations_1.ultraSlow,
        timingFunction: timingFunctions_1.linear,
        fillMode: 'forwards',
    },
    // Fade Outs
    // Basic Fade Out Animation --UltraFast
    fadeExitUltraFast: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': { opacity: 1 },
                '100%': { opacity: delta },
            });
        },
        keyframeParams: { delta: 0 },
        duration: durations_1.ultraFast,
        timingFunction: timingFunctions_1.linear,
        fillMode: 'forwards',
    },
    // Basic Fade Out Animation --Faster
    fadeExitFaster: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': { opacity: 1 },
                '100%': { opacity: delta },
            });
        },
        keyframeParams: { delta: 0 },
        duration: durations_1.faster,
        timingFunction: timingFunctions_1.linear,
        fillMode: 'forwards',
    },
    // Basic Fade Out Animation --Fast
    fadeExitFast: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': { opacity: 1 },
                '100%': { opacity: delta },
            });
        },
        keyframeParams: { delta: 0 },
        duration: durations_1.fast,
        timingFunction: timingFunctions_1.linear,
        fillMode: 'forwards',
    },
    // Basic Fade Out Animation --Normal
    fadeExitNormal: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': { opacity: 1 },
                '100%': { opacity: delta },
            });
        },
        keyframeParams: { delta: 0 },
        duration: durations_1.normal,
        timingFunction: timingFunctions_1.linear,
        fillMode: 'forwards',
    },
    // Basic Fade Out Animation --Slow
    fadeExitSlow: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': { opacity: 1 },
                '100%': { opacity: delta },
            });
        },
        keyframeParams: { delta: 0 },
        duration: durations_1.slow,
        timingFunction: timingFunctions_1.linear,
        fillMode: 'forwards',
    },
    // Basic Fade Out Animation --Slower
    fadeExitSlower: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': { opacity: 1 },
                '100%': { opacity: delta },
            });
        },
        keyframeParams: { delta: 0 },
        duration: durations_1.slower,
        timingFunction: timingFunctions_1.linear,
        fillMode: 'forwards',
    },
    // Basic Fade Out Animation --UltraSlow
    fadeExitUltraSlow: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': { opacity: 1 },
                '100%': { opacity: delta },
            });
        },
        keyframeParams: { delta: 0 },
        duration: durations_1.ultraSlow,
        timingFunction: timingFunctions_1.linear,
        fillMode: 'forwards',
    },
};
