// @fluentui/react-northstar/dist/dts/src/themes/teams/animations/scale.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.scaleAnimations = void 0;
var timingFunctions_1 = require("./timingFunctions");
var durations_1 = require("./durations");
exports.scaleAnimations = {
    // Scale in w/ fade- UltraFast
    scaleEnterUltraFast: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': {
                    transform: "scale(" + delta + ")",
                    opacity: 0,
                },
                '100%': {
                    transform: 'scale(1)',
                    opacity: 1,
                },
            });
        },
        keyframeParams: { delta: 0.88 },
        duration: durations_1.ultraFast,
        timingFunction: timingFunctions_1.decelerateMax,
        fillMode: 'forwards',
    },
    // Scale in w/ fade- Faster
    scaleEnterFaster: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': {
                    transform: "scale(" + delta + ")",
                    opacity: 0,
                },
                '100%': {
                    transform: 'scale(1)',
                    opacity: 1,
                },
            });
        },
        keyframeParams: { delta: 0.88 },
        duration: durations_1.faster,
        timingFunction: timingFunctions_1.decelerateMax,
        fillMode: 'forwards',
    },
    // Scale in w/ fade- Fast
    scaleEnterFast: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': {
                    transform: "scale(" + delta + ")",
                    opacity: 0,
                },
                '100%': {
                    transform: 'scale(1)',
                    opacity: 1,
                },
            });
        },
        keyframeParams: { delta: 0.88 },
        duration: durations_1.fast,
        timingFunction: timingFunctions_1.decelerateMax,
        fillMode: 'forwards',
    },
    // Scale in  w/ fade- Normal
    scaleEnterNormal: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': {
                    transform: "scale(" + delta + ")",
                    opacity: 0,
                },
                '100%': {
                    transform: 'scale(1)',
                    opacity: 1,
                },
            });
        },
        keyframeParams: { delta: 0.88 },
        duration: durations_1.normal,
        timingFunction: timingFunctions_1.decelerateMax,
        fillMode: 'forwards',
    },
    // Scale in w/ fade - Slow
    scaleEnterSlow: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': {
                    transform: "scale(" + delta + ")",
                    opacity: 0,
                },
                '100%': {
                    transform: 'scale(1)',
                    opacity: 1,
                },
            });
        },
        keyframeParams: { delta: 0.88 },
        duration: durations_1.slow,
        timingFunction: timingFunctions_1.decelerateMax,
        fillMode: 'forwards',
    },
    // Scale in w/ fade - Slower
    scaleEnterSlower: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': {
                    transform: "scale(" + delta + ")",
                    opacity: 0,
                },
                '100%': {
                    transform: 'scale(1)',
                    opacity: 1,
                },
            });
        },
        keyframeParams: { delta: 0.88 },
        duration: durations_1.slower,
        timingFunction: timingFunctions_1.decelerateMax,
        fillMode: 'forwards',
    },
    // Scale in w/ fade - UlrtaSlow
    scaleEnterUltraSlow: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': {
                    transform: "scale(" + delta + ")",
                    opacity: 0,
                },
                '100%': {
                    transform: 'scale(1)',
                    opacity: 1,
                },
            });
        },
        keyframeParams: { delta: 0.88 },
        duration: durations_1.ultraSlow,
        timingFunction: timingFunctions_1.decelerateMax,
        fillMode: 'forwards',
    },
    // Scale Out w/ Fade - UltraFast
    scaleExitUltraFast: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': {
                    transform: 'scale(1)',
                    opacity: 1,
                },
                '100%': {
                    transform: "scale(" + delta + ")",
                    opacity: 0,
                },
            });
        },
        keyframeParams: { delta: 0.9 },
        duration: durations_1.ultraFast,
        timingFunction: timingFunctions_1.accelerateMax,
        fillMode: 'forwards',
    },
    // Scale Out w/ Fade - Faster
    scaleExitFaster: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': {
                    transform: 'scale(1)',
                    opacity: 1,
                },
                '100%': {
                    transform: "scale(" + delta + ")",
                    opacity: 0,
                },
            });
        },
        keyframeParams: { delta: 0.9 },
        duration: durations_1.faster,
        timingFunction: timingFunctions_1.accelerateMax,
        fillMode: 'forwards',
    },
    // Scale Out w/ Fade - Fast
    scaleExitFast: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': {
                    transform: 'scale(1)',
                    opacity: 1,
                },
                '100%': {
                    transform: "scale(" + delta + ")",
                    opacity: 0,
                },
            });
        },
        keyframeParams: { delta: 0.9 },
        duration: durations_1.fast,
        timingFunction: timingFunctions_1.accelerateMax,
        fillMode: 'forwards',
    },
    // Scale Out w/ Fade - Normal
    scaleExitNormal: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': {
                    transform: 'scale(1)',
                    opacity: 1,
                },
                '100%': {
                    transform: "scale(" + delta + ")",
                    opacity: 0,
                },
            });
        },
        keyframeParams: { delta: 0.9 },
        duration: durations_1.normal,
        timingFunction: timingFunctions_1.accelerateMax,
        fillMode: 'forwards',
    },
    // Scale Out w/ Fade - Slow
    scaleExitSlow: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': {
                    transform: 'scale(1)',
                    opacity: 1,
                },
                '100%': {
                    transform: "scale(" + delta + ")",
                    opacity: 0,
                },
            });
        },
        keyframeParams: { delta: 0.9 },
        duration: durations_1.slow,
        timingFunction: timingFunctions_1.accelerateMax,
        fillMode: 'forwards',
    },
    // Scale Out w/ Fade - Slower
    scaleExitSlower: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': {
                    transform: 'scale(1)',
                    opacity: 1,
                },
                '100%': {
                    transform: "scale(" + delta + ")",
                    opacity: 0,
                },
            });
        },
        keyframeParams: { delta: 0.9 },
        duration: durations_1.slower,
        timingFunction: timingFunctions_1.accelerateMax,
        fillMode: 'forwards',
    },
    // Scale Out w/ Fade - UltraSlow
    scaleExitUltraSlow: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': {
                    transform: 'scale(1)',
                    opacity: 1,
                },
                '100%': {
                    transform: "scale(" + delta + ")",
                    opacity: 0,
                },
            });
        },
        keyframeParams: { delta: 0.9 },
        duration: durations_1.ultraSlow,
        timingFunction: timingFunctions_1.accelerateMax,
        fillMode: 'forwards',
    },
    scaleDownSoft: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': {
                    transform: 'scale(1)',
                    opacity: 1,
                },
                '100%': {
                    transform: "scale(" + delta + ")",
                    opacity: 0,
                },
            });
        },
        keyframeParams: { delta: 0.9 },
        duration: '100ms',
        timingFunction: 'cubic-bezier(.78, 0, .22, 1)',
        fillMode: 'both',
    },
};
