// @fluentui/react-northstar/dist/dts/src/themes/teams/animations/slide.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.slideAnimations = void 0;
var timingFunctions_1 = require("./timingFunctions");
var durations_1 = require("./durations");
exports.slideAnimations = {
    // Slide Down Animations
    // Slides object down to original position --ultrafast
    slideDownEnterUltraFast: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': {
                    transform: "translateY(-" + delta + ")",
                    opacity: 0,
                },
                '100%': {
                    transform: 'translateY(0px)',
                    opacity: 1,
                },
            });
        },
        keyframeParams: { delta: '20px' },
        duration: durations_1.ultraFast,
        timingFunction: timingFunctions_1.decelerateMax,
        direction: 'forward',
        fillMode: 'forwards',
    },
    // Slides object down to original position -Faster
    slideDownEnterFaster: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': {
                    transform: "translateY(-" + delta + ")",
                    opacity: 0,
                },
                '100%': {
                    transform: 'translateY(0px)',
                    opacity: 1,
                },
            });
        },
        keyframeParams: { delta: '20px' },
        duration: durations_1.faster,
        timingFunction: timingFunctions_1.decelerateMax,
        direction: 'forward',
        fillMode: 'forwards',
    },
    // Slides object down to original position -fast
    slideDownEnterFast: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': {
                    transform: "translateY(-" + delta + ")",
                    opacity: 0,
                },
                '100%': {
                    transform: 'translateY(0px)',
                    opacity: 1,
                },
            });
        },
        keyframeParams: { delta: '20px' },
        duration: durations_1.fast,
        timingFunction: timingFunctions_1.decelerateMax,
        direction: 'forward',
        fillMode: 'forwards',
    },
    // Slides object in from top to bottom --normal
    slideDownEnterNormal: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': {
                    transform: "translateY(-" + delta + ")",
                    opacity: 0,
                },
                '100%': {
                    transform: 'translateY(0px)',
                    opacity: 1,
                },
            });
        },
        keyframeParams: { delta: '20px' },
        duration: durations_1.normal,
        timingFunction: timingFunctions_1.decelerateMax,
        direction: 'forward',
        fillMode: 'forwards',
    },
    // Slides object in from top to bottom --slow
    slideDownEnterSlow: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': {
                    transform: "translateY(-" + delta + ")",
                    opacity: 0,
                },
                '100%': {
                    transform: 'translateY(0px)',
                    opacity: 1,
                },
            });
        },
        keyframeParams: { delta: '20px' },
        duration: durations_1.slow,
        timingFunction: timingFunctions_1.decelerateMax,
        direction: 'forward',
        fillMode: 'forwards',
    },
    // Slides object in from top to bottom --slower
    slideDownEnterSlower: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': {
                    transform: "translateY(-" + delta + ")",
                    opacity: 0,
                },
                '100%': {
                    transform: 'translateY(0px)',
                    opacity: 1,
                },
            });
        },
        keyframeParams: { delta: '20px' },
        duration: durations_1.slower,
        timingFunction: timingFunctions_1.decelerateMax,
        direction: 'forward',
        fillMode: 'forwards',
    },
    // Slides object in from top to bottom - Slow
    slideDownEnterUltraSlow: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': {
                    transform: "translateY(-" + delta + ")",
                    opacity: 0,
                },
                '100%': {
                    transform: 'translateY(0px)',
                    opacity: 1,
                },
            });
        },
        keyframeParams: { delta: '20px' },
        duration: durations_1.ultraSlow,
        timingFunction: timingFunctions_1.decelerateMax,
        direction: 'forward',
        fillMode: 'forwards',
    },
    // Slide Up Animations
    // Slides object in from top to bottom --ultrafast
    slideUpEnterUltraFast: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': {
                    transform: "translateY(" + delta + ")",
                    opacity: 0,
                },
                '100%': {
                    transform: 'translateY(0px)',
                    opacity: 1,
                },
            });
        },
        keyframeParams: { delta: '20px' },
        duration: durations_1.ultraFast,
        timingFunction: timingFunctions_1.decelerateMax,
        direction: 'forward',
        fillMode: 'forwards',
    },
    // Slides object in from top to bottom --faster
    slideUpEnterFaster: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': {
                    transform: "translateY(" + delta + ")",
                    opacity: 0,
                },
                '100%': {
                    transform: 'translateY(0px)',
                    opacity: 1,
                },
            });
        },
        keyframeParams: { delta: '20px' },
        duration: durations_1.faster,
        timingFunction: timingFunctions_1.decelerateMax,
        direction: 'forward',
        fillMode: 'forwards',
    },
    // Slides object in from top to bottom --fast
    slideUpEnterFast: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': {
                    transform: "translateY(" + delta + ")",
                    opacity: 0,
                },
                '100%': {
                    transform: 'translateY(0px)',
                    opacity: 1,
                },
            });
        },
        keyframeParams: { delta: '20px' },
        duration: durations_1.fast,
        timingFunction: timingFunctions_1.decelerateMax,
        direction: 'forward',
        fillMode: 'forwards',
    },
    // Slides object in from down to up-normal
    slideUpEnterNormal: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': {
                    transform: "translateY(" + delta + ")",
                    opacity: 0,
                },
                '100%': {
                    transform: 'translateY(0px)',
                    opacity: 1,
                },
            });
        },
        keyframeParams: { delta: '20px' },
        duration: durations_1.normal,
        timingFunction: timingFunctions_1.decelerateMax,
        direction: 'forward',
        fillMode: 'forwards',
    },
    // Slides object in from top to bottom --slow
    slideUpEnterSlow: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': {
                    transform: "translateY(" + delta + ")",
                    opacity: 0,
                },
                '100%': {
                    transform: 'translateY(0px)',
                    opacity: 1,
                },
            });
        },
        keyframeParams: { delta: '20px' },
        duration: durations_1.slow,
        timingFunction: timingFunctions_1.decelerateMax,
        direction: 'forward',
        fillMode: 'forwards',
    },
    // Slides object in from bottom to top - slower
    slideUpEnterSlower: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': {
                    transform: "translateY(" + delta + ")",
                    opacity: 0,
                },
                '100%': {
                    transform: 'translateY(0px)',
                    opacity: 1,
                },
            });
        },
        keyframeParams: { delta: '20px' },
        duration: durations_1.slower,
        timingFunction: timingFunctions_1.decelerateMax,
        direction: 'forward',
        fillMode: 'forwards',
    },
    // Slides object in from bottom to top - ultraslow
    slideUpEnterUltraSlow: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': {
                    transform: "translateY(" + delta + ")",
                    opacity: 0,
                },
                '100%': {
                    transform: 'translateY(0px)',
                    opacity: 1,
                },
            });
        },
        keyframeParams: { delta: '20px' },
        duration: durations_1.ultraSlow,
        timingFunction: timingFunctions_1.decelerateMax,
        direction: 'forward',
        fillMode: 'forwards',
    },
    // ---Horizontal Slide Animations---- //
    // Slide Left Animations
    // Slides object in from right to left --ultrafast
    slideLeftEnterUltraFast: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': {
                    transform: "translateX(" + delta + ")",
                    opacity: 0,
                },
                '100%': {
                    transform: 'translateX(0px)',
                    opacity: 1,
                },
            });
        },
        keyframeParams: { delta: '20px' },
        duration: durations_1.ultraFast,
        timingFunction: timingFunctions_1.decelerateMax,
        direction: 'forward',
        fillMode: 'forwards',
    },
    // Slides object in from right to left -faster
    slideLeftEnterFaster: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': {
                    transform: "translateX(" + delta + ")",
                    opacity: 0,
                },
                '100%': {
                    transform: 'translateX(0px)',
                    opacity: 1,
                },
            });
        },
        keyframeParams: { delta: '20px' },
        duration: durations_1.faster,
        timingFunction: timingFunctions_1.decelerateMax,
        direction: 'forward',
        fillMode: 'forwards',
    },
    // Slides object in from right to left -Fast
    slideLeftEnterFast: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': {
                    transform: "translateX(" + delta + ")",
                    opacity: 0,
                },
                '100%': {
                    transform: 'translateX(0px)',
                    opacity: 1,
                },
            });
        },
        keyframeParams: { delta: '20px' },
        duration: durations_1.fast,
        timingFunction: timingFunctions_1.decelerateMax,
        direction: 'forward',
        fillMode: 'forwards',
    },
    // Slides object in from right to left -Normal
    slideLeftEnterNormal: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': {
                    transform: "translateX(" + delta + ")",
                    opacity: 0,
                },
                '100%': {
                    transform: 'translateX(0px)',
                    opacity: 1,
                },
            });
        },
        keyframeParams: { delta: '200px' },
        duration: durations_1.normal,
        timingFunction: timingFunctions_1.decelerateMax,
        direction: 'forward',
        fillMode: 'forwards',
    },
    // Slides object in from right to left -Slow
    slideLeftEnterSlow: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': {
                    transform: "translateX(" + delta + ")",
                    opacity: 0,
                },
                '100%': {
                    transform: 'translateX(0px)',
                    opacity: 1,
                },
            });
        },
        keyframeParams: { delta: '20px' },
        duration: durations_1.slow,
        timingFunction: timingFunctions_1.decelerateMax,
        direction: 'forward',
        fillMode: 'forwards',
    },
    // Slides object in from right to left - slower
    slideLeftEnterSlower: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': {
                    transform: "translateX(" + delta + ")",
                    opacity: 0,
                },
                '100%': {
                    transform: 'translateX(0px)',
                    opacity: 1,
                },
            });
        },
        keyframeParams: { delta: '20px' },
        duration: durations_1.slower,
        timingFunction: timingFunctions_1.decelerateMax,
        direction: 'forward',
        fillMode: 'forwards',
    },
    // Slides object in from right to left - ultraslow
    slideLeftEnterUltraSlow: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': {
                    transform: "translateX(" + delta + ")",
                    opacity: 0,
                },
                '100%': {
                    transform: 'translateX(0px)',
                    opacity: 1,
                },
            });
        },
        keyframeParams: { delta: '20px' },
        duration: durations_1.ultraSlow,
        timingFunction: timingFunctions_1.decelerateMax,
        direction: 'forward',
        fillMode: 'forwards',
    },
    // Slide Right Animations
    // Slides object in from left to right -ultrafast
    slideRightEnterUltraFast: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': {
                    transform: "translateX(-" + delta + ")",
                    opacity: 0,
                },
                '100%': {
                    transform: 'translateX(0px)',
                    opacity: 1,
                },
            });
        },
        keyframeParams: { delta: '20px' },
        duration: durations_1.ultraFast,
        timingFunction: timingFunctions_1.decelerateMax,
        direction: 'forward',
        fillMode: 'forwards',
    },
    // Slides object in from left to right - Faster
    slideRightEnterFaster: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': {
                    transform: "translateX(-" + delta + ")",
                    opacity: 0,
                },
                '100%': {
                    transform: 'translateX(0px)',
                    opacity: 1,
                },
            });
        },
        keyframeParams: { delta: '20px' },
        duration: durations_1.faster,
        timingFunction: timingFunctions_1.decelerateMax,
        direction: 'forward',
        fillMode: 'forwards',
    },
    // Slides object in from left to right -fast
    slideRightEnterFast: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': {
                    transform: "translateX(-" + delta + ")",
                    opacity: 0,
                },
                '100%': {
                    transform: 'translateX(0px)',
                    opacity: 1,
                },
            });
        },
        keyframeParams: { delta: '20px' },
        duration: durations_1.fast,
        timingFunction: timingFunctions_1.decelerateMax,
        direction: 'forward',
        fillMode: 'forwards',
    },
    // Slides object in from left to right -normal
    slideRightEnterNormal: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': {
                    transform: "translateX(-" + delta + ")",
                    opacity: 0,
                },
                '100%': {
                    transform: 'translateX(0px)',
                    opacity: 1,
                },
            });
        },
        keyframeParams: { delta: '200px' },
        duration: durations_1.normal,
        timingFunction: timingFunctions_1.decelerateMax,
        direction: 'forward',
        fillMode: 'forwards',
    },
    // Slides object in from left to right -slow
    slideRightEnterSlow: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': {
                    transform: "translateX(-" + delta + ")",
                    opacity: 0,
                },
                '100%': {
                    transform: 'translateX(0px)',
                    opacity: 1,
                },
            });
        },
        keyframeParams: { delta: '20px' },
        duration: durations_1.slow,
        timingFunction: timingFunctions_1.decelerateMax,
        direction: 'forward',
        fillMode: 'forwards',
    },
    // Slides object in from left to right- slower
    slideRightEnterSlower: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': {
                    transform: "translateX(-" + delta + ")",
                    opacity: 0,
                },
                '100%': {
                    transform: 'translateX(0px)',
                    opacity: 1,
                },
            });
        },
        keyframeParams: { delta: '20px' },
        duration: durations_1.slower,
        timingFunction: timingFunctions_1.decelerateMax,
        direction: 'forward',
        fillMode: 'forwards',
    },
    // Slides object in from left to right- ultraslow
    slideRightEnterUltraSlow: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': {
                    transform: "translateX(-" + delta + ")",
                    opacity: 0,
                },
                '100%': {
                    transform: 'translateX(0px)',
                    opacity: 1,
                },
            });
        },
        keyframeParams: { delta: '20px' },
        duration: durations_1.ultraSlow,
        timingFunction: timingFunctions_1.decelerateMax,
        direction: 'forward',
        fillMode: 'forwards',
    },
    // Slide Exit Animation///
    // Slide Down - Exit - UltraFast
    slideDownExitUltraFast: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': {
                    transform: 'translateY(0px)',
                    opacity: 1,
                },
                '100%': {
                    transform: "translateY(" + delta + ")",
                    opacity: 0,
                },
            });
        },
        keyframeParams: { delta: '20px' },
        duration: durations_1.ultraFast,
        timingFunction: timingFunctions_1.accelerateMax,
        direction: 'forward',
        fillMode: 'forwards',
    },
    // Slide Down - Exit - Faster
    slideDownExitFaster: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': {
                    transform: 'translateY(0px)',
                    opacity: 1,
                },
                '100%': {
                    transform: "translateY(" + delta + ")",
                    opacity: 0,
                },
            });
        },
        keyframeParams: { delta: '20px' },
        duration: durations_1.faster,
        timingFunction: timingFunctions_1.accelerateMax,
        direction: 'forward',
        fillMode: 'forwards',
    },
    // Slide Down - Exit - Fast
    slideDownExitFast: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': {
                    transform: 'translateY(0px)',
                    opacity: 1,
                },
                '100%': {
                    transform: "translateY(" + delta + ")",
                    opacity: 0,
                },
            });
        },
        keyframeParams: { delta: '20px' },
        duration: durations_1.fast,
        timingFunction: timingFunctions_1.accelerateMax,
        direction: 'forward',
        fillMode: 'forwards',
    },
    // Slide Down - Exit - Normal
    slideDownExitNormal: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': {
                    transform: 'translateY(0px)',
                    opacity: 1,
                },
                '100%': {
                    transform: "translateY(" + delta + ")",
                    opacity: 0,
                },
            });
        },
        keyframeParams: { delta: '20px' },
        duration: durations_1.normal,
        timingFunction: timingFunctions_1.accelerateMax,
        direction: 'forward',
        fillMode: 'forwards',
    },
    // Slide Down - Exit - Slow
    slideDownExitSlow: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': {
                    transform: 'translateY(0px)',
                    opacity: 1,
                },
                '100%': {
                    transform: "translateY(" + delta + ")",
                    opacity: 0,
                },
            });
        },
        keyframeParams: { delta: '20px' },
        duration: durations_1.slow,
        timingFunction: timingFunctions_1.accelerateMax,
        direction: 'forward',
        fillMode: 'forwards',
    },
    // Slide Down - Exit - Slower
    slideDownExitSlower: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': {
                    transform: 'translateY(0px)',
                    opacity: 1,
                },
                '100%': {
                    transform: "translateY(" + delta + ")",
                    opacity: 0,
                },
            });
        },
        keyframeParams: { delta: '20px' },
        duration: durations_1.slower,
        timingFunction: timingFunctions_1.accelerateMax,
        direction: 'forward',
        fillMode: 'forwards',
    },
    // Slide Down - Exit - Ultraslow
    slideDownExitUltraSlow: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': {
                    transform: 'translateY(0px)',
                    opacity: 1,
                },
                '100%': {
                    transform: "translateY(" + delta + ")",
                    opacity: 0,
                },
            });
        },
        keyframeParams: { delta: '20px' },
        duration: durations_1.ultraSlow,
        timingFunction: timingFunctions_1.accelerateMax,
        direction: 'forward',
        fillMode: 'forwards',
    },
    // Slide Up - Exit- Animations
    // Slide Up - Exit - UltraFast
    slideUpExitUltraFast: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': {
                    transform: 'translateY(0px)',
                    opacity: 1,
                },
                '100%': {
                    transform: "translateY(-" + delta + ")",
                    opacity: 0,
                },
            });
        },
        keyframeParams: { delta: '20px' },
        duration: durations_1.ultraFast,
        timingFunction: timingFunctions_1.accelerateMax,
        direction: 'forward',
        fillMode: 'forwards',
    },
    // Slide Up - Exit - Faster
    slideUpExitFaster: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': {
                    transform: 'translateY(0px)',
                    opacity: 1,
                },
                '100%': {
                    transform: "translateY(-" + delta + ")",
                    opacity: 0,
                },
            });
        },
        keyframeParams: { delta: '20px' },
        duration: durations_1.faster,
        timingFunction: timingFunctions_1.accelerateMax,
        direction: 'forward',
        fillMode: 'forwards',
    },
    // Slide Up - Exit - Fast
    slideUpExitFast: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': {
                    transform: 'translateY(0px)',
                    opacity: 1,
                },
                '100%': {
                    transform: "translateY(-" + delta + ")",
                    opacity: 0,
                },
            });
        },
        keyframeParams: { delta: '20px' },
        duration: durations_1.fast,
        timingFunction: timingFunctions_1.accelerateMax,
        direction: 'forward',
        fillMode: 'forwards',
    },
    // Slide Up - Exit - Normal
    slideUpExitNormal: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': {
                    transform: 'translateY(0px)',
                    opacity: 1,
                },
                '100%': {
                    transform: "translateY(-" + delta + ")",
                    opacity: 0,
                },
            });
        },
        keyframeParams: { delta: '20px' },
        duration: durations_1.normal,
        timingFunction: timingFunctions_1.accelerateMax,
        direction: 'forward',
        fillMode: 'forwards',
    },
    // Slide Up - Exit - Slow
    slideUpExitSlow: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': {
                    transform: 'translateY(0px)',
                    opacity: 1,
                },
                '100%': {
                    transform: "translateY(-" + delta + ")",
                    opacity: 0,
                },
            });
        },
        keyframeParams: { delta: '20px' },
        duration: durations_1.slow,
        timingFunction: timingFunctions_1.accelerateMax,
        direction: 'forward',
        fillMode: 'forwards',
    },
    // Slide Up - Exit - Slower
    slideUpExitSlower: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': {
                    transform: 'translateY(0px)',
                    opacity: 1,
                },
                '100%': {
                    transform: "translateY(-" + delta + ")",
                    opacity: 0,
                },
            });
        },
        keyframeParams: { delta: '20px' },
        duration: durations_1.slower,
        timingFunction: timingFunctions_1.accelerateMax,
        direction: 'forward',
        fillMode: 'forwards',
    },
    // Slide Up - Exit - Ultra Slow
    slideUpExitUltraSlow: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': {
                    transform: 'translateY(0px)',
                    opacity: 1,
                },
                '100%': {
                    transform: "translateY(-" + delta + ")",
                    opacity: 0,
                },
            });
        },
        keyframeParams: { delta: '20px' },
        duration: durations_1.ultraSlow,
        timingFunction: timingFunctions_1.accelerateMax,
        direction: 'forward',
        fillMode: 'forwards',
    },
    // Slide Right Exit Animations
    // Slide Right Exit - Ultrafast
    slideRightExitUltraFast: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': {
                    transform: 'translateX(0px)',
                    opacity: 1,
                },
                '100%': {
                    transform: "translateX(" + delta + ")",
                    opacity: 0,
                },
            });
        },
        keyframeParams: { delta: '20px' },
        duration: durations_1.ultraFast,
        timingFunction: timingFunctions_1.accelerateMax,
        direction: 'forward',
        fillMode: 'forwards',
    },
    // Slide Right Exit - Faster
    slideRightExitFaster: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': {
                    transform: 'translateX(0px)',
                    opacity: 1,
                },
                '100%': {
                    transform: "translateX(" + delta + ")",
                    opacity: 0,
                },
            });
        },
        keyframeParams: { delta: '20px' },
        duration: durations_1.faster,
        timingFunction: timingFunctions_1.accelerateMax,
        direction: 'forward',
        fillMode: 'forwards',
    },
    // Slide Right Exit - Fast
    slideRightExitFast: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': {
                    transform: 'translateX(0px)',
                    opacity: 1,
                },
                '100%': {
                    transform: "translateX(" + delta + ")",
                    opacity: 0,
                },
            });
        },
        keyframeParams: { delta: '20px' },
        duration: durations_1.fast,
        timingFunction: timingFunctions_1.accelerateMax,
        direction: 'forward',
        fillMode: 'forwards',
    },
    // Slide Right Exit - Normal
    slideRightExitNormal: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': {
                    transform: 'translateX(0px)',
                    opacity: 1,
                },
                '100%': {
                    transform: "translateX(" + delta + ")",
                    opacity: 0,
                },
            });
        },
        keyframeParams: { delta: '20px' },
        duration: durations_1.normal,
        timingFunction: timingFunctions_1.accelerateMax,
        direction: 'forward',
        fillMode: 'forwards',
    },
    // Slide Right Exit - Slow
    slideRightExitSlow: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': {
                    transform: 'translateX(0px)',
                    opacity: 1,
                },
                '100%': {
                    transform: "translateX(" + delta + ")",
                    opacity: 0,
                },
            });
        },
        keyframeParams: { delta: '20px' },
        duration: durations_1.slow,
        timingFunction: timingFunctions_1.accelerateMax,
        direction: 'forward',
        fillMode: 'forwards',
    },
    // Slide Right Exit - Ultrafast
    slideRightExitSlower: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': {
                    transform: 'translateX(0px)',
                    opacity: 1,
                },
                '100%': {
                    transform: "translateX(" + delta + ")",
                    opacity: 0,
                },
            });
        },
        keyframeParams: { delta: '20px' },
        duration: durations_1.slower,
        timingFunction: timingFunctions_1.accelerateMax,
        direction: 'forward',
        fillMode: 'forwards',
    },
    // Slide Right Exit - Ultraslow
    slideRightExitUltraSlow: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': {
                    transform: 'translateX(0px)',
                    opacity: 1,
                },
                '100%': {
                    transform: "translateX(" + delta + ")",
                    opacity: 0,
                },
            });
        },
        keyframeParams: { delta: '20px' },
        duration: durations_1.ultraSlow,
        timingFunction: timingFunctions_1.accelerateMax,
        direction: 'forward',
        fillMode: 'forwards',
    },
    // Slide Left Exit Aniamtions
    // Slide Left Exit -ultrafast
    slideLeftExitUltraFast: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': {
                    transform: 'translateX(0px)',
                    opacity: 1,
                },
                '100%': {
                    transform: "translateX(-" + delta + ")",
                    opacity: 0,
                },
            });
        },
        keyframeParams: { delta: '20px' },
        duration: durations_1.ultraFast,
        timingFunction: timingFunctions_1.accelerateMax,
        direction: 'forward',
        fillMode: 'forwards',
    },
    // Slide Left Exit -Faster
    slideLeftExitFaster: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': {
                    transform: 'translateX(0px)',
                    opacity: 1,
                },
                '100%': {
                    transform: "translateX(-" + delta + ")",
                    opacity: 0,
                },
            });
        },
        keyframeParams: { delta: '20px' },
        duration: durations_1.faster,
        timingFunction: timingFunctions_1.accelerateMax,
        direction: 'forward',
        fillMode: 'forwards',
    },
    // Slide Left Exit -Fast
    slideLeftExitFast: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': {
                    transform: 'translateX(0px)',
                    opacity: 1,
                },
                '100%': {
                    transform: "translateX(-" + delta + ")",
                    opacity: 0,
                },
            });
        },
        keyframeParams: { delta: '20px' },
        duration: durations_1.fast,
        timingFunction: timingFunctions_1.accelerateMax,
        direction: 'forward',
        fillMode: 'forwards',
    },
    // Slide Left Exit -Normal
    slideLeftExitNormal: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': {
                    transform: 'translateX(0px)',
                    opacity: 1,
                },
                '100%': {
                    transform: "translateX(-" + delta + ")",
                    opacity: 0,
                },
            });
        },
        keyframeParams: { delta: '20px' },
        duration: durations_1.normal,
        timingFunction: timingFunctions_1.accelerateMax,
        direction: 'forward',
        fillMode: 'forwards',
    },
    // Slide Left Exit -Slow
    slideLeftExitSlow: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': {
                    transform: 'translateX(0px)',
                    opacity: 1,
                },
                '100%': {
                    transform: "translateX(-" + delta + ")",
                    opacity: 0,
                },
            });
        },
        keyframeParams: { delta: '20px' },
        duration: durations_1.slow,
        timingFunction: timingFunctions_1.accelerateMax,
        direction: 'forward',
        fillMode: 'forwards',
    },
    // Slide Left Exit -Slower
    slideLeftExitSlower: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': {
                    transform: 'translateX(0px)',
                    opacity: 1,
                },
                '100%': {
                    transform: "translateX(-" + delta + ")",
                    opacity: 0,
                },
            });
        },
        keyframeParams: { delta: '20px' },
        duration: durations_1.slower,
        timingFunction: timingFunctions_1.accelerateMax,
        direction: 'forward',
        fillMode: 'forwards',
    },
    // Slide Left Exit -Ultraslow
    slideLeftExitUltraSlow: {
        keyframe: function (_a) {
            var delta = _a.delta;
            return ({
                '0%': {
                    transform: 'translateX(0px)',
                    opacity: 1,
                },
                '100%': {
                    transform: "translateX(-" + delta + ")",
                    opacity: 0,
                },
            });
        },
        keyframeParams: { delta: '20px' },
        duration: durations_1.ultraSlow,
        timingFunction: timingFunctions_1.accelerateMax,
        direction: 'forward',
        fillMode: 'forwards',
    },
};
