// @fluentui/react-northstar/dist/commonjs/themes/teams/components/Datepicker/datepickerVariables.js
"use strict";

exports.__esModule = true;
exports.datepickerVariables = void 0;

var _utils = require("../../../../utils");

var _dateTimeUtilities = require("../../../../utils/date-time-utilities");

var datepickerVariables = function datepickerVariables(siteVars) {
  return {
    calendarCellBorder: 'none',
    calendarCellHeight: (0, _utils.pxToRem)(34),
    calendarCellWidth: (0, _utils.pxToRem)(34),
    calendarCellPadding: (0, _utils.pxToRem)(0),
    calendarCellMargin: (0, _utils.pxToRem)(0),
    calendarCellBackgroundColor: siteVars.colorScheme.default.background,
    calendarCellColor: 'inherit',
    calendarCellQuietColor: siteVars.colorScheme.default.foreground2,
    calendarCellSelectedColor: siteVars.colorScheme.brand.background4,
    calendarCellSelectedBackgroundColor: siteVars.colorScheme.brand.borderActive1,
    calendarCellTodayBackgroundColor: siteVars.colorScheme.brand.backgroundFocus,
    calendarCellTodayColor: siteVars.colorScheme.white.foreground,
    calendarCellTodayBorderRadius: '50%',
    calendarCellHoverBackgroundColor: siteVars.colorScheme.brand.backgroundHover2,
    calendarCellHoverColor: 'inherit',
    calendarCellDisabledColor: siteVars.colorScheme.brand.foregroundDisabled,
    calendarCellDisabledBackgroundColor: siteVars.colorScheme.default.background,
    calendarHeaderCellFontWeight: siteVars.fontWeightRegular,
    calendarHeaderCellHeight: (0, _utils.pxToRem)(32),
    calendarHeaderCellPadding: (0, _utils.pxToRem)(2) + " " + (0, _utils.pxToRem)(0) + " " + (0, _utils.pxToRem)(2) + " " + (0, _utils.pxToRem)(0),
    calendarHeaderCellWidth: "calc(100%/" + _dateTimeUtilities.DAYS_IN_WEEK + ")",
    calendarHeaderPaddingTop: (0, _utils.pxToRem)(5),
    calendarHeaderPaddingBottom: (0, _utils.pxToRem)(5),
    calendarHeaderLabelPaddingLeft: (0, _utils.pxToRem)(10),
    calendarHeaderLabelFontWeight: siteVars.fontWeightBold,
    calendarMinHeight: (0, _utils.pxToRem)(282)
  };
};

exports.datepickerVariables = datepickerVariables;
