// @fluentui/react-northstar/dist/commonjs/themes/teams-high-contrast/componentVariables.js
"use strict";

exports.__esModule = true;
exports.SplitButtonDivider = exports.Card = exports.DatepickerCalendarCellButton = exports.DatepickerCalendarCell = exports.TooltipContent = exports.Label = exports.DropdownSelectedItem = exports.DropdownItem = exports.Dropdown = exports.Provider = exports.Reaction = exports.ToolbarMenuRadioGroupWrapper = exports.ToolbarMenuRadioGroup = exports.ToolbarMenuItemActiveIndicator = exports.ToolbarMenuItemSubmenuIndicator = exports.ToolbarMenuItemIcon = exports.ToolbarMenuItem = exports.ToolbarMenuDivider = exports.ToolbarMenu = exports.ToolbarDivider = exports.ToolbarItem = exports.ToolbarCustomItem = exports.Toolbar = exports.TextArea = exports.Text = exports.Segment = exports.RadioGroupItem = exports.MenuItemWrapper = exports.MenuItemIndicator = exports.MenuItemContent = exports.MenuItemIcon = exports.MenuItem = exports.MenuDivider = exports.Menu = exports.Input = exports.Header = exports.Embed = exports.Divider = exports.Dialog = exports.Checkbox = exports.ChatMessage = exports.Chat = exports.Button = exports.AvatarImage = exports.Avatar = void 0;

var _avatarVariables = require("./components/Avatar/avatarVariables");

exports.Avatar = _avatarVariables.avatarVariables;
exports.AvatarImage = _avatarVariables.avatarVariables;

var _buttonVariables = require("./components/Button/buttonVariables");

exports.Button = _buttonVariables.buttonVariables;

var _chatVariables = require("./components/Chat/chatVariables");

exports.Chat = _chatVariables.chatVariables;

var _chatMessageVariables = require("./components/Chat/chatMessageVariables");

exports.ChatMessage = _chatMessageVariables.chatMessageVariables;

var _checkboxVariables = require("./components/Checkbox/checkboxVariables");

exports.Checkbox = _checkboxVariables.checkboxVariables;

var _dialogVariables = require("./components/Dialog/dialogVariables");

exports.Dialog = _dialogVariables.dialogVariables;

var _dividerVariables = require("./components/Divider/dividerVariables");

exports.Divider = _dividerVariables.dividerVariables;

var _embedVariables = require("./components/Embed/embedVariables");

exports.Embed = _embedVariables.embedVariables;

var _headerVariables = require("./components/Header/headerVariables");

exports.Header = _headerVariables.headerVariables;

var _inputVariables = require("./components/Input/inputVariables");

exports.Input = _inputVariables.inputVariables;

var _menuVariables = require("./components/Menu/menuVariables");

exports.Menu = _menuVariables.menuVariables;

var _menuDividerVariables = require("./components/Menu/menuDividerVariables");

exports.MenuDivider = _menuDividerVariables.menuDividerVariables;

var _menuItemVariables = require("./components/Menu/menuItemVariables");

exports.MenuItem = _menuItemVariables.menuItemVariables;

var _menuItemIconVariables = require("./components/Menu/menuItemIconVariables");

exports.MenuItemIcon = _menuItemIconVariables.menuItemIconVariables;

var _menuItemContentVariables = require("./components/Menu/menuItemContentVariables");

exports.MenuItemContent = _menuItemContentVariables.menuItemContentVariables;
exports.MenuItemIndicator = _menuItemContentVariables.menuItemContentVariables;

var _menuItemWrapperVariables = require("./components/Menu/menuItemWrapperVariables");

exports.MenuItemWrapper = _menuItemWrapperVariables.menuItemWrapperVariables;

var _radioGroupItemVariables = require("./components/RadioGroup/radioGroupItemVariables");

exports.RadioGroupItem = _radioGroupItemVariables.radioGroupItemVariables;

var _segmentVariables = require("./components/Segment/segmentVariables");

exports.Segment = _segmentVariables.segmentVariables;

var _textVariables = require("./components/Text/textVariables");

exports.Text = _textVariables.textVariables;

var _textAreaVariables = require("./components/TextArea/textAreaVariables");

exports.TextArea = _textAreaVariables.textAreaVariables;

var _toolbarVariables = require("./components/Toolbar/toolbarVariables");

exports.Toolbar = _toolbarVariables.toolbarVariables;

var _toolbarCustomItemVariables = require("./components/Toolbar/toolbarCustomItemVariables");

exports.ToolbarCustomItem = _toolbarCustomItemVariables.toolbarCustomItemVariables;

var _toolbarItemVariables = require("./components/Toolbar/toolbarItemVariables");

exports.ToolbarItem = _toolbarItemVariables.toolbarItemVariables;

var _toolbarDividerVariables = require("./components/Toolbar/toolbarDividerVariables");

exports.ToolbarDivider = _toolbarDividerVariables.toolbarDividerVariables;

var _toolbarMenuVariables = require("./components/Toolbar/toolbarMenuVariables");

exports.ToolbarMenu = _toolbarMenuVariables.toolbarMenuVariables;

var _toolbarMenuDividerVariables = require("./components/Toolbar/toolbarMenuDividerVariables");

exports.ToolbarMenuDivider = _toolbarMenuDividerVariables.toolbarMenuDividerVariables;

var _toolbarMenuItemVariables = require("./components/Toolbar/toolbarMenuItemVariables");

exports.ToolbarMenuItem = _toolbarMenuItemVariables.toolbarMenuItemVariables;

var _toolbarMenuItemIconVariables = require("./components/Toolbar/toolbarMenuItemIconVariables");

exports.ToolbarMenuItemIcon = _toolbarMenuItemIconVariables.toolbarMenuItemIconVariables;

var _toolbarMenuItemSubmenuIndicatorVariables = require("./components/Toolbar/toolbarMenuItemSubmenuIndicatorVariables");

exports.ToolbarMenuItemSubmenuIndicator = _toolbarMenuItemSubmenuIndicatorVariables.toolbarMenuItemSubmenuIndicatorVariables;

var _toolbarMenuItemActiveIndicatorVariables = require("./components/Toolbar/toolbarMenuItemActiveIndicatorVariables");

exports.ToolbarMenuItemActiveIndicator = _toolbarMenuItemActiveIndicatorVariables.toolbarMenuItemActiveIndicatorVariables;

var _toolbarMenuRadioGroupVariables = require("./components/Toolbar/toolbarMenuRadioGroupVariables");

exports.ToolbarMenuRadioGroup = _toolbarMenuRadioGroupVariables.toolbarMenuRadioGroupVariables;

var _toolbarMenuRadioGroupWrapperVariables = require("./components/Toolbar/toolbarMenuRadioGroupWrapperVariables");

exports.ToolbarMenuRadioGroupWrapper = _toolbarMenuRadioGroupWrapperVariables.toolbarMenuRadioGroupWrapperVariables;

var _reactionVariables = require("./components/Reaction/reactionVariables");

exports.Reaction = _reactionVariables.reactionVariables;

var _providerVariables = require("./components/Provider/providerVariables");

exports.Provider = _providerVariables.providerVariables;

var _dropdownVariables = require("./components/Dropdown/dropdownVariables");

exports.Dropdown = _dropdownVariables.dropdownVariables;

var _dropdownItemVariables = require("./components/Dropdown/dropdownItemVariables");

exports.DropdownItem = _dropdownItemVariables.dropdownItemVariables;

var _dropdownSelectedItemVariables = require("./components/Dropdown/dropdownSelectedItemVariables");

exports.DropdownSelectedItem = _dropdownSelectedItemVariables.dropdownSelectedItemVariables;

var _labelVariables = require("./components/Label/labelVariables");

exports.Label = _labelVariables.labelVariables;

var _tooltipContentVariables = require("./components/Tooltip/tooltipContentVariables");

exports.TooltipContent = _tooltipContentVariables.tooltipContentVariables;

var _datepickerCalendarCellVariables = require("./components/Datepicker/datepickerCalendarCellVariables");

exports.DatepickerCalendarCell = _datepickerCalendarCellVariables.datepickerCalendarCellVariables;

var _datepickerCalendarCellButtonVariables = require("./components/Datepicker/datepickerCalendarCellButtonVariables");

exports.DatepickerCalendarCellButton = _datepickerCalendarCellButtonVariables.datepickerCalendarCellButtonVariables;

var _cardVariables = require("./components/Card/cardVariables");

exports.Card = _cardVariables.cardVariables;

var _splitButtonVariables = require("./components/SplitButton/splitButtonVariables");

exports.SplitButtonDivider = _splitButtonVariables.splitButtonVariables;
