// @fluentui/react-northstar/dist/commonjs/themes/teams-high-contrast/siteVariables.js
"use strict";

exports.__esModule = true;
var _exportNames = {
  focusInnerBorderColor: true,
  focusOuterBorderColor: true,
  shadow2: true,
  shadow4: true,
  shadow8: true,
  shadow16: true,
  shadow28: true,
  shadow64: true,
  bodyBackground: true,
  bodyColor: true
};
exports.bodyColor = exports.bodyBackground = exports.shadow64 = exports.shadow28 = exports.shadow16 = exports.shadow8 = exports.shadow4 = exports.shadow2 = exports.focusOuterBorderColor = exports.focusInnerBorderColor = void 0;

var _siteVariables = require("../teams/siteVariables");

var _colors = require("./colors");

Object.keys(_colors).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _colors[key]) return;
  exports[key] = _colors[key];
});
//
// BORDER STYLES
//
var focusInnerBorderColor = _siteVariables.colors.black;
exports.focusInnerBorderColor = focusInnerBorderColor;
var focusOuterBorderColor = _colors.accessibleCyan; //
// SHADOW LEVELS
//

exports.focusOuterBorderColor = focusOuterBorderColor;
var shadow2 = '0 0 2px rgba(0, 0, 0, .24), 0 1px 2px rgba(0, 0, 0, .28)';
exports.shadow2 = shadow2;
var shadow4 = '0 0 2px rgba(0, 0, 0, .24), 0 2px 4px rgba(0, 0, 0, .28)';
exports.shadow4 = shadow4;
var shadow8 = '0 0 2px rgba(0, 0, 0, .24), 0 4px 8px rgba(0, 0, 0, .28)';
exports.shadow8 = shadow8;
var shadow16 = '0 0 2px rgba(0, 0, 0, .24), 0 8px 16px rgba(0, 0, 0, .28)';
exports.shadow16 = shadow16;
var shadow28 = '0 0 8px rgba(0, 0, 0, .40), 0 14px 28px rgba(0, 0, 0, .48)';
exports.shadow28 = shadow28;
var shadow64 = '0 0 8px rgba(0, 0, 0, .40), 0 32px 64px rgba(0, 0, 0, .48)'; //
// SEMANTIC ASSIGNMENTS
//

exports.shadow64 = shadow64;
var bodyBackground = _siteVariables.colors.black;
exports.bodyBackground = bodyBackground;
var bodyColor = _siteVariables.colors.white;
exports.bodyColor = bodyColor;
