// @fluentui/react-northstar/dist/dts/src/themes/teams/components/Carousel/carouselNavigationStyles.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.carouselNavigationStyles = void 0;
var utils_1 = require("../../../../utils");
var colors_1 = require("../../colors");
exports.carouselNavigationStyles = {
    root: function (_a) {
        var p = _a.props, v = _a.variables, theme = _a.theme;
        var iconOnly = p.iconOnly, primary = p.primary, vertical = p.vertical, thumbnails = p.thumbnails;
        var colors = colors_1.getColorScheme(v.colorScheme, null, primary);
        var siteVariables = theme.siteVariables;
        return __assign(__assign(__assign(__assign({ display: 'flex', minHeight: utils_1.pxToRem(24), margin: 0, padding: 0, color: v.color, backgroundColor: v.backgroundColor || 'inherit', listStyleType: 'none', justifyContent: 'center', position: 'relative', zIndex: 2 }, (!vertical &&
            thumbnails && {
            justifyContent: 'start',
            transform: "translateX(" + utils_1.pxToRem(v.width / 2 - v.thumbnailWidth / 2 - +p.activeIndex * v.thumbnailWidth) + ")",
            transition: 'transform .5s ease',
        })), (iconOnly && { alignItems: 'center' })), (vertical && __assign({ flexDirection: 'column', backgroundColor: v.verticalBackgroundColor, width: 'fit-content', padding: utils_1.pxToRem(8) + " 0" }, (iconOnly && {
            display: 'inline-block',
            width: 'auto',
        })))), (!iconOnly &&
            !vertical && {
            // primary has hardcoded grey border color
            border: v.borderWidth + " solid " + (primary ? v.primaryBorderColor : v.borderColor || colors.border),
            borderRadius: siteVariables.borderRadiusMedium,
        }));
    },
};
