// @fluentui/react-northstar/dist/dts/src/themes/teams/components/Carousel/carouselVariables.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.carouselVariables = void 0;
var utils_1 = require("../../../../utils");
var carouselVariables = function (siteVars) { return ({
    width: 300,
    height: 300,
    focusOuterBorderColor: siteVars.focusOuterBorderColor,
    focusOuterBorderRadius: siteVars.borderRadiusMedium,
    paddleColor: siteVars.colorScheme.default.foreground4,
    paddleColorDisabled: siteVars.colorScheme.brand.foregroundDisabled,
    paddleBackgroundColor: siteVars.colorScheme.onyx.background3,
    paddleBackgroundColorActive: siteVars.colorScheme.onyx.backgroundPressed,
    paddleBackgroundColorHover: siteVars.colorScheme.onyx.backgroundHover,
    paddleBackgroundColorDisabled: siteVars.colorScheme.default.backgroundDisabled,
    paddleBorderColor: 'transparent',
    paddleBorderRadius: siteVars.borderRadiusMedium,
    paddleHeight: utils_1.pxToRem(32),
    paddleIndicatorSize: utils_1.pxToRem(16),
    focusBorderRadius: siteVars.borderRadiusMedium,
    focusBorderWidth: siteVars.borderWidth,
    focusInnerBorderColor: siteVars.focusInnerBorderColor,
    focusBorderZIndex: siteVars.zIndexes.foreground,
}); };
exports.carouselVariables = carouselVariables;
