// @fluentui/react-northstar/dist/dts/src/themes/teams/components/Carousel/carouselNavigationVariables.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.carouselNavigationVariables = exports.menuColorAreas = void 0;
var utils_1 = require("../../../../utils");
var colorUtils_1 = require("../../../colorUtils");
exports.menuColorAreas = utils_1.stringLiteralsArray('border', 'borderActive', 'foregroundActive', 'foregroundFocus', 'foregroundHover', 'backgroundHover', 'backgroundActive', 'backgroundFocus', 'foregroundDisabled');
var carouselNavigationVariables = function (siteVars) {
    return {
        width: 300,
        colorScheme: colorUtils_1.pickValuesFromColorScheme(colorUtils_1.extendColorScheme(siteVars.colorScheme, {
            default: {
                borderActive: siteVars.colorScheme.default.border2,
                backgroundActive: siteVars.colorScheme.default.backgroundActive1,
                backgroundFocus: siteVars.colorScheme.default.backgroundFocus1,
                foregroundDisabled: siteVars.colorScheme.default.foregroundDisabled1,
            },
            brand: {
                foregroundHover: siteVars.colors.white,
                backgroundHover: siteVars.colors.brand[300],
                foregroundActive: siteVars.colors.white,
                borderActive: siteVars.colors.brand[600],
                backgroundActive: siteVars.colors.brand[500],
                foregroundFocus: siteVars.colors.white,
                backgroundFocus: siteVars.colors.brand[300],
                foregroundDisabled: siteVars.colorScheme.brand.foregroundDisabled1,
            },
        }), exports.menuColorAreas),
        color: siteVars.colors.grey[500],
        colorActive: siteVars.colors.black,
        borderColor: undefined,
        borderColorActive: undefined,
        borderColorFocus: siteVars.colors.white,
        outlineColorFocus: siteVars.colors.black,
        backgroundColor: undefined,
        backgroundColorFocus: undefined,
        backgroundColorHover: undefined,
        backgroundColorActive: undefined,
        iconOnlyColorActive: siteVars.colors.brand[600],
        lineHeightBase: siteVars.lineHeightMedium,
        horizontalPadding: utils_1.pxToRem(14) + " " + utils_1.pxToRem(18),
        verticalBackgroundColor: siteVars.colors.white,
        verticalItemPadding: utils_1.pxToRem(7) + " " + utils_1.pxToRem(14),
        verticalItemBorderWidth: utils_1.pxToRem(2),
        verticalItemBorderColor: 'transparent',
        verticalBackgroundColorFocus: siteVars.colors.grey[150],
        primaryBorderColor: siteVars.colorScheme.default.border2,
        indicatorBackgroundColor: siteVars.colors.grey[500],
        indicatorActiveBackgroundColor: siteVars.colors.brand[600],
        borderWidth: utils_1.pxToRem(1),
        thumbnailWidth: 75,
    };
};
exports.carouselNavigationVariables = carouselNavigationVariables;
