// @fluentui/react-northstar/dist/dts/src/themes/teams/components/Carousel/carouselStyles.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.carouselStyles = void 0;
var utils_1 = require("../../../../utils");
exports.carouselStyles = {
    root: function () { return ({
        display: 'inline-block',
        position: 'relative',
    }); },
    itemsContainerWrapper: function (_a) {
        var v = _a.variables, p = _a.props;
        return (__assign({ display: 'flex', width: utils_1.pxToRem(v.width), overflowX: 'hidden' }, (p.shouldFocusContainer &&
            p.isFromKeyboard && {
            border: "1px solid " + v.focusOuterBorderColor,
            borderRadius: v.focusOuterBorderRadius,
        })));
    },
    itemsContainer: function () { return ({
        padding: 0,
        margin: 0,
        display: 'flex',
        listStyle: 'none',
        willChange: 'transform',
    }); },
};
