// @fluentui/react-northstar/dist/dts/src/themes/teams-dark/colors.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.categoryColorScheme = exports.colorScheme = void 0;
var siteVariables_1 = require("../teams/siteVariables");
var colors_1 = require("../teams/colors");
exports.colorScheme = {
    default: colors_1.createColorScheme({
        foreground: siteVariables_1.colors.white,
        foreground1: siteVariables_1.colors.grey[250],
        foreground2: siteVariables_1.colors.grey[300],
        foreground3: siteVariables_1.colors.white,
        foreground4: siteVariables_1.colors.white,
        foreground5: siteVariables_1.colors.grey[450],
        foreground6: siteVariables_1.colors.grey[550],
        foreground7: siteVariables_1.colors.white,
        foreground8: siteVariables_1.colors.grey[200],
        foreground9: siteVariables_1.colors.grey[340],
        background: siteVariables_1.colors.grey[650],
        background1: siteVariables_1.colors.grey[700],
        background2: siteVariables_1.colors.grey[800],
        background3: siteVariables_1.colors.grey[550],
        background4: siteVariables_1.colors.grey[600],
        background5: siteVariables_1.colors.grey[250],
        background6: siteVariables_1.colors.grey[500],
        border: siteVariables_1.colors.grey[450],
        border1: siteVariables_1.colors.grey[850],
        border2: siteVariables_1.colors.grey[900],
        border3: siteVariables_1.colors.grey[650],
        borderTransparent: 'transparent',
        borderTransparentActive: 'transparent',
        shadow: siteVariables_1.colors.black,
        shadowHover: siteVariables_1.colors.black,
        foregroundHover: siteVariables_1.colors.white,
        foregroundHover1: siteVariables_1.colors.white,
        foregroundHover2: siteVariables_1.colors.white,
        backgroundHover: siteVariables_1.colors.grey[550],
        backgroundHover1: siteVariables_1.colors.grey[550],
        backgroundHover2: 'transparent',
        backgroundHover3: siteVariables_1.colors.grey[850],
        backgroundHover4: siteVariables_1.colors.grey[600],
        borderHover: siteVariables_1.colors.grey[400],
        foregroundPressed: siteVariables_1.colors.white,
        backgroundPressed: siteVariables_1.colors.grey[500],
        backgroundPressed3: siteVariables_1.colors.grey[450],
        borderPressed: siteVariables_1.colors.grey[400],
        foregroundActive: siteVariables_1.colors.white,
        foregroundActive1: siteVariables_1.colors.white,
        backgroundActive: siteVariables_1.colors.grey[800],
        backgroundActive1: siteVariables_1.colors.grey[500],
        backgroundActive2: siteVariables_1.colors.grey[600],
        borderActive: siteVariables_1.colors.grey[450],
        borderActive1: siteVariables_1.colors.grey[850],
        borderActive2: siteVariables_1.colors.grey[900],
        borderActive3: siteVariables_1.colors.grey[650],
        borderActive4: siteVariables_1.colors.grey[440],
        foregroundFocus: siteVariables_1.colors.white,
        foregroundFocus1: siteVariables_1.colors.grey[250],
        foregroundFocus2: siteVariables_1.colors.grey[300],
        foregroundFocus3: siteVariables_1.colors.white,
        backgroundFocus: siteVariables_1.colors.grey[650],
        backgroundFocus1: siteVariables_1.colors.grey[700],
        backgroundFocus2: siteVariables_1.colors.grey[800],
        backgroundFocus3: siteVariables_1.colors.grey[550],
        borderFocusWithin: siteVariables_1.colors.black,
        borderFocus: siteVariables_1.colors.white,
        foregroundDisabled1: siteVariables_1.colors.grey[450],
        foregroundDisabled: siteVariables_1.colors.grey[450],
        backgroundDisabled: siteVariables_1.colors.grey[550],
        backgroundDisabled1: siteVariables_1.colors.grey[550],
        backgroundDisabled2: siteVariables_1.colors.grey[650],
        backgroundDisabled3: siteVariables_1.colors.grey[850],
        borderDisabled: siteVariables_1.colors.grey[550],
    }),
    brand: colors_1.createColorScheme({
        foreground: siteVariables_1.colors.brand[400],
        foreground1: siteVariables_1.colors.brand[400],
        foreground2: siteVariables_1.colors.brand[400],
        foreground3: siteVariables_1.colors.brand[200],
        foreground4: siteVariables_1.colors.white,
        foreground5: siteVariables_1.colors.grey[750],
        background: siteVariables_1.colors.brand[600],
        background1: siteVariables_1.colors.brand[900],
        background2: siteVariables_1.colors.brand[900],
        background3: siteVariables_1.colors.brand[1000],
        background4: siteVariables_1.colors.grey[650],
        background5: siteVariables_1.colors.brand[900],
        background6: siteVariables_1.colors.brand[450],
        border: siteVariables_1.colors.grey[450],
        border1: siteVariables_1.colors.brand[800],
        border2: siteVariables_1.colors.brand[800],
        shadow: siteVariables_1.colors.black,
        shadowHover: siteVariables_1.colors.black,
        foregroundHover: siteVariables_1.colors.brand[400],
        foregroundHover1: siteVariables_1.colors.white,
        foregroundHover2: siteVariables_1.colors.brand[200],
        borderHover: siteVariables_1.colors.brand[600],
        backgroundHover: siteVariables_1.colors.brand[500],
        backgroundHover1: siteVariables_1.colors.brand[900],
        backgroundHover2: siteVariables_1.colors.brand[900],
        backgroundHover3: siteVariables_1.colors.brand[900],
        foregroundPressed: siteVariables_1.colors.brand[200],
        foregroundPressed1: siteVariables_1.colors.white,
        backgroundPressed: siteVariables_1.colors.brand[700],
        backgroundPressed1: siteVariables_1.colors.brand[800],
        backgroundPressed2: siteVariables_1.colors.brand[800],
        borderPressed: siteVariables_1.colors.brand[800],
        foregroundActive: siteVariables_1.colors.brand[400],
        foregroundActive1: siteVariables_1.colors.brand[400],
        foregroundActive2: siteVariables_1.colors.brand[200],
        backgroundActive: siteVariables_1.colors.brand[400],
        backgroundActive1: siteVariables_1.colors.brand[400],
        // active border no change (just copied)
        borderActive: siteVariables_1.colors.grey[450],
        borderActive1: siteVariables_1.colors.brand[800],
        borderActive2: siteVariables_1.colors.brand[800],
        foregroundFocus: siteVariables_1.colors.brand[400],
        foregroundFocus1: siteVariables_1.colors.brand[400],
        foregroundFocus2: siteVariables_1.colors.brand[400],
        foregroundFocus3: siteVariables_1.colors.brand[200],
        foregroundFocus4: siteVariables_1.colors.white,
        backgroundFocus: siteVariables_1.colors.brand[600],
        backgroundFocus1: siteVariables_1.colors.brand[900],
        backgroundFocus2: siteVariables_1.colors.brand[900],
        backgroundFocus3: siteVariables_1.colors.brand[1000],
        borderFocus: siteVariables_1.colors.white,
        borderFocusWithin: siteVariables_1.colors.black,
        borderFocus1: siteVariables_1.colors.brand[400],
        foregroundDisabled1: siteVariables_1.colors.grey[450],
        foregroundDisabled: siteVariables_1.colors.grey[450],
        backgroundDisabled1: siteVariables_1.colors.grey[550],
        backgroundDisabled: siteVariables_1.colors.grey[550],
        backgroundDisabled2: siteVariables_1.colors.grey[650],
        borderDisabled: siteVariables_1.colors.grey[550],
    }),
    black: {
        foreground: siteVariables_1.colors.black,
        foreground1: siteVariables_1.colors.grey[750],
        background: siteVariables_1.colors.white,
        background1: siteVariables_1.colors.white,
        border: siteVariables_1.colors.black,
        shadow: siteVariables_1.colors.black,
        foregroundHover: siteVariables_1.colors.white,
        backgroundHover: siteVariables_1.colors.black,
        borderHover: siteVariables_1.colors.black,
        shadowHover: siteVariables_1.colors.black,
        foregroundActive: siteVariables_1.colors.white,
        backgroundActive: siteVariables_1.colors.black,
        borderActive: siteVariables_1.colors.black,
        shadowActive: siteVariables_1.colors.black,
        foregroundFocus: siteVariables_1.colors.white,
        backgroundFocus: siteVariables_1.colors.black,
        borderFocus: siteVariables_1.colors.black,
        shadowFocus: siteVariables_1.colors.black,
        foregroundPressed: siteVariables_1.colors.white,
        backgroundPressed: siteVariables_1.colors.black,
        borderPressed: siteVariables_1.colors.black,
        shadowPressed: siteVariables_1.colors.black,
        foregroundDisabled: siteVariables_1.colors.white,
        backgroundDisabled: siteVariables_1.colors.black,
        borderDisabled: siteVariables_1.colors.black,
        shadowDisabled: siteVariables_1.colors.black,
    },
    white: {
        foreground: siteVariables_1.colors.white,
        foreground1: siteVariables_1.colors.white,
        background: siteVariables_1.colors.black,
        background1: siteVariables_1.colors.grey[700],
        border: siteVariables_1.colors.white,
        shadow: siteVariables_1.colors.white,
        foregroundHover: siteVariables_1.colors.black,
        backgroundHover: siteVariables_1.colors.white,
        borderHover: siteVariables_1.colors.white,
        shadowHover: siteVariables_1.colors.white,
        foregroundActive: siteVariables_1.colors.black,
        backgroundActive: siteVariables_1.colors.white,
        borderActive: siteVariables_1.colors.white,
        shadowActive: siteVariables_1.colors.white,
        foregroundFocus: siteVariables_1.colors.black,
        backgroundFocus: siteVariables_1.colors.white,
        borderFocus: siteVariables_1.colors.white,
        shadowFocus: siteVariables_1.colors.white,
        foregroundPressed: siteVariables_1.colors.black,
        backgroundPressed: siteVariables_1.colors.white,
        borderPressed: siteVariables_1.colors.white,
        shadowPressed: siteVariables_1.colors.white,
        foregroundDisabled: siteVariables_1.colors.black,
        backgroundDisabled: siteVariables_1.colors.white,
        borderDisabled: siteVariables_1.colors.white,
        shadowDisabled: siteVariables_1.colors.white,
    },
    green: colors_1.createColorScheme({
        foreground: siteVariables_1.colors.green[200],
        foreground1: siteVariables_1.colors.grey[800],
        foreground2: siteVariables_1.colors.green[200],
        foreground3: siteVariables_1.colors.green[750],
        background: siteVariables_1.colors.green[200],
        background1: siteVariables_1.colors.green[200],
        background2: siteVariables_1.colors.green[800],
        border: siteVariables_1.colors.green[900],
        border1: siteVariables_1.colors.green[200],
    }),
    orange: colors_1.createColorScheme({
        foreground: siteVariables_1.colors.orange[300],
        foreground1: siteVariables_1.colors.orange[300],
        foreground2: siteVariables_1.colors.grey[750],
        background: siteVariables_1.colors.orange[300],
        background1: siteVariables_1.colors.orange[300],
        border: siteVariables_1.colors.grey[400],
    }),
    pink: colors_1.createColorScheme({
        foreground: siteVariables_1.colors.pink[200],
        foreground1: siteVariables_1.colors.pink[400],
        background: siteVariables_1.colors.pink[800],
        border: siteVariables_1.colors.pink[900],
    }),
    red: colors_1.createColorScheme({
        foreground: siteVariables_1.colors.red[200],
        foreground1: siteVariables_1.colors.white,
        foreground2: siteVariables_1.colors.grey[800],
        background: siteVariables_1.colors.red[300],
        background1: siteVariables_1.colors.red[800],
        background2: siteVariables_1.colors.ruby[500],
        background3: siteVariables_1.colors.red[400],
        border: siteVariables_1.colors.red[900],
        border1: siteVariables_1.colors.red[200],
        foregroundHover: siteVariables_1.colors.white,
        backgroundHover: siteVariables_1.colors.ruby[600],
        backgroundHover1: siteVariables_1.colors.red[400],
        foregroundPressed: siteVariables_1.colors.white,
        backgroundPressed: siteVariables_1.colors.ruby[700],
    }),
    yellow: colors_1.createColorScheme({
        foreground: siteVariables_1.colors.yellow[300],
        foreground1: siteVariables_1.colors.grey[800],
        foreground2: siteVariables_1.colors.grey[800],
        foreground3: siteVariables_1.colors.grey[750],
        foreground4: siteVariables_1.colors.yellow[200],
        background: siteVariables_1.colors.yellow[400],
        background1: siteVariables_1.colors.grey[450],
        background2: siteVariables_1.colors.yellow[500],
        background3: siteVariables_1.colors.yellow[800],
        border: siteVariables_1.colors.yellow[900],
        border1: siteVariables_1.colors.yellow[200],
    }),
    silver: colors_1.createColorScheme({
        foreground: siteVariables_1.colors.white,
        foreground1: siteVariables_1.colors.silver[200],
        foregroundHover: siteVariables_1.colors.white,
        foregroundPressed: siteVariables_1.colors.white,
        border: siteVariables_1.colors.silver[600],
        background: 'transparent',
        backgroundHover: siteVariables_1.colors.silver[800],
        borderHover: siteVariables_1.colors.silver[600],
        backgroundPressed: siteVariables_1.colors.silver[700],
        borderPressed: siteVariables_1.colors.silver[600],
        foregroundDisabled: siteVariables_1.colors.silver[600],
        backgroundDisabled: siteVariables_1.colors.silver[900],
    }),
    onyx: colors_1.createColorScheme({
        background: siteVariables_1.colors.onyx[500],
        background1: siteVariables_1.colors.onyx[100],
        background2: siteVariables_1.colors.onyx[500],
        background3: siteVariables_1.colors.onyx[700],
        backgroundHover: siteVariables_1.colors.onyx[900],
        backgroundPressed: siteVariables_1.colors.onyx[900],
        border: siteVariables_1.colors.onyx[800],
        border1: 'transparent',
        border2: siteVariables_1.colors.silver[400],
    }),
    amethyst: colors_1.createColorScheme({
        background: siteVariables_1.colors.onyx[200],
        backgroundHover: siteVariables_1.colors.amethyst[700],
        backgroundHover1: siteVariables_1.colors.grey[750],
        backgroundActive: siteVariables_1.colors.amethyst[700],
    }),
};
exports.colorScheme.grey = exports.colorScheme.default;
var createCategoryColorScheme = function (color, customValues) {
    if (customValues === void 0) { customValues = {}; }
    return __assign({ foreground: siteVariables_1.categoryColors[color][250], foreground1: siteVariables_1.categoryColors[color][550], background: siteVariables_1.categoryColors[color][800], borderActive: siteVariables_1.categoryColors[color][350] }, customValues);
};
exports.categoryColorScheme = {
    redDark: createCategoryColorScheme('redDark'),
    red: createCategoryColorScheme('red'),
    orangeDark: createCategoryColorScheme('orangeDark'),
    orange: createCategoryColorScheme('orange'),
    orangeLight: createCategoryColorScheme('orangeLight'),
    yellowDark: createCategoryColorScheme('yellowDark'),
    yellow: createCategoryColorScheme('yellow'),
    brown: createCategoryColorScheme('brown'),
    oliveDark: createCategoryColorScheme('oliveDark'),
    olive: createCategoryColorScheme('olive'),
    greenDark: createCategoryColorScheme('greenDark'),
    green: createCategoryColorScheme('green'),
    tealDark: createCategoryColorScheme('tealDark'),
    teal: createCategoryColorScheme('teal'),
    tealLight: createCategoryColorScheme('tealLight'),
    blueDark: createCategoryColorScheme('blueDark'),
    blue: createCategoryColorScheme('blue'),
    purpleDark: createCategoryColorScheme('purpleDark'),
    purple: createCategoryColorScheme('purple'),
    maroon: createCategoryColorScheme('maroon'),
    pink: createCategoryColorScheme('pink'),
    smokeDark: createCategoryColorScheme('smokeDark'),
    smokeLight: createCategoryColorScheme('smokeLight'),
    steelDark: createCategoryColorScheme('steelDark'),
    steelLight: createCategoryColorScheme('steelLight'),
    neon: createCategoryColorScheme('neon'),
    formatting: {
        foreground1: siteVariables_1.categoryColors.red[250],
        background1: siteVariables_1.categoryColors.red[700],
        foreground2: siteVariables_1.categoryColors.orangeDark[350],
        background2: siteVariables_1.categoryColors.orange[700],
        foreground3: siteVariables_1.categoryColors.yellow[250],
        background3: siteVariables_1.categoryColors.yellow[800],
        foreground4: siteVariables_1.categoryColors.neon[150],
        background4: siteVariables_1.categoryColors.neon[800],
        foreground5: siteVariables_1.categoryColors.green[250],
        background5: siteVariables_1.categoryColors.green[700],
        foreground6: siteVariables_1.categoryColors.tealLight[250],
        background6: siteVariables_1.categoryColors.tealLight[700],
        foreground7: siteVariables_1.categoryColors.blueDark[250],
        background7: siteVariables_1.categoryColors.blueDark[450],
        foreground8: siteVariables_1.categoryColors.maroon[350],
        background8: siteVariables_1.categoryColors.maroon[800],
    },
};
