// @fluentui/react-northstar/dist/dts/src/themes/teams-dark/siteVariables.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.bodyColor = exports.bodyBackground = exports.shadow64 = exports.shadow28 = exports.shadow16 = exports.shadow8 = exports.shadow4 = exports.shadow2 = exports.shadowLevel4 = exports.shadowLevel3 = exports.shadowLevel2 = exports.shadowLevel1 = exports.focusOuterBorderColor = exports.focusInnerBorderColor = exports.categoryColorScheme = exports.colorScheme = void 0;
var siteVariables_1 = require("../teams/siteVariables");
var colors_1 = require("./colors");
Object.defineProperty(exports, "colorScheme", { enumerable: true, get: function () { return colors_1.colorScheme; } });
Object.defineProperty(exports, "categoryColorScheme", { enumerable: true, get: function () { return colors_1.categoryColorScheme; } });
//
// BORDER STYLES
//
exports.focusInnerBorderColor = siteVariables_1.colors.black;
exports.focusOuterBorderColor = siteVariables_1.colors.white;
//
// SHADOW LEVELS
//
exports.shadowLevel1 = '0 .2rem .4rem -.075rem rgba(0, 0, 0, .25)';
exports.shadowLevel2 = '0 .4rem 0.7rem -0.1rem rgba(0, 0, 0, .25)';
exports.shadowLevel3 = '0 .8rem 1rem -0.2rem rgba(0, 0, 0, .25)';
exports.shadowLevel4 = '0 .6rem 1.8rem -0.4rem rgba(0, 0, 0, .25)';
exports.shadow2 = '0 0 2px rgba(0, 0, 0, .24), 0 1px 2px rgba(0, 0, 0, .28)';
exports.shadow4 = '0 0 2px rgba(0, 0, 0, .24), 0 2px 4px rgba(0, 0, 0, .28)';
exports.shadow8 = '0 0 2px rgba(0, 0, 0, .24), 0 4px 8px rgba(0, 0, 0, .28)';
exports.shadow16 = '0 0 2px rgba(0, 0, 0, .24), 0 8px 16px rgba(0, 0, 0, .28)';
exports.shadow28 = '0 0 8px rgba(0, 0, 0, .40), 0 14px 28px rgba(0, 0, 0, .48)';
exports.shadow64 = '0 0 8px rgba(0, 0, 0, .40), 0 32px 64px rgba(0, 0, 0, .48)';
//
// SEMANTIC ASSIGNMENTS
//
exports.bodyBackground = siteVariables_1.colors.grey[650];
exports.bodyColor = siteVariables_1.colors.white;
