// @fluentui/react-northstar/dist/commonjs/components/Form/FormFieldCustom.js
"use strict";

exports.__esModule = true;
exports.FormFieldCustom = exports.formFieldCustomClassName = void 0;

var _accessibility = require("@fluentui/accessibility");

var PropTypes = _interopRequireWildcard(require("prop-types"));

var React = _interopRequireWildcard(require("react"));

var _utils = require("../../utils");

var _reactBindings = require("@fluentui/react-bindings");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var formFieldCustomClassName = 'ui-form__fieldcustom';
exports.formFieldCustomClassName = formFieldCustomClassName;

/**
 * A FormFieldCustom represents a Form element containing a label and an input.
 */
var FormFieldCustom = /*#__PURE__*/React.forwardRef(function (props, ref) {
  var context = (0, _reactBindings.useFluentContext)();

  var _useTelemetry = (0, _reactBindings.useTelemetry)(FormFieldCustom.displayName, context.telemetry),
      setStart = _useTelemetry.setStart,
      setEnd = _useTelemetry.setEnd;

  setStart();
  var children = props.children,
      required = props.required,
      type = props.type,
      className = props.className,
      design = props.design,
      styles = props.styles,
      variables = props.variables,
      inline = props.inline;
  var ElementType = (0, _reactBindings.getElementType)(props);
  var unhandledProps = (0, _reactBindings.useUnhandledProps)(FormFieldCustom.handledProps, props);
  var getA11yProps = (0, _reactBindings.useAccessibility)(props.accessibility, {
    debugName: FormFieldCustom.displayName,
    rtl: context.rtl
  });

  var _useStyles = (0, _reactBindings.useStyles)(FormFieldCustom.displayName, {
    className: formFieldCustomClassName,
    mapPropsToStyles: function mapPropsToStyles() {
      return {
        type: type,
        inline: inline,
        required: required
      };
    },
    mapPropsToInlineStyles: function mapPropsToInlineStyles() {
      return {
        className: className,
        design: design,
        styles: styles,
        variables: variables
      };
    },
    rtl: context.rtl
  }),
      classes = _useStyles.classes;

  var element = /*#__PURE__*/React.createElement(ElementType, getA11yProps('root', Object.assign({
    className: classes.root,
    ref: ref
  }, unhandledProps)), children);
  setEnd();
  return element;
});
exports.FormFieldCustom = FormFieldCustom;
FormFieldCustom.displayName = 'FormFieldCustom';
FormFieldCustom.propTypes = Object.assign({}, _utils.commonPropTypes.createCommon({
  content: false,
  children: false
}), {
  id: PropTypes.string,
  inline: PropTypes.bool,
  name: PropTypes.string,
  required: PropTypes.bool,
  type: PropTypes.string
});
FormFieldCustom.handledProps = Object.keys(FormFieldCustom.propTypes);
FormFieldCustom.defaultProps = {
  accessibility: _accessibility.formFieldBehavior
};
