// @fluentui/react-northstar/dist/commonjs/components/Form/FormTextArea.js
"use strict";

exports.__esModule = true;
exports.FormTextArea = exports.formTextAreaClassName = void 0;

var _reactBindings = require("@fluentui/react-bindings");

var _TextArea = require("../TextArea/TextArea");

var _formFieldBase = require("./utils/formFieldBase");

var _utils = require("../../utils");

var formTextAreaClassName = 'ui-form__textarea';
/**
 * An FormTextArea renders a TextArea wrapped by FormField.
 */

exports.formTextAreaClassName = formTextAreaClassName;
var FormTextArea = (0, _reactBindings.compose)(_formFieldBase._FormFieldBase, {
  className: formTextAreaClassName,
  displayName: 'FormTextArea',
  overrideStyles: true,
  slots: {
    control: _TextArea.TextArea
  },
  slotProps: function slotProps(_ref) {
    var errorMessage = _ref.errorMessage,
        required = _ref.required;
    return {
      label: {
        required: required
      },
      control: {
        error: !!errorMessage
      },
      message: {
        error: !!errorMessage
      }
    };
  }
});
exports.FormTextArea = FormTextArea;
FormTextArea.propTypes = _utils.commonPropTypes.createCommon({
  children: false,
  content: false
});
