// @fluentui/react-northstar/dist/commonjs/components/Form/FormRadioGroup.js
"use strict";

exports.__esModule = true;
exports.FormRadioGroup = exports.formRadioGroupClassName = void 0;

var _reactBindings = require("@fluentui/react-bindings");

var _utils = require("../../utils");

var _formFieldBase = require("./utils/formFieldBase");

var _RadioGroup = require("../RadioGroup/RadioGroup");

var formRadioGroupClassName = 'ui-forms';
/**
 * An FormRadioGroup renders a RadioGroup wrapped by FormField.
 */

exports.formRadioGroupClassName = formRadioGroupClassName;
var FormRadioGroup = (0, _reactBindings.compose)(_formFieldBase._FormFieldBase, {
  className: formRadioGroupClassName,
  displayName: 'FormRadioGroup',
  overrideStyles: true,
  slots: {
    control: _RadioGroup.RadioGroup
  },
  slotProps: function slotProps(_ref) {
    var errorMessage = _ref.errorMessage;
    return {
      message: {
        error: !!errorMessage
      }
    };
  }
});
exports.FormRadioGroup = FormRadioGroup;
FormRadioGroup.propTypes = _utils.commonPropTypes.createCommon({
  children: false,
  content: false
});
