// @fluentui/react-northstar/dist/commonjs/components/Form/FormDropdown.js
"use strict";

exports.__esModule = true;
exports.FormDropdown = exports.formDropdownClassName = void 0;

var _reactBindings = require("@fluentui/react-bindings");

var _utils = require("../../utils");

var _Dropdown = require("../Dropdown/Dropdown");

var _formFieldBase = require("./utils/formFieldBase");

var formDropdownClassName = 'ui-form__dropdown';
/**
 * An FormDropdown renders a Dropdown wrapped by FormField.
 */

exports.formDropdownClassName = formDropdownClassName;
var FormDropdown = (0, _reactBindings.compose)(_formFieldBase._FormFieldBase, {
  className: formDropdownClassName,
  displayName: 'FormDropdown',
  overrideStyles: true,
  slots: {
    control: _Dropdown.Dropdown
  },
  slotProps: function slotProps(_ref) {
    var errorMessage = _ref.errorMessage,
        inline = _ref.inline;
    return {
      control: {
        inline: inline
      },
      message: {
        error: !!errorMessage
      }
    };
  }
});
exports.FormDropdown = FormDropdown;
FormDropdown.propTypes = _utils.commonPropTypes.createCommon({
  accessibility: false,
  children: false,
  content: false
});
