// @fluentui/react-northstar/dist/commonjs/components/Form/FormInput.js
"use strict";

exports.__esModule = true;
exports.FormInput = exports.formInputClassName = void 0;

var _reactBindings = require("@fluentui/react-bindings");

var _Input = require("../Input/Input");

var _formFieldBase = require("./utils/formFieldBase");

var _utils = require("../../utils");

var formInputClassName = 'ui-form__input';
/**
 * An FormInput renders a Input wrapped by FormField.
 */

exports.formInputClassName = formInputClassName;
var FormInput = (0, _reactBindings.compose)(_formFieldBase._FormFieldBase, {
  className: formInputClassName,
  displayName: 'FormInput',
  handledProps: ['label', 'labelPosition', 'required'],
  overrideStyles: true,
  slots: {
    control: _Input.Input
  },
  slotProps: function slotProps(_ref) {
    var errorMessage = _ref.errorMessage,
        required = _ref.required,
        labelPosition = _ref.labelPosition;
    return {
      label: {
        required: required
      },
      control: {
        error: !!errorMessage,
        labelPosition: labelPosition,
        required: required
      },
      message: {
        error: !!errorMessage
      }
    };
  }
});
exports.FormInput = FormInput;
FormInput.propTypes = _utils.commonPropTypes.createCommon({
  children: false,
  content: false
});
