// @fluentui/react-northstar/dist/commonjs/components/Form/FormCheckbox.js
"use strict";

exports.__esModule = true;
exports.FormCheckbox = exports.formCheckboxClassName = void 0;

var _reactBindings = require("@fluentui/react-bindings");

var _utils = require("../../utils");

var _Checkbox = require("../Checkbox/Checkbox");

var _formFieldBase = require("./utils/formFieldBase");

var formCheckboxClassName = 'ui-form__checkbox';
/**
 * An FormCheckbox renders a Checkbox wrapped by FormField.
 */

exports.formCheckboxClassName = formCheckboxClassName;
var FormCheckbox = (0, _reactBindings.compose)(_formFieldBase._FormFieldBase, {
  className: formCheckboxClassName,
  displayName: 'FormCheckbox',
  overrideStyles: true,
  handledProps: ['label'],
  slots: {
    label: function label() {
      return null;
    },
    control: _Checkbox.Checkbox
  },
  slotProps: function slotProps(_ref) {
    var label = _ref.label,
        errorMessage = _ref.errorMessage;
    return {
      control: {
        label: label
      },
      message: {
        error: !!errorMessage
      }
    };
  }
});
exports.FormCheckbox = FormCheckbox;
FormCheckbox.propTypes = _utils.commonPropTypes.createCommon({
  content: false,
  children: false
});
