// @fluentui/react-northstar/dist/commonjs/components/Form/FormDatepicker.js
"use strict";

exports.__esModule = true;
exports.FormDatepicker = exports.formDatepickerClassName = void 0;

var _reactBindings = require("@fluentui/react-bindings");

var _utils = require("../../utils");

var _Datepicker = require("../Datepicker/Datepicker");

var _formFieldBase = require("./utils/formFieldBase");

var formDatepickerClassName = 'ui-form__datepicker';
/**
 * An FormDatepicker renders a Datepicker wrapped by FormField.
 */

exports.formDatepickerClassName = formDatepickerClassName;
var FormDatepicker = (0, _reactBindings.compose)(_formFieldBase._FormFieldBase, {
  className: formDatepickerClassName,
  displayName: 'FormDatepicker',
  overrideStyles: true,
  slots: {
    control: _Datepicker.Datepicker
  },
  slotProps: function slotProps(_ref) {
    var errorMessage = _ref.errorMessage,
        required = _ref.required;
    return {
      control: {
        required: required
      },
      message: {
        error: !!errorMessage
      }
    };
  }
});
exports.FormDatepicker = FormDatepicker;
FormDatepicker.propTypes = _utils.commonPropTypes.createCommon({
  accessibility: false,
  children: false,
  content: false
});
