// @fluentui/react-northstar/dist/dts/src/themes/teams-v2/siteVariables.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.colorScheme = exports.colors = void 0;
exports.colors = {
    black: '#000',
    white: '#fff',
    grey: {
        25: '#fafafa',
        50: '#f5f5f5',
        100: '#f0f0f0',
        150: '#ebebeb',
        200: '#e0e0e0',
        220: '#d6d6d6',
        230: '#d1d1d1',
        250: '#c7c7c7',
        270: '#bdbdbd',
        300: '#b3b3b3',
        310: '#adadad',
        350: '#949494',
        400: '#8a8a8a',
        430: '#707070',
        440: '#666',
        450: '#616161',
        460: '#5c5c5c',
        500: '#424242',
        550: '#3d3d3d',
        600: '#333',
        650: '#2e2e2e',
        700: '#292929',
        750: '#242424',
        800: '#1f1f1f',
        850: '#1a1a1a',
        870: '#141414',
        900: '#0f0f0f',
        910: '#0a0a0a',
    },
    brand: {
        50: '#e8ebfa',
        100: '#dce0fa',
        200: '#c5cbfa',
        300: '#9299f7',
        400: '#7f85f5',
        450: '#7f85f5',
        500: '#7579eb',
        600: '#5b5fc7',
        700: '#4f52b2',
        800: '#444791',
        900: '#3d3e78',
        1000: '#2f2f4a',
    },
};
exports.colorScheme = {
    default: {
        foreground: exports.colors.grey['750'],
        foreground1: exports.colors.grey['500'],
        foreground2: exports.colors.grey['450'],
        foreground3: exports.colors.white,
        foreground4: exports.colors.white,
        foreground5: exports.colors.grey['100'],
        foreground6: exports.colors.grey['200'],
        foreground7: exports.colors.grey['750'],
        background: exports.colors.white,
        background1: exports.colors.grey['25'],
        background2: exports.colors.grey['50'],
        background3: exports.colors.grey['100'],
        background4: exports.colors.grey['150'],
        background5: exports.colors.grey['200'],
        border: exports.colors.grey['230'],
        border1: exports.colors.grey['100'],
        border2: exports.colors.grey['200'],
        border3: exports.colors.grey['100'],
        foregroundHover: exports.colors.grey['750'],
        foregroundHover1: exports.colors.white,
        foregroundHover2: exports.colors.white,
        backgroundHover: exports.colors.grey['50'],
        backgroundHover1: exports.colors.grey['25'],
        backgroundHover2: 'transparent',
        backgroundHover3: exports.colors.grey['150'],
        backgroundHover4: exports.colors.grey['25'],
        borderHover: exports.colors.grey['250'],
        foregroundPressed: exports.colors.grey['750'],
        backgroundPressed: exports.colors.grey['200'],
        foregroundActive: exports.colors.grey['750'],
        foregroundActive1: exports.colors.white,
        backgroundActive: exports.colors.grey['150'],
        backgroundActive1: exports.colors.white,
        borderActive: exports.colors.grey['270'],
        // foregroundFocus: not specified,
        // backgroundFocus: not specified,
        borderFocus: exports.colors.black,
        borderFocusWithin: exports.colors.white,
        foregroundDisabled: exports.colors.grey['250'],
        foregroundDisabled1: exports.colors.grey['250'],
        borderDisabled: exports.colors.grey['200'],
        backgroundDisabled: exports.colors.grey['100'],
        backgroundDisabled1: exports.colors.grey['100'],
    },
    brand: {
        background: exports.colors.brand['600'],
        background1: exports.colors.brand['50'],
        background2: exports.colors.brand['900'],
        background3: exports.colors.brand['1000'],
        background4: exports.colors.brand['800'],
        background5: exports.colors.brand['100'],
        foreground: exports.colors.brand['600'],
        foreground1: exports.colors.brand['600'],
        foreground2: exports.colors.brand['700'],
        foreground3: exports.colors.brand['200'],
        foreground4: exports.colors.white,
        border: exports.colors.grey['200'],
        border1: exports.colors.brand['300'],
        border2: exports.colors.brand['200'],
        foregroundHover: exports.colors.brand['600'],
        foregroundHover1: exports.colors.white,
        foregroundHover2: exports.colors.brand['200'],
        borderHover: exports.colors.brand['300'],
        backgroundHover: exports.colors.brand['600'],
        backgroundHover1: exports.colors.brand['50'],
        foregroundPressed: exports.colors.brand['700'],
        foregroundPressed1: exports.colors.white,
        backgroundPressed: exports.colors.brand['800'],
        borderPressed: exports.colors.brand['300'],
        foregroundActive: exports.colors.brand['600'],
        foregroundActive1: exports.colors.brand['600'],
        foregroundActive2: exports.colors.brand['50'],
        backgroundActive: exports.colors.brand['600'],
        backgroundActive1: exports.colors.brand['600'],
        borderActive: exports.colors.grey['200'],
        borderActive1: exports.colors.brand['50'],
        borderActive2: exports.colors.brand['300'],
        foregroundFocus: exports.colors.brand['600'],
        foregroundFocus1: exports.colors.brand['600'],
        foregroundFocus2: exports.colors.brand['700'],
        foregroundFocus3: exports.colors.brand['50'],
        foregroundFocus4: exports.colors.white,
        backgroundFocus: exports.colors.brand['600'],
        backgroundFocus1: exports.colors.brand['50'],
        backgroundFocus2: exports.colors.brand['900'],
        backgroundFocus3: exports.colors.brand['1000'],
        borderFocus: exports.colors.black,
        borderFocus1: exports.colors.brand['600'],
        borderFocusWithin: exports.colors.white,
        foregroundDisabled: exports.colors.grey['250'],
        foregroundDisabled1: exports.colors.grey['250'],
        borderDisabled: exports.colors.grey['550'],
        backgroundDisabled: exports.colors.grey['100'],
        backgroundDisabled1: exports.colors.grey['100'],
    },
};
