// @fluentui/react-northstar/dist/commonjs/utils/positioner/index.js
"use strict";

exports.__esModule = true;
var _exportNames = {
  createReferenceFromClick: true,
  getBasePlacement: true,
  getScrollParent: true,
  partitionPopperPropsFromShorthand: true,
  Popper: true,
  usePopper: true
};
exports.usePopper = exports.Popper = exports.partitionPopperPropsFromShorthand = exports.getScrollParent = exports.getBasePlacement = exports.createReferenceFromClick = void 0;

var _types = require("./types");

Object.keys(_types).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _types[key]) return;
  exports[key] = _types[key];
});

var _types2 = require("./types.internal");

Object.keys(_types2).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _types2[key]) return;
  exports[key] = _types2[key];
});

var _createReferenceFromClick = require("./createReferenceFromClick");

exports.createReferenceFromClick = _createReferenceFromClick.createReferenceFromClick;

var _getBasePlacement = require("./getBasePlacement");

exports.getBasePlacement = _getBasePlacement.getBasePlacement;

var _getScrollParent = require("./getScrollParent");

exports.getScrollParent = _getScrollParent.getScrollParent;

var _partitionPopperPropsFromShorthand = require("./partitionPopperPropsFromShorthand");

exports.partitionPopperPropsFromShorthand = _partitionPopperPropsFromShorthand.partitionPopperPropsFromShorthand;

var _Popper = require("./Popper");

exports.Popper = _Popper.Popper;

var _usePopper = require("./usePopper");

exports.usePopper = _usePopper.usePopper;
