// @fluentui/react-northstar/dist/dts/src/components/Form/FormInput.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FormInput = exports.formInputClassName = void 0;
var react_bindings_1 = require("@fluentui/react-bindings");
var Input_1 = require("../Input/Input");
var formFieldBase_1 = require("./utils/formFieldBase");
var utils_1 = require("../../utils");
exports.formInputClassName = 'ui-form__input';
/**
 * An FormInput renders a Input wrapped by FormField.
 */
exports.FormInput = react_bindings_1.compose(formFieldBase_1._FormFieldBase, {
    className: exports.formInputClassName,
    displayName: 'FormInput',
    handledProps: ['label', 'labelPosition', 'required'],
    overrideStyles: true,
    slots: {
        control: Input_1.Input,
    },
    slotProps: function (_a) {
        var errorMessage = _a.errorMessage, required = _a.required, labelPosition = _a.labelPosition;
        return ({
            label: {
                required: required,
            },
            control: {
                error: !!errorMessage,
                labelPosition: labelPosition,
                required: required,
            },
            message: {
                error: !!errorMessage,
            },
        });
    },
});
exports.FormInput.propTypes = utils_1.commonPropTypes.createCommon({
    children: false,
    content: false,
});
