// @fluentui/react-northstar/dist/dts/src/components/Form/FormDropdown.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FormDropdown = exports.formDropdownClassName = void 0;
var react_bindings_1 = require("@fluentui/react-bindings");
var utils_1 = require("../../utils");
var Dropdown_1 = require("../Dropdown/Dropdown");
var formFieldBase_1 = require("./utils/formFieldBase");
exports.formDropdownClassName = 'ui-form__dropdown';
/**
 * An FormDropdown renders a Dropdown wrapped by FormField.
 */
exports.FormDropdown = react_bindings_1.compose(formFieldBase_1._FormFieldBase, {
    className: exports.formDropdownClassName,
    displayName: 'FormDropdown',
    overrideStyles: true,
    slots: {
        control: Dropdown_1.Dropdown,
    },
    slotProps: function (_a) {
        var errorMessage = _a.errorMessage, inline = _a.inline;
        return ({
            control: {
                inline: inline,
            },
            message: {
                error: !!errorMessage,
            },
        });
    },
});
exports.FormDropdown.propTypes = utils_1.commonPropTypes.createCommon({
    accessibility: false,
    children: false,
    content: false,
});
