// @fluentui/react-northstar/dist/dts/src/components/Form/FormRadioGroup.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FormRadioGroup = exports.formRadioGroupClassName = void 0;
var react_bindings_1 = require("@fluentui/react-bindings");
var utils_1 = require("../../utils");
var formFieldBase_1 = require("./utils/formFieldBase");
var RadioGroup_1 = require("../RadioGroup/RadioGroup");
exports.formRadioGroupClassName = 'ui-forms';
/**
 * An FormRadioGroup renders a RadioGroup wrapped by FormField.
 */
exports.FormRadioGroup = react_bindings_1.compose(formFieldBase_1._FormFieldBase, {
    className: exports.formRadioGroupClassName,
    displayName: 'FormRadioGroup',
    overrideStyles: true,
    slots: {
        control: RadioGroup_1.RadioGroup,
    },
    slotProps: function (_a) {
        var errorMessage = _a.errorMessage;
        return ({
            message: {
                error: !!errorMessage,
            },
        });
    },
});
exports.FormRadioGroup.propTypes = utils_1.commonPropTypes.createCommon({
    children: false,
    content: false,
});
