// @fluentui/react-northstar/dist/dts/src/components/Form/FormDatepicker.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FormDatepicker = exports.formDatepickerClassName = void 0;
var react_bindings_1 = require("@fluentui/react-bindings");
var utils_1 = require("../../utils");
var Datepicker_1 = require("../Datepicker/Datepicker");
var formFieldBase_1 = require("./utils/formFieldBase");
exports.formDatepickerClassName = 'ui-form__datepicker';
/**
 * An FormDatepicker renders a Datepicker wrapped by FormField.
 */
exports.FormDatepicker = react_bindings_1.compose(formFieldBase_1._FormFieldBase, {
    className: exports.formDatepickerClassName,
    displayName: 'FormDatepicker',
    overrideStyles: true,
    slots: {
        control: Datepicker_1.Datepicker,
    },
    slotProps: function (_a) {
        var errorMessage = _a.errorMessage, required = _a.required;
        return ({
            control: {
                required: required,
            },
            message: {
                error: !!errorMessage,
            },
        });
    },
});
exports.FormDatepicker.propTypes = utils_1.commonPropTypes.createCommon({
    accessibility: false,
    children: false,
    content: false,
});
