// @fluentui/react-northstar/dist/dts/src/components/Form/FormTextArea.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FormTextArea = exports.formTextAreaClassName = void 0;
var react_bindings_1 = require("@fluentui/react-bindings");
var TextArea_1 = require("../TextArea/TextArea");
var formFieldBase_1 = require("./utils/formFieldBase");
var utils_1 = require("../../utils");
exports.formTextAreaClassName = 'ui-form__textarea';
/**
 * An FormTextArea renders a TextArea wrapped by FormField.
 */
exports.FormTextArea = react_bindings_1.compose(formFieldBase_1._FormFieldBase, {
    className: exports.formTextAreaClassName,
    displayName: 'FormTextArea',
    overrideStyles: true,
    slots: {
        control: TextArea_1.TextArea,
    },
    slotProps: function (_a) {
        var errorMessage = _a.errorMessage, required = _a.required;
        return ({
            label: {
                required: required,
            },
            control: {
                error: !!errorMessage,
            },
            message: {
                error: !!errorMessage,
            },
        });
    },
});
exports.FormTextArea.propTypes = utils_1.commonPropTypes.createCommon({
    children: false,
    content: false,
});
