// @fluentui/react-northstar/dist/dts/src/components/Form/FormCheckbox.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FormCheckbox = exports.formCheckboxClassName = void 0;
var react_bindings_1 = require("@fluentui/react-bindings");
var utils_1 = require("../../utils");
var Checkbox_1 = require("../Checkbox/Checkbox");
var formFieldBase_1 = require("./utils/formFieldBase");
exports.formCheckboxClassName = 'ui-form__checkbox';
/**
 * An FormCheckbox renders a Checkbox wrapped by FormField.
 */
exports.FormCheckbox = react_bindings_1.compose(formFieldBase_1._FormFieldBase, {
    className: exports.formCheckboxClassName,
    displayName: 'FormCheckbox',
    overrideStyles: true,
    handledProps: ['label'],
    slots: {
        label: function () { return null; },
        control: Checkbox_1.Checkbox,
    },
    slotProps: function (_a) {
        var label = _a.label, errorMessage = _a.errorMessage;
        return ({
            control: {
                label: label,
            },
            message: {
                error: !!errorMessage,
            },
        });
    },
});
exports.FormCheckbox.propTypes = utils_1.commonPropTypes.createCommon({
    content: false,
    children: false,
});
