// @fluentui/react-northstar/dist/dts/src/themes/teams/components/Button/buttonVariables.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.buttonVariables = void 0;
var utils_1 = require("../../../../utils");
var buttonVariables = function (siteVars) { return ({
    padding: "0 " + utils_1.pxToRem(20),
    height: utils_1.pxToRem(32),
    minWidth: utils_1.pxToRem(96),
    loadingMinWidth: utils_1.pxToRem(118),
    maxWidth: utils_1.pxToRem(280),
    borderRadius: siteVars.borderRadiusMedium,
    contentFontSize: siteVars.fontSizes.medium,
    contentFontWeight: siteVars.fontWeightSemibold,
    contentLineHeight: siteVars.lineHeightMedium,
    color: siteVars.colorScheme.default.foreground,
    colorHover: siteVars.colorScheme.default.foregroundHover,
    colorActive: siteVars.colorScheme.default.foregroundPressed,
    colorDisabled: siteVars.colorScheme.brand.foregroundDisabled,
    colorFocus: undefined,
    iconSize: utils_1.pxToRem(16),
    backgroundColor: siteVars.colorScheme.default.background,
    backgroundColorActive: siteVars.colorScheme.default.backgroundPressed,
    backgroundColorHover: siteVars.colorScheme.default.backgroundHover1,
    backgroundColorFocus: undefined,
    backgroundColorDisabled: siteVars.colorScheme.default.backgroundDisabled,
    borderColor: siteVars.colorScheme.default.border,
    borderColorHover: siteVars.colorScheme.default.borderHover,
    borderColorFocus: undefined,
    borderColorActive: siteVars.colorScheme.default.borderPressed,
    borderColorDisabled: 'transparent',
    backgroundColorIconOnlyHover: siteVars.colorScheme.default.backgroundHover2,
    primaryColor: siteVars.colorScheme.brand.foreground4,
    primaryColorHover: siteVars.colorScheme.brand.foreground4,
    primaryBackgroundColor: siteVars.colorScheme.brand.background,
    primaryBackgroundColorActive: siteVars.colorScheme.brand.backgroundPressed,
    primaryBackgroundColorHover: siteVars.colorScheme.brand.backgroundHover,
    primaryBackgroundColorDisabled: siteVars.colorScheme.default.backgroundDisabled,
    primaryBackgroundColorFocus: undefined,
    primaryBorderColor: 'transparent',
    tintedColor: siteVars.colorScheme.brand.foreground,
    tintedColorHover: siteVars.colorScheme.brand.foregroundHover,
    tintedBackgroundColor: siteVars.colorScheme.default.background,
    tintedBackgroundColorActive: siteVars.colorScheme.brand.backgroundHover1,
    tintedBackgroundColorHover: siteVars.colorScheme.brand.backgroundHover1,
    tintedBorderColor: siteVars.colorScheme.brand.border1,
    tintedBorderColorHover: siteVars.colorScheme.brand.borderHover,
    circularBorderRadius: utils_1.pxToRem(999),
    textColor: siteVars.colorScheme.default.foreground1,
    textColorHover: siteVars.colorScheme.brand.foreground1,
    textPrimaryColor: siteVars.colorScheme.brand.foreground,
    textPrimaryColorHover: siteVars.colorScheme.brand.foreground1,
    textColorDisabled: siteVars.colorScheme.brand.foregroundDisabled1,
    textColorIconOnlyHover: siteVars.colorScheme.brand.foregroundHover,
    primaryBoxShadow: siteVars.shadowLevel1Dark,
    boxShadow: siteVars.shadowLevel1,
    loaderBorderSize: utils_1.pxToRem(2),
    loaderSize: utils_1.pxToRem(20),
    loaderSvgHeight: utils_1.pxToRem(1220),
    loaderSvgAnimationHeight: utils_1.pxToRem(-1200),
    sizeSmallContentFontSize: siteVars.fontSizes.small,
    sizeSmallContentLineHeight: siteVars.lineHeightSmall,
    sizeSmallHeight: utils_1.pxToRem(24),
    sizeSmallMinWidth: utils_1.pxToRem(72),
    sizeSmallPadding: "0 " + utils_1.pxToRem(8),
    sizeSmallLoaderBorderSize: utils_1.pxToRem(2),
    sizeSmallLoaderSize: utils_1.pxToRem(15),
    sizeSmallLoaderSvgHeight: utils_1.pxToRem(895),
    sizeSmallLoaderSvgAnimationHeight: utils_1.pxToRem(-880),
}); };
exports.buttonVariables = buttonVariables;
