// @fluentui/react-northstar/dist/dts/src/themes/teams/components/Button/buttonGroupStyles.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.buttonGroupStyles = void 0;
var utils_1 = require("../../../../utils");
var commonButtonsStyles = function (circular) { return (circular ? { marginRight: utils_1.pxToRem(8) } : { borderRadius: 0 }); };
exports.buttonGroupStyles = {
    root: function () { return ({}); },
    middleButton: function (_a) {
        var p = _a.props;
        return (__assign({}, commonButtonsStyles(p.circular)));
    },
    firstButton: function (_a) {
        var p = _a.props, v = _a.variables;
        return (__assign(__assign({}, commonButtonsStyles(p.circular)), (!p.circular && {
            borderTopLeftRadius: v.borderRadius,
            borderBottomLeftRadius: v.borderRadius,
        })));
    },
    lastButton: function (_a) {
        var p = _a.props, v = _a.variables;
        return (__assign(__assign({}, commonButtonsStyles(p.circular)), (!p.circular && {
            borderTopRightRadius: v.borderRadius,
            borderBottomRightRadius: v.borderRadius,
        })));
    },
};
