// @fluentui/react-northstar/dist/dts/src/themes/teams/components/Card/cardVariables.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.cardVariables = void 0;
var utils_1 = require("../../../../utils");
var cardVariables = function (siteVars) {
    return {
        backgroundColor: siteVars.colorScheme.default.background1,
        backgroundColorDisabled: siteVars.colorScheme.default.backgroundDisabled1,
        backgroundColorFocus: siteVars.colorScheme.default.backgroundFocus1,
        backgroundColorHover: siteVars.colorScheme.default.backgroundHover1,
        backgroundColorPressed: siteVars.colorScheme.default.backgroundPressed,
        ghostBackgroundColor: 'none',
        ghostBackgroundColorDisabled: siteVars.colorScheme.default.backgroundDisabled,
        ghostBackgroundColorFocus: 'none',
        ghostBackgroundColorHover: siteVars.colorScheme.default.backgroundHover,
        ghostBackgroundColorPressed: siteVars.colorScheme.default.backgroundPressed,
        invertedBackgroundColor: siteVars.colorScheme.default.background2,
        invertedBackgroundColorDisabled: siteVars.colorScheme.default.backgroundDisabled2,
        invertedBackgroundColorFocus: siteVars.colorScheme.default.background2,
        invertedBackgroundColorHover: siteVars.colorScheme.default.backgroundHover2,
        invertedBackgroundColorPressed: siteVars.colorScheme.default.backgroundPressed,
        borderStyle: 'solid',
        borderColor: 'transparent',
        borderColorDisabled: 'transparent',
        borderColorHover: 'transparent',
        borderColorPressed: 'transparent',
        borderWidth: siteVars.borderWidth,
        borderRadius: siteVars.borderRadiusMedium,
        boxShadow: 'none',
        boxShadowDisabled: 'none',
        boxShadowHover: 'none',
        boxShadowFocus: 'none',
        boxShadowPressed: 'none',
        elevatedBoxShadow: siteVars.shadowLevel1,
        elevatedBoxShadowDisabled: siteVars.shadowLevel1,
        elevatedBoxShadowHover: siteVars.shadowLevel3,
        elevatedBoxShadowFocus: siteVars.shadowLevel1,
        elevatedBoxShadowPressed: siteVars.shadowLevel1,
        padding: utils_1.pxToRem(16),
        margin: utils_1.pxToRem(0),
        colorDisabled: siteVars.colorScheme.default.foregroundDisabled,
        compactPadding: utils_1.pxToRem(0),
        previewMargin: "0 0 " + utils_1.pxToRem(10) + " 0",
        headerMargin: "0 0 " + utils_1.pxToRem(10) + " 0",
        bodyMargin: "0 0 " + utils_1.pxToRem(10) + " 0",
        footerMargin: "0 0 " + utils_1.pxToRem(10) + " 0",
        fittedPreviewMargin: utils_1.pxToRem(0),
        fittedHeaderMargin: utils_1.pxToRem(0),
        fittedBodyMargin: utils_1.pxToRem(0),
        fittedFooterMargin: utils_1.pxToRem(0),
        previewMarginHorizontal: "0 " + utils_1.pxToRem(10) + " 0 0",
        topControlsTop: utils_1.pxToRem(10),
        topControlsRight: utils_1.pxToRem(0),
        // TODO: update with latest values from design
        width: utils_1.pxToRem(300),
        height: '100%',
        sizeSmallWidth: utils_1.pxToRem(200),
        sizeSmallHeight: '100%',
        sizeSmallPadding: utils_1.pxToRem(0),
        sizeLargeWidth: utils_1.pxToRem(500),
        sizeLargeHeight: '100%',
        sizeLargePadding: utils_1.pxToRem(16),
        fluidWidth: '100%',
        fluidHeight: '100%',
        expandableBoxStartMaxHeight: utils_1.pxToRem(20),
        expandableBoxEndMaxHeight: utils_1.pxToRem(500),
        expandableBoxShrinkTransition: 'max-height 0.5s ease-in',
        expandableBoxExpandTransition: 'max-height 1s ease-in',
        selectedBackgroundColor: siteVars.colorScheme.default.backgroundFocus3,
        selectedBorderColor: siteVars.colorScheme.default.borderActive,
        selectedBoxShadow: siteVars.shadowLevel4,
        selectedBackgroundColorDisabled: siteVars.colorScheme.default.backgroundDisabled3,
        selectedBackgroundColorFocus: siteVars.colorScheme.default.backgroundFocus3,
        selectedBackgroundColorHover: siteVars.colorScheme.default.backgroundHover3,
        selectedBackgroundColorPressed: siteVars.colorScheme.default.backgroundPressed3,
    };
};
exports.cardVariables = cardVariables;
