// @fluentui/react-northstar/dist/dts/src/themes/teams/components/Loader/loaderVariables.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.loaderVariables = void 0;
var utils_1 = require("../../../../utils");
var loaderSvgDataUrl_1 = require("./loaderSvgDataUrl");
var loaderSecondarySvgDataUrl_1 = require("./loaderSecondarySvgDataUrl");
var loaderVariables = function () { return ({
    containerHeights: {
        smallest: utils_1.pxToRem(24),
        smaller: utils_1.pxToRem(24),
        small: utils_1.pxToRem(24),
        medium: utils_1.pxToRem(36),
        large: utils_1.pxToRem(72),
        larger: utils_1.pxToRem(72),
        largest: utils_1.pxToRem(72),
    },
    containerWidths: {
        smallest: utils_1.pxToRem(24),
        smaller: utils_1.pxToRem(24),
        small: utils_1.pxToRem(24),
        medium: utils_1.pxToRem(36),
        large: utils_1.pxToRem(72),
        larger: utils_1.pxToRem(72),
        largest: utils_1.pxToRem(72),
    },
    svgContent: loaderSvgDataUrl_1.loaderSvgDataUrl,
    secondarySvgContent: loaderSecondarySvgDataUrl_1.loaderSecondarySvgDataUrl,
    svgHeights: {
        smaller: utils_1.pxToRem(1464),
        smallest: utils_1.pxToRem(1464),
        small: utils_1.pxToRem(1464),
        medium: utils_1.pxToRem(2196),
        large: utils_1.pxToRem(4392),
        larger: utils_1.pxToRem(4392),
        largest: utils_1.pxToRem(4392),
    },
    svgTranslatePosition: {
        smallest: utils_1.pxToRem(-1440),
        smaller: utils_1.pxToRem(-1440),
        small: utils_1.pxToRem(-1440),
        medium: utils_1.pxToRem(-2160),
        large: utils_1.pxToRem(-4320),
        larger: utils_1.pxToRem(-4320),
        largest: utils_1.pxToRem(-4320),
    },
    svgWidths: {
        smallest: utils_1.pxToRem(24),
        smaller: utils_1.pxToRem(24),
        small: utils_1.pxToRem(24),
        medium: utils_1.pxToRem(36),
        large: utils_1.pxToRem(72),
        larger: utils_1.pxToRem(72),
        largest: utils_1.pxToRem(72),
    },
}); };
exports.loaderVariables = loaderVariables;
