// botframework-streaming/lib/index.js
"use strict";
/**
 * @module botframework-streaming
 */
/**
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebSocketServer = exports.WebSocketClient = exports.NodeWebSocketFactoryBase = exports.NodeWebSocketFactory = exports.NodeWebSocket = exports.SubscribableStream = exports.StreamingResponse = exports.StreamingRequest = exports.RequestHandler = exports.NamedPipeServer = exports.NamedPipeClient = exports.HttpContent = exports.ContentStream = void 0;
var contentStream_1 = require("./contentStream");
Object.defineProperty(exports, "ContentStream", { enumerable: true, get: function () { return contentStream_1.ContentStream; } });
var httpContentStream_1 = require("./httpContentStream");
Object.defineProperty(exports, "HttpContent", { enumerable: true, get: function () { return httpContentStream_1.HttpContent; } });
var namedPipe_1 = require("./namedPipe");
Object.defineProperty(exports, "NamedPipeClient", { enumerable: true, get: function () { return namedPipe_1.NamedPipeClient; } });
Object.defineProperty(exports, "NamedPipeServer", { enumerable: true, get: function () { return namedPipe_1.NamedPipeServer; } });
var requestHandler_1 = require("./requestHandler");
Object.defineProperty(exports, "RequestHandler", { enumerable: true, get: function () { return requestHandler_1.RequestHandler; } });
var streamingRequest_1 = require("./streamingRequest");
Object.defineProperty(exports, "StreamingRequest", { enumerable: true, get: function () { return streamingRequest_1.StreamingRequest; } });
var streamingResponse_1 = require("./streamingResponse");
Object.defineProperty(exports, "StreamingResponse", { enumerable: true, get: function () { return streamingResponse_1.StreamingResponse; } });
var subscribableStream_1 = require("./subscribableStream");
Object.defineProperty(exports, "SubscribableStream", { enumerable: true, get: function () { return subscribableStream_1.SubscribableStream; } });
var webSocket_1 = require("./webSocket");
Object.defineProperty(exports, "NodeWebSocket", { enumerable: true, get: function () { return webSocket_1.NodeWebSocket; } });
Object.defineProperty(exports, "NodeWebSocketFactory", { enumerable: true, get: function () { return webSocket_1.NodeWebSocketFactory; } });
Object.defineProperty(exports, "NodeWebSocketFactoryBase", { enumerable: true, get: function () { return webSocket_1.NodeWebSocketFactoryBase; } });
Object.defineProperty(exports, "WebSocketClient", { enumerable: true, get: function () { return webSocket_1.WebSocketClient; } });
Object.defineProperty(exports, "WebSocketServer", { enumerable: true, get: function () { return webSocket_1.WebSocketServer; } });
