// @fluentui/react-component-nesting-registry/dist/es/hooks/useNestingRoot.js
import * as React from 'react';
import { NestingContext } from '../NestingContext';
import { RefStack } from '../utils/RefStack';
var registrySet = new RefStack(); // These hooks are not used currently

/* eslint-disable */

export var useNestingRoot = function useNestingRoot() {
  var _React$useState = React.useState(registrySet),
      registry = _React$useState[0];

  var parentRef = React.useRef(null);
  var nestedProps = React.useMemo(function () {
    return {
      value: {
        getContextRefs: registry.getContextRefs,
        register: registry.register,
        unregister: registry.unregister
      }
    };
  }, []);
  var getRefs = React.useCallback(function () {
    return registry.getContextRefs(parentRef);
  }, []);
  React.useEffect(function () {
    registry.register(parentRef);
    return function () {
      return registry.unregister(parentRef);
    };
  }, []);
  return {
    NestedComponent: NestingContext.Provider,
    nestedProps: nestedProps,
    getRefs: getRefs,
    isRoot: true,
    ref: parentRef
  };
};
