// @fluentui/react-northstar/dist/dts/src/themes/teams/components/SplitButton/splitButtonStyles.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.splitButtonStyles = void 0;
var getBorderFocusStyles_1 = require("../../getBorderFocusStyles");
exports.splitButtonStyles = {
    menuButton: function (_a) {
        var p = _a.props, v = _a.variables;
        return (__assign(__assign(__assign({ borderTopRightRadius: 0, borderBottomRightRadius: 0, borderRightWidth: 0, padding: v.padding, minWidth: v.minWidth }, (p.size === 'small' && {
            height: v.smallDimension,
            padding: v.smallPadding,
            minWidth: v.smallMinWidth,
        })), ((p.flat || p.size === 'small') && {
            boxShadow: 'none',
        })), { ':focus-visible': {
                borderRightWidth: 0,
                ':before': {
                    borderRightWidth: 0,
                },
                ':after': {
                    borderRightWidth: 0,
                },
            }, ':active': {
                animationName: 'unset',
                animationDuration: 'unset',
            } }));
    },
    root: function (_a) {
        var p = _a.props, v = _a.variables, siteVariables = _a.theme.siteVariables;
        var borderFocusStyles = getBorderFocusStyles_1.getBorderFocusStyles({
            variables: {
                borderRadius: v.focusBorderRadius,
                borderWidth: v.focusBorderWidth,
                focusInnerBorderColor: v.focusInnerBorderColor,
                focusOuterBorderColor: v.focusOuterBorderColor,
                zIndexes: { foreground: v.focusBorderZIndex },
            },
        });
        return {
            borderRadius: v.borderRadius,
            position: 'relative',
            whiteSpace: 'nowrap',
            display: 'inline-block',
            width: 'fit-content',
            ':focus-within': __assign({ boxShadow: 'none' }, (p.isFromKeyboard && __assign({}, borderFocusStyles[':focus-visible']))),
        };
    },
};
