// @fluentui/react-northstar/dist/dts/src/themes/teams/components/SplitButton/splitButtonDividerStyles.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.splitButtonDividerStyles = void 0;
exports.splitButtonDividerStyles = {
    root: function (_a) {
        var props = _a.props, variables = _a.variables;
        return {
            height: 'auto',
            alignItems: 'center',
            display: 'inline-block',
            '::before': __assign({ content: '""', position: 'absolute', top: 0, bottom: 0, width: "1px", height: '100%', zIndex: 1, background: variables.dividerColor }, (props.primary && {
                background: variables.dividerPrimaryColor,
            })),
        };
    },
};
