// @fluentui/react-northstar/dist/commonjs/themes/teams/components/List/listItemVariables.js
"use strict";

exports.__esModule = true;
exports.listItemVariables = void 0;

var _utils = require("../../../../utils");

var listItemVariables = function listItemVariables(siteVariables) {
  return {
    minHeight: (0, _utils.pxToRem)(48),
    rootPadding: "0 " + (0, _utils.pxToRem)(18) + " 0 " + (0, _utils.pxToRem)(20),
    gap: (0, _utils.pxToRem)(8),
    zIndex: siteVariables.zIndexes.foreground,
    // Header
    // TODO: prod app uses 17.5px here, it should be 16px per the design guide!
    headerLineHeight: siteVariables.lineHeightSmall,
    headerFontSize: siteVariables.fontSizes.medium,
    // Header Media
    headerMediaFontSize: siteVariables.fontSizes.small,
    // TODO: prod app uses 17.5px here, it should be 16px per the design guide!
    headerMediaLineHeight: siteVariables.lineHeightSmall,
    importantFontWeight: siteVariables.fontWeightBold,
    // Content
    contentFontSize: siteVariables.fontSizes.small,
    contentLineHeight: siteVariables.lineHeightSmall,
    // Content Media
    contentMediaFontSize: siteVariables.fontSizes.small,
    contentMediaLineHeight: siteVariables.lineHeightSmall,
    // Selectable
    selectableFocusHoverColor: siteVariables.colorScheme.default.foregroundHover,
    selectableFocusHoverBackgroundColor: siteVariables.colorScheme.default.backgroundHover,
    // Selected
    selectedColor: siteVariables.colorScheme.default.foregroundPressed,
    selectedBackgroundColor: siteVariables.colorScheme.default.backgroundActive1
  };
};

exports.listItemVariables = listItemVariables;
