// @fluentui/styles/dist/commonjs/index.js
"use strict";

exports.__esModule = true;
var _exportNames = {
  callable: true,
  createTheme: true,
  isDebugEnabled: true,
  deepmerge: true,
  mergeThemes: true,
  emptyTheme: true,
  mergeComponentVariables: true,
  mergeComponentStyles: true,
  mergeFontFaces: true,
  mergeStaticStyles: true,
  mergeStyles: true,
  mergeThemeStyles: true,
  mergeThemeVariables: true,
  mergeSiteVariables: true,
  objectKeyToValues: true,
  withDebugId: true
};
exports.withDebugId = exports.objectKeyToValues = exports.mergeSiteVariables = exports.mergeThemeVariables = exports.mergeThemeStyles = exports.mergeStyles = exports.mergeStaticStyles = exports.mergeFontFaces = exports.mergeComponentStyles = exports.mergeComponentVariables = exports.emptyTheme = exports.mergeThemes = exports.deepmerge = exports.isDebugEnabled = exports.createTheme = exports.callable = void 0;

var _callable = require("./callable");

exports.callable = _callable.callable;

var _createTheme = require("./createTheme");

exports.createTheme = _createTheme.createTheme;

var _debugEnabled = require("./debugEnabled");

exports.isDebugEnabled = _debugEnabled.isEnabled;

var _deepmerge = require("./deepmerge");

exports.deepmerge = _deepmerge.deepmerge;

var _mergeThemes = require("./mergeThemes");

exports.mergeThemes = _mergeThemes.mergeThemes;
exports.emptyTheme = _mergeThemes.emptyTheme;
exports.mergeComponentVariables = _mergeThemes.mergeComponentVariables;
exports.mergeComponentStyles = _mergeThemes.mergeComponentStyles;
exports.mergeFontFaces = _mergeThemes.mergeFontFaces;
exports.mergeStaticStyles = _mergeThemes.mergeStaticStyles;
exports.mergeStyles = _mergeThemes.mergeStyles;
exports.mergeThemeStyles = _mergeThemes.mergeThemeStyles;
exports.mergeThemeVariables = _mergeThemes.mergeThemeVariables;
exports.mergeSiteVariables = _mergeThemes.mergeSiteVariables;

var _objectKeysToValues = require("./objectKeysToValues");

exports.objectKeyToValues = _objectKeysToValues.objectKeyToValues;

var _withDebugId = require("./withDebugId");

exports.withDebugId = _withDebugId.withDebugId;

var _types = require("./types");

Object.keys(_types).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _types[key]) return;
  exports[key] = _types[key];
});
