// @fluentui/accessibility/dist/dts/src/behaviors/Menu/menuBehavior.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.menuBehavior = void 0;
var types_1 = require("../../focusZone/types");
var menuItemBehavior_1 = require("./menuItemBehavior");
var menuDividerBehavior_1 = require("./menuDividerBehavior");
/**
 * @description
 * Implements ARIA Menu design pattern.
 * The 'menu' role is used to identify an element that creates a list of common actions or functions that a user can invoke.
 *
 * @specification
 * Adds role='menu'.
 * Provides arrow key navigation in horizontal direction.
 * When 'vertical' prop is used, provides keyboard navigation in vertical direction.
 * Keyboard navigation is circular.
 */
var menuBehavior = function (props) { return ({
    attributes: {
        root: {
            role: 'menu',
        },
    },
    focusZone: {
        props: {
            isCircularNavigation: true,
            shouldFocusInnerElementWhenReceivedFocus: true,
            direction: props.vertical ? types_1.FocusZoneDirection.vertical : types_1.FocusZoneDirection.horizontal,
        },
    },
    childBehaviors: {
        item: menuItemBehavior_1.menuItemBehavior,
        divider: menuDividerBehavior_1.menuDividerBehavior,
    },
}); };
exports.menuBehavior = menuBehavior;
