// @fluentui/react-northstar/dist/dts/src/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FocusZoneUtilities = exports.usePopper = exports.UNSTABLE_Popper = exports.createReferenceFromClick = exports.animationClassName = exports.Input = exports.teamsDarkV2Theme = exports.teamsV2Theme = exports.teamsHighContrastTheme = exports.teamsDarkTheme = exports.teamsTheme = void 0;
__exportStar(require("@fluentui/accessibility"), exports);
__exportStar(require("@fluentui/react-component-ref"), exports);
__exportStar(require("@fluentui/react-bindings"), exports);
__exportStar(require("@fluentui/react-icons-northstar"), exports);
__exportStar(require("@fluentui/styles"), exports);
//
// Themes
//
var teams_1 = require("./themes/teams");
Object.defineProperty(exports, "teamsTheme", { enumerable: true, get: function () { return teams_1.teamsTheme; } });
var teams_dark_1 = require("./themes/teams-dark");
Object.defineProperty(exports, "teamsDarkTheme", { enumerable: true, get: function () { return teams_dark_1.teamsDarkTheme; } });
var teams_high_contrast_1 = require("./themes/teams-high-contrast");
Object.defineProperty(exports, "teamsHighContrastTheme", { enumerable: true, get: function () { return teams_high_contrast_1.teamsHighContrastTheme; } });
var teams_v2_1 = require("./themes/teams-v2");
Object.defineProperty(exports, "teamsV2Theme", { enumerable: true, get: function () { return teams_v2_1.teamsV2Theme; } });
var teams_dark_v2_1 = require("./themes/teams-dark-v2");
Object.defineProperty(exports, "teamsDarkV2Theme", { enumerable: true, get: function () { return teams_dark_v2_1.teamsDarkV2Theme; } });
__exportStar(require("./themes/types"), exports);
__exportStar(require("./themes/colorUtils"), exports);
//
// Teams theme
//
__exportStar(require("./themes/teams/types"), exports);
//
// Components
//
__exportStar(require("./components/Accordion/Accordion"), exports);
__exportStar(require("./components/Accordion/AccordionTitle"), exports);
__exportStar(require("./components/Accordion/AccordionContent"), exports);
__exportStar(require("./components/Alert/Alert"), exports);
__exportStar(require("./components/Alert/AlertDismissAction"), exports);
__exportStar(require("./components/Attachment/Attachment"), exports);
__exportStar(require("./components/Attachment/AttachmentAction"), exports);
__exportStar(require("./components/Attachment/AttachmentBody"), exports);
__exportStar(require("./components/Attachment/AttachmentDescription"), exports);
__exportStar(require("./components/Attachment/AttachmentHeader"), exports);
__exportStar(require("./components/Attachment/AttachmentIcon"), exports);
__exportStar(require("./components/Avatar/Avatar"), exports);
__exportStar(require("./components/Avatar/AvatarStatus"), exports);
__exportStar(require("./components/Avatar/AvatarStatusIcon"), exports);
__exportStar(require("./components/Avatar/AvatarImage"), exports);
__exportStar(require("./components/Avatar/AvatarIcon"), exports);
__exportStar(require("./components/Avatar/AvatarLabel"), exports);
__exportStar(require("./components/Box/Box"), exports);
__exportStar(require("./components/Button/Button"), exports);
__exportStar(require("./components/Button/ButtonGroup"), exports);
__exportStar(require("./components/Button/ButtonContent"), exports);
__exportStar(require("./components/Breadcrumb/Breadcrumb"), exports);
__exportStar(require("./components/Breadcrumb/BreadcrumbItem"), exports);
__exportStar(require("./components/Breadcrumb/BreadcrumbDivider"), exports);
__exportStar(require("./components/Breadcrumb/BreadcrumbLink"), exports);
__exportStar(require("./components/Chat/Chat"), exports);
__exportStar(require("./components/Chat/ChatItem"), exports);
__exportStar(require("./components/Chat/ChatMessage"), exports);
__exportStar(require("./components/Chat/ChatMessageDetails"), exports);
__exportStar(require("./components/Chat/ChatMessageHeader"), exports);
__exportStar(require("./components/Chat/ChatMessageReadStatus"), exports);
__exportStar(require("./components/Chat/chatDensity"), exports);
__exportStar(require("./components/Checkbox/Checkbox"), exports);
__exportStar(require("./components/Debug/Debug"), exports);
__exportStar(require("./components/Debug/DebugSelector"), exports);
__exportStar(require("./components/Design/Design"), exports);
__exportStar(require("./components/MenuButton/MenuButton"), exports);
__exportStar(require("./components/Pill/PillGroup"), exports);
__exportStar(require("./components/Pill/Pill"), exports);
__exportStar(require("./components/Pill/PillContent"), exports);
__exportStar(require("./components/Pill/PillAction"), exports);
__exportStar(require("./components/Pill/PillImage"), exports);
__exportStar(require("./components/Pill/PillIcon"), exports);
__exportStar(require("./components/Divider/Divider"), exports);
__exportStar(require("./components/Divider/DividerContent"), exports);
__exportStar(require("./components/Dialog/Dialog"), exports);
__exportStar(require("./components/Dialog/DialogFooter"), exports);
__exportStar(require("./components/Dropdown/Dropdown"), exports);
__exportStar(require("./components/Dropdown/DropdownItem"), exports);
__exportStar(require("./components/Dropdown/DropdownSelectedItem"), exports);
__exportStar(require("./components/Dropdown/DropdownSearchInput"), exports);
__exportStar(require("./components/Embed/Embed"), exports);
__exportStar(require("./components/Debug/FiberNavigator"), exports);
__exportStar(require("./components/Flex/Flex"), exports);
__exportStar(require("./components/Flex/FlexItem"), exports);
__exportStar(require("./components/Form/Form"), exports);
__exportStar(require("./components/Form/FormField"), exports);
__exportStar(require("./components/Form/FormInput"), exports);
__exportStar(require("./components/Form/FormTextArea"), exports);
__exportStar(require("./components/Form/FormInput"), exports);
__exportStar(require("./components/Form/FormLabel"), exports);
__exportStar(require("./components/Form/FormMessage"), exports);
__exportStar(require("./components/Form/FormButton"), exports);
__exportStar(require("./components/Form/FormDropdown"), exports);
__exportStar(require("./components/Form/FormDatepicker"), exports);
__exportStar(require("./components/Form/FormRadioGroup"), exports);
__exportStar(require("./components/Form/FormSlider"), exports);
__exportStar(require("./components/Form/FormCheckbox"), exports);
__exportStar(require("./components/Form/FormFieldCustom"), exports);
__exportStar(require("./components/Grid/Grid"), exports);
__exportStar(require("./components/Header/Header"), exports);
__exportStar(require("./components/Header/HeaderDescription"), exports);
__exportStar(require("./components/SvgIcon/SvgIcon"), exports);
__exportStar(require("./components/Image/Image"), exports);
__exportStar(require("./components/Input/Input"), exports);
__exportStar(require("./components/Input/InputLabel"), exports);
// Ouch, @fluentui/react-compose already exports `Input` type :(
var Input_1 = require("./components/Input/Input");
Object.defineProperty(exports, "Input", { enumerable: true, get: function () { return Input_1.Input; } });
__exportStar(require("./components/ItemLayout/ItemLayout"), exports);
__exportStar(require("./components/Label/Label"), exports);
__exportStar(require("./components/Loader/Loader"), exports);
__exportStar(require("./components/Layout/Layout"), exports);
__exportStar(require("./components/List/List"), exports);
__exportStar(require("./components/List/ListItem"), exports);
__exportStar(require("./components/List/ListItemContent"), exports);
__exportStar(require("./components/List/ListItemContentMedia"), exports);
__exportStar(require("./components/List/ListItemEndMedia"), exports);
__exportStar(require("./components/List/ListItemHeader"), exports);
__exportStar(require("./components/List/ListItemHeaderMedia"), exports);
__exportStar(require("./components/List/ListItemMedia"), exports);
__exportStar(require("./components/Menu/Menu"), exports);
__exportStar(require("./components/Menu/MenuItem"), exports);
__exportStar(require("./components/Menu/MenuItemIcon"), exports);
__exportStar(require("./components/Menu/MenuItemContent"), exports);
__exportStar(require("./components/Menu/MenuItemIndicator"), exports);
__exportStar(require("./components/Menu/MenuItemWrapper"), exports);
__exportStar(require("./components/Menu/MenuDivider"), exports);
__exportStar(require("./components/Popup/Popup"), exports);
__exportStar(require("./components/Popup/PopupContent"), exports);
__exportStar(require("./components/Portal/Portal"), exports);
__exportStar(require("./components/Provider/Provider"), exports);
__exportStar(require("./components/Provider/ProviderConsumer"), exports);
__exportStar(require("./components/RadioGroup/RadioGroup"), exports);
__exportStar(require("./components/RadioGroup/RadioGroupItem"), exports);
__exportStar(require("./components/Segment/Segment"), exports);
__exportStar(require("./components/Skeleton/Skeleton"), exports);
__exportStar(require("./components/Skeleton/SkeletonLine"), exports);
__exportStar(require("./components/Skeleton/SkeletonShape"), exports);
__exportStar(require("./components/Skeleton/SkeletonButton"), exports);
__exportStar(require("./components/Skeleton/SkeletonText"), exports);
__exportStar(require("./components/Skeleton/SkeletonInput"), exports);
__exportStar(require("./components/Skeleton/SkeletonAvatar"), exports);
__exportStar(require("./components/Slider/Slider"), exports);
__exportStar(require("./components/Status/Status"), exports);
__exportStar(require("./components/Text/Text"), exports);
__exportStar(require("./components/Animation/Animation"), exports);
var useAnimationStyles_1 = require("./components/Animation/useAnimationStyles");
Object.defineProperty(exports, "animationClassName", { enumerable: true, get: function () { return useAnimationStyles_1.animationClassName; } });
__exportStar(require("./components/TextArea/TextArea"), exports);
__exportStar(require("./components/Toolbar/Toolbar"), exports);
__exportStar(require("./components/Toolbar/ToolbarCustomItem"), exports);
__exportStar(require("./components/Toolbar/ToolbarDivider"), exports);
__exportStar(require("./components/Toolbar/ToolbarItem"), exports);
__exportStar(require("./components/Toolbar/ToolbarItemWrapper"), exports);
__exportStar(require("./components/Toolbar/ToolbarItemIcon"), exports);
__exportStar(require("./components/Toolbar/ToolbarMenu"), exports);
__exportStar(require("./components/Toolbar/ToolbarMenuDivider"), exports);
__exportStar(require("./components/Toolbar/ToolbarMenuItem"), exports);
__exportStar(require("./components/Toolbar/ToolbarMenuItemContent"), exports);
__exportStar(require("./components/Toolbar/ToolbarMenuItemIcon"), exports);
__exportStar(require("./components/Toolbar/ToolbarMenuItemSubmenuIndicator"), exports);
__exportStar(require("./components/Toolbar/ToolbarMenuItemActiveIndicator"), exports);
__exportStar(require("./components/Toolbar/ToolbarMenuRadioGroup"), exports);
__exportStar(require("./components/Toolbar/ToolbarMenuRadioGroupWrapper"), exports);
__exportStar(require("./components/Toolbar/ToolbarRadioGroup"), exports);
__exportStar(require("./components/Tree/Tree"), exports);
__exportStar(require("./components/Tree/TreeItem"), exports);
__exportStar(require("./components/Tree/TreeTitle"), exports);
__exportStar(require("./components/Tree/context"), exports);
__exportStar(require("./components/Tree/hooks/useTree"), exports);
__exportStar(require("./components/Tree/hooks/useVirtualTree"), exports);
__exportStar(require("./components/Reaction/Reaction"), exports);
__exportStar(require("./components/Reaction/ReactionGroup"), exports);
__exportStar(require("./components/SplitButton/SplitButton"), exports);
__exportStar(require("./components/SplitButton/SplitButtonToggle"), exports);
__exportStar(require("./components/Video/Video"), exports);
__exportStar(require("./components/Tooltip/Tooltip"), exports);
__exportStar(require("./components/Tooltip/TooltipContent"), exports);
__exportStar(require("./components/Carousel/Carousel"), exports);
__exportStar(require("./components/Carousel/CarouselItem"), exports);
__exportStar(require("./components/Carousel/CarouselNavigation"), exports);
__exportStar(require("./components/Carousel/CarouselNavigationItem"), exports);
__exportStar(require("./components/Carousel/CarouselPaddle"), exports);
__exportStar(require("./components/Carousel/CarouselPaddlesContainer"), exports);
__exportStar(require("./components/Table/Table"), exports);
__exportStar(require("./components/Table/TableRow"), exports);
__exportStar(require("./components/Table/TableCell"), exports);
__exportStar(require("./components/Card/Card"), exports);
__exportStar(require("./components/Card/CardPreview"), exports);
__exportStar(require("./components/Card/CardTopControls"), exports);
__exportStar(require("./components/Card/CardHeader"), exports);
__exportStar(require("./components/Card/CardBody"), exports);
__exportStar(require("./components/Card/CardFooter"), exports);
__exportStar(require("./components/Card/CardColumn"), exports);
__exportStar(require("./components/Card/CardExpandableBox"), exports);
__exportStar(require("./components/Datepicker/Datepicker"), exports);
__exportStar(require("./components/Datepicker/DatepickerCalendar"), exports);
__exportStar(require("./components/Datepicker/DatepickerCalendarHeader"), exports);
__exportStar(require("./components/Datepicker/DatepickerCalendarHeaderAction"), exports);
__exportStar(require("./components/Datepicker/DatepickerCalendarHeaderCell"), exports);
__exportStar(require("./components/Datepicker/DatepickerCalendarCell"), exports);
__exportStar(require("./components/Datepicker/DatepickerCalendarCellButton"), exports);
__exportStar(require("./components/Datepicker/DatepickerCalendarGrid"), exports);
__exportStar(require("./components/Datepicker/DatepickerCalendarGridRow"), exports);
//
// Utilities
//
__exportStar(require("./utils/createComponent"), exports);
__exportStar(require("./utils/date-time-utilities"), exports);
__exportStar(require("./utils"), exports);
__exportStar(require("./types"), exports);
var positioner_1 = require("./utils/positioner");
Object.defineProperty(exports, "createReferenceFromClick", { enumerable: true, get: function () { return positioner_1.createReferenceFromClick; } });
Object.defineProperty(exports, "UNSTABLE_Popper", { enumerable: true, get: function () { return positioner_1.Popper; } });
Object.defineProperty(exports, "usePopper", { enumerable: true, get: function () { return positioner_1.usePopper; } });
__exportStar(require("./utils/positioner/types"), exports);
//
// FocusZone
//
var react_bindings_1 = require("@fluentui/react-bindings");
exports.FocusZoneUtilities = {
    getFirstTabbable: react_bindings_1.getFirstTabbable,
    getLastTabbable: react_bindings_1.getLastTabbable,
    getFirstFocusable: react_bindings_1.getFirstFocusable,
    getLastFocusable: react_bindings_1.getLastFocusable,
    getNextElement: react_bindings_1.getNextElement,
    getPreviousElement: react_bindings_1.getPreviousElement,
    focusAsync: react_bindings_1.focusAsync,
};
