// @fluentui/react-northstar/dist/commonjs/themes/teams/components/Menu/menuItemIndicatorStyles.js
"use strict";

exports.__esModule = true;
exports.menuItemIndicatorStyles = void 0;

var _utils = require("../../../../utils");

var _submenuIndicatorDirection = require("./submenuIndicatorDirection");

var menuItemIndicatorStyles = {
  root: function root(_ref) {
    var p = _ref.props,
        v = _ref.variables;
    return Object.assign({
      position: 'relative',
      float: 'right',
      left: (0, _utils.pxToRem)(12),
      userSelect: 'none',
      marginRight: (0, _utils.pxToRem)(4)
    }, p.inSubmenu && {
      position: 'absolute',
      top: (0, _utils.pxToRem)(6),
      right: (0, _utils.pxToRem)(2),
      left: 'unset'
    }, {
      content: '" "',
      display: 'block',
      overflow: 'hidden',
      height: (0, _utils.pxToRem)(16),
      width: (0, _utils.pxToRem)(16),
      backgroundSize: (0, _utils.pxToRem)(16),
      color: v.indicatorColor
    }, p.active && Object.assign({
      color: v.activeIndicatorColor
    }, p.primary && Object.assign({
      color: v.activePrimaryIndicatorColor
    }, p.vertical && {
      color: v.activePrimaryVerticalIndicatorColor
    })), p.underlined && {
      color: v.indicatorColor
    }, p.iconOnly && {
      color: v.indicatorColor
    }, (0, _submenuIndicatorDirection.submenuIndicatorDirection)(p.vertical));
  }
};
exports.menuItemIndicatorStyles = menuItemIndicatorStyles;
