// @fluentui/react-northstar/dist/commonjs/themes/teams/components/Menu/menuDividerStyles.js
"use strict";

exports.__esModule = true;
exports.menuDividerStyles = void 0;

var _colors = require("../../colors");

var _menuItemStyles = require("./menuItemStyles");

var menuDividerStyles = {
  root: function root(_ref) {
    var _Object$assign;

    var p = _ref.props,
        v = _ref.variables;
    var colors = (0, _colors.getColorScheme)(v.colorScheme, null, p.primary);
    var borderColor = p.primary ? v.primaryBorderColor : v.borderColor || colors.border;
    var borderType = p.vertical ? 'borderTop' : 'borderLeft';
    return Object.assign({}, p.pointing && p.vertical && {
      marginBottom: _menuItemStyles.verticalPointingBottomMargin
    }, p.pills && Object.assign({}, p.vertical ? {
      margin: "0 0 " + _menuItemStyles.verticalPillsBottomMargin + " 0"
    } : {
      margin: "0 " + _menuItemStyles.horizontalPillsRightMargin + " 0 0"
    }), p.hasContent ? {
      display: 'flex',
      justifyContent: 'center',
      flexDirection: 'column',
      textAlign: 'center'
    } : Object.assign((_Object$assign = {}, _Object$assign[borderType] = "1px solid " + borderColor, _Object$assign), !p.vertical && {
      alignSelf: 'stretch'
    }, p.vertical && p.inSubmenu && {
      margin: '8px 0'
    }));
  }
};
exports.menuDividerStyles = menuDividerStyles;
