// @fluentui/react-northstar/dist/commonjs/themes/teams/components/Menu/menuStyles.js
"use strict";

exports.__esModule = true;
exports.menuStyles = void 0;

var _utils = require("../../../../utils");

var _colors = require("../../colors");

var menuStyles = {
  root: function root(_ref) {
    var p = _ref.props,
        v = _ref.variables,
        theme = _ref.theme;
    var iconOnly = p.iconOnly,
        fluid = p.fluid,
        pointing = p.pointing,
        pills = p.pills,
        primary = p.primary,
        underlined = p.underlined,
        vertical = p.vertical,
        submenu = p.submenu;
    var colors = (0, _colors.getColorScheme)(v.colorScheme, null, primary);
    var siteVariables = theme.siteVariables;
    return Object.assign({
      display: 'flex',
      minHeight: (0, _utils.pxToRem)(24),
      margin: 0,
      padding: 0,
      color: v.color,
      backgroundColor: v.backgroundColor || 'inherit',
      listStyleType: 'none'
    }, iconOnly && {
      alignItems: 'center'
    }, vertical && Object.assign({
      flexDirection: 'column',
      backgroundColor: v.verticalBackgroundColor,
      padding: (0, _utils.pxToRem)(8) + " 0"
    }, submenu && {
      boxShadow: v.verticalBoxShadow
    }, !fluid && !submenu && {
      width: 'fit-content'
    }, iconOnly && {
      display: 'inline-block',
      width: 'auto'
    }), !pills && !iconOnly && !(pointing && vertical) && !underlined && {
      // primary has hardcoded grey border color
      border: v.borderWidth + " solid " + (primary ? v.primaryBorderColor : v.borderColor || colors.border),
      borderRadius: siteVariables.borderRadiusMedium
    }, underlined && {
      borderBottom: v.underlinedBottomBorderWidth + " solid " + v.underlinedBorderColor
    });
  }
};
exports.menuStyles = menuStyles;
