// botframework-connector/lib/connectorApi/models/parameters.js
"use strict";
/**
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.viewId = exports.pageSize = exports.memberId = exports.conversationId = exports.continuationToken = exports.attachmentId = exports.activityId = void 0;
exports.activityId = {
    parameterPath: "activityId",
    mapper: {
        required: true,
        serializedName: "activityId",
        type: {
            name: "String"
        }
    }
};
exports.attachmentId = {
    parameterPath: "attachmentId",
    mapper: {
        required: true,
        serializedName: "attachmentId",
        type: {
            name: "String"
        }
    }
};
exports.continuationToken = {
    parameterPath: [
        "options",
        "continuationToken"
    ],
    mapper: {
        serializedName: "continuationToken",
        type: {
            name: "String"
        }
    }
};
exports.conversationId = {
    parameterPath: "conversationId",
    mapper: {
        required: true,
        serializedName: "conversationId",
        type: {
            name: "String"
        }
    }
};
exports.memberId = {
    parameterPath: "memberId",
    mapper: {
        required: true,
        serializedName: "memberId",
        type: {
            name: "String"
        }
    }
};
exports.pageSize = {
    parameterPath: [
        "options",
        "pageSize"
    ],
    mapper: {
        serializedName: "pageSize",
        type: {
            name: "Number"
        }
    }
};
exports.viewId = {
    parameterPath: "viewId",
    mapper: {
        required: true,
        serializedName: "viewId",
        type: {
            name: "String"
        }
    }
};
