// @fluentui/react-northstar/dist/dts/src/themes/teams/components/Skeleton/utils/animations.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getAnimations = void 0;
var keyframes_1 = require("./keyframes");
var SkeletonLine_1 = require("../../../../../components/Skeleton/SkeletonLine");
var SkeletonShape_1 = require("../../../../../components/Skeleton/SkeletonShape");
var SkeletonButton_1 = require("../../../../../components/Skeleton/SkeletonButton");
var SkeletonInput_1 = require("../../../../../components/Skeleton/SkeletonInput");
var SkeletonAvatar_1 = require("../../../../../components/Skeleton/SkeletonAvatar");
var SkeletonText_1 = require("../../../../../components/Skeleton/SkeletonText");
var waveStyles = function (primaryColor, secondoryColor) { return ({
    position: 'static',
    overflow: 'hidden',
    animationDuration: '4.5s',
    animationName: keyframes_1.wave,
    animationIterationCount: 'infinite',
    backgroundColor: '#FFFFFF',
    animationTimingFunction: 'linear',
    backgroundImage: "linear-gradient(to right, " + secondoryColor + " 0%, " + primaryColor + " 20%, " + secondoryColor + " 40%)",
    backgroundSize: '1200px 100%',
}); };
var pulseStyles = {
    animationName: keyframes_1.pulse,
    animationDuration: '1.5s',
    animationIterationCount: 'infinite',
};
var getAnimations = function (primaryColor, secondoryColor) {
    var _a, _b;
    return ({
        pulse: (_a = {},
            _a["& ." + SkeletonLine_1.skeletonLineClassName] = pulseStyles,
            _a["& ." + SkeletonShape_1.skeletonShapeClassName] = pulseStyles,
            _a["& ." + SkeletonButton_1.skeletonButtonClassName] = pulseStyles,
            _a["& ." + SkeletonAvatar_1.skeletonAvatarClassName] = pulseStyles,
            _a["& ." + SkeletonInput_1.skeletonInputClassName] = pulseStyles,
            _a["& ." + SkeletonText_1.skeletonTextClassName] = pulseStyles,
            _a),
        wave: (_b = {},
            _b["& ." + SkeletonLine_1.skeletonLineClassName] = waveStyles(primaryColor, secondoryColor),
            _b["& ." + SkeletonShape_1.skeletonShapeClassName] = waveStyles(primaryColor, secondoryColor),
            _b["& ." + SkeletonButton_1.skeletonButtonClassName] = waveStyles(primaryColor, secondoryColor),
            _b["& ." + SkeletonAvatar_1.skeletonAvatarClassName] = waveStyles(primaryColor, secondoryColor),
            _b["& ." + SkeletonInput_1.skeletonInputClassName] = waveStyles(primaryColor, secondoryColor),
            _b["& ." + SkeletonText_1.skeletonTextClassName] = waveStyles(primaryColor, secondoryColor),
            _b),
    });
};
exports.getAnimations = getAnimations;
