// @fluentui/react-northstar/dist/dts/src/themes/teams-dark-v2/siteVariables.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.colorScheme = exports.colors = void 0;
var siteVariables_1 = require("../teams-v2/siteVariables");
exports.colors = siteVariables_1.colors;
exports.colorScheme = {
    default: {
        foreground: exports.colors.white,
        foreground1: exports.colors.grey['220'],
        foreground2: exports.colors.grey['310'],
        foreground3: exports.colors.white,
        foreground4: exports.colors.white,
        foreground5: exports.colors.grey['450'],
        foreground6: exports.colors.grey['550'],
        foreground7: exports.colors.white,
        background: exports.colors.grey['700'],
        background1: exports.colors.grey['750'],
        background2: exports.colors.grey['800'],
        background3: exports.colors.grey['870'],
        background4: exports.colors.grey['550'],
        background5: exports.colors.grey['600'],
        border: exports.colors.grey['450'],
        border1: exports.colors.grey['850'],
        border2: exports.colors.grey['900'],
        border3: exports.colors.grey['550'],
        foregroundHover: exports.colors.white,
        foregroundHover1: exports.colors.white,
        foregroundHover2: exports.colors.white,
        backgroundHover: exports.colors.grey['550'],
        backgroundHover1: exports.colors.grey['750'],
        backgroundHover2: 'transparent',
        backgroundHover3: exports.colors.grey['650'],
        backgroundHover4: exports.colors.grey['750'],
        backgroundHover6: exports.colors.grey['550'],
        borderHover: exports.colors.grey['430'],
        foregroundPressed: exports.colors.white,
        backgroundPressed: exports.colors.grey['650'],
        foregroundActive: exports.colors.white,
        foregroundActive1: exports.colors.white,
        backgroundActive: exports.colors.grey['600'],
        backgroundActive1: exports.colors.grey['800'],
        borderActive: exports.colors.grey['440'],
        // foregroundFocus: not specified,
        // backgroundFocus: not specified,
        borderFocus: exports.colors.white,
        borderFocusWithin: exports.colors.black,
        foregroundDisabled: exports.colors.grey['460'],
        foregroundDisabled1: exports.colors.grey['460'],
        borderDisabled: exports.colors.grey['500'],
        backgroundDisabled: exports.colors.grey['800'],
        backgroundDisabled1: exports.colors.grey['800'],
    },
    brand: {
        background: exports.colors.brand['700'],
        background1: exports.colors.brand['1000'],
        background2: exports.colors.brand['900'],
        background3: exports.colors.brand['1000'],
        background4: exports.colors.grey['910'],
        background5: exports.colors.brand['900'],
        foreground: exports.colors.brand['450'],
        foreground1: exports.colors.brand['450'],
        foreground2: exports.colors.brand['450'],
        foreground3: exports.colors.brand['200'],
        foreground4: exports.colors.white,
        border: exports.colors.grey['450'],
        border1: exports.colors.brand['800'],
        border2: exports.colors.brand['800'],
        foregroundHover: exports.colors.brand['450'],
        foregroundHover1: exports.colors.white,
        foregroundHover2: exports.colors.brand['200'],
        borderHover: exports.colors.brand['600'],
        backgroundHover: exports.colors.brand['600'],
        backgroundHover1: exports.colors.brand['900'],
        foregroundPressed: exports.colors.brand['200'],
        foregroundPressed1: exports.colors.white,
        backgroundPressed: exports.colors.brand['800'],
        borderPressed: exports.colors.brand['800'],
        foregroundActive: exports.colors.brand['450'],
        foregroundActive1: exports.colors.brand['450'],
        foregroundActive2: exports.colors.brand['50'],
        backgroundActive: exports.colors.brand['450'],
        backgroundActive1: exports.colors.brand['450'],
        borderActive: exports.colors.grey['450'],
        borderActive1: exports.colors.brand['800'],
        borderActive2: exports.colors.brand['800'],
        foregroundFocus: exports.colors.brand['450'],
        foregroundFocus1: exports.colors.brand['450'],
        foregroundFocus2: exports.colors.brand['450'],
        foregroundFocus3: exports.colors.brand['50'],
        foregroundFocus4: exports.colors.white,
        backgroundFocus: exports.colors.brand['450'],
        backgroundFocus1: exports.colors.brand['1000'],
        backgroundFocus2: exports.colors.brand['900'],
        backgroundFocus3: exports.colors.brand['1000'],
        borderFocus: exports.colors.white,
        borderFocus1: exports.colors.brand['450'],
        borderFocusWithin: exports.colors.black,
        foregroundDisabled: exports.colors.grey['460'],
        foregroundDisabled1: exports.colors.grey['460'],
        borderDisabled: exports.colors.grey['500'],
        backgroundDisabled: exports.colors.grey['800'],
        backgroundDisabled1: exports.colors.grey['800'],
    },
};
