// @fluentui/react-northstar/dist/dts/src/themes/teams/components/Datepicker/datepickerVariables.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.datepickerVariables = void 0;
var utils_1 = require("../../../../utils");
var date_time_utilities_1 = require("../../../../utils/date-time-utilities");
var datepickerVariables = function (siteVars) { return ({
    calendarCellBorder: 'none',
    calendarCellHeight: utils_1.pxToRem(34),
    calendarCellWidth: utils_1.pxToRem(34),
    calendarCellPadding: utils_1.pxToRem(0),
    calendarCellMargin: utils_1.pxToRem(0),
    calendarCellBackgroundColor: siteVars.colorScheme.default.background,
    calendarCellColor: 'inherit',
    calendarCellQuietColor: siteVars.colorScheme.default.foreground2,
    calendarCellSelectedColor: siteVars.colorScheme.brand.background4,
    calendarCellSelectedBackgroundColor: siteVars.colorScheme.brand.borderActive1,
    calendarCellTodayBackgroundColor: siteVars.colorScheme.brand.backgroundFocus,
    calendarCellTodayColor: siteVars.colorScheme.white.foreground,
    calendarCellTodayBorderRadius: '50%',
    calendarCellHoverBackgroundColor: siteVars.colorScheme.brand.backgroundHover2,
    calendarCellHoverColor: 'inherit',
    calendarCellDisabledColor: siteVars.colorScheme.brand.foregroundDisabled,
    calendarCellDisabledBackgroundColor: siteVars.colorScheme.default.background,
    calendarHeaderCellFontWeight: siteVars.fontWeightRegular,
    calendarHeaderCellHeight: utils_1.pxToRem(32),
    calendarHeaderCellPadding: utils_1.pxToRem(2) + " " + utils_1.pxToRem(0) + " " + utils_1.pxToRem(2) + " " + utils_1.pxToRem(0),
    calendarHeaderCellWidth: "calc(100%/" + date_time_utilities_1.DAYS_IN_WEEK + ")",
    calendarHeaderPaddingTop: utils_1.pxToRem(5),
    calendarHeaderPaddingBottom: utils_1.pxToRem(5),
    calendarHeaderLabelPaddingLeft: utils_1.pxToRem(10),
    calendarHeaderLabelFontWeight: siteVars.fontWeightBold,
    calendarMinHeight: utils_1.pxToRem(282),
}); };
exports.datepickerVariables = datepickerVariables;
