// @fluentui/react-northstar/dist/dts/src/themes/teams/components/Datepicker/datepickerCalendarCellButtonStyles.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.datepickerCalendarCellButtonStyles = void 0;
var getBorderFocusStyles_1 = require("../../getBorderFocusStyles");
exports.datepickerCalendarCellButtonStyles = {
    root: function (_a) {
        var p = _a.props, v = _a.variables, theme = _a.theme;
        var siteVariables = theme.siteVariables;
        var borderWidth = siteVariables.borderWidth;
        var borderFocusStyles = getBorderFocusStyles_1.getBorderFocusStyles({
            variables: siteVariables,
            borderPadding: borderWidth,
        });
        return __assign(__assign(__assign(__assign(__assign(__assign({ textAlign: 'center', height: '100%', width: '100%', cursor: 'pointer', border: v.calendarCellBorder, padding: v.calendarCellPadding, margin: v.calendarCellMargin, display: 'table-cell', position: 'relative' }, borderFocusStyles), { backgroundColor: v.calendarCellBackgroundColor, color: v.calendarCellColor, ':hover': {
                backgroundColor: v.calendarCellHoverBackgroundColor,
                color: v.calendarCellHoverColor,
            } }), (p.quiet && {
            color: v.calendarCellQuietColor,
        })), (p.selected && {
            color: v.calendarCellSelectedColor,
            backgroundColor: v.calendarCellSelectedBackgroundColor,
        })), (p.disabled && {
            color: v.calendarCellDisabledColor,
            cursor: 'default',
            backgroundColor: v.calendarCellDisabledBackgroundColor,
            ':hover': {},
        })), (p.today && {
            backgroundColor: v.calendarCellTodayBackgroundColor,
            color: v.calendarCellTodayColor,
            borderRadius: v.calendarCellTodayBorderRadius,
        }));
    },
};
