// @uifabric/utilities/lib-commonjs/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
tslib_1.__exportStar(require("./Async"), exports);
tslib_1.__exportStar(require("./AutoScroll"), exports);
tslib_1.__exportStar(require("./BaseComponent"), exports);
tslib_1.__exportStar(require("./DelayedRender"), exports);
tslib_1.__exportStar(require("./EventGroup"), exports);
tslib_1.__exportStar(require("./FabricPerformance"), exports);
tslib_1.__exportStar(require("./GlobalSettings"), exports);
tslib_1.__exportStar(require("./KeyCodes"), exports);
tslib_1.__exportStar(require("./Rectangle"), exports);
tslib_1.__exportStar(require("./appendFunction"), exports);
tslib_1.__exportStar(require("./aria"), exports);
tslib_1.__exportStar(require("./array"), exports);
tslib_1.__exportStar(require("./asAsync"), exports);
tslib_1.__exportStar(require("./assertNever"), exports);
tslib_1.__exportStar(require("./classNamesFunction"), exports);
tslib_1.__exportStar(require("./componentAs/composeComponentAs"), exports);
tslib_1.__exportStar(require("./controlled"), exports);
tslib_1.__exportStar(require("./css"), exports);
tslib_1.__exportStar(require("./customizations/Customizations"), exports);
tslib_1.__exportStar(require("./customizations/Customizer"), exports);
tslib_1.__exportStar(require("./customizations/CustomizerContext"), exports);
tslib_1.__exportStar(require("./customizations/customizable"), exports);
tslib_1.__exportStar(require("./customizations/useCustomizationSettings"), exports);
tslib_1.__exportStar(require("./customizations/mergeCustomizations"), exports);
tslib_1.__exportStar(require("./customizations/mergeSettings"), exports);
tslib_1.__exportStar(require("./dom"), exports);
tslib_1.__exportStar(require("./extendComponent"), exports);
tslib_1.__exportStar(require("./focus"), exports);
tslib_1.__exportStar(require("./getId"), exports);
tslib_1.__exportStar(require("./getNativeElementProps"), exports);
tslib_1.__exportStar(require("./hoist"), exports);
tslib_1.__exportStar(require("./hoistStatics"), exports);
tslib_1.__exportStar(require("./initializeComponentRef"), exports);
tslib_1.__exportStar(require("./initializeFocusRects"), exports);
tslib_1.__exportStar(require("./useFocusRects"), exports);
tslib_1.__exportStar(require("./initials"), exports);
tslib_1.__exportStar(require("./keyboard"), exports);
tslib_1.__exportStar(require("./language"), exports);
tslib_1.__exportStar(require("./math"), exports);
tslib_1.__exportStar(require("./memoize"), exports);
tslib_1.__exportStar(require("./merge"), exports);
tslib_1.__exportStar(require("./mobileDetector"), exports);
tslib_1.__exportStar(require("./modalize"), exports);
var object_1 = require("./object");
exports.assign = object_1.assign;
exports.filteredAssign = object_1.filteredAssign;
exports.mapEnumByName = object_1.mapEnumByName;
exports.shallowCompare = object_1.shallowCompare;
exports.values = object_1.values;
exports.omit = object_1.omit;
tslib_1.__exportStar(require("./osDetector"), exports);
tslib_1.__exportStar(require("./overflow"), exports);
tslib_1.__exportStar(require("./properties"), exports);
tslib_1.__exportStar(require("./renderFunction/composeRenderFunction"), exports);
tslib_1.__exportStar(require("./resources"), exports);
tslib_1.__exportStar(require("./rtl"), exports);
tslib_1.__exportStar(require("./safeRequestAnimationFrame"), exports);
tslib_1.__exportStar(require("./safeSetTimeout"), exports);
tslib_1.__exportStar(require("./scroll"), exports);
tslib_1.__exportStar(require("./selection/index"), exports);
tslib_1.__exportStar(require("./string"), exports);
tslib_1.__exportStar(require("./styled"), exports);
tslib_1.__exportStar(require("./warn"), exports);
tslib_1.__exportStar(require("./ie11Detector"), exports);
tslib_1.__exportStar(require("./getPropsWithDefaults"), exports);
var setFocusVisibility_1 = require("./setFocusVisibility");
exports.setFocusVisibility = setFocusVisibility_1.setFocusVisibility;
exports.IsFocusVisibleClassName = setFocusVisibility_1.IsFocusVisibleClassName;
var setSSR_1 = require("./dom/setSSR");
exports.setSSR = setSSR_1.setSSR;
var createMergedRef_1 = require("./createMergedRef");
exports.createMergedRef = createMergedRef_1.createMergedRef;
require("./version");
