// @fluentui/react-northstar/dist/dts/src/themes/teams/components/Avatar/avatarVariables.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.avatarVariables = exports.labelColorAreas = void 0;
var utils_1 = require("../../../../utils");
exports.labelColorAreas = utils_1.stringLiteralsArray('foreground', 'background');
var avatarVariables = function (siteVariables) { return ({
    avatarBorderColor: '',
    avatarBorderWidth: '0',
    squareAvatarBorderRadius: siteVariables.borderRadiusMedium,
    iconColor: siteVariables.colors.white,
    iconBackgroundColor: siteVariables.colors.brand[600],
    statusBorderWidth: '2px',
    statusIconSize: utils_1.pxToRem(7),
    statusBorderColor: siteVariables.bodyBackground,
    statusSuccessBackgroundColor: siteVariables.colorScheme.green.background,
    statusSuccessColor: siteVariables.colorScheme.green.foreground1,
    statusInfoBackgroundColor: siteVariables.colorScheme.brand.background,
    statusInfoColor: siteVariables.colorScheme.default.foreground2,
    statusWarningBackgroundColor: siteVariables.colorScheme.yellow.background,
    statusWarningColor: siteVariables.colorScheme.yellow.foreground2,
    statusErrorBackgroundColor: siteVariables.colorScheme.red.background,
    statusErrorColor: siteVariables.colorScheme.red.foreground2,
    statusBackgroundColor: siteVariables.colorScheme.default.background5,
    statusColor: siteVariables.colorScheme.default.foreground4,
    imageWidth: undefined,
    imageHeight: undefined,
    imageAvatarRadius: utils_1.pxToRem(9999),
    imageAvatarSize: utils_1.pxToRem(32),
    imageCircularRadius: utils_1.pxToRem(9999),
    labelCircularRadius: utils_1.pxToRem(9999),
    labelColor: 'rgba(0, 0, 0, 0.6)',
    labelBackground: 'rgb(232, 232, 232)',
}); };
exports.avatarVariables = avatarVariables;
