// @fluentui/react-northstar/dist/dts/src/themes/teams/components/Avatar/avatarLabelStyles.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.avatarLabelStyles = void 0;
var utils_1 = require("../../../../utils");
var sizeToPxValue = {
    smallest: 20,
    smaller: 24,
    small: 28,
    medium: 32,
    large: 44,
    larger: 64,
    largest: 96,
};
exports.avatarLabelStyles = {
    root: function (_a) {
        var p = _a.props, v = _a.variables;
        var sizeInRem = utils_1.pxToRem(sizeToPxValue[p.size]);
        return __assign(__assign({ alignItems: 'center', overflow: 'hidden', color: v.labelColor, backgroundColor: v.labelBackground, borderRadius: '50%', display: 'inline-block', width: sizeInRem, height: sizeInRem, lineHeight: sizeInRem, fontSize: utils_1.pxToRem(sizeToPxValue[p.size] / 2.333), verticalAlign: 'top', textAlign: 'center', padding: '0' }, (p.square && {
            borderRadius: v.squareAvatarBorderRadius,
        })), (p.circular && {
            borderRadius: v.labelCircularRadius,
        }));
    },
};
