// @fluentui/react-northstar/dist/dts/src/themes/teams/components/Avatar/avatarStatusStyles.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.avatarStatusStyles = exports.getSizeStyles = void 0;
var utils_1 = require("../../../../utils");
var sizeToPxValue = {
    smallest: 6,
    smaller: 10,
    small: 10,
    medium: 10,
    large: 10,
    larger: 16,
    largest: 0,
};
var getSizeStyles = function (sizeInPx) {
    var sizeInRem = utils_1.pxToRem(sizeInPx);
    return {
        height: sizeInRem,
        width: sizeInRem,
    };
};
exports.getSizeStyles = getSizeStyles;
exports.avatarStatusStyles = {
    root: function (_a) {
        var v = _a.variables, _b = _a.props, color = _b.color, size = _b.size, state = _b.state;
        return (__assign(__assign(__assign(__assign(__assign(__assign(__assign(__assign({ display: 'inline-flex', alignItems: 'center', justifyContent: 'center' }, exports.getSizeStyles(sizeToPxValue[size])), { verticalAlign: 'middle', borderRadius: '9999px', position: 'absolute', bottom: 0, right: 0, boxShadow: "0 0 0 " + v.statusBorderWidth + " " + v.statusBorderColor }), (state === 'success' && {
            backgroundColor: v.statusSuccessBackgroundColor,
        })), (state === 'info' && {
            backgroundColor: v.statusInfoBackgroundColor,
        })), (state === 'warning' && {
            backgroundColor: v.statusWarningBackgroundColor,
        })), (state === 'error' && {
            backgroundColor: v.statusErrorBackgroundColor,
        })), (state === 'unknown' && {
            backgroundColor: v.statusBackgroundColor,
        })), (!!color && {
            backgroundColor: color,
        })));
    },
};
