// @fluentui/accessibility/dist/dts/src/behaviors/List/navigableListItemBehavior.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.navigableListItemBehavior = void 0;
var keyboard_key_1 = require("../../keyboard-key");
var attributes_1 = require("../../attributes");
/**
 * @specification
 * Adds role='menuitem'.
 * Adds attribute 'data-is-focusable=true' to 'root' slot.
 * Triggers 'performClick' action with 'Enter' or 'Spacebar' on 'root'.
 */
var navigableListItemBehavior = function (props) {
    var _a;
    return ({
        attributes: {
            root: (_a = {
                    role: 'menuitem'
                },
                _a[attributes_1.IS_FOCUSABLE_ATTRIBUTE] = true,
                _a),
        },
        keyActions: {
            root: {
                performClick: {
                    keyCombinations: [{ keyCode: keyboard_key_1.keyboardKey.Enter }, { keyCode: keyboard_key_1.SpacebarKey }],
                },
            },
        },
    });
};
exports.navigableListItemBehavior = navigableListItemBehavior;
