// @fluentui/accessibility/dist/dts/src/behaviors/List/selectableListItemBehavior.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.selectableListItemBehavior = void 0;
var keyboard_key_1 = require("../../keyboard-key");
var attributes_1 = require("../../attributes");
/**
 * @specification
 * Adds role='option'. This role is used for a selectable item in a list.
 * Adds attribute 'data-is-focusable=true' to 'root' slot.
 * Adds attribute 'aria-selected=true' based on the property 'selected'. Based on this screen readers will recognize the selected state of the item.
 * Triggers 'performClick' action with 'Enter' or 'Spacebar' on 'root'.
 */
var selectableListItemBehavior = function (props) {
    var _a;
    return ({
        attributes: {
            root: (_a = {
                    role: 'option',
                    'aria-selected': !!props.selected
                },
                _a[attributes_1.IS_FOCUSABLE_ATTRIBUTE] = true,
                _a),
        },
        keyActions: {
            root: {
                performClick: {
                    keyCombinations: [{ keyCode: keyboard_key_1.keyboardKey.Enter }, { keyCode: keyboard_key_1.SpacebarKey }],
                },
            },
        },
    });
};
exports.selectableListItemBehavior = selectableListItemBehavior;
