// @fluentui/react-northstar/dist/commonjs/components/Menu/MenuItemIndicator.js
"use strict";

exports.__esModule = true;
exports.MenuItemIndicator = exports.menuItemIndicatorClassName = void 0;

var React = _interopRequireWildcard(require("react"));

var _reactBindings = require("@fluentui/react-bindings");

var _accessibility = require("@fluentui/accessibility");

var PropTypes = _interopRequireWildcard(require("prop-types"));

var _utils = require("../../utils");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var menuItemIndicatorClassName = 'ui-menu__itemindicator';
/**
 * A MenuItemIndicator allows a user to have a dedicated component that can be targeted from the theme.
 */

exports.menuItemIndicatorClassName = menuItemIndicatorClassName;
var MenuItemIndicator = /*#__PURE__*/React.forwardRef(function (props, ref) {
  var context = (0, _reactBindings.useFluentContext)();

  var _useTelemetry = (0, _reactBindings.useTelemetry)(MenuItemIndicator.displayName, context.telemetry),
      setStart = _useTelemetry.setStart,
      setEnd = _useTelemetry.setEnd;

  setStart();
  var className = props.className,
      children = props.children,
      design = props.design,
      styles = props.styles,
      variables = props.variables,
      content = props.content,
      iconOnly = props.iconOnly,
      vertical = props.vertical,
      inSubmenu = props.inSubmenu,
      active = props.active,
      primary = props.primary,
      underlined = props.underlined;

  var _useStyles = (0, _reactBindings.useStyles)(MenuItemIndicator.displayName, {
    className: menuItemIndicatorClassName,
    mapPropsToStyles: function mapPropsToStyles() {
      return {
        iconOnly: iconOnly,
        vertical: vertical,
        inSubmenu: inSubmenu,
        active: active,
        primary: primary,
        underlined: underlined
      };
    },
    mapPropsToInlineStyles: function mapPropsToInlineStyles() {
      return {
        className: className,
        design: design,
        styles: styles,
        variables: variables
      };
    },
    rtl: context.rtl
  }),
      classes = _useStyles.classes;

  var getA11Props = (0, _reactBindings.useAccessibility)(props.accessibility, {
    debugName: MenuItemIndicator.displayName,
    rtl: context.rtl
  });
  var ElementType = (0, _reactBindings.getElementType)(props);
  var unhandledProps = (0, _reactBindings.useUnhandledProps)(MenuItemIndicator.handledProps, props);
  var element = /*#__PURE__*/React.createElement(ElementType, getA11Props('root', Object.assign({
    className: classes.root,
    ref: ref
  }, unhandledProps)), (0, _reactBindings.childrenExist)(children) ? children : content);
  setEnd();
  return element;
});
exports.MenuItemIndicator = MenuItemIndicator;
MenuItemIndicator.displayName = 'MenuItemIndicator';
MenuItemIndicator.defaultProps = {
  as: 'span',
  accessibility: _accessibility.indicatorBehavior
};
MenuItemIndicator.propTypes = Object.assign({}, _utils.commonPropTypes.createCommon(), {
  iconOnly: PropTypes.bool,
  vertical: PropTypes.bool,
  inSubmenu: PropTypes.bool,
  active: PropTypes.bool,
  primary: PropTypes.bool,
  underlined: PropTypes.bool
});
MenuItemIndicator.handledProps = Object.keys(MenuItemIndicator.propTypes);
MenuItemIndicator.shorthandConfig = {
  mappedProp: 'content'
};
