// @fluentui/react-northstar/dist/dts/src/themes/teams/components/Breadcrumb/breadcrumbLinkStyles.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.breadcrumbLinkStyles = void 0;
var getBorderFocusStyles_1 = require("../../getBorderFocusStyles");
exports.breadcrumbLinkStyles = {
    root: function (_a) {
        var p = _a.props, v = _a.variables, siteVariables = _a.theme.siteVariables;
        var borderFocusStyles = getBorderFocusStyles_1.getBorderFocusStyles(__assign(__assign(__assign(__assign({ variables: siteVariables }, (p.size === 'smaller' && {
            borderPadding: {
                top: v.linkPaddingTop,
                right: v.linkPaddingRightSmaller,
                bottom: v.linkPaddingBottom,
                left: v.linkPaddingLeftSmaller,
            },
        })), (p.size === 'small' && {
            borderPadding: {
                top: v.linkPaddingTop,
                right: v.linkPaddingRightSmall,
                bottom: v.linkPaddingBottom,
                left: v.linkPaddingLeftSmall,
            },
        })), (p.size === 'medium' && {
            borderPadding: {
                top: v.linkPaddingTop,
                right: v.linkPaddingRightMedium,
                bottom: v.linkPaddingBottom,
                left: v.linkPaddingLeftMedium,
            },
        })), (p.size === 'large' && {
            borderPadding: {
                top: v.linkPaddingTop,
                right: v.linkPaddingRightLarge,
                bottom: v.linkPaddingBottom,
                left: v.linkPaddingLeftLarge,
            },
        })));
        return __assign({ userSelect: 'none', display: 'flex', alignItems: 'center', position: 'relative', ':visited': {
                color: 'inherit',
            } }, borderFocusStyles);
    },
};
