// @fluentui/react-northstar/dist/dts/src/themes/teams/components/Breadcrumb/breadcrumbVariables.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.breadcrumbVariables = void 0;
var utils_1 = require("../../../../utils");
var breadcrumbVariables = function (siteVars) { return ({
    dividerMargin: "0 " + utils_1.pxToRem(8),
    fontSizeSmaller: utils_1.pxToRem(12),
    fontSizeSmall: utils_1.pxToRem(12),
    fontSizeMedium: utils_1.pxToRem(14),
    fontSizeLarge: utils_1.pxToRem(18),
    // Breadcrumb Item
    itemCurrentFontWeight: siteVars.fontWeightBold,
    // Breadcrumb Link
    linkPaddingTop: utils_1.pxToRem(6),
    linkPaddingBottom: utils_1.pxToRem(6),
    linkPaddingLeftSmaller: utils_1.pxToRem(4),
    linkPaddingRightSmaller: utils_1.pxToRem(4),
    linkPaddingLeftSmall: utils_1.pxToRem(8),
    linkPaddingRightSmall: utils_1.pxToRem(8),
    linkPaddingLeftMedium: utils_1.pxToRem(12),
    linkPaddingRightMedium: utils_1.pxToRem(12),
    linkPaddingLeftLarge: utils_1.pxToRem(12),
    linkPaddingRightLarge: utils_1.pxToRem(12),
    linkSmallerGap: utils_1.pxToRem(2),
    linkSmallGap: utils_1.pxToRem(8),
    linkMediumGap: utils_1.pxToRem(8),
    linkLargeGap: utils_1.pxToRem(8),
    disabledColor: siteVars.colorScheme.default.foregroundDisabled,
}); };
exports.breadcrumbVariables = breadcrumbVariables;
