// @fluentui/react-northstar/dist/dts/src/themes/teams/components/Breadcrumb/breadcrumbItemStyles.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.breadcrumbItemStyles = void 0;
exports.breadcrumbItemStyles = {
    root: function (_a) {
        var p = _a.props, v = _a.variables;
        return (__assign(__assign(__assign(__assign(__assign(__assign({ display: 'inline-flex', alignItems: 'center', verticalAlign: 'middle' }, (p.active && {
            fontWeight: v.itemCurrentFontWeight,
        })), (p.disabled && {
            color: v.disabledColor,
        })), (p.size === 'smaller' && {
            paddingLeft: v.linkPaddingLeftSmaller,
            paddingRight: v.linkPaddingRightSmaller,
            gap: v.linkSmallerGap,
        })), (p.size === 'small' && {
            paddingLeft: v.linkPaddingLeftSmall,
            paddingRight: v.linkPaddingRightSmall,
            gap: v.linkSmallGap,
        })), (p.size === 'medium' && {
            paddingLeft: v.linkPaddingLeftMedium,
            paddingRight: v.linkPaddingRightMedium,
            gap: v.linkMediumGap,
        })), (p.size === 'large' && {
            paddingLeft: v.linkPaddingLeftLarge,
            paddingRight: v.linkPaddingRightLarge,
            gap: v.linkLargeGap,
        })));
    },
};
