// @fluentui/react-northstar/dist/commonjs/themes/teams/components/Table/tableVariables.js
"use strict";

exports.__esModule = true;
exports.tableVariables = void 0;

var _utils = require("../../../../utils");

var tableVariables = function tableVariables(siteVariables) {
  return {
    color: siteVariables.colorScheme.default.foreground,
    hoverColor: siteVariables.colorScheme.default.foregroundHover,
    backgroundColor: siteVariables.colorScheme.default.background,
    backgroundHoverColor: siteVariables.colorScheme.default.backgroundHover1,
    defaultRowHeight: (0, _utils.pxToRem)(48),
    compactRowHeight: (0, _utils.pxToRem)(36),
    minCellWidth: '0',
    cellPadding: "0 " + (0, _utils.pxToRem)(8),
    rowPadding: '0',
    headerFontSize: (0, _utils.pxToRem)(12),
    bodyFontSize: (0, _utils.pxToRem)(13),
    borderWidth: (0, _utils.pxToRem)(1),
    headerBorderColor: siteVariables.colorScheme.default.backgroundHover1,
    headerBorderHoverColor: 'transparent',
    headerBorderFocusColor: 'transparent',
    rowBorderColor: siteVariables.colorScheme.default.backgroundHover1,
    rowBorderHoverColor: siteVariables.colorScheme.default.backgroundHover1,
    rowBorderFocusColor: siteVariables.colorScheme.default.borderFocus,
    cellBorderColor: 'transparent',
    cellBorderHoverColor: 'transparent',
    cellBorderFocusColor: siteVariables.colorScheme.default.borderFocus
  };
};

exports.tableVariables = tableVariables;
