// @fluentui/react-northstar/dist/commonjs/themes/teams/components/Table/tableCellStyles.js
"use strict";

exports.__esModule = true;
exports.tableCellStyles = void 0;

var _getBorderFocusStyles = require("../../getBorderFocusStyles");

var tableCellStyles = {
  root: function root(_ref) {
    var v = _ref.variables,
        siteVariables = _ref.theme.siteVariables;
    var borderFocusStyles = (0, _getBorderFocusStyles.getBorderFocusStyles)({
      variables: siteVariables
    });
    return Object.assign({
      display: 'flex',
      flexFlow: 'row nowrap',
      flexGrow: 1,
      flexBasis: 0,
      minWidth: v.minCellWidth,
      outline: 0,
      borderWidth: v.borderWidth,
      borderStyle: 'solid',
      borderColor: 'transparent'
    }, borderFocusStyles, {
      padding: v.cellPadding,
      position: 'relative',
      height: '100%'
    });
  },
  content: function content(_ref2) {
    var p = _ref2.props;
    return Object.assign({
      alignSelf: 'center'
    }, p.truncateContent && {
      display: 'block',
      overflow: 'hidden',
      textOverflow: 'ellipsis',
      whiteSpace: 'nowrap'
    });
  }
};
exports.tableCellStyles = tableCellStyles;
