// @fluentui/react-northstar/dist/commonjs/themes/teams/components/Table/tableRowStyles.js
"use strict";

exports.__esModule = true;
exports.tableRowStyles = void 0;

var _getBorderFocusStyles = require("../../getBorderFocusStyles");

var tableRowStyles = {
  root: function root(_ref) {
    var p = _ref.props,
        v = _ref.variables,
        siteVariables = _ref.theme.siteVariables;
    var borderFocusStyles = (0, _getBorderFocusStyles.getBorderFocusStyles)({
      variables: siteVariables
    });
    return Object.assign({
      display: 'flex',
      flexDirection: 'row',
      alignItems: 'center',
      fontSize: v.bodyFontSize,
      height: v.defaultRowHeight,
      color: v.color,
      backgroundColor: v.backgroundColor,
      borderWidth: v.borderWidth,
      borderStyle: 'solid',
      borderColor: 'transparent',
      borderBottomColor: v.rowBorderColor,
      padding: v.rowPadding,
      position: 'relative',
      width: '100%',
      ':hover': {
        color: v.hoverColor,
        backgroundColor: v.backgroundHoverColor,
        borderColor: v.rowBorderHoverColor
      }
    }, borderFocusStyles, p.header && {
      fontSize: v.headerFontSize,
      ':hover': {
        color: v.color,
        backgroundColor: v.backgroundColor
      }
    }, p.compact && {
      height: v.compactRowHeight
    });
  }
};
exports.tableRowStyles = tableRowStyles;
