// @fluentui/react-northstar-fela-renderer/dist/dts/test/felaExpandCssShorthandsPlugin-test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var felaExpandCssShorthandsPlugin_1 = require("../src/felaExpandCssShorthandsPlugin");
describe('felaExpandCssShorthandsPlugin', function () {
    test('should expand margin prop', function () {
        var style = {
            display: 'block',
            margin: '0px 10px',
        };
        expect(felaExpandCssShorthandsPlugin_1.felaExpandCssShorthandsPlugin(style)).toMatchObject({
            display: 'block',
            marginTop: '0px',
            marginRight: '10px',
            marginBottom: '0px',
            marginLeft: '10px',
        });
    });
    test('should handle "undefined" and "null"', function () {
        var style = {
            margin: '10px',
            marginLeft: null,
            marginRight: undefined,
        };
        expect(felaExpandCssShorthandsPlugin_1.felaExpandCssShorthandsPlugin(style)).toMatchObject({
            marginTop: '10px',
            marginRight: undefined,
            marginBottom: '10px',
            marginLeft: null,
        });
    });
    test('should handle arrays', function () {
        var style = {
            margin: ['10px', '0px'],
        };
        expect(felaExpandCssShorthandsPlugin_1.felaExpandCssShorthandsPlugin(style)).toMatchObject({
            marginTop: ['10px', '0px'],
            marginRight: ['10px', '0px'],
            marginBottom: ['10px', '0px'],
            marginLeft: ['10px', '0px'],
        });
    });
    test('should expand pseudo object', function () {
        var style = {
            display: 'block',
            '::before': {
                margin: '0px',
            },
        };
        expect(felaExpandCssShorthandsPlugin_1.felaExpandCssShorthandsPlugin(style)).toMatchObject({
            display: 'block',
            '::before': {
                marginTop: '0px',
                marginRight: '0px',
                marginBottom: '0px',
                marginLeft: '0px',
            },
        });
    });
    test('should expand nested pseudo object', function () {
        var style = {
            display: 'block',
            '::before': {
                margin: '0px',
                ':hover': {
                    padding: '10px',
                },
            },
        };
        expect(felaExpandCssShorthandsPlugin_1.felaExpandCssShorthandsPlugin(style)).toMatchObject({
            display: 'block',
            '::before': {
                marginTop: '0px',
                marginRight: '0px',
                marginBottom: '0px',
                marginLeft: '0px',
                ':hover': {
                    paddingTop: '10px',
                    paddingRight: '10px',
                    paddingBottom: '10px',
                    paddingLeft: '10px',
                },
            },
        });
    });
    test('should merge expanded prop with its shorthand', function () {
        var style = {
            marginTop: '3px',
            margin: '10px',
            marginRight: '15px',
        };
        expect(felaExpandCssShorthandsPlugin_1.felaExpandCssShorthandsPlugin(style)).toMatchObject({
            marginTop: '10px',
            marginRight: '15px',
            marginBottom: '10px',
            marginLeft: '10px',
        });
    });
});
