// @fluentui/react-northstar-fela-renderer/dist/dts/test/felaRenderer-test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var react_northstar_fela_renderer_1 = require("@fluentui/react-northstar-fela-renderer");
// @ts-ignore
var jest_react_fela_1 = require("jest-react-fela");
var React = require("react");
var react_fela_1 = require("react-fela");
var felaRenderer = react_northstar_fela_renderer_1.createFelaRenderer().getOriginalRenderer();
describe('felaRenderer', function () {
    test('basic styles are rendered', function () {
        var snapshot = jest_react_fela_1.createSnapshot(React.createElement(react_fela_1.FelaComponent, { style: { color: 'red' } }), {}, felaRenderer);
        expect(snapshot).toMatchSnapshot();
    });
    test('CSS fallback values are rendered', function () {
        var snapshot = jest_react_fela_1.createSnapshot(React.createElement(react_fela_1.FelaComponent, { style: { color: ['red', 'blue'] } }), {}, felaRenderer);
        expect(snapshot).toMatchSnapshot();
    });
    test('keyframe colors are rendered', function () {
        var styles = {
            animationName: {
                keyframe: function (_a) {
                    var fromColor = _a.fromColor, toColor = _a.toColor;
                    return ({
                        from: {
                            color: fromColor,
                        },
                        to: {
                            color: toColor,
                        },
                    });
                },
                params: {
                    fromColor: 'red',
                    toColor: 'blue',
                },
            },
            animationDuration: '5s',
        };
        var snapshot = jest_react_fela_1.createSnapshot(React.createElement(react_fela_1.FelaComponent, { style: styles }), {}, felaRenderer);
        expect(snapshot).toMatchSnapshot();
    });
    test('array returned by keyframe results in CSS fallback values', function () {
        var styles = {
            animationName: {
                keyframe: function (_a) {
                    var steps = _a.steps;
                    var obj = {};
                    steps.forEach(function (step) {
                        obj[step] = { color: ['blue', 'red', 'yellow'] };
                    });
                    return obj;
                },
                params: { steps: ['0%', '100%'] },
            },
        };
        var snapshot = jest_react_fela_1.createSnapshot(React.createElement(react_fela_1.FelaComponent, { style: styles }), {}, felaRenderer);
        expect(snapshot).toMatchSnapshot();
    });
    test('animations are not applied if animations are disabled', function () {
        var styles = {
            animationName: {
                keyframe: {
                    from: {
                        transform: 'rotate(0deg)',
                    },
                    to: {
                        transform: 'rotate(360deg)',
                    },
                },
            },
        };
        var snapshot = jest_react_fela_1.createSnapshot(React.createElement(react_fela_1.FelaComponent, { style: styles, disableAnimations: true }), {}, felaRenderer);
        expect(snapshot).toMatchSnapshot();
    });
    test('marginLeft is rendered into marginRight due to RTL', function () {
        var snapshot = jest_react_fela_1.createSnapshot(React.createElement(react_fela_1.FelaComponent, { style: { marginLeft: '10px' } }), { rtl: true }, felaRenderer);
        expect(snapshot).toMatchSnapshot();
    });
    test('marginLeft is rendered into marginLeft due to RTL with `noFlip`', function () {
        var snapshot = jest_react_fela_1.createSnapshot(React.createElement(react_fela_1.FelaComponent, { style: { marginLeft: '10px /* @noflip */' } }), {}, felaRenderer);
        expect(snapshot).toMatchSnapshot();
    });
    test('styles are expanded to longhand values', function () {
        var snapshot = jest_react_fela_1.createSnapshot(React.createElement(react_fela_1.FelaComponent, { style: {
                borderStyle: 'solid',
                // spaces in color value are important
                borderColor: 'rgba(51,204, 51, 1) rgba(51,0,204, 1)',
            } }), {}, felaRenderer);
        expect(snapshot).toMatchSnapshot();
    });
    test('prefixes required styles', function () {
        var snapshot = jest_react_fela_1.createSnapshot(React.createElement(react_fela_1.FelaComponent, { style: {
                cursor: 'zoom-in',
                display: 'flex',
                filter: 'blur(5px)',
                height: 'min-content',
                position: 'sticky',
                transition: 'width 2s, height 2s, background-color 2s, transform 2s',
                ':hover': {
                    backgroundImage: 'image-set("cat.png" 1x, "cat-2x.png" 2x)',
                    display: 'inline-flex',
                    height: 'fit-content',
                },
            } }), {}, felaRenderer);
        expect(snapshot).toMatchSnapshot();
    });
});
