// @fluentui/react-northstar-fela-renderer/dist/dts/test/felaDisableAnimationsPlugin-test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var felaDisableAnimationsPlugin_1 = require("../src/felaDisableAnimationsPlugin");
var stylesWithAnimationShorthand = {
    animation: 'k1',
    margin: '0px 10px',
};
var stylesWithFallbackValues = {
    display: ['grid', '-ms-grid'],
};
var stylesWithAnimationProps = {
    animationName: 'k1',
    animationDuration: '1s',
    margin: '0px 10px',
};
describe('felaDisableAnimationsPlugin', function () {
    test('does not disable animations if the props are not provided', function () {
        expect(felaDisableAnimationsPlugin_1.felaDisableAnimationsPlugin(stylesWithAnimationShorthand, 'RULE')).toMatchObject(stylesWithAnimationShorthand);
    });
    test('does not disable animations if the disableAnimations flag is undefined', function () {
        expect(felaDisableAnimationsPlugin_1.felaDisableAnimationsPlugin(stylesWithAnimationShorthand, 'RULE', undefined, {
            disableAnimations: undefined,
        })).toMatchObject(stylesWithAnimationShorthand);
    });
    test('does not disable animations if the disableAnimations flag is false', function () {
        expect(felaDisableAnimationsPlugin_1.felaDisableAnimationsPlugin(stylesWithAnimationProps, 'RULE', undefined, {
            disableAnimations: false,
        })).toMatchObject(stylesWithAnimationProps);
    });
    test('disables animations if the disableAnimations flag is true', function () {
        expect(felaDisableAnimationsPlugin_1.felaDisableAnimationsPlugin(stylesWithAnimationProps, 'RULE', undefined, {
            disableAnimations: true,
        })).toMatchObject({ margin: '0px 10px' });
    });
    test('disables animations if the disableAnimations flag is true and the animation css shorthand is used', function () {
        expect(felaDisableAnimationsPlugin_1.felaDisableAnimationsPlugin(stylesWithAnimationShorthand, 'RULE', undefined, {
            disableAnimations: true,
        })).toMatchObject({ margin: '0px 10px' });
    });
    test('keeps unrelated fallback values', function () {
        expect(felaDisableAnimationsPlugin_1.felaDisableAnimationsPlugin(stylesWithFallbackValues, 'RULE', undefined, {
            disableAnimations: true,
        })).toMatchObject({ display: ['grid', '-ms-grid'] });
    });
});
