// @fluentui/react-northstar/dist/dts/src/utils/commonPropTypes.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.commonPropTypes = void 0;
var customPropTypes = require("@fluentui/react-proptypes");
var PropTypes = require("prop-types");
exports.commonPropTypes = {
    createCommon: function (config) {
        if (config === void 0) { config = {}; }
        var _a = config.accessibility, accessibility = _a === void 0 ? true : _a, _b = config.as, as = _b === void 0 ? true : _b, _c = config.children, children = _c === void 0 ? 'node' : _c, _d = config.className, className = _d === void 0 ? true : _d, _e = config.color, color = _e === void 0 ? false : _e, _f = config.content, content = _f === void 0 ? 'node' : _f, _g = config.styled, styled = _g === void 0 ? true : _g;
        return __assign(__assign(__assign(__assign(__assign(__assign(__assign({}, (accessibility && {
            accessibility: customPropTypes.accessibility,
        })), (as && {
            as: PropTypes.elementType,
        })), (children && {
            children: children === 'element' ? PropTypes.element : PropTypes.node,
        })), (className && {
            className: PropTypes.string,
        })), (color && {
            color: PropTypes.string,
        })), (content && {
            content: content === 'shorthand' ? customPropTypes.itemShorthand : customPropTypes.nodeContent,
        })), (styled && {
            styles: PropTypes.oneOfType([PropTypes.object, PropTypes.func]),
            variables: PropTypes.oneOfType([PropTypes.object, PropTypes.func]),
            design: customPropTypes.design,
        }));
    },
};
