// @fluentui/react-northstar/dist/dts/src/utils/createComponent.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createComponent = void 0;
var _ = require("lodash");
var createComponentInternal_1 = require("./createComponentInternal");
var createComponent = function (_a) {
    var displayName = _a.displayName, className = _a.className, render = _a.render, defaultProps = _a.defaultProps, actionHandlers = _a.actionHandlers;
    return createComponentInternal_1.createComponentInternal({
        displayName: displayName,
        className: className,
        render: function (config, props) {
            var filteredConfig = _.pick(config, ['accessibility', 'classes', 'rtl', 'styles']);
            return render(__assign({ config: filteredConfig }, props));
        },
        defaultProps: defaultProps,
        actionHandlers: actionHandlers,
    });
};
exports.createComponent = createComponent;
