// @fluentui/react-northstar/dist/dts/src/utils/factories.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createShorthand = exports.createShorthandInternal = exports.createShorthandFactory = void 0;
var styles_1 = require("@fluentui/styles");
var classnames_1 = require("classnames");
var _ = require("lodash");
var React = require("react");
var ReactIs = require("react-is");
// It's only necessary to map props that don't use 'children' as value ('children' is the default)
var mappedProps = {
    iframe: 'src',
    img: 'src',
    input: 'type',
};
function createShorthandFactory(_a) {
    var Component = _a.Component, mappedProp = _a.mappedProp, mappedArrayProp = _a.mappedArrayProp, allowsJSX = _a.allowsJSX;
    if (!ReactIs.isValidElementType(Component)) {
        throw new Error('createShorthandFactory() Component must be a string or function.');
    }
    return function (value, options) {
        return createShorthandInternal({
            allowsJSX: allowsJSX,
            Component: Component,
            mappedProp: mappedProp,
            mappedArrayProp: mappedArrayProp,
            value: value,
            options: options,
        });
    };
}
exports.createShorthandFactory = createShorthandFactory;
// ============================================================
// Factories
// ============================================================
function createShorthandInternal(_a) {
    var Component = _a.Component, mappedProp = _a.mappedProp, mappedArrayProp = _a.mappedArrayProp, value = _a.value, _b = _a.options, options = _b === void 0 ? {} : _b, _c = _a.allowsJSX, allowsJSX = _c === void 0 ? true : _c;
    if (!ReactIs.isValidElementType(Component)) {
        throw new Error('createShorthand() Component must be a string or function.');
    }
    // short circuit noop values
    var valIsNoop = _.isNil(value) || typeof value === 'boolean';
    if (valIsNoop && !options.render)
        return null;
    var valIsPrimitive = typeof value === 'string' || typeof value === 'number';
    var valIsPropsObject = _.isPlainObject(value);
    var valIsArray = _.isArray(value);
    var valIsReactElement = React.isValidElement(value);
    // unhandled type warning
    if (process.env.NODE_ENV !== 'production') {
        var displayName = typeof Component === 'string' ? Component : Component.displayName;
        if (!valIsPrimitive && !valIsPropsObject && !valIsArray && !valIsReactElement && !valIsNoop) {
            /* eslint-disable-next-line no-console */
            console.error([
                "The shorthand prop for \"" + displayName + "\" component was passed a JSX element but this slot only supports string|number|object|array|ReactElements.",
                ' Use null|undefined|boolean for none.',
                " Received: " + value,
            ].join(''));
        }
        if (!allowsJSX && valIsReactElement) {
            /* eslint-disable-next-line no-console */
            console.error([
                "The shorthand prop for \"" + displayName + "\" component was passed a JSX element but this slot only supports string|number|object|array.",
                ' Use null|undefined|boolean for none.',
                " Received: " + value,
            ].join(''));
        }
    }
    // ----------------------------------------
    // Build up props
    // ----------------------------------------
    var defaultProps = options.defaultProps ? options.defaultProps() || {} : {};
    // User's props
    var usersProps = (valIsReactElement && {}) || (valIsPropsObject && value) || {};
    // Override props
    var overrideProps = typeof options.overrideProps === 'function'
        ? options.overrideProps(__assign(__assign({}, defaultProps), usersProps))
        : options.overrideProps || {};
    // Merge props
    var props = __assign(__assign(__assign({}, defaultProps), usersProps), overrideProps);
    var mappedHTMLProps = mappedProps[overrideProps.as || defaultProps.as];
    // Map prop for primitive value
    if (valIsPrimitive || valIsReactElement) {
        props[mappedHTMLProps || mappedProp || 'children'] = value;
    }
    // Map prop for array value
    if (valIsArray) {
        props[mappedHTMLProps || mappedArrayProp || 'children'] = value;
    }
    // Merge className
    if (defaultProps.className || overrideProps.className || usersProps.className) {
        var mergedClassesNames = classnames_1.default(defaultProps.className, overrideProps.className, usersProps.className);
        props.className = _.uniq(mergedClassesNames.split(' ')).join(' ');
    }
    // Merge style
    if (defaultProps.style || overrideProps.style || usersProps.style) {
        props.style = __assign(__assign(__assign({}, defaultProps.style), usersProps.style), overrideProps.style);
    }
    // Merge styles
    if (defaultProps.styles || overrideProps.styles || usersProps.styles) {
        props.styles = styles_1.mergeStyles(defaultProps.styles, usersProps.styles, overrideProps.styles);
    }
    // ----------------------------------------
    // Get key
    // ----------------------------------------
    var _d = options.generateKey, generateKey = _d === void 0 ? true : _d;
    // Use key or generate key
    if (generateKey && _.isNil(props.key)) {
        if (valIsPrimitive) {
            // use string/number shorthand values as the key
            props.key = value;
        }
        if (valIsReactElement) {
            // use the key from React Element
            var elementKey = value.key;
            // <div /> - key is not passed as will be `null`
            // <div key={null} /> - key is passed as `null` and will be stringified
            var isNullKey = elementKey === null;
            if (!isNullKey) {
                props.key = elementKey;
            }
        }
    }
    // Remove the kind prop from the props object
    delete props.kind;
    // ----------------------------------------
    // Create Element
    // ----------------------------------------
    var render = options.render;
    if (render) {
        return render(Component, props);
    }
    if (typeof props.children === 'function') {
        return props.children(Component, __assign(__assign({}, props), { children: undefined }));
    }
    if (!allowsJSX && valIsReactElement) {
        return null;
    }
    // Create ReactElements from built up props
    if (valIsPrimitive || valIsPropsObject || valIsArray || valIsReactElement) {
        return React.createElement(Component, props);
    }
    return null;
}
exports.createShorthandInternal = createShorthandInternal;
function createShorthand(Component, value, options) {
    var _a;
    var _b = Component.shorthandConfig || ((_a = Component.fluentComposeConfig) === null || _a === void 0 ? void 0 : _a.shorthandConfig) || {}, _c = _b.mappedProp, mappedProp = _c === void 0 ? 'children' : _c, _d = _b.allowsJSX, allowsJSX = _d === void 0 ? true : _d, mappedArrayProp = _b.mappedArrayProp;
    return createShorthandInternal({
        Component: Component,
        mappedProp: mappedProp,
        allowsJSX: allowsJSX,
        mappedArrayProp: mappedArrayProp,
        value: value,
        options: options || {},
    });
}
exports.createShorthand = createShorthand;
