// @fluentui/react-northstar/dist/dts/src/utils/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.screenReaderContainerStyles = exports.shouldPreventDefaultOnKeyDown = exports.getKindProp = exports.createComponent = exports.pxToRem = exports.doesNodeContainClick = exports.isBrowser = exports.partitionHTMLProps = exports.htmlInputProps = exports.htmlInputEvents = exports.htmlInputAttrs = exports.htmlImageProps = exports.renderComponent = exports.mergeProviderContexts = exports.getOrGenerateIdFromShorthand = exports.stringLiteralsArray = exports.rtlTextContainer = exports.childrenExist = void 0;
var react_bindings_1 = require("@fluentui/react-bindings");
Object.defineProperty(exports, "childrenExist", { enumerable: true, get: function () { return react_bindings_1.childrenExist; } });
var rtlTextContainer_1 = require("./rtlTextContainer");
Object.defineProperty(exports, "rtlTextContainer", { enumerable: true, get: function () { return rtlTextContainer_1.rtlTextContainer; } });
var stringLiteralsArray_1 = require("./stringLiteralsArray");
Object.defineProperty(exports, "stringLiteralsArray", { enumerable: true, get: function () { return stringLiteralsArray_1.stringLiteralsArray; } });
var getOrGenerateIdFromShorthand_1 = require("./getOrGenerateIdFromShorthand");
Object.defineProperty(exports, "getOrGenerateIdFromShorthand", { enumerable: true, get: function () { return getOrGenerateIdFromShorthand_1.getOrGenerateIdFromShorthand; } });
__exportStar(require("./factories"), exports);
var mergeProviderContexts_1 = require("./mergeProviderContexts");
Object.defineProperty(exports, "mergeProviderContexts", { enumerable: true, get: function () { return mergeProviderContexts_1.mergeProviderContexts; } });
__exportStar(require("./renderComponent"), exports);
var renderComponent_1 = require("./renderComponent");
Object.defineProperty(exports, "renderComponent", { enumerable: true, get: function () { return renderComponent_1.renderComponent; } });
var htmlPropsUtils_1 = require("./htmlPropsUtils");
Object.defineProperty(exports, "htmlImageProps", { enumerable: true, get: function () { return htmlPropsUtils_1.htmlImageProps; } });
Object.defineProperty(exports, "htmlInputAttrs", { enumerable: true, get: function () { return htmlPropsUtils_1.htmlInputAttrs; } });
Object.defineProperty(exports, "htmlInputEvents", { enumerable: true, get: function () { return htmlPropsUtils_1.htmlInputEvents; } });
Object.defineProperty(exports, "htmlInputProps", { enumerable: true, get: function () { return htmlPropsUtils_1.htmlInputProps; } });
Object.defineProperty(exports, "partitionHTMLProps", { enumerable: true, get: function () { return htmlPropsUtils_1.partitionHTMLProps; } });
var isBrowser_1 = require("./isBrowser");
Object.defineProperty(exports, "isBrowser", { enumerable: true, get: function () { return isBrowser_1.isBrowser; } });
var doesNodeContainClick_1 = require("./doesNodeContainClick");
Object.defineProperty(exports, "doesNodeContainClick", { enumerable: true, get: function () { return doesNodeContainClick_1.doesNodeContainClick; } });
var fontSizeUtility_1 = require("./fontSizeUtility");
Object.defineProperty(exports, "pxToRem", { enumerable: true, get: function () { return fontSizeUtility_1.pxToRem; } });
var createComponent_1 = require("./createComponent");
Object.defineProperty(exports, "createComponent", { enumerable: true, get: function () { return createComponent_1.createComponent; } });
var getKindProp_1 = require("./getKindProp");
Object.defineProperty(exports, "getKindProp", { enumerable: true, get: function () { return getKindProp_1.getKindProp; } });
var shouldPreventDefaultOnKeyDown_1 = require("./shouldPreventDefaultOnKeyDown");
Object.defineProperty(exports, "shouldPreventDefaultOnKeyDown", { enumerable: true, get: function () { return shouldPreventDefaultOnKeyDown_1.shouldPreventDefaultOnKeyDown; } });
__exportStar(require("./whatInput"), exports);
var accessibilityStyles_1 = require("./accessibility/Styles/accessibilityStyles");
Object.defineProperty(exports, "screenReaderContainerStyles", { enumerable: true, get: function () { return accessibilityStyles_1.screenReaderContainerStyles; } });
__exportStar(require("./commonPropTypes"), exports);
