// @fluentui/react-northstar/dist/dts/src/utils/renderComponent.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.renderComponent = void 0;
var react_bindings_1 = require("@fluentui/react-bindings");
var react_northstar_styles_renderer_1 = require("@fluentui/react-northstar-styles-renderer");
var styles_1 = require("@fluentui/styles");
var _ = require("lodash");
var providerMissingHandler_1 = require("./providerMissingHandler");
var renderComponent = function (config, context) {
    var className = config.className, displayName = config.displayName, handledProps = config.handledProps, props = config.props, state = config.state, actionHandlers = config.actionHandlers, render = config.render, _a = config.saveDebug, saveDebug = _a === void 0 ? function () { } : _a;
    if (_.isEmpty(context)) {
        providerMissingHandler_1.logProviderMissingWarning();
    }
    var _b = react_bindings_1.deprecated_getTelemetry(displayName, context.telemetry, config.isFirstRenderRef), setStart = _b.setStart, setEnd = _b.setEnd;
    var rtl = context.rtl || false;
    setStart();
    var ElementType = react_bindings_1.getElementType(props);
    var unhandledProps = react_bindings_1.getUnhandledProps(handledProps, props);
    var stateAndProps = __assign(__assign({}, state), props);
    var accessibility = react_bindings_1.unstable_getAccessibility(displayName, props.accessibility, stateAndProps, rtl, actionHandlers);
    var _c = react_bindings_1.unstable_getStyles({
        allDisplayNames: [displayName],
        className: className,
        disableAnimations: context.disableAnimations || false,
        primaryDisplayName: displayName,
        componentProps: stateAndProps,
        inlineStylesProps: stateAndProps,
        renderer: context.renderer || react_northstar_styles_renderer_1.noopRenderer,
        rtl: rtl,
        saveDebug: saveDebug,
        theme: context.theme || styles_1.emptyTheme,
        performance: __assign(__assign({}, context.performance), { 
            // we cannot enable caching for class components
            enableStylesCaching: false, enableBooleanVariablesCaching: false }),
        telemetry: context.telemetry,
    }), classes = _c.classes, variables = _c.variables, styles = _c.styles, theme = _c.theme;
    var resolvedConfig = {
        ElementType: ElementType,
        unhandledProps: unhandledProps,
        classes: classes,
        variables: variables,
        styles: styles,
        accessibility: accessibility,
        rtl: rtl,
        theme: theme,
    };
    if (accessibility.focusZone) {
        var originalElementType = resolvedConfig.ElementType;
        resolvedConfig.ElementType = react_bindings_1.FocusZone;
        resolvedConfig.unhandledProps = __assign(__assign({}, resolvedConfig.unhandledProps), accessibility.focusZone.props);
        resolvedConfig.unhandledProps.as = originalElementType;
        resolvedConfig.unhandledProps.isRtl = resolvedConfig.rtl;
    }
    var element = render(resolvedConfig);
    setEnd();
    return element;
};
exports.renderComponent = renderComponent;
