// @fluentui/react-northstar-fela-renderer/dist/commonjs/felaExpandCssShorthandsPlugin.js
"use strict";

exports.__esModule = true;
exports.felaExpandCssShorthandsPlugin = void 0;

var _inlineStyleExpandShorthand = require("inline-style-expand-shorthand");

// @ts-ignore
// https://jsperf.com/array-indexof-vs-object-key-lookup2/12
var handledCssProps = {
  // 'font', Oops, is not supported by inline-style-expand-shorthand
  padding: true,
  margin: true,
  border: true,
  borderWidth: true,
  borderStyle: true,
  borderColor: true,
  borderTop: true,
  borderRight: true,
  borderBottom: true,
  borderLeft: true,
  borderRadius: true,
  background: true,
  outline: true
};

var felaExpandCssShorthandsPlugin = function felaExpandCssShorthandsPlugin(styles) {
  return Object.keys(styles).reduce(function (acc, cssPropertyName) {
    var _Object$assign4;

    var cssPropertyValue = styles[cssPropertyName];

    if (cssPropertyValue === null || typeof cssPropertyValue === 'undefined') {
      var _Object$assign;

      return Object.assign({}, acc, (_Object$assign = {}, _Object$assign[cssPropertyName] = cssPropertyValue, _Object$assign));
    }

    if (handledCssProps[cssPropertyName]) {
      var expandedProps = (0, _inlineStyleExpandShorthand.expandProperty)(cssPropertyName, cssPropertyValue);

      if (expandedProps) {
        return Object.assign({}, acc, expandedProps);
      }
    }

    if (Array.isArray(cssPropertyValue)) {
      var _Object$assign2;

      return Object.assign({}, acc, (_Object$assign2 = {}, _Object$assign2[cssPropertyName] = cssPropertyValue, _Object$assign2));
    }

    if (typeof cssPropertyValue === 'object') {
      var _Object$assign3;

      return Object.assign({}, acc, (_Object$assign3 = {}, _Object$assign3[cssPropertyName] = felaExpandCssShorthandsPlugin(cssPropertyValue), _Object$assign3));
    }

    return Object.assign({}, acc, (_Object$assign4 = {}, _Object$assign4[cssPropertyName] = cssPropertyValue, _Object$assign4));
  }, {});
};

exports.felaExpandCssShorthandsPlugin = felaExpandCssShorthandsPlugin;
