// @fluentui/react-northstar-fela-renderer/dist/commonjs/felaStylisEnhancer.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.felaStylisEnhancer = void 0;

var _felaUtils = require("fela-utils");

var _stylis = _interopRequireDefault(require("stylis"));

// @ts-ignore
// `stylis@3` is a CJS library, there are known issues with them:
// https://github.com/rollup/rollup/issues/1267#issuecomment-446681320
var Stylis = _stylis.default.default || _stylis.default; // We use Stylis only for vendor prefixing, all other capabilities are disabled

var stylis = new Stylis({
  cascade: false,
  compress: false,
  global: false,
  keyframe: false,
  preserve: false,
  semicolon: false
});

var felaStylisEnhancer = function felaStylisEnhancer(renderer) {
  var existingEmitChange = renderer._emitChange.bind(renderer);

  renderer._emitChange = function (change) {
    if (change.type === _felaUtils.RULE_TYPE) {
      var prefixed = stylis('', change.declaration); // Fela uses objects by references, it's safe to override properties

      change.declaration = prefixed.slice(1, -1);
    }

    existingEmitChange(change);
  };

  return renderer;
};

exports.felaStylisEnhancer = felaStylisEnhancer;
