// @fluentui/react-northstar/dist/commonjs/themes/teams/components/Skeleton/skeletonVariables.js
"use strict";

exports.__esModule = true;
exports.skeletonVariables = void 0;

var _utils = require("../../../../utils");

var skeletonVariables = function skeletonVariables(siteVariables) {
  return {
    lineBackground: siteVariables.colorScheme.default.foreground6,
    shapeBackground: siteVariables.colorScheme.default.foreground6,
    animationBackground: siteVariables.colorScheme.default.foreground6,
    animationBackgroundSecondary: siteVariables.colorScheme.default.foreground6,
    lineMargin: "0 0 " + (0, _utils.pxToRem)(2) + " 0",
    shapeMargin: "0 0 " + (0, _utils.pxToRem)(2) + " 0",
    // Skeleton Button
    buttonHeight: (0, _utils.pxToRem)(32),
    buttonWidth: (0, _utils.pxToRem)(96),
    buttonSmallHeight: (0, _utils.pxToRem)(24),
    buttonSmallWidth: (0, _utils.pxToRem)(72),
    buttonBackground: siteVariables.colorScheme.default.foreground6,
    buttonCircularBorderRadius: (0, _utils.pxToRem)(999),
    // Skeleton Text
    textBackground: siteVariables.colorScheme.default.foreground6,
    textWidth: '100%',
    textSmallerHeight: (0, _utils.pxToRem)(14),
    textSmallHeight: (0, _utils.pxToRem)(16),
    textMediumHeight: (0, _utils.pxToRem)(19),
    textLargeHeight: (0, _utils.pxToRem)(24),
    textLargerHeight: (0, _utils.pxToRem)(32),
    // Skeleton Input
    inputHeight: (0, _utils.pxToRem)(32),
    inputWidth: (0, _utils.pxToRem)(154),
    inputBackground: siteVariables.colorScheme.default.foreground6,
    // Skeleton Avatar
    avatarBackground: siteVariables.colorScheme.default.foreground6,
    avatarSmallest: (0, _utils.pxToRem)(20),
    avatarSmaller: (0, _utils.pxToRem)(24),
    avatarSmall: (0, _utils.pxToRem)(28),
    avatarMedium: (0, _utils.pxToRem)(32),
    avatarLarge: (0, _utils.pxToRem)(44),
    avatarLarger: (0, _utils.pxToRem)(64),
    avatarLargest: (0, _utils.pxToRem)(96)
  };
};

exports.skeletonVariables = skeletonVariables;
