// @fluentui/react-northstar/dist/es/themes/teams/components/Carousel/carouselVariables.js
import { pxToRem } from '../../../../utils';
export var carouselVariables = function carouselVariables(siteVars) {
  return {
    width: 300,
    height: 300,
    focusOuterBorderColor: siteVars.focusOuterBorderColor,
    focusOuterBorderRadius: siteVars.borderRadiusMedium,
    paddleColor: siteVars.colorScheme.default.foreground4,
    paddleColorDisabled: siteVars.colorScheme.brand.foregroundDisabled,
    paddleBackgroundColor: siteVars.colorScheme.onyx.background3,
    paddleBackgroundColorActive: siteVars.colorScheme.onyx.backgroundPressed,
    paddleBackgroundColorHover: siteVars.colorScheme.onyx.backgroundHover,
    paddleBackgroundColorDisabled: siteVars.colorScheme.default.backgroundDisabled,
    paddleBorderColor: 'transparent',
    paddleBorderRadius: siteVars.borderRadiusMedium,
    paddleHeight: pxToRem(32),
    paddleIndicatorSize: pxToRem(16),
    focusBorderRadius: siteVars.borderRadiusMedium,
    focusBorderWidth: siteVars.borderWidth,
    focusInnerBorderColor: siteVars.focusInnerBorderColor,
    focusBorderZIndex: siteVars.zIndexes.foreground
  };
};
