// @fluentui/react-northstar/dist/es/components/Datepicker/DatepickerCalendarGridRow.js
import { compose } from '@fluentui/react-bindings';
import { commonPropTypes } from '../../utils';
import { Box } from '../Box/Box';
export var datepickerCalendarGridRowClassName = 'ui-datepicker__calendargridrow';
/**
 * A DatepickerCalendarGridRow is used to display the calendar grid row.
 */

export var DatepickerCalendarGridRow = /*#__PURE__*/function () {
  var DatepickerCalendarGridRow = compose(Box, {
    className: datepickerCalendarGridRowClassName,
    displayName: 'DatepickerCalendarGridRow',
    handledProps: ['isRowSelectionActive'],
    overrideStyles: true,
    mapPropsToStylesProps: function mapPropsToStylesProps(_ref) {
      var isRowSelectionActive = _ref.isRowSelectionActive;
      return {
        isRowSelectionActive: isRowSelectionActive
      };
    },
    shorthandConfig: {
      mappedProp: 'content'
    }
  });
  DatepickerCalendarGridRow.defaultProps = {
    as: 'tr'
  };
  DatepickerCalendarGridRow.propTypes = commonPropTypes.createCommon();
  return DatepickerCalendarGridRow;
}();
