// @fluentui/react-northstar/dist/dts/test/specs/components/Attachment/Attachment-test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var commonTests_1 = require("test/specs/commonTests");
var utils_1 = require("test/utils");
var accessibility_1 = require("@fluentui/accessibility");
var Attachment_1 = require("src/components/Attachment/Attachment");
var AttachmentAction_1 = require("src/components/Attachment/AttachmentAction");
var AttachmentDescription_1 = require("src/components/Attachment/AttachmentDescription");
var AttachmentHeader_1 = require("src/components/Attachment/AttachmentHeader");
var AttachmentIcon_1 = require("src/components/Attachment/AttachmentIcon");
var attachmentImplementsShorthandProp = commonTests_1.implementsShorthandProp(Attachment_1.Attachment);
var getAttachment = function (onClickAttachment, onClickButton) {
    return utils_1.mountWithProvider(React.createElement(Attachment_1.Attachment, { actionable: true, action: {
            icon: 'more',
            onClick: onClickButton,
        }, onClick: onClickAttachment }));
};
describe('Attachment', function () {
    commonTests_1.isConformant(Attachment_1.Attachment, { testPath: __filename, constructorName: 'Attachment' });
    attachmentImplementsShorthandProp('action', AttachmentAction_1.AttachmentAction);
    attachmentImplementsShorthandProp('description', AttachmentDescription_1.AttachmentDescription);
    attachmentImplementsShorthandProp('icon', AttachmentIcon_1.AttachmentIcon);
    attachmentImplementsShorthandProp('header', AttachmentHeader_1.AttachmentHeader);
    describe('accessibility', function () {
        commonTests_1.handlesAccessibility(Attachment_1.Attachment, {
            defaultRootRole: undefined,
        });
        test('handleClick is executed when Enter is pressed on attachment element', function () {
            var onClickAttachment = jest.fn();
            var onClickButton = jest.fn();
            var attachment = getAttachment(onClickAttachment, onClickButton);
            attachment.find("." + Attachment_1.attachmentClassName).simulate('keydown', { keyCode: accessibility_1.keyboardKey.Enter });
            expect(onClickAttachment).toHaveBeenCalledTimes(1);
        });
        test('handleClick is Not executed on attachment element, when Enter is pressed on more option button', function () {
            var onClickAttachment = jest.fn();
            var onClickButton = jest.fn();
            var attachment = getAttachment(onClickAttachment, onClickButton);
            utils_1.findIntrinsicElement(attachment, "." + AttachmentAction_1.attachmentActionClassName).simulate('keydown', {
                keyCode: accessibility_1.keyboardKey.Enter,
            });
            expect(onClickAttachment).not.toBeCalled();
        });
    });
});
