// @fluentui/accessibility/dist/dts/src/behaviors/Radio/radioGroupBehavior.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.radioGroupBehavior = void 0;
var keyboard_key_1 = require("../../keyboard-key");
/**
 * @description
 * Implements ARIA Radio Group design pattern.
 * @specification
 *  Adds role='radiogroup'. This allows screen readers to handle the component as a radio group.
 *  Triggers 'nextItem' action with 'ArrowDown' or 'ArrowRight' on 'root'.
 *  Triggers 'prevItem' action with 'ArrowUp' or 'ArrowLeft' on 'root'.
 */
var radioGroupBehavior = function () { return ({
    attributes: {
        root: {
            role: 'radiogroup',
        },
    },
    keyActions: {
        root: {
            nextItem: {
                keyCombinations: [{ keyCode: keyboard_key_1.keyboardKey.ArrowDown }, { keyCode: keyboard_key_1.keyboardKey.ArrowRight }],
            },
            prevItem: {
                keyCombinations: [{ keyCode: keyboard_key_1.keyboardKey.ArrowUp }, { keyCode: keyboard_key_1.keyboardKey.ArrowLeft }],
            },
        },
    },
}); };
exports.radioGroupBehavior = radioGroupBehavior;
