// @fluentui/react-northstar/dist/dts/src/themes/teams/components/Reaction/reactionStyles.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.reactionStyles = void 0;
var Reaction_1 = require("../../../../components/Reaction/Reaction");
var utils_1 = require("../../../../utils");
var contentClassNameSelector = "& ." + Reaction_1.reactionSlotClassNames.content;
exports.reactionStyles = {
    root: function (_a) {
        var _b;
        var p = _a.props, v = _a.variables, siteVariables = _a.theme.siteVariables;
        return ({
            cursor: 'pointer',
            background: 'transparent',
            border: utils_1.pxToRem(0),
            padding: utils_1.pxToRem(0),
            display: 'inline-flex',
            justifyContent: 'center',
            alignItems: 'center',
            color: v.meReacting ? v.meReactingColor : v.otherReactingColor,
            ':hover': (_b = {
                    color: v.meReacting ? v.meReactingColorHover : v.otherReactingColorHover
                },
                _b[contentClassNameSelector] = {
                    fontWeight: v.fontWeightHover,
                },
                _b),
            position: 'relative',
            ':focus': {
                outline: 'none',
            },
            ':focus-visible': {
                ':after': {
                    content: '""',
                    position: 'absolute',
                    top: "-" + utils_1.pxToRem(2),
                    right: "-" + utils_1.pxToRem(2),
                    bottom: "-" + utils_1.pxToRem(2),
                    left: "-" + utils_1.pxToRem(2),
                    borderWidth: '1px',
                    borderStyle: 'solid',
                    borderColor: v.borderColorFocus,
                    borderRadius: siteVariables.borderRadiusMedium,
                    boxShadow: "0px 0px 0px 1px " + v.boxShadowColor + " inset",
                },
            },
        });
    },
    icon: function (_a) {
        var p = _a.props;
        return ({
            display: 'inline-flex',
            alignItems: 'center',
            justifyContent: 'center',
            marginRight: p.hasContent ? utils_1.pxToRem(4) : utils_1.pxToRem(0),
        });
    },
    content: function (_a) {
        var v = _a.variables;
        return ({
            fontSize: v.contentFontSize,
        });
    },
};
