// @fluentui/react-northstar/dist/dts/src/themes/teams/colors.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getColorScheme = exports.getColorSchemeKey = exports.isValidColor = exports.availableColors = exports.colorScheme = exports.createColorScheme = exports.colors = exports.transparentColors = exports.primitiveColors = exports.naturalColors = exports.contextualColors = void 0;
exports.contextualColors = {
    brand: {
        50: '#e8ebfa',
        100: '#dcdffa',
        200: '#c5cbfa',
        300: '#acb3fa',
        400: '#9399f5',
        450: '#9499f5',
        500: '#7a80eb',
        600: '#5b5fc7',
        700: '#444691',
        800: '#3d3e78',
        900: '#383966',
        1000: '#2f2f4a',
    },
};
exports.naturalColors = {
    grey: {
        0: '#FFFFFF',
        25: '#FCFCFB',
        50: '#FAF9F8',
        100: '#F3F2F1',
        150: '#EDEBE9',
        200: '#E1DFDD',
        250: '#C8C6C4',
        300: '#B3B0AD',
        340: '#999',
        350: '#979593',
        400: '#8A8886',
        430: '#707070',
        440: '#666666',
        450: '#605E5C',
        500: '#484644',
        550: '#3B3A39',
        600: '#323131',
        650: '#2D2C2C',
        700: '#292828',
        750: '#252423',
        800: '#201F1F',
        850: '#1B1A1A',
        900: '#11100F',
        1000: '#000000',
    },
    orange: {
        50: '#F9ECEA',
        100: '#EFDBD3',
        200: '#EDC2A7',
        300: '#E97548',
        400: '#CC4A31',
        500: '#BD432C',
        600: '#A33D2A',
        700: '#833122',
        800: '#664134',
        900: '#51332C',
    },
    pink: {
        50: '#FCF2FA',
        100: '#F1DFEE',
        200: '#EC6FAE',
        300: '#DE569A',
        400: '#E959D9',
        500: '#B4009E',
        600: '#943670',
        700: undefined,
        800: '#3E2D3B',
        900: '#1F191D',
    },
    red: {
        50: '#FCF4F6',
        100: '#F3D6D8',
        200: '#F75E75',
        300: '#E73550',
        400: '#C4314B',
        500: '#A72037',
        600: '#8E192E',
        700: '#4F222B',
        800: '#3E1F25',
        900: '#1E040A',
    },
    green: {
        50: '#E7F2DA',
        100: '#BDDA9B',
        200: '#92C353',
        300: '#6BB700',
        400: '#13A40E',
        500: undefined,
        600: '#237B4B',
        700: undefined,
        800: '#0D2E0D',
        900: '#032003',
    },
    yellow: {
        50: undefined,
        100: '#FBF6D9',
        200: '#F2E384',
        300: '#F9EC02',
        400: '#F8D22A',
        500: '#FFB900',
        600: '#FFAA44',
        700: '#835C00',
        800: '#463100',
        900: '#261A00',
    },
};
var contextualAndNaturalColors = __assign(__assign({}, exports.contextualColors), exports.naturalColors);
exports.primitiveColors = {
    black: '#000',
    white: '#fff',
};
exports.transparentColors = {
    silver: {
        100: 'rgba(255,255,255,0.85)',
        200: 'rgba(255,255,255,0.75)',
        300: 'rgba(255,255,255,0.65)',
        400: 'rgba(255,255,255,0.5)',
        500: 'rgba(255,255,255,0.4)',
        600: 'rgba(255,255,255,0.3)',
        700: 'rgba(255,255,255,0.2)',
        800: 'rgba(255,255,255,0.1)',
        900: 'rgba(255,255,255,0.05)',
    },
    ruby: {
        100: undefined,
        200: undefined,
        300: undefined,
        400: undefined,
        500: 'rgba(196,49,75,0.9)',
        600: 'rgba(167,32,55,0.9)',
        700: 'rgba(142,25,46,0.9)',
        800: undefined,
        900: undefined,
    },
    onyx: {
        100: 'rgba(59,58,57,0.9)',
        200: 'rgba(45,44,44,0.4)',
        300: 'rgba(37,36,35,0.2)',
        400: 'rgba(37,36,35,0.65)',
        500: 'rgba(41,40,40,0.9)',
        600: 'rgba(0,0,0,0.05)',
        700: 'rgba(0,0,0,0.5)',
        800: 'rgba(27,26,26,0.9)',
        900: 'rgba(0,0,0,0.8)',
    },
    amethyst: {
        100: undefined,
        200: undefined,
        300: undefined,
        400: 'rgba(98,100,167,0.75)',
        500: 'rgba(51,52,74,0.5)',
        600: 'rgba(70,71,117,0.4)',
        700: 'rgba(98,100,167,0.15)',
        800: undefined,
        900: undefined,
    },
};
exports.colors = __assign(__assign(__assign({}, contextualAndNaturalColors), exports.primitiveColors), exports.transparentColors);
var createColorScheme = function (customValues) {
    if (customValues === void 0) { customValues = {}; }
    return __assign({ foreground: undefined, background: undefined, border: undefined, shadow: undefined, foregroundHover: undefined, backgroundHover: undefined, borderHover: undefined, shadowHover: undefined, foregroundActive: undefined, backgroundActive: undefined, borderActive: undefined, shadowActive: undefined, foregroundFocus: undefined, backgroundFocus: undefined, borderFocus: undefined, shadowFocus: undefined, foregroundPressed: undefined, backgroundPressed: undefined, borderPressed: undefined, shadowPressed: undefined, foregroundDisabled: undefined, backgroundDisabled: undefined, borderDisabled: undefined, shadowDisabled: undefined }, customValues);
};
exports.createColorScheme = createColorScheme;
exports.colorScheme = {
    default: exports.createColorScheme({
        foreground: exports.colors.grey[750],
        foreground1: exports.colors.grey[500],
        foreground2: exports.colors.grey[450],
        foreground3: exports.colors.white,
        foreground4: exports.colors.white,
        foreground5: exports.colors.grey[100],
        foreground6: exports.colors.grey[200],
        foreground7: exports.colors.grey[750],
        foreground8: exports.colors.grey[750],
        foreground9: exports.colors.grey[430],
        background: exports.colors.white,
        background1: exports.colors.grey[50],
        background2: exports.colors.grey[100],
        background3: exports.colors.grey[150],
        background4: exports.colors.grey[100],
        background5: exports.colors.grey[350],
        background6: exports.colors.grey[550],
        border: exports.colors.grey[200],
        border1: exports.colors.grey[150],
        border2: exports.colors.grey[200],
        border3: exports.colors.grey[150],
        borderTransparent: 'transparent',
        borderTransparentActive: 'transparent',
        shadow: exports.colors.black,
        shadowHover: exports.colors.black,
        foregroundHover: exports.colors.grey[750],
        foregroundHover1: exports.colors.white,
        foregroundHover2: exports.colors.white,
        backgroundHover: exports.colors.grey[100],
        backgroundHover1: exports.colors.grey[150],
        backgroundHover2: 'transparent',
        backgroundHover3: exports.colors.grey[150],
        backgroundHover4: exports.colors.grey[50],
        borderHover: exports.colors.grey[250],
        foregroundPressed: exports.colors.grey[750],
        backgroundPressed: exports.colors.grey[200],
        backgroundPressed3: exports.colors.grey[150],
        borderPressed: exports.colors.grey[250],
        foregroundActive: exports.colors.grey[750],
        foregroundActive1: exports.colors.white,
        backgroundActive: exports.colors.grey[100],
        backgroundActive1: exports.colors.grey[150],
        backgroundActive2: exports.colors.grey[150],
        borderActive: exports.colors.grey[200],
        borderActive1: exports.colors.grey[150],
        borderActive2: exports.colors.grey[200],
        borderActive3: exports.colors.grey[150],
        borderActive4: exports.colors.grey[400],
        foregroundFocus: exports.colors.grey[750],
        foregroundFocus1: exports.colors.grey[500],
        foregroundFocus2: exports.colors.grey[450],
        foregroundFocus3: exports.colors.white,
        backgroundFocus: exports.colors.white,
        backgroundFocus1: exports.colors.grey[50],
        backgroundFocus2: exports.colors.grey[100],
        backgroundFocus3: exports.colors.grey[150],
        borderFocusWithin: exports.colors.white,
        borderFocus: exports.colors.black,
        foregroundDisabled1: exports.colors.grey[250],
        foregroundDisabled: exports.colors.grey[250],
        backgroundDisabled: exports.colors.grey[150],
        backgroundDisabled1: exports.colors.grey[150],
        backgroundDisabled2: exports.colors.grey[50],
        backgroundDisabled3: exports.colors.grey[50],
        borderDisabled: exports.colors.grey[150],
    }),
    brand: exports.createColorScheme({
        foreground: exports.colors.brand[600],
        foreground1: exports.colors.brand[600],
        foreground2: exports.colors.brand[700],
        foreground3: exports.colors.brand[200],
        foreground4: exports.colors.white,
        foreground5: exports.colors.white,
        background: exports.colors.brand[600],
        background1: exports.colors.brand[100],
        background2: exports.colors.brand[900],
        background3: exports.colors.brand[1000],
        background4: exports.colors.brand[800],
        background5: exports.colors.brand[100],
        background6: exports.colors.brand[600],
        border: exports.colors.grey[200],
        border1: exports.colors.brand[200],
        border2: exports.colors.brand[300],
        shadow: exports.colors.black,
        shadowHover: exports.colors.black,
        foregroundHover: exports.colors.brand[600],
        foregroundHover1: exports.colors.white,
        foregroundHover2: exports.colors.brand[200],
        borderHover: exports.colors.brand[300],
        backgroundHover: exports.colors.brand[700],
        backgroundHover1: exports.colors.brand[50],
        backgroundHover2: exports.colors.brand[100],
        backgroundHover3: exports.colors.brand[100],
        foregroundPressed: exports.colors.brand[800],
        foregroundPressed1: exports.colors.white,
        backgroundPressed: exports.colors.brand[800],
        backgroundPressed1: exports.colors.brand[100],
        backgroundPressed2: exports.colors.brand[100],
        borderPressed: exports.colors.brand[300],
        foregroundActive: exports.colors.brand[600],
        foregroundActive1: exports.colors.brand[600],
        foregroundActive2: exports.colors.brand[200],
        backgroundActive: exports.colors.brand[600],
        backgroundActive1: exports.colors.brand[600],
        borderActive: exports.colors.grey[200],
        borderActive1: exports.colors.brand[200],
        borderActive2: exports.colors.brand[300],
        foregroundFocus: exports.colors.brand[600],
        foregroundFocus1: exports.colors.brand[600],
        foregroundFocus2: exports.colors.brand[700],
        foregroundFocus3: exports.colors.brand[200],
        foregroundFocus4: exports.colors.white,
        backgroundFocus: exports.colors.brand[600],
        backgroundFocus1: exports.colors.brand[100],
        backgroundFocus2: exports.colors.brand[900],
        backgroundFocus3: exports.colors.brand[1000],
        borderFocus: exports.colors.black,
        borderFocusWithin: exports.colors.white,
        borderFocus1: exports.colors.brand[600],
        foregroundDisabled: exports.colors.grey[250],
        foregroundDisabled1: exports.colors.grey[250],
        backgroundDisabled: exports.colors.grey[150],
        backgroundDisabled1: exports.colors.grey[150],
        borderDisabled: exports.colors.grey[150],
    }),
    black: {
        foreground: exports.colors.black,
        foreground1: exports.colors.white,
        background: exports.colors.white,
        background1: exports.colors.grey[750],
        border: exports.colors.black,
        shadow: exports.colors.black,
        foregroundHover: exports.colors.white,
        backgroundHover: exports.colors.black,
        borderHover: exports.colors.black,
        shadowHover: exports.colors.black,
        foregroundActive: exports.colors.white,
        backgroundActive: exports.colors.black,
        borderActive: exports.colors.black,
        shadowActive: exports.colors.black,
        foregroundFocus: exports.colors.white,
        backgroundFocus: exports.colors.black,
        borderFocus: exports.colors.black,
        shadowFocus: exports.colors.black,
        foregroundPressed: exports.colors.white,
        backgroundPressed: exports.colors.black,
        borderPressed: exports.colors.black,
        shadowPressed: exports.colors.black,
        foregroundDisabled: exports.colors.white,
        backgroundDisabled: exports.colors.black,
        borderDisabled: exports.colors.black,
        shadowDisabled: exports.colors.black,
    },
    white: {
        foreground: exports.colors.white,
        foreground1: exports.colors.grey[750],
        background: exports.colors.black,
        background1: exports.colors.white,
        border: exports.colors.white,
        shadow: exports.colors.white,
        foregroundHover: exports.colors.black,
        backgroundHover: exports.colors.white,
        borderHover: exports.colors.white,
        shadowHover: exports.colors.white,
        foregroundActive: exports.colors.black,
        backgroundActive: exports.colors.white,
        borderActive: exports.colors.white,
        shadowActive: exports.colors.white,
        foregroundFocus: exports.colors.black,
        backgroundFocus: exports.colors.white,
        borderFocus: exports.colors.white,
        shadowFocus: exports.colors.white,
        foregroundPressed: exports.colors.black,
        backgroundPressed: exports.colors.white,
        borderPressed: exports.colors.white,
        shadowPressed: exports.colors.white,
        foregroundDisabled: exports.colors.black,
        backgroundDisabled: exports.colors.white,
        borderDisabled: exports.colors.white,
        shadowDisabled: exports.colors.white,
    },
    green: exports.createColorScheme({
        foreground: exports.colors.green[600],
        foreground1: exports.colors.white,
        foreground2: exports.colors.green[400],
        foreground3: exports.colors.white,
        background: exports.colors.green[300],
        background1: exports.colors.green[600],
        background2: exports.colors.green[50],
        border: exports.colors.green[100],
        border1: exports.colors.green[600],
    }),
    orange: exports.createColorScheme({
        foreground: exports.colors.orange[400],
        foreground1: exports.colors.orange[300],
        foreground2: exports.colors.white,
        background: exports.colors.orange[400],
        background1: exports.colors.orange[400],
        border: exports.colors.orange[200],
    }),
    pink: exports.createColorScheme({
        foreground: exports.colors.pink[600],
        foreground1: exports.colors.pink[500],
        background: exports.colors.pink[50],
        border: exports.colors.pink[100],
    }),
    red: exports.createColorScheme({
        foreground: exports.colors.red[400],
        foreground1: exports.colors.white,
        foreground2: exports.colors.white,
        background: exports.colors.red[400],
        background1: exports.colors.red[50],
        background2: exports.colors.ruby[500],
        background3: exports.colors.red[400],
        border: exports.colors.red[100],
        border1: exports.colors.red[300],
        foregroundHover: exports.colors.white,
        backgroundHover: exports.colors.ruby[600],
        backgroundHover1: exports.colors.red[400],
        foregroundPressed: exports.colors.white,
        backgroundPressed: exports.colors.ruby[700],
    }),
    yellow: exports.createColorScheme({
        foreground: exports.colors.yellow[300],
        foreground1: exports.colors.grey[800],
        foreground2: exports.colors.white,
        foreground3: exports.colors.grey[750],
        foreground4: exports.colors.yellow[700],
        background: exports.colors.yellow[600],
        background1: exports.colors.yellow[100],
        background2: exports.colors.yellow[500],
        background3: exports.colors.yellow[100],
        border: exports.colors.yellow[200],
        border1: exports.colors.yellow[700],
    }),
    silver: exports.createColorScheme({
        foreground: exports.colors.white,
        foreground1: exports.colors.silver[200],
        foregroundHover: exports.colors.white,
        foregroundPressed: exports.colors.white,
        border: exports.colors.silver[600],
        background: 'transparent',
        backgroundHover: exports.colors.silver[800],
        borderHover: exports.colors.silver[600],
        backgroundPressed: exports.colors.silver[700],
        borderPressed: exports.colors.silver[600],
        foregroundDisabled: exports.colors.silver[600],
        backgroundDisabled: exports.colors.silver[900],
    }),
    onyx: exports.createColorScheme({
        background: exports.colors.onyx[500],
        background1: exports.colors.onyx[100],
        background2: exports.colors.onyx[500],
        background3: exports.colors.onyx[400],
        backgroundHover: exports.colors.onyx[500],
        backgroundPressed: exports.colors.onyx[800],
        border: exports.colors.onyx[800],
        border1: 'transparent',
        border2: exports.colors.onyx[300],
    }),
    amethyst: exports.createColorScheme({
        background: exports.colors.amethyst[600],
        backgroundHover: exports.colors.amethyst[700],
        backgroundHover1: exports.colors.amethyst[500],
        backgroundActive: exports.colors.amethyst[700],
    }),
};
exports.colorScheme.grey = exports.colorScheme.default;
exports.availableColors = [
    'default',
    'black',
    'white',
    'brand',
    'grey',
    'red',
    'yellow',
    'green',
    'pink',
    'orange',
];
var isValidColor = function (color) {
    return color && exports.availableColors.indexOf(color) >= 0;
};
exports.isValidColor = isValidColor;
var getColorSchemeKey = function (color, primary) {
    return color && exports.isValidColor(color) ? color : primary ? 'brand' : 'default';
};
exports.getColorSchemeKey = getColorSchemeKey;
var getColorScheme = function (colorScheme, color, primary) {
    return colorScheme[exports.getColorSchemeKey(color, primary)];
};
exports.getColorScheme = getColorScheme;
