// @fluentui/react-northstar/dist/dts/src/themes/teams/components/Dialog/dialogVariables.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.dialogVariables = void 0;
var utils_1 = require("../../../../utils");
var dialogVariables = function (siteVariables) { return ({
    border: 'none',
    rootBackground: siteVariables.colors.white,
    rootBorderRadius: siteVariables.borderRadiusXLarge,
    rootWidth: '600px',
    rootPadding: utils_1.pxToRem(27) + " " + utils_1.pxToRem(32) + " " + utils_1.pxToRem(32) + " " + utils_1.pxToRem(32),
    rootPaddingZoom: "" + utils_1.pxToRem(12),
    contentMargin: "0 0 " + utils_1.pxToRem(20) + " 0",
    boxShadow: siteVariables.shadowLevel4,
    foregroundColor: siteVariables.colors.grey[900],
    headerFontSize: siteVariables.fontSizes.large,
    headerFontWeight: siteVariables.fontWeightBold,
    headerMargin: "0 0 " + utils_1.pxToRem(8) + " 0",
    overlayBackground: 'rgba(37, 36, 36, .75)',
    overlayZIndex: siteVariables.zIndexes.overlay,
    headerActionMargin: utils_1.pxToRem(-3) + " " + utils_1.pxToRem(-8) + " 0 0",
    // in case of customized footer this var allows to align styles
    footerActionsBreakpoint: '400px',
}); };
exports.dialogVariables = dialogVariables;
