// @fluentui/react-northstar/dist/dts/src/themes/teams/components/TextArea/textAreaStyles.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.textAreaStyles = void 0;
var utils_1 = require("../../../../utils");
exports.textAreaStyles = {
    root: function (_a) {
        var p = _a.props, v = _a.variables;
        return (__assign(__assign(__assign(__assign(__assign(__assign({ margin: v.margin, height: v.height, backgroundColor: v.backgroundColor }, (p.inverted && {
            backgroundColor: v.invertedBackgroundColor,
        })), { color: v.fontColor, borderColor: v.borderColor, borderRadius: v.borderRadius, borderStyle: 'solid', borderWidth: v.borderWidth, outline: 0, padding: v.padding, resize: p.resize || 'none' }), (p.fluid && {
            width: '100%',
        })), (p.disabled && {
            pointerEvents: 'none',
            color: v.disabledColor,
            boxShadow: 'none',
        })), (p.error && { border: utils_1.pxToRem(1) + " solid " + v.borderColorError })), { '::placeholder': __assign({ color: v.placeholderColor, opacity: 1 }, (p.disabled && {
                color: v.disabledColor,
            })), ':focus': {
                borderColor: v.borderColorFocus,
            } }));
    },
};
