// @fluentui/react-northstar/dist/dts/src/themes/teams/components/TextArea/textAreaVariables.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.textAreaVariables = void 0;
var utils_1 = require("../../../../utils");
var textAreaVariables = function (siteVars) { return ({
    margin: '0',
    padding: utils_1.pxToRem(7) + " " + utils_1.pxToRem(12),
    borderColor: 'transparent',
    borderRadius: siteVars.borderRadiusMedium + " " + siteVars.borderRadiusMedium + " " + siteVars.borderRadiusSmall + " " + siteVars.borderRadiusSmall,
    borderWidth: "0 0 " + utils_1.pxToRem(2) + " 0",
    backgroundColor: siteVars.colorScheme.default.background2,
    invertedBackgroundColor: siteVars.colorScheme.default.background,
    placeholderColor: siteVars.colorScheme.default.foreground1,
    disabledColor: siteVars.colorScheme.brand.foregroundDisabled,
    fontColor: siteVars.colorScheme.default.foreground,
    fontSize: siteVars.fontSizes.medium,
    borderColorFocus: "transparent transparent " + siteVars.colorScheme.brand.borderFocus1 + " transparent",
    height: 'auto',
    borderColorError: siteVars.colorScheme.red.background,
}); };
exports.textAreaVariables = textAreaVariables;
