// @fluentui/react-northstar/dist/dts/src/themes/teams/components/Menu/menuItemContentStyles.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.menuItemContentStyles = void 0;
var utils_1 = require("../../../../utils");
exports.menuItemContentStyles = {
    root: function (_a) {
        var p = _a.props;
        var widthAdjust = (p.hasIcon ? 26 : 0) + (p.hasMenu ? 16 : 0);
        return __assign({ whiteSpace: 'normal', lineHeight: 1.5, marginTop: utils_1.pxToRem(-4), marginBottom: utils_1.pxToRem(-4), display: 'inline-block', userSelect: 'none' }, ((p.inSubmenu || p.vertical) && {
            width: 'max-content',
            minWidth: utils_1.pxToRem(46 - widthAdjust),
            maxWidth: utils_1.pxToRem(262 - widthAdjust),
            marginRight: utils_1.pxToRem(16),
        }));
    },
};
