// @fluentui/react-northstar/dist/dts/src/themes/teams/components/Menu/menuItemIndicatorStyles.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.menuItemIndicatorStyles = void 0;
var utils_1 = require("../../../../utils");
var submenuIndicatorDirection_1 = require("./submenuIndicatorDirection");
exports.menuItemIndicatorStyles = {
    root: function (_a) {
        var p = _a.props, v = _a.variables;
        return __assign(__assign(__assign(__assign(__assign(__assign({ position: 'relative', float: 'right', left: utils_1.pxToRem(12), userSelect: 'none', marginRight: utils_1.pxToRem(4) }, (p.inSubmenu && {
            position: 'absolute',
            top: utils_1.pxToRem(6),
            right: utils_1.pxToRem(2),
            left: 'unset',
        })), { content: '" "', display: 'block', overflow: 'hidden', height: utils_1.pxToRem(16), width: utils_1.pxToRem(16), backgroundSize: utils_1.pxToRem(16), color: v.indicatorColor }), (p.active && __assign({ color: v.activeIndicatorColor }, (p.primary && __assign({ color: v.activePrimaryIndicatorColor }, (p.vertical && {
            color: v.activePrimaryVerticalIndicatorColor,
        })))))), (p.underlined && { color: v.indicatorColor })), (p.iconOnly && { color: v.indicatorColor })), submenuIndicatorDirection_1.submenuIndicatorDirection(p.vertical));
    },
};
