// @fluentui/react-northstar/dist/dts/src/themes/teams/components/Menu/menuVariables.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.menuVariables = exports.menuColorAreas = void 0;
var utils_1 = require("../../../../utils");
var colorUtils_1 = require("../../../colorUtils");
exports.menuColorAreas = utils_1.stringLiteralsArray('border', 'borderActive', 'foregroundActive', 'foregroundFocus', 'foregroundHover', 'backgroundHover', 'backgroundActive', 'backgroundFocus', 'foregroundDisabled');
var menuVariables = function (siteVars) {
    return {
        colorScheme: colorUtils_1.pickValuesFromColorScheme(colorUtils_1.extendColorScheme(siteVars.colorScheme, {
            default: {
                borderActive: siteVars.colorScheme.default.border2,
                backgroundActive: siteVars.colorScheme.default.backgroundActive2,
                backgroundFocus: siteVars.colorScheme.default.backgroundFocus1,
                foregroundDisabled: siteVars.colorScheme.default.foregroundDisabled1,
            },
            brand: {
                foregroundHover: siteVars.colors.white,
                backgroundHover: siteVars.colors.brand[300],
                foregroundActive: siteVars.colors.white,
                borderActive: siteVars.colors.brand[600],
                backgroundActive: siteVars.colors.brand[500],
                foregroundFocus: siteVars.colors.white,
                backgroundFocus: siteVars.colors.brand[300],
                foregroundDisabled: siteVars.colorScheme.brand.foregroundDisabled1,
            },
        }), exports.menuColorAreas),
        color: siteVars.colors.grey[500],
        colorActive: siteVars.colors.black,
        colorFocus: undefined,
        colorHover: 'inherit',
        colorDisabled: undefined,
        borderColor: undefined,
        borderColorHover: undefined,
        borderColorActive: undefined,
        borderColorFocus: siteVars.colors.white,
        outlineColorFocus: siteVars.colors.black,
        backgroundColor: undefined,
        backgroundColorFocus: undefined,
        backgroundColorHover: undefined,
        backgroundColorActive: undefined,
        iconSize: utils_1.pxToRem(16),
        iconOnlyBorderRadius: siteVars.borderRadiusMedium,
        iconOnlyColorActive: siteVars.colors.brand[600],
        iconOnlyColorFocus: siteVars.colors.brand[600],
        iconOnlyWrapperBackgroundColorFocus: undefined,
        activeIconOnlyWrapperBackgroundColor: undefined,
        iconOnlyBackgroundColorHover: undefined,
        iconOnlyColorHover: siteVars.colors.brand[600],
        underlinedBorderColor: siteVars.colors.grey[200],
        lineHeightBase: siteVars.lineHeightMedium,
        horizontalPadding: utils_1.pxToRem(14) + " " + utils_1.pxToRem(18),
        verticalBackgroundColor: siteVars.colorScheme.default.background,
        verticalItemPadding: utils_1.pxToRem(7) + " " + utils_1.pxToRem(12),
        verticalBoxShadow: siteVars.shadow16,
        verticalDividerMargin: utils_1.pxToRem(8) + " 0",
        verticalItemBorderWidth: utils_1.pxToRem(2),
        verticalItemBorderColor: 'transparent',
        verticalPointingBorderColor: siteVars.colorScheme.brand.borderActive,
        verticalBackgroundColorFocus: undefined,
        verticalBorderColor: siteVars.colorScheme.default.borderTransparent,
        activeUnderlinedColor: undefined,
        activeUnderlinedPrimaryColor: siteVars.colors.brand[600],
        activeUnderlinedBorderBottomColor: siteVars.colors.black,
        activeUnderlinedWrapperColor: undefined,
        pointingIndicatorBackgroundColor: siteVars.colors.brand[600],
        underlinedBottomBorderWidth: utils_1.pxToRem(2),
        underlinedColorHover: 'inherit',
        primaryBorderColor: siteVars.colorScheme.default.border2,
        primaryWrapperColorFocus: siteVars.colors.white,
        primaryWrapperBackgroundColorFocus: siteVars.colors.brand[500],
        underlinedWrapperColorHover: siteVars.colors.black,
        underlinedWrapperBackgroundHover: 'transparent',
        dividerHeight: utils_1.pxToRem(1),
        borderWidth: utils_1.pxToRem(1),
        menuZIndex: siteVars.zIndexes.overlay,
        beakZIndex: siteVars.zIndexes.menuItem,
        wrapperColorActive: siteVars.colors.grey[500],
        wrapperColorFocus: siteVars.colors.grey[500],
        wrapperColorHover: undefined,
        primaryWrapperColorHover: siteVars.colors.white,
        wrapperBackgroundColorFocus: siteVars.colors.grey[150],
        activeWrapperBackgroundColorHover: undefined,
        indicatorColor: siteVars.colors.grey[500],
        activeIndicatorColor: siteVars.colors.grey[750],
        activePrimaryIndicatorColor: siteVars.colors.grey[750],
        activePrimaryVerticalIndicatorColor: siteVars.colors.white,
        primaryIndicatorColorHover: siteVars.colors.white,
        indicatorColorHover: siteVars.colors.grey[500],
    };
};
exports.menuVariables = menuVariables;
