// @fluentui/react-northstar/dist/dts/src/themes/teams/components/Menu/menuDividerStyles.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.menuDividerStyles = void 0;
var colors_1 = require("../../colors");
var menuItemStyles_1 = require("./menuItemStyles");
exports.menuDividerStyles = {
    root: function (_a) {
        var _b;
        var p = _a.props, v = _a.variables;
        var colors = colors_1.getColorScheme(v.colorScheme, null, p.primary);
        var borderColor = p.primary ? v.primaryBorderColor : v.borderColor || colors.border;
        var borderType = p.vertical ? 'borderTop' : 'borderLeft';
        return __assign(__assign(__assign({}, (p.pointing &&
            p.vertical && {
            marginBottom: menuItemStyles_1.verticalPointingBottomMargin,
        })), (p.pills && __assign({}, (p.vertical
            ? { margin: "0 0 " + menuItemStyles_1.verticalPillsBottomMargin + " 0" }
            : { margin: "0 " + menuItemStyles_1.horizontalPillsRightMargin + " 0 0" })))), (p.hasContent
            ? {
                display: 'flex',
                justifyContent: 'center',
                flexDirection: 'column',
                textAlign: 'center',
            }
            : __assign(__assign((_b = {}, _b[borderType] = "1px solid " + borderColor, _b), (!p.vertical && {
                alignSelf: 'stretch',
            })), (p.vertical &&
                p.inSubmenu && {
                margin: '8px 0',
            }))));
    },
};
