// @fluentui/accessibility/dist/commonjs/behaviors/Tree/treeItemBehavior.js
"use strict";

exports.__esModule = true;
exports.treeItemBehavior = void 0;

var _keyboardKey = require("../../keyboard-key");

var _attributes = require("../../attributes");

var _treeTitleBehavior = require("./treeTitleBehavior");

/**
 * @description
 * Triggers 'performClick' action with 'Spacebar' on 'root', when tree item is selectable and has no subtree. In other cases 'performClick' is triggered with 'Spacebar' or 'Enter'.
 * Triggers 'performSelection' action with 'Spacebar' on 'root', when has a opened subtree.
 * Adds attribute 'aria-checked=true' based on the properties 'selectable' & 'selected' if the component has 'hasSubtree' property true. Does not set anything if false or undefined.
 *
 * @specification
 * Adds attribute 'aria-expanded=true' based on the property 'expanded' if the component has 'hasSubtree' property.
 * Adds attribute 'tabIndex=-1' to 'root' slot if 'hasSubtree' property is true. Does not set the attribute otherwise.
 * Adds attribute 'aria-setsize=3' based on the property 'treeSize' if the component has 'hasSubtree' property.
 * Adds attribute 'aria-posinset=2' based on the property 'index' if the component has 'hasSubtree' property.
 * Adds attribute 'aria-level=1' based on the property 'level' if the component has 'hasSubtree' property.
 * Adds attribute 'role=treeitem' to 'root' slot if 'hasSubtree' property is true. Sets the attribute to 'none' otherwise.
 * Triggers 'performClick' action with 'Enter' or 'Spacebar' on 'root'.
 * Triggers 'expandSiblings' action with '*' on 'root'.
 * Triggers 'focusParent' action with 'ArrowLeft' on 'root', when has a closed subtree.
 * Triggers 'focusParent' action with 'ArrowLeft' on 'root', when has no subtree.
 * Triggers 'collapse' action with 'ArrowLeft' on 'root', when has an opened subtree.
 * Triggers 'expand' action with 'ArrowRight' on 'root', when has a closed subtree.
 * Triggers 'focusFirstChild' action with 'ArrowRight' on 'root', when has an opened subtree.
 */
var treeItemBehavior = function treeItemBehavior(props) {
  var _Object$assign;

  var definition = {
    attributes: {
      root: Object.assign({
        role: 'none'
      }, props.hasSubtree && Object.assign((_Object$assign = {
        'aria-expanded': props.expanded,
        tabIndex: -1
      }, _Object$assign[_attributes.IS_FOCUSABLE_ATTRIBUTE] = true, _Object$assign.role = 'treeitem', _Object$assign['aria-setsize'] = props.treeSize, _Object$assign['aria-posinset'] = props.index, _Object$assign['aria-level'] = props.level, _Object$assign), props.selectable && {
        'aria-checked': props.indeterminate ? 'mixed' : !!props.selected
      }))
    },
    keyActions: {
      root: Object.assign({
        performClick: {
          keyCombinations: [{
            keyCode: _keyboardKey.EnterKey
          }, {
            keyCode: _keyboardKey.SpacebarKey
          }]
        }
      }, props.hasSubtree ? props.expanded ? {
        collapse: {
          keyCombinations: [{
            keyCode: _keyboardKey.keyboardKey.ArrowLeft
          }]
        },
        focusFirstChild: {
          keyCombinations: [{
            keyCode: _keyboardKey.keyboardKey.ArrowRight
          }]
        }
      } : {
        expand: {
          keyCombinations: [{
            keyCode: _keyboardKey.keyboardKey.ArrowRight
          }]
        },
        focusParent: {
          keyCombinations: [{
            keyCode: _keyboardKey.keyboardKey.ArrowLeft
          }]
        }
      } : {
        focusParent: {
          keyCombinations: [{
            keyCode: _keyboardKey.keyboardKey.ArrowLeft
          }]
        }
      }, {
        expandSiblings: {
          keyCombinations: [{
            keyCode: _keyboardKey.keyboardKey['*']
          }]
        }
      }, props.selectable && {
        performClick: {
          keyCombinations: props.hasSubtree ? [{
            keyCode: _keyboardKey.keyboardKey.Enter
          }] : [{
            keyCode: _keyboardKey.SpacebarKey
          }]
        },
        performSelection: {
          keyCombinations: [{
            keyCode: _keyboardKey.SpacebarKey
          }]
        }
      })
    },
    childBehaviors: {
      title: _treeTitleBehavior.treeTitleBehavior
    }
  };

  if (process.env.NODE_ENV !== 'production' && !props.hasSubtree) {
    // Override the default trigger's accessibility schema class.
    definition.attributes.root['data-aa-class'] = 'SingleTreeItem';
  }

  return definition;
};

exports.treeItemBehavior = treeItemBehavior;
