// @fluentui/accessibility/dist/commonjs/behaviors/Tree/treeItemAsListItemBehavior.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.treeItemAsListItemBehavior = void 0;

var _merge2 = _interopRequireDefault(require("lodash/merge"));

var _treeItemBehavior = require("./treeItemBehavior");

var _treeTitleAsListItemTitleBehavior = require("./treeTitleAsListItemTitleBehavior");

/**
 * @description
 * Adds role 'listitem' to a non-leaf item and 'none' to a leaf item.
 */
var treeItemAsListItemBehavior = function treeItemAsListItemBehavior(props) {
  var behavior = (0, _treeItemBehavior.treeItemBehavior)(props);
  var definition = (0, _merge2.default)(behavior, {
    attributes: {
      root: Object.assign({}, props.hasSubtree && {
        role: 'listitem'
      })
    },
    childBehaviors: {
      title: _treeTitleAsListItemTitleBehavior.treeTitleAsListItemTitleBehavior
    }
  });

  if (process.env.NODE_ENV !== 'production' && props.hasSubtree) {
    // Override the default trigger's accessibility schema class.
    definition.attributes.root['data-aa-class'] = 'TreeItemList';
  }

  return definition;
};

exports.treeItemAsListItemBehavior = treeItemAsListItemBehavior;
