// @fluentui/accessibility/dist/commonjs/behaviors/Tree/treeAsListboxBehavior.js
"use strict";

exports.__esModule = true;
exports.treeAsListboxBehavior = void 0;

var _treeBehavior = require("./treeBehavior");

var _treeItemAsOptionBehavior = require("./treeItemAsOptionBehavior");

var _treeTitleAsOptionBehavior = require("./treeTitleAsOptionBehavior");

/**
 * @specification
 * Adds role='listbox'.
 * Adds attribute 'aria-multiselectable=true' to 'root' slot if 'selectable' property is true. Does not set the attribute otherwise.
 */
var treeAsListboxBehavior = function treeAsListboxBehavior(props) {
  var behavior = (0, _treeBehavior.treeBehavior)(props);
  var definition = {
    attributes: {
      root: Object.assign({}, behavior.attributes.root, {
        role: 'listbox'
      })
    },
    keyActions: behavior.keyActions,
    focusZone: behavior.focusZone,
    childBehaviors: {
      item: _treeItemAsOptionBehavior.treeItemAsOptionBehavior,
      title: _treeTitleAsOptionBehavior.treeTitleAsOptionBehavior
    }
  };

  if (process.env.NODE_ENV !== 'production') {
    definition.attributes.root['data-aa-class'] = 'TreeListbox';
  }

  return definition;
};

exports.treeAsListboxBehavior = treeAsListboxBehavior;
