// @fluentui/react-northstar/dist/dts/src/themes/teams/components/Pill/pillVariables.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.pillVariables = void 0;
var utils_1 = require("../../../../utils");
var pillVariables = function (siteVars) { return ({
    background: siteVars.colorScheme.default.background3,
    backgroundHover: siteVars.colorScheme.default.background1,
    borderRadius: '9999px',
    smallerRoundedBorderRadius: utils_1.pxToRem(2),
    // Disabled
    disabledBackground: siteVars.colorScheme.default.backgroundDisabled,
    disabledColor: siteVars.colorScheme.default.foregroundDisabled,
    // Inverted
    invertedBackground: siteVars.colorScheme.default.background,
    // Outline
    outlineBackground: 'transparent',
    // TODO: The design spec maps to Neutral Stroke 1 that is equivalent to gre[440]
    // but we don't have this token
    outlineBorderColor: siteVars.colorScheme.default.borderActive4,
    outlineDisabledborder: siteVars.colorScheme.default.borderDisabled,
    // Smaller
    smallerHeight: utils_1.pxToRem(20),
    smallerMinWidth: utils_1.pxToRem(80),
    smallerMargin: utils_1.pxToRem(6) + " " + utils_1.pxToRem(2),
    // Small
    smallHeight: utils_1.pxToRem(24),
    smallMinWidth: utils_1.pxToRem(80),
    smallMargin: utils_1.pxToRem(4),
    // Medium (default)
    height: utils_1.pxToRem(32),
    minWidth: utils_1.pxToRem(90),
    margin: utils_1.pxToRem(6) + " " + utils_1.pxToRem(4),
    roundedBorderRadius: siteVars.borderRadiusMedium,
    // Content Smaller
    contentPaddingSmaller: utils_1.pxToRem(2) + " " + utils_1.pxToRem(8),
    contentFontSizeSmaller: utils_1.pxToRem(12),
    // Content Small
    contentFontSizeSmall: utils_1.pxToRem(12),
    contentPaddingSmall: utils_1.pxToRem(4) + " " + utils_1.pxToRem(8),
    // Content Medium
    contentFontSize: utils_1.pxToRem(14),
    contentPadding: utils_1.pxToRem(6) + " " + utils_1.pxToRem(8),
    // Action Pill
    actionMargin: "0 " + utils_1.pxToRem(8),
    actionWidth: utils_1.pxToRem(16),
    smallOrSmallerActionWidth: utils_1.pxToRem(12),
    // Icon Pill
    iconMargin: "0 0 0 " + utils_1.pxToRem(8),
    iconWidth: utils_1.pxToRem(20),
    smallOrSmallerIconWidth: utils_1.pxToRem(16),
    selectedIconColor: siteVars.colorScheme.default.border,
    selectedIconCheckColor: siteVars.colorScheme.brand.background2,
    selectedImageIconWidth: utils_1.pxToRem(32),
    smallSelectedImageIconWidth: utils_1.pxToRem(24),
    smallerSelectedImageIconWidth: utils_1.pxToRem(20),
    // Image Pill
    imageWidth: utils_1.pxToRem(32),
    imageHeight: utils_1.pxToRem(32),
    smallImageWidth: utils_1.pxToRem(24),
    smallImageHeight: utils_1.pxToRem(24),
    smallerImageWidth: utils_1.pxToRem(20),
    smallerImageHeight: utils_1.pxToRem(20),
}); };
exports.pillVariables = pillVariables;
