// @fluentui/react-bindings/dist/dts/src/styles/createAnimationStyles.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createAnimationStyles = void 0;
// Notice:
// This temporary lives here, will be remove once `animation` prop will be dropped
var createAnimationStyles = function (animation, theme) {
    var animationCSSProp;
    var _a = theme.animations, animations = _a === void 0 ? {} : _a;
    var animationName = animation.name;
    if (animations[animationName]) {
        var _b = animations[animationName], keyframe = _b.keyframe, duration = _b.duration, delay = _b.delay, direction = _b.direction, fillMode = _b.fillMode, iterationCount = _b.iterationCount, playState = _b.playState, timingFunction = _b.timingFunction, keyframeParams = _b.keyframeParams;
        var animationThemeKeyframeParams = keyframeParams || {};
        var animationPropKeyframeParams = animation.keyframeParams;
        var mergedKeyframeParams = animationPropKeyframeParams
            ? __assign(__assign({}, animationThemeKeyframeParams), (animationPropKeyframeParams || {})) : animationThemeKeyframeParams;
        var keyframeDefinition = typeof keyframe === 'string' ? keyframe : { keyframe: keyframe, params: mergedKeyframeParams };
        animationCSSProp = {
            animationName: keyframeDefinition,
            animationDelay: animation.delay || delay,
            animationDirection: animation.direction || direction,
            animationDuration: animation.duration || duration,
            animationFillMode: animation.fillMode || fillMode,
            animationIterationCount: animation.iterationCount || iterationCount,
            animationPlayState: animation.playState || playState,
            animationTimingFunction: animation.timingFunction || timingFunction,
        };
    }
    else {
        // animations was not found in the theme object
        animationCSSProp = __assign(__assign(__assign(__assign(__assign(__assign(__assign({ animationName: animation.name }, (animation.delay && { animationDelay: animation.delay })), (animation.direction && { animationDirection: animation.direction })), (animation.duration && { animationDuration: animation.duration })), (animation.fillMode && { animationFillMode: animation.fillMode })), (animation.iterationCount && {
            animationIterationCount: animation.iterationCount,
        })), (animation.playState && { animationPlayState: animation.playState })), (animation.timingFunction && {
            animationTimingFunction: animation.timingFunction,
        }));
    }
    return animationCSSProp;
};
exports.createAnimationStyles = createAnimationStyles;
