// @fluentui/react-component-nesting-registry/dist/dts/src/hooks/useNestingRoot.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useNestingRoot = void 0;
var React = require("react");
var NestingContext_1 = require("../NestingContext");
var RefStack_1 = require("../utils/RefStack");
var registrySet = new RefStack_1.RefStack();
// These hooks are not used currently
/* eslint-disable */
var useNestingRoot = function () {
    var registry = React.useState(registrySet)[0];
    var parentRef = React.useRef(null);
    var nestedProps = React.useMemo(function () { return ({
        value: {
            getContextRefs: registry.getContextRefs,
            register: registry.register,
            unregister: registry.unregister,
        },
    }); }, []);
    var getRefs = React.useCallback(function () {
        return registry.getContextRefs(parentRef);
    }, []);
    React.useEffect(function () {
        registry.register(parentRef);
        return function () { return registry.unregister(parentRef); };
    }, []);
    return {
        NestedComponent: NestingContext_1.NestingContext.Provider,
        nestedProps: nestedProps,
        getRefs: getRefs,
        isRoot: true,
        ref: parentRef,
    };
};
exports.useNestingRoot = useNestingRoot;
