// @uifabric/merge-styles/lib-amd/index.js
define(["require", "exports", "tslib", "./mergeStyles", "./mergeStyleSets", "./concatStyleSets", "./concatStyleSetsWithProps", "./fontFace", "./keyframes", "./Stylesheet", "./StyleOptionsState", "./version"], function (require, exports, tslib_1, mergeStyles_1, mergeStyleSets_1, concatStyleSets_1, concatStyleSetsWithProps_1, fontFace_1, keyframes_1, Stylesheet_1, StyleOptionsState_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.mergeStyles = mergeStyles_1.mergeStyles;
    exports.mergeCss = mergeStyles_1.mergeCss;
    exports.mergeStyleSets = mergeStyleSets_1.mergeStyleSets;
    exports.mergeCssSets = mergeStyleSets_1.mergeCssSets;
    exports.concatStyleSets = concatStyleSets_1.concatStyleSets;
    exports.concatStyleSetsWithProps = concatStyleSetsWithProps_1.concatStyleSetsWithProps;
    exports.fontFace = fontFace_1.fontFace;
    exports.keyframes = keyframes_1.keyframes;
    tslib_1.__exportStar(Stylesheet_1, exports);
    exports.setRTL = StyleOptionsState_1.setRTL;
});
