// @fluentui/react-northstar/dist/dts/src/components/Menu/MenuItemWrapper.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MenuItemWrapper = exports.menuItemWrapperClassName = void 0;
var React = require("react");
var react_bindings_1 = require("@fluentui/react-bindings");
var customPropTypes = require("@fluentui/react-proptypes");
var PropTypes = require("prop-types");
var utils_1 = require("../../utils");
exports.menuItemWrapperClassName = 'ui-menu__itemwrapper';
/**
 * A MenuItemWrapper allows a user to have a dedicated component that can be targeted from the theme.
 */
exports.MenuItemWrapper = React.forwardRef(function (props, ref) {
    var context = react_bindings_1.useFluentContext();
    var _a = react_bindings_1.useTelemetry(exports.MenuItemWrapper.displayName, context.telemetry), setStart = _a.setStart, setEnd = _a.setEnd;
    setStart();
    var className = props.className, children = props.children, design = props.design, styles = props.styles, variables = props.variables, content = props.content, active = props.active, disabled = props.disabled, iconOnly = props.iconOnly, isFromKeyboard = props.isFromKeyboard, pills = props.pills, pointing = props.pointing, secondary = props.secondary, underlined = props.underlined, vertical = props.vertical, primary = props.primary, on = props.on;
    var classes = react_bindings_1.useStyles(exports.MenuItemWrapper.displayName, {
        className: exports.menuItemWrapperClassName,
        mapPropsToStyles: function () { return ({
            active: active,
            disabled: disabled,
            iconOnly: iconOnly,
            isFromKeyboard: isFromKeyboard,
            pills: pills,
            pointing: pointing,
            secondary: secondary,
            underlined: underlined,
            vertical: vertical,
            primary: primary,
            on: on,
        }); },
        mapPropsToInlineStyles: function () { return ({
            className: className,
            design: design,
            styles: styles,
            variables: variables,
        }); },
        rtl: context.rtl,
    }).classes;
    var getA11Props = react_bindings_1.useAccessibility(props.accessibility, {
        debugName: exports.MenuItemWrapper.displayName,
        rtl: context.rtl,
    });
    var ElementType = react_bindings_1.getElementType(props);
    var unhandledProps = react_bindings_1.useUnhandledProps(exports.MenuItemWrapper.handledProps, props);
    var element = (React.createElement(ElementType, __assign({}, getA11Props('root', __assign({ className: classes.root, ref: ref }, unhandledProps))), react_bindings_1.childrenExist(children) ? children : content));
    setEnd();
    return element;
});
exports.MenuItemWrapper.displayName = 'MenuItemWrapper';
exports.MenuItemWrapper.defaultProps = {
    as: 'li',
};
exports.MenuItemWrapper.propTypes = __assign(__assign({}, utils_1.commonPropTypes.createCommon()), { active: PropTypes.bool, disabled: PropTypes.bool, iconOnly: PropTypes.bool, isFromKeyboard: PropTypes.bool, pills: PropTypes.bool, pointing: PropTypes.oneOf(['start', 'end', true, false]), primary: customPropTypes.every([customPropTypes.disallow(['secondary']), PropTypes.bool]), secondary: customPropTypes.every([customPropTypes.disallow(['primary']), PropTypes.bool]), underlined: PropTypes.bool, vertical: PropTypes.bool, on: PropTypes.oneOf(['hover']) });
exports.MenuItemWrapper.handledProps = Object.keys(exports.MenuItemWrapper.propTypes);
exports.MenuItemWrapper.shorthandConfig = {
    mappedProp: 'content',
};
