// @fluentui/react-northstar/dist/dts/src/themes/teams/components/Text/textVariables.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.textVariables = exports.textColorAreas = void 0;
var utils_1 = require("../../../../utils");
var colorUtils_1 = require("../../../colorUtils");
exports.textColorAreas = utils_1.stringLiteralsArray('foreground');
var textVariables = function (siteVariables) {
    return {
        colorScheme: colorUtils_1.pickValuesFromColorScheme(siteVariables.colorScheme, exports.textColorAreas),
        fontSizeSmallest: siteVariables.fontSizes.smallest,
        fontLineHeightSmallest: siteVariables.lineHeightSmallest,
        fontSizeSmaller: siteVariables.fontSizes.smaller,
        fontLineHeightSmaller: siteVariables.lineHeightSmaller,
        fontSizeSmall: siteVariables.fontSizes.small,
        fontLineHeightSmall: siteVariables.lineHeightSmall,
        fontSizeMedium: siteVariables.fontSizes.medium,
        fontLineHeightMedium: siteVariables.lineHeightMedium,
        fontSizeLarge: siteVariables.fontSizes.large,
        fontLineHeightLarge: siteVariables.lineHeightLarge,
        fontSizeLarger: siteVariables.fontSizes.larger,
        fontLineHeightLarger: siteVariables.lineHeightLarger,
        fontSizeLargest: siteVariables.fontSizes.largest,
        fontLineHeightLargest: siteVariables.lineHeightLargest,
        fontWeightLight: siteVariables.fontWeightLight,
        fontWeightSemilight: siteVariables.fontWeightSemilight,
        fontWeightRegular: siteVariables.fontWeightRegular,
        fontWeightSemibold: siteVariables.fontWeightSemibold,
        fontWeightBold: siteVariables.fontWeightBold,
        atMentionOtherColor: siteVariables.colors.brand[600],
        atMentionMeColor: siteVariables.colors.orange[400],
        atMentionMeFontWeight: siteVariables.fontWeightBold,
        disabledColor: siteVariables.colors.grey[250],
        errorColor: siteVariables.colorScheme.red.foreground,
        importantWeight: siteVariables.fontWeightBold,
        importantColor: siteVariables.colors.red[400],
        successColor: siteVariables.colors.green[600],
        timestampColor: siteVariables.colorScheme.default.foreground1,
    };
};
exports.textVariables = textVariables;
