// @fluentui/accessibility/dist/commonjs/behaviors/Toolbar/menuItemAsToolbarButtonBehavior.js
"use strict";

exports.__esModule = true;
exports.menuItemAsToolbarButtonBehavior = void 0;

var _keyboardKey = require("../../keyboard-key");

var _attributes = require("../../attributes");

/**
 * @description
 * The behavior is designed for particular structure of menu item. The item consists of root element and anchor inside the root element.
 *
 * @specification
 * Adds role 'presentation' to 'wrapper' slot.
 * Adds role 'button' to 'root' slot.
 * Adds attribute 'tabIndex=0' to 'root' slot.
 * Adds attribute 'aria-label' based on the property 'aria-label' to 'root' slot.
 * Adds attribute 'aria-labelledby' based on the property 'aria-labelledby' to 'root' slot.
 * Adds attribute 'aria-describedby' based on the property 'aria-describedby' to 'root' slot.
 * Adds attribute 'aria-disabled=true' based on the property 'disabled'. This can be overriden by providing 'aria-disabled' property directly to the component.
 * Adds attribute 'aria-haspopup=true' to 'root' slot if 'hasMenu' property is set.
 * Triggers 'performClick' action with 'Enter' or 'Spacebar' on 'wrapper'.
 * Triggers 'closeMenuAndFocusTrigger' action with 'Escape' on 'wrapper'.
 * Triggers 'openMenu' action with 'ArrowDown' on 'wrapper', when orientation is horizontal.
 * Triggers 'doNotNavigateNextParentItem' action with 'ArrowLeft' or 'ArrowRight' on 'wrapper', when toolbar button has submenu and it is opened.
 * Adds attribute 'data-is-focusable=true' to 'root' slot.
 */
var menuItemAsToolbarButtonBehavior = function menuItemAsToolbarButtonBehavior(props) {
  var _root;

  return {
    attributes: {
      wrapper: {
        role: 'presentation'
      },
      root: (_root = {
        role: 'button',
        tabIndex: 0,
        'aria-haspopup': props.hasMenu ? 'true' : undefined,
        'aria-disabled': props['disabled'],
        'aria-label': props['aria-label'],
        'aria-labelledby': props['aria-labelledby'],
        'aria-describedby': props['aria-describedby']
      }, _root[_attributes.IS_FOCUSABLE_ATTRIBUTE] = true, _root)
    },
    keyActions: {
      wrapper: {
        performClick: {
          keyCombinations: [{
            keyCode: _keyboardKey.keyboardKey.Enter
          }, {
            keyCode: _keyboardKey.SpacebarKey
          }]
        },
        closeMenuAndFocusTrigger: {
          keyCombinations: [{
            keyCode: _keyboardKey.keyboardKey.Escape
          }]
        },
        openMenu: !props.vertical && {
          keyCombinations: [{
            keyCode: _keyboardKey.keyboardKey.ArrowDown
          }]
        },
        doNotNavigateNextParentItem: {
          keyCombinations: props.hasMenu && props.menuOpen ? [{
            keyCode: _keyboardKey.keyboardKey.ArrowLeft
          }, {
            keyCode: _keyboardKey.keyboardKey.ArrowRight
          }] : null
        }
      }
    }
  };
};

exports.menuItemAsToolbarButtonBehavior = menuItemAsToolbarButtonBehavior;
