// @fluentui/accessibility/dist/commonjs/behaviors/Toolbar/menuAsToolbarBehavior.js
"use strict";

exports.__esModule = true;
exports.menuAsToolbarBehavior = void 0;

var _types = require("../../focusZone/types");

var _menuItemAsToolbarButtonBehavior = require("./menuItemAsToolbarButtonBehavior");

/**
 * @description
 * Implements ARIA Toolbar design pattern.
 * Child item components need to have menuItemAsToolbarButtonBehavior assigned.
 * @specification
 * Adds role 'toolbar' to 'root' slot.
 * Provides arrow key navigation in bidirectionalDomOrder direction.
 * When component's container element receives focus, focus will be set to the default focusable child element of the component.
 */
var menuAsToolbarBehavior = function menuAsToolbarBehavior() {
  return {
    attributes: {
      root: {
        role: 'toolbar'
      }
    },
    focusZone: {
      props: {
        shouldFocusInnerElementWhenReceivedFocus: true,
        direction: _types.FocusZoneDirection.bidirectionalDomOrder
      }
    },
    childBehaviors: {
      item: _menuItemAsToolbarButtonBehavior.menuItemAsToolbarButtonBehavior
    }
  };
};

exports.menuAsToolbarBehavior = menuAsToolbarBehavior;
