// @fluentui/accessibility/dist/commonjs/behaviors/Toolbar/toolbarMenuItemCheckboxBehavior.js
"use strict";

exports.__esModule = true;
exports.toolbarMenuItemCheckboxBehavior = void 0;

var _keyboardKey = require("../../keyboard-key");

/**
 * @specification
 * Adds attribute 'aria-checked=true' based on the property 'active'.
 * Adds attribute 'aria-disabled=true' based on the property 'disabled'.
 * Adds role='menuitemcheckbox'.
 * Adds role 'presentation' to 'wrapper' slot.
 * Triggers 'performClick' action with 'Enter' or 'Spacebar' on 'root'.
 */
var toolbarMenuItemCheckboxBehavior = function toolbarMenuItemCheckboxBehavior(props) {
  return {
    attributes: {
      wrapper: {
        role: 'presentation'
      },
      root: {
        'aria-checked': props.active,
        'aria-disabled': props.disabled,
        role: 'menuitemcheckbox'
      }
    },
    keyActions: {
      root: {
        performClick: {
          keyCombinations: [{
            keyCode: _keyboardKey.keyboardKey.Enter
          }, {
            keyCode: _keyboardKey.SpacebarKey
          }]
        }
      }
    }
  };
};

exports.toolbarMenuItemCheckboxBehavior = toolbarMenuItemCheckboxBehavior;
