// @fluentui/accessibility/dist/commonjs/behaviors/Toolbar/toolbarMenuBehavior.js
"use strict";

exports.__esModule = true;
exports.toolbarMenuBehavior = void 0;

var _types = require("../../focusZone/types");

var _toolbarMenuItemBehavior = require("./toolbarMenuItemBehavior");

var _keyboardKey = require("../../keyboard-key");

/**
 * @description
 * The 'menu' role is used to identify an element that creates a list of common actions or functions that a user can invoke.
 *
 * @specification
 * Adds role='menu'.
 * Provides arrow key navigation in vertical direction.
 * Keyboard navigation is circular.
 * Component will get focus when mounted.
 */
var toolbarMenuBehavior = function toolbarMenuBehavior() {
  return {
    attributes: {
      root: {
        role: 'menu'
      }
    },
    keyActions: {
      root: {
        performClick: {
          keyCombinations: [{
            keyCode: _keyboardKey.keyboardKey.Enter
          }, {
            keyCode: _keyboardKey.SpacebarKey
          }]
        }
      }
    },
    focusZone: {
      props: {
        isCircularNavigation: true,
        shouldFocusOnMount: true,
        direction: _types.FocusZoneDirection.vertical
      }
    },
    childBehaviors: {
      item: _toolbarMenuItemBehavior.toolbarMenuItemBehavior
    }
  };
};

exports.toolbarMenuBehavior = toolbarMenuBehavior;
