// @fluentui/accessibility/dist/commonjs/behaviors/Toolbar/toolbarMenuItemRadioBehavior.js
"use strict";

exports.__esModule = true;
exports.toolbarMenuItemRadioBehavior = void 0;

var _keyboardKey = require("../../keyboard-key");

var _attributes = require("../../attributes");

/**
 * @specification
 *  Adds attribute 'aria-checked=true' based on the property 'active'.
 *  Adds attribute 'aria-disabled=true' based on the property 'disabled'.
 *  Adds role='menuitemradio'.
 */
var toolbarMenuItemRadioBehavior = function toolbarMenuItemRadioBehavior(props) {
  var _root;

  return {
    attributes: {
      root: (_root = {}, _root[_attributes.IS_FOCUSABLE_ATTRIBUTE] = true, _root['aria-checked'] = props.active, _root['aria-disabled'] = props.disabled, _root.role = 'menuitemradio', _root)
    },
    keyActions: {
      root: {
        performClick: {
          keyCombinations: [{
            keyCode: _keyboardKey.keyboardKey.Enter
          }, {
            keyCode: _keyboardKey.SpacebarKey
          }]
        }
      }
    }
  };
};

exports.toolbarMenuItemRadioBehavior = toolbarMenuItemRadioBehavior;
