// @fluentui/accessibility/dist/commonjs/behaviors/Toolbar/toolbarItemBehavior.js
"use strict";

exports.__esModule = true;
exports.toolbarItemBehavior = void 0;

var _keyboardKey = require("../../keyboard-key");

var _buttonBehavior = require("../Button/buttonBehavior");

/**
 * @specification
 * Adds role='button' if element type is other than 'button'. This allows screen readers to handle the component as a button.
 * Adds attribute 'tabIndex=0' if element type is other than 'button'.
 * Adds attribute 'aria-haspopup=menu' to 'root' slot if 'hasMenu' property is set.
 * Adds attribute 'aria-haspopup=dialog' to 'root' slot if 'hasPopup' property is set.
 * Triggers 'performClick' action with 'Enter' or 'Spacebar' on 'root'.
 * Triggers 'closeMenuAndFocusTrigger' action with 'Escape' on 'wrapper', when toolbar button has submenu and it is opened.
 * Triggers 'doNotNavigateNextToolbarItem' action with 'ArrowLeft' or 'ArrowRight' on 'wrapper', when toolbar button has submenu and it is opened.
 * Adds attribute 'disabled=true' based on the property 'disabled'.
 * Adds attribute 'aria-disabled=true' based on the property 'disabledFocusable'.
 */
var toolbarItemBehavior = function toolbarItemBehavior(props) {
  var behaviorData = (0, _buttonBehavior.buttonBehavior)(props);
  behaviorData.attributes.root = Object.assign({}, behaviorData.attributes.root, {
    'aria-haspopup': props.hasPopup ? 'dialog' : props.hasMenu ? 'menu' : undefined,
    disabled: props.disabled,
    'aria-disabled': props.disabledFocusable
  });
  behaviorData.keyActions.wrapper = Object.assign({}, behaviorData.keyActions.wrapper, {
    performWrapperClick: {
      keyCombinations: [{
        keyCode: _keyboardKey.keyboardKey.Enter
      }, {
        keyCode: _keyboardKey.SpacebarKey
      }]
    },
    closeMenuAndFocusTrigger: {
      keyCombinations: props.hasMenu && props.menuOpen ? [{
        keyCode: _keyboardKey.keyboardKey.Escape
      }, {
        keyCode: _keyboardKey.keyboardKey.Tab,
        shiftKey: true
      }] : null
    },
    doNotNavigateNextToolbarItem: {
      keyCombinations: props.hasMenu && props.menuOpen ? [{
        keyCode: _keyboardKey.keyboardKey.ArrowLeft
      }, {
        keyCode: _keyboardKey.keyboardKey.ArrowRight
      }] : null
    }
  });
  return behaviorData;
};

exports.toolbarItemBehavior = toolbarItemBehavior;
