// @fluentui/accessibility/dist/commonjs/behaviors/Toolbar/toolbarRadioGroupItemBehavior.js
"use strict";

exports.__esModule = true;
exports.toolbarRadioGroupItemBehavior = void 0;

var _buttonBehavior = require("../Button/buttonBehavior");

/**
 * @specification
 *  Adds role='radio'. This allows screen readers to handle the component as a radio button.
 *  Adds attribute 'aria-checked=true' based on the property 'active'.
 *  Triggers 'performClick' action with 'Enter' or 'Spacebar' on 'root'.
 *  Adds attribute 'disabled=true' based on the property 'disabled'.
 *  Adds attribute 'aria-disabled=true' based on the property 'disabledFocusable'.
 */
var toolbarRadioGroupItemBehavior = function toolbarRadioGroupItemBehavior(props) {
  var definition = {
    attributes: {
      root: {
        role: 'radio',
        'aria-checked': props.active,
        disabled: props.disabled,
        'aria-disabled': props.disabledFocusable
      }
    },
    keyActions: (0, _buttonBehavior.buttonBehavior)(props).keyActions
  };

  if (process.env.NODE_ENV !== 'production') {
    // Override the default trigger's accessibility schema class.
    definition.attributes.root['data-aa-class'] = 'RadioGroupItem';
  }

  return definition;
};

exports.toolbarRadioGroupItemBehavior = toolbarRadioGroupItemBehavior;
