// @fluentui/accessibility/dist/dts/src/behaviors/Button/buttonBehavior.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.buttonBehavior = void 0;
var keyboard_key_1 = require("../../keyboard-key");
var buttonBehavior = function (props) {
    var definition = {
        attributes: {
            root: {
                role: props.as === 'button' ? undefined : 'button',
                tabIndex: props.as === 'button' || props.disabled ? undefined : 0,
                disabled: props.as === 'button' ? props.disabled : undefined,
                'aria-disabled': props.disabledFocusable,
            },
        },
        keyActions: {
            root: __assign({}, (props.as !== 'button' &&
                props.as !== 'a' && {
                performClick: {
                    keyCombinations: [{ keyCode: keyboard_key_1.keyboardKey.Enter }, { keyCode: keyboard_key_1.SpacebarKey }],
                },
            })),
        },
    };
    if (process.env.NODE_ENV !== 'production') {
        // Override the default trigger's accessibility schema class.
        if (props.disabledFocusable) {
            definition.attributes.root['data-aa-class'] = 'DisabledFocusableButton';
        }
    }
    return definition;
};
exports.buttonBehavior = buttonBehavior;
