// @fluentui/react-northstar/dist/commonjs/themes/teams/components/Dialog/dialogVariables.js
"use strict";

exports.__esModule = true;
exports.dialogVariables = void 0;

var _utils = require("../../../../utils");

var dialogVariables = function dialogVariables(siteVariables) {
  return {
    border: 'none',
    rootBackground: siteVariables.colors.white,
    rootBorderRadius: siteVariables.borderRadiusXLarge,
    rootWidth: '600px',
    rootPadding: (0, _utils.pxToRem)(27) + " " + (0, _utils.pxToRem)(32) + " " + (0, _utils.pxToRem)(32) + " " + (0, _utils.pxToRem)(32),
    rootPaddingZoom: "" + (0, _utils.pxToRem)(12),
    contentMargin: "0 0 " + (0, _utils.pxToRem)(20) + " 0",
    boxShadow: siteVariables.shadowLevel4,
    foregroundColor: siteVariables.colors.grey[900],
    headerFontSize: siteVariables.fontSizes.large,
    headerFontWeight: siteVariables.fontWeightBold,
    headerMargin: "0 0 " + (0, _utils.pxToRem)(8) + " 0",
    overlayBackground: 'rgba(37, 36, 36, .75)',
    // todo: update to a palette value when daisy has mapped one
    overlayZIndex: siteVariables.zIndexes.overlay,
    headerActionMargin: (0, _utils.pxToRem)(-3) + " " + (0, _utils.pxToRem)(-8) + " 0 0",
    // in case of customized footer this var allows to align styles
    footerActionsBreakpoint: '400px'
  };
};

exports.dialogVariables = dialogVariables;
