// @fluentui/accessibility/dist/dts/src/behaviors/Carousel/carouselItemBehavior.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.carouselItemBehavior = void 0;
var keyboard_key_1 = require("../../keyboard-key");
/**
 * @description
 * Adds attribute 'tabIndex=0' to 'root' slot if 'active' property and 'navigation' property is true. Sets the attribute to '-1' otherwise.
 * @specification
 * Adds attribute 'role=tabpanel' based on the property 'navigation' to 'root' slot.
 * Adds attribute 'aria-hidden=false' to 'root' slot if 'active' property is true. Sets the attribute to 'true' otherwise.
 * Triggers 'arrowKeysNavigationStopPropagation' action with 'ArrowRight' or 'ArrowLeft' on 'root'.
 */
var carouselItemBehavior = function (props) { return ({
    attributes: {
        root: {
            role: props.navigation ? 'tabpanel' : undefined,
            'aria-hidden': props.active ? 'false' : 'true',
            tabIndex: props.navigation ? (props.active ? 0 : -1) : undefined,
        },
    },
    keyActions: {
        root: {
            arrowKeysNavigationStopPropagation: {
                keyCombinations: [{ keyCode: keyboard_key_1.keyboardKey.ArrowRight }, { keyCode: keyboard_key_1.keyboardKey.ArrowLeft }],
            },
        },
    },
}); };
exports.carouselItemBehavior = carouselItemBehavior;
