// botframework-connector/lib/teams/models/mappers.js
"use strict";
/**
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.AppBasedLinkQuery = exports.TaskModuleRequestContext = exports.TaskModuleResponse = exports.TaskModuleMessageResponse = exports.TaskModuleContinueResponse = exports.TaskModuleTaskInfo = exports.FileConsentCardResponse = exports.FileUploadInfo = exports.FileInfoCard = exports.FileDownloadInfo = exports.FileConsentCard = exports.MessagingExtensionResponse = exports.MessagingExtensionActionResponse = exports.CacheInfo = exports.MessagingExtensionResult = exports.MessagingExtensionSuggestedAction = exports.MessagingExtensionAttachment = exports.Attachment = exports.TaskModuleResponseBase = exports.MessagingExtensionAction = exports.TaskModuleRequest = exports.MessageActionsPayload = exports.MessageActionsPayloadReaction = exports.MessageActionsPayloadMention = exports.MessageActionsPayloadAttachment = exports.MessageActionsPayloadBody = exports.MessageActionsPayloadFrom = exports.MessageActionsPayloadConversation = exports.MessageActionsPayloadApp = exports.MessageActionsPayloadUser = exports.Activity = exports.MessagingExtensionQuery = exports.MessagingExtensionParameter = exports.MessagingExtensionQueryOptions = exports.SigninStateVerificationQuery = exports.O365ConnectorCardActionQuery = exports.O365ConnectorCardMultichoiceInput = exports.O365ConnectorCardMultichoiceInputChoice = exports.O365ConnectorCardDateInput = exports.O365ConnectorCardTextInput = exports.O365ConnectorCardActionCard = exports.O365ConnectorCardInputBase = exports.O365ConnectorCardHttpPOST = exports.O365ConnectorCardOpenUri = exports.O365ConnectorCardOpenUriTarget = exports.O365ConnectorCardViewAction = exports.O365ConnectorCard = exports.O365ConnectorCardSection = exports.O365ConnectorCardActionBase = exports.O365ConnectorCardImage = exports.O365ConnectorCardFact = exports.CardImage = exports.CardAction = exports.TeamsMeetingDetails = exports.TeamsMeetingInfo = exports.TeamsMeetingParticipant = exports.Meeting = exports.TeamsChannelAccount = exports.ChannelAccount = exports.TeamsChannelData = exports.TenantInfo = exports.NotificationInfo = exports.TeamInfo = exports.TeamDetails = exports.ConversationList = exports.ChannelInfo = void 0;
exports.ChannelInfo = {
    serializedName: 'ChannelInfo',
    type: {
        name: 'Composite',
        className: 'ChannelInfo',
        modelProperties: {
            id: {
                serializedName: 'id',
                type: {
                    name: 'String',
                },
            },
            name: {
                serializedName: 'name',
                type: {
                    name: 'String',
                },
            },
        },
    },
};
exports.ConversationList = {
    serializedName: 'ConversationList',
    type: {
        name: 'Composite',
        className: 'ConversationList',
        modelProperties: {
            conversations: {
                serializedName: 'conversations',
                type: {
                    name: 'Sequence',
                    element: {
                        type: {
                            name: 'Composite',
                            className: 'ChannelInfo',
                        },
                    },
                },
            },
        },
    },
};
exports.TeamDetails = {
    serializedName: 'TeamDetails',
    type: {
        name: 'Composite',
        className: 'TeamDetails',
        modelProperties: {
            id: {
                serializedName: 'id',
                type: {
                    name: 'String',
                },
            },
            name: {
                serializedName: 'name',
                type: {
                    name: 'String',
                },
            },
            aadGroupId: {
                serializedName: 'aadGroupId',
                type: {
                    name: 'String',
                },
            },
        },
    },
};
exports.TeamInfo = {
    serializedName: 'TeamInfo',
    type: {
        name: 'Composite',
        className: 'TeamInfo',
        modelProperties: {
            id: {
                serializedName: 'id',
                type: {
                    name: 'String',
                },
            },
            name: {
                serializedName: 'name',
                type: {
                    name: 'String',
                },
            },
            aadGroupId: {
                serializedName: 'aadGroupId',
                type: {
                    name: 'String',
                },
            },
        },
    },
};
exports.NotificationInfo = {
    serializedName: 'NotificationInfo',
    type: {
        name: 'Composite',
        className: 'NotificationInfo',
        modelProperties: {
            alert: {
                serializedName: 'alert',
                type: {
                    name: 'Boolean',
                },
            },
            alertInMeeting: {
                serializedName: 'alertInMeeting',
                type: {
                    name: 'Boolean',
                },
            },
            externalResourceUrl: {
                serializedName: 'externalResourceUrl',
                type: {
                    name: 'String',
                },
            },
        },
    },
};
exports.TenantInfo = {
    serializedName: 'TenantInfo',
    type: {
        name: 'Composite',
        className: 'TenantInfo',
        modelProperties: {
            id: {
                serializedName: 'id',
                type: {
                    name: 'String',
                },
            },
        },
    },
};
exports.TeamsChannelData = {
    serializedName: 'TeamsChannelData',
    type: {
        name: 'Composite',
        className: 'TeamsChannelData',
        modelProperties: {
            channel: {
                serializedName: 'channel',
                type: {
                    name: 'Composite',
                    className: 'ChannelInfo',
                },
            },
            eventType: {
                serializedName: 'eventType',
                type: {
                    name: 'String',
                },
            },
            team: {
                serializedName: 'team',
                type: {
                    name: 'Composite',
                    className: 'TeamInfo',
                },
            },
            notification: {
                serializedName: 'notification',
                type: {
                    name: 'Composite',
                    className: 'NotificationInfo',
                },
            },
            tenant: {
                serializedName: 'tenant',
                type: {
                    name: 'Composite',
                    className: 'TenantInfo',
                },
            },
        },
    },
};
exports.ChannelAccount = {
    serializedName: 'ChannelAccount',
    type: {
        name: 'Composite',
        className: 'ChannelAccount',
        modelProperties: {
            id: {
                serializedName: 'id',
                type: {
                    name: 'String',
                },
            },
            name: {
                serializedName: 'name',
                type: {
                    name: 'String',
                },
            },
        },
    },
};
exports.TeamsChannelAccount = {
    serializedName: 'TeamsChannelAccount',
    type: {
        name: 'Composite',
        className: 'TeamsChannelAccount',
        modelProperties: Object.assign(Object.assign({}, exports.ChannelAccount.type.modelProperties), { givenName: {
                serializedName: 'givenName',
                type: {
                    name: 'String',
                },
            }, surname: {
                serializedName: 'surname',
                type: {
                    name: 'String',
                },
            }, email: {
                serializedName: 'email',
                type: {
                    name: 'String',
                },
            }, userPrincipalName: {
                serializedName: 'userPrincipalName',
                type: {
                    name: 'String',
                },
            }, tenantId: {
                serializedName: 'tenantId',
                type: {
                    name: 'String',
                },
            }, userRole: {
                serializedName: 'userRole',
                type: {
                    name: 'String',
                },
            } }),
    },
};
exports.Meeting = {
    serializedName: 'meeting',
    type: {
        name: 'Composite',
        className: 'Meeting',
        modelProperties: {
            role: {
                serializedName: 'role',
                type: {
                    name: 'String',
                },
            },
            inMeeting: {
                serializedName: 'inMeeting',
                type: {
                    name: 'Boolean',
                },
            },
        },
    },
};
exports.TeamsMeetingParticipant = {
    serializedName: 'TeamsMeetingParticipant',
    type: {
        name: 'Composite',
        className: 'TeamsMeetingParticipant',
        modelProperties: {
            user: {
                serializedName: 'user',
                type: {
                    name: 'Composite',
                    className: 'TeamsChannelAccount',
                },
            },
            meeting: {
                serializedName: 'meeting',
                type: {
                    name: 'Composite',
                    className: 'Meeting',
                },
            },
            conversation: {
                serializedName: 'conversation',
                type: {
                    name: 'Composite',
                    className: 'MessageActionsPayloadConversation',
                },
            },
        },
    },
};
exports.TeamsMeetingInfo = {
    serializedName: 'TeamsMeetingInfo',
    type: {
        name: 'Composite',
        className: 'TeamsMeetingInfo',
        modelProperties: {
            details: {
                serializedName: 'details',
                type: {
                    name: 'Composite',
                    className: 'TeamsMeetingDetails',
                },
            },
            conversation: {
                serializedName: 'conversation',
                type: {
                    name: 'Composite',
                    className: 'MessageActionsPayloadConversation',
                },
            },
            organizer: {
                serializedName: 'organizer',
                type: {
                    name: 'Composite',
                    className: 'TeamsChannelAccount',
                },
            },
        },
    },
};
exports.TeamsMeetingDetails = {
    serializedName: 'TeamsMeetingDetails',
    type: {
        name: 'Composite',
        className: 'TeamsMeetingDetails',
        modelProperties: {
            id: {
                serializedName: 'id',
                type: {
                    name: 'String',
                },
            },
            msGraphResourceId: {
                serializedName: 'msGraphResourceId',
                type: {
                    name: 'String',
                },
            },
            scheduledStartTime: {
                serializedName: 'scheduledStartTime',
                type: {
                    name: 'DateTime',
                },
            },
            scheduledEndTime: {
                serializedName: 'scheduledEndTime',
                type: {
                    name: 'DateTime',
                },
            },
            joinUrl: {
                serializedName: 'joinUrl',
                type: {
                    name: 'String',
                },
            },
            title: {
                serializedName: 'title',
                type: {
                    name: 'String',
                },
            },
            type: {
                serializedName: 'type',
                type: {
                    name: 'String',
                },
            },
        },
    },
};
exports.CardAction = {
    serializedName: 'CardAction',
    type: {
        name: 'Composite',
        className: 'CardAction',
        modelProperties: {
            type: {
                serializedName: 'type',
                type: {
                    name: 'String',
                },
            },
            title: {
                serializedName: 'title',
                type: {
                    name: 'String',
                },
            },
            image: {
                serializedName: 'image',
                type: {
                    name: 'String',
                },
            },
            value: {
                serializedName: 'value',
                type: {
                    name: 'Object',
                },
            },
        },
    },
};
exports.CardImage = {
    serializedName: 'CardImage',
    type: {
        name: 'Composite',
        className: 'CardImage',
        modelProperties: {
            url: {
                serializedName: 'url',
                type: {
                    name: 'String',
                },
            },
            alt: {
                serializedName: 'alt',
                type: {
                    name: 'String',
                },
            },
            tap: {
                serializedName: 'tap',
                type: {
                    name: 'Composite',
                    className: 'CardAction',
                },
            },
        },
    },
};
exports.O365ConnectorCardFact = {
    serializedName: 'O365ConnectorCardFact',
    type: {
        name: 'Composite',
        className: 'O365ConnectorCardFact',
        modelProperties: {
            name: {
                serializedName: 'name',
                type: {
                    name: 'String',
                },
            },
            value: {
                serializedName: 'value',
                type: {
                    name: 'String',
                },
            },
        },
    },
};
exports.O365ConnectorCardImage = {
    serializedName: 'O365ConnectorCardImage',
    type: {
        name: 'Composite',
        className: 'O365ConnectorCardImage',
        modelProperties: {
            image: {
                serializedName: 'image',
                type: {
                    name: 'String',
                },
            },
            title: {
                serializedName: 'title',
                type: {
                    name: 'String',
                },
            },
        },
    },
};
exports.O365ConnectorCardActionBase = {
    serializedName: 'O365ConnectorCardActionBase',
    type: {
        name: 'Composite',
        className: 'O365ConnectorCardActionBase',
        modelProperties: {
            type: {
                serializedName: '@type',
                type: {
                    name: 'String',
                },
            },
            name: {
                serializedName: 'name',
                type: {
                    name: 'String',
                },
            },
            id: {
                serializedName: '@id',
                type: {
                    name: 'String',
                },
            },
        },
    },
};
exports.O365ConnectorCardSection = {
    serializedName: 'O365ConnectorCardSection',
    type: {
        name: 'Composite',
        className: 'O365ConnectorCardSection',
        modelProperties: {
            title: {
                serializedName: 'title',
                type: {
                    name: 'String',
                },
            },
            text: {
                serializedName: 'text',
                type: {
                    name: 'String',
                },
            },
            activityTitle: {
                serializedName: 'activityTitle',
                type: {
                    name: 'String',
                },
            },
            activitySubtitle: {
                serializedName: 'activitySubtitle',
                type: {
                    name: 'String',
                },
            },
            activityText: {
                serializedName: 'activityText',
                type: {
                    name: 'String',
                },
            },
            activityImage: {
                serializedName: 'activityImage',
                type: {
                    name: 'String',
                },
            },
            activityImageType: {
                serializedName: 'activityImageType',
                type: {
                    name: 'String',
                },
            },
            markdown: {
                serializedName: 'markdown',
                type: {
                    name: 'Boolean',
                },
            },
            facts: {
                serializedName: 'facts',
                type: {
                    name: 'Sequence',
                    element: {
                        type: {
                            name: 'Composite',
                            className: 'O365ConnectorCardFact',
                        },
                    },
                },
            },
            images: {
                serializedName: 'images',
                type: {
                    name: 'Sequence',
                    element: {
                        type: {
                            name: 'Composite',
                            className: 'O365ConnectorCardImage',
                        },
                    },
                },
            },
            potentialAction: {
                serializedName: 'potentialAction',
                type: {
                    name: 'Sequence',
                    element: {
                        type: {
                            name: 'Composite',
                            className: 'O365ConnectorCardActionBase',
                        },
                    },
                },
            },
        },
    },
};
exports.O365ConnectorCard = {
    serializedName: 'O365ConnectorCard',
    type: {
        name: 'Composite',
        className: 'O365ConnectorCard',
        modelProperties: {
            title: {
                serializedName: 'title',
                type: {
                    name: 'String',
                },
            },
            text: {
                serializedName: 'text',
                type: {
                    name: 'String',
                },
            },
            summary: {
                serializedName: 'summary',
                type: {
                    name: 'String',
                },
            },
            themeColor: {
                serializedName: 'themeColor',
                type: {
                    name: 'String',
                },
            },
            sections: {
                serializedName: 'sections',
                type: {
                    name: 'Sequence',
                    element: {
                        type: {
                            name: 'Composite',
                            className: 'O365ConnectorCardSection',
                        },
                    },
                },
            },
            potentialAction: {
                serializedName: 'potentialAction',
                type: {
                    name: 'Sequence',
                    element: {
                        type: {
                            name: 'Composite',
                            className: 'O365ConnectorCardActionBase',
                        },
                    },
                },
            },
        },
    },
};
exports.O365ConnectorCardViewAction = {
    serializedName: 'O365ConnectorCardViewAction',
    type: {
        name: 'Composite',
        className: 'O365ConnectorCardViewAction',
        modelProperties: Object.assign(Object.assign({}, exports.O365ConnectorCardActionBase.type.modelProperties), { target: {
                serializedName: 'target',
                type: {
                    name: 'Sequence',
                    element: {
                        type: {
                            name: 'String',
                        },
                    },
                },
            } }),
    },
};
exports.O365ConnectorCardOpenUriTarget = {
    serializedName: 'O365ConnectorCardOpenUriTarget',
    type: {
        name: 'Composite',
        className: 'O365ConnectorCardOpenUriTarget',
        modelProperties: {
            os: {
                serializedName: 'os',
                type: {
                    name: 'String',
                },
            },
            uri: {
                serializedName: 'uri',
                type: {
                    name: 'String',
                },
            },
        },
    },
};
exports.O365ConnectorCardOpenUri = {
    serializedName: 'O365ConnectorCardOpenUri',
    type: {
        name: 'Composite',
        className: 'O365ConnectorCardOpenUri',
        modelProperties: Object.assign(Object.assign({}, exports.O365ConnectorCardActionBase.type.modelProperties), { targets: {
                serializedName: 'targets',
                type: {
                    name: 'Sequence',
                    element: {
                        type: {
                            name: 'Composite',
                            className: 'O365ConnectorCardOpenUriTarget',
                        },
                    },
                },
            } }),
    },
};
exports.O365ConnectorCardHttpPOST = {
    serializedName: 'O365ConnectorCardHttpPOST',
    type: {
        name: 'Composite',
        className: 'O365ConnectorCardHttpPOST',
        modelProperties: Object.assign(Object.assign({}, exports.O365ConnectorCardActionBase.type.modelProperties), { body: {
                serializedName: 'body',
                type: {
                    name: 'String',
                },
            } }),
    },
};
exports.O365ConnectorCardInputBase = {
    serializedName: 'O365ConnectorCardInputBase',
    type: {
        name: 'Composite',
        className: 'O365ConnectorCardInputBase',
        modelProperties: {
            type: {
                serializedName: '@type',
                type: {
                    name: 'String',
                },
            },
            id: {
                serializedName: 'id',
                type: {
                    name: 'String',
                },
            },
            isRequired: {
                serializedName: 'isRequired',
                type: {
                    name: 'Boolean',
                },
            },
            title: {
                serializedName: 'title',
                type: {
                    name: 'String',
                },
            },
            value: {
                serializedName: 'value',
                type: {
                    name: 'String',
                },
            },
        },
    },
};
exports.O365ConnectorCardActionCard = {
    serializedName: 'O365ConnectorCardActionCard',
    type: {
        name: 'Composite',
        className: 'O365ConnectorCardActionCard',
        modelProperties: Object.assign(Object.assign({}, exports.O365ConnectorCardActionBase.type.modelProperties), { inputs: {
                serializedName: 'inputs',
                type: {
                    name: 'Sequence',
                    element: {
                        type: {
                            name: 'Composite',
                            className: 'O365ConnectorCardInputBase',
                        },
                    },
                },
            }, actions: {
                serializedName: 'actions',
                type: {
                    name: 'Sequence',
                    element: {
                        type: {
                            name: 'Composite',
                            className: 'O365ConnectorCardActionBase',
                        },
                    },
                },
            } }),
    },
};
exports.O365ConnectorCardTextInput = {
    serializedName: 'O365ConnectorCardTextInput',
    type: {
        name: 'Composite',
        className: 'O365ConnectorCardTextInput',
        modelProperties: Object.assign(Object.assign({}, exports.O365ConnectorCardInputBase.type.modelProperties), { isMultiline: {
                serializedName: 'isMultiline',
                type: {
                    name: 'Boolean',
                },
            }, maxLength: {
                serializedName: 'maxLength',
                type: {
                    name: 'Number',
                },
            } }),
    },
};
exports.O365ConnectorCardDateInput = {
    serializedName: 'O365ConnectorCardDateInput',
    type: {
        name: 'Composite',
        className: 'O365ConnectorCardDateInput',
        modelProperties: Object.assign(Object.assign({}, exports.O365ConnectorCardInputBase.type.modelProperties), { includeTime: {
                serializedName: 'includeTime',
                type: {
                    name: 'Boolean',
                },
            } }),
    },
};
exports.O365ConnectorCardMultichoiceInputChoice = {
    serializedName: 'O365ConnectorCardMultichoiceInputChoice',
    type: {
        name: 'Composite',
        className: 'O365ConnectorCardMultichoiceInputChoice',
        modelProperties: {
            display: {
                serializedName: 'display',
                type: {
                    name: 'String',
                },
            },
            value: {
                serializedName: 'value',
                type: {
                    name: 'String',
                },
            },
        },
    },
};
exports.O365ConnectorCardMultichoiceInput = {
    serializedName: 'O365ConnectorCardMultichoiceInput',
    type: {
        name: 'Composite',
        className: 'O365ConnectorCardMultichoiceInput',
        modelProperties: Object.assign(Object.assign({}, exports.O365ConnectorCardInputBase.type.modelProperties), { choices: {
                serializedName: 'choices',
                type: {
                    name: 'Sequence',
                    element: {
                        type: {
                            name: 'Composite',
                            className: 'O365ConnectorCardMultichoiceInputChoice',
                        },
                    },
                },
            }, style: {
                serializedName: 'style',
                type: {
                    name: 'String',
                },
            }, isMultiSelect: {
                serializedName: 'isMultiSelect',
                type: {
                    name: 'Boolean',
                },
            } }),
    },
};
exports.O365ConnectorCardActionQuery = {
    serializedName: 'O365ConnectorCardActionQuery',
    type: {
        name: 'Composite',
        className: 'O365ConnectorCardActionQuery',
        modelProperties: {
            body: {
                serializedName: 'body',
                type: {
                    name: 'String',
                },
            },
            actionId: {
                serializedName: 'actionId',
                type: {
                    name: 'String',
                },
            },
        },
    },
};
exports.SigninStateVerificationQuery = {
    serializedName: 'SigninStateVerificationQuery',
    type: {
        name: 'Composite',
        className: 'SigninStateVerificationQuery',
        modelProperties: {
            state: {
                serializedName: 'state',
                type: {
                    name: 'String',
                },
            },
        },
    },
};
exports.MessagingExtensionQueryOptions = {
    serializedName: 'MessagingExtensionQueryOptions',
    type: {
        name: 'Composite',
        className: 'MessagingExtensionQueryOptions',
        modelProperties: {
            skip: {
                serializedName: 'skip',
                type: {
                    name: 'Number',
                },
            },
            count: {
                serializedName: 'count',
                type: {
                    name: 'Number',
                },
            },
        },
    },
};
exports.MessagingExtensionParameter = {
    serializedName: 'MessagingExtensionParameter',
    type: {
        name: 'Composite',
        className: 'MessagingExtensionParameter',
        modelProperties: {
            name: {
                serializedName: 'name',
                type: {
                    name: 'String',
                },
            },
            value: {
                serializedName: 'value',
                type: {
                    name: 'Object',
                },
            },
        },
    },
};
exports.MessagingExtensionQuery = {
    serializedName: 'MessagingExtensionQuery',
    type: {
        name: 'Composite',
        className: 'MessagingExtensionQuery',
        modelProperties: {
            commandId: {
                serializedName: 'commandId',
                type: {
                    name: 'String',
                },
            },
            parameters: {
                serializedName: 'parameters',
                type: {
                    name: 'Sequence',
                    element: {
                        type: {
                            name: 'Composite',
                            className: 'MessagingExtensionParameter',
                        },
                    },
                },
            },
            queryOptions: {
                serializedName: 'queryOptions',
                type: {
                    name: 'Composite',
                    className: 'MessagingExtensionQueryOptions',
                },
            },
            state: {
                serializedName: 'state',
                type: {
                    name: 'String',
                },
            },
        },
    },
};
exports.Activity = {
    serializedName: 'Activity',
    type: {
        name: 'Composite',
        className: 'Activity',
        modelProperties: {
            dummyProperty: {
                serializedName: 'dummyProperty',
                type: {
                    name: 'String',
                },
            },
        },
    },
};
exports.MessageActionsPayloadUser = {
    serializedName: 'MessageActionsPayloadUser',
    type: {
        name: 'Composite',
        className: 'MessageActionsPayloadUser',
        modelProperties: {
            userIdentityType: {
                serializedName: 'userIdentityType',
                type: {
                    name: 'String',
                },
            },
            id: {
                serializedName: 'id',
                type: {
                    name: 'String',
                },
            },
            displayName: {
                serializedName: 'displayName',
                type: {
                    name: 'String',
                },
            },
        },
    },
};
exports.MessageActionsPayloadApp = {
    serializedName: 'MessageActionsPayloadApp',
    type: {
        name: 'Composite',
        className: 'MessageActionsPayloadApp',
        modelProperties: {
            applicationIdentityType: {
                serializedName: 'applicationIdentityType',
                type: {
                    name: 'String',
                },
            },
            id: {
                serializedName: 'id',
                type: {
                    name: 'String',
                },
            },
            displayName: {
                serializedName: 'displayName',
                type: {
                    name: 'String',
                },
            },
        },
    },
};
exports.MessageActionsPayloadConversation = {
    serializedName: 'MessageActionsPayloadConversation',
    type: {
        name: 'Composite',
        className: 'MessageActionsPayloadConversation',
        modelProperties: {
            conversationIdentityType: {
                serializedName: 'conversationIdentityType',
                type: {
                    name: 'String',
                },
            },
            id: {
                serializedName: 'id',
                type: {
                    name: 'String',
                },
            },
            displayName: {
                serializedName: 'displayName',
                type: {
                    name: 'String',
                },
            },
        },
    },
};
exports.MessageActionsPayloadFrom = {
    serializedName: 'MessageActionsPayloadFrom',
    type: {
        name: 'Composite',
        className: 'MessageActionsPayloadFrom',
        modelProperties: {
            user: {
                serializedName: 'user',
                type: {
                    name: 'Composite',
                    className: 'MessageActionsPayloadUser',
                },
            },
            application: {
                serializedName: 'application',
                type: {
                    name: 'Composite',
                    className: 'MessageActionsPayloadApp',
                },
            },
            conversation: {
                serializedName: 'conversation',
                type: {
                    name: 'Composite',
                    className: 'MessageActionsPayloadConversation',
                },
            },
        },
    },
};
exports.MessageActionsPayloadBody = {
    serializedName: 'MessageActionsPayload_body',
    type: {
        name: 'Composite',
        className: 'MessageActionsPayloadBody',
        modelProperties: {
            contentType: {
                serializedName: 'contentType',
                type: {
                    name: 'String',
                },
            },
            content: {
                serializedName: 'content',
                type: {
                    name: 'String',
                },
            },
            textContent: {
                serializedName: 'textContent',
                type: {
                    name: 'String',
                },
            },
        },
    },
};
exports.MessageActionsPayloadAttachment = {
    serializedName: 'MessageActionsPayloadAttachment',
    type: {
        name: 'Composite',
        className: 'MessageActionsPayloadAttachment',
        modelProperties: {
            id: {
                serializedName: 'id',
                type: {
                    name: 'String',
                },
            },
            contentType: {
                serializedName: 'contentType',
                type: {
                    name: 'String',
                },
            },
            contentUrl: {
                serializedName: 'contentUrl',
                type: {
                    name: 'String',
                },
            },
            content: {
                serializedName: 'content',
                type: {
                    name: 'Object',
                },
            },
            name: {
                serializedName: 'name',
                type: {
                    name: 'String',
                },
            },
            thumbnailUrl: {
                serializedName: 'thumbnailUrl',
                type: {
                    name: 'String',
                },
            },
        },
    },
};
exports.MessageActionsPayloadMention = {
    serializedName: 'MessageActionsPayloadMention',
    type: {
        name: 'Composite',
        className: 'MessageActionsPayloadMention',
        modelProperties: {
            id: {
                serializedName: 'id',
                type: {
                    name: 'Number',
                },
            },
            mentionText: {
                serializedName: 'mentionText',
                type: {
                    name: 'String',
                },
            },
            mentioned: {
                serializedName: 'mentioned',
                type: {
                    name: 'Composite',
                    className: 'MessageActionsPayloadFrom',
                },
            },
        },
    },
};
exports.MessageActionsPayloadReaction = {
    serializedName: 'MessageActionsPayloadReaction',
    type: {
        name: 'Composite',
        className: 'MessageActionsPayloadReaction',
        modelProperties: {
            reactionType: {
                serializedName: 'reactionType',
                type: {
                    name: 'String',
                },
            },
            createdDateTime: {
                serializedName: 'createdDateTime',
                type: {
                    name: 'String',
                },
            },
            user: {
                serializedName: 'user',
                type: {
                    name: 'Composite',
                    className: 'MessageActionsPayloadFrom',
                },
            },
        },
    },
};
exports.MessageActionsPayload = {
    serializedName: 'MessageActionsPayload',
    type: {
        name: 'Composite',
        className: 'MessageActionsPayload',
        modelProperties: {
            id: {
                serializedName: 'id',
                type: {
                    name: 'String',
                },
            },
            replyToId: {
                serializedName: 'replyToId',
                type: {
                    name: 'String',
                },
            },
            messageType: {
                serializedName: 'messageType',
                type: {
                    name: 'String',
                },
            },
            createdDateTime: {
                serializedName: 'createdDateTime',
                type: {
                    name: 'String',
                },
            },
            lastModifiedDateTime: {
                serializedName: 'lastModifiedDateTime',
                type: {
                    name: 'String',
                },
            },
            deleted: {
                serializedName: 'deleted',
                type: {
                    name: 'Boolean',
                },
            },
            subject: {
                serializedName: 'subject',
                type: {
                    name: 'String',
                },
            },
            summary: {
                serializedName: 'summary',
                type: {
                    name: 'String',
                },
            },
            importance: {
                serializedName: 'importance',
                type: {
                    name: 'String',
                },
            },
            locale: {
                serializedName: 'locale',
                type: {
                    name: 'String',
                },
            },
            from: {
                serializedName: 'from',
                type: {
                    name: 'Composite',
                    className: 'MessageActionsPayloadFrom',
                },
            },
            body: {
                serializedName: 'body',
                type: {
                    name: 'Composite',
                    className: 'MessageActionsPayloadBody',
                },
            },
            attachmentLayout: {
                serializedName: 'attachmentLayout',
                type: {
                    name: 'String',
                },
            },
            attachments: {
                serializedName: 'attachments',
                type: {
                    name: 'Sequence',
                    element: {
                        type: {
                            name: 'Composite',
                            className: 'MessageActionsPayloadAttachment',
                        },
                    },
                },
            },
            mentions: {
                serializedName: 'mentions',
                type: {
                    name: 'Sequence',
                    element: {
                        type: {
                            name: 'Composite',
                            className: 'MessageActionsPayloadMention',
                        },
                    },
                },
            },
            reactions: {
                serializedName: 'reactions',
                type: {
                    name: 'Sequence',
                    element: {
                        type: {
                            name: 'Composite',
                            className: 'MessageActionsPayloadReaction',
                        },
                    },
                },
            },
            linkToMessage: {
                serializedName: 'linkToMessage',
                type: {
                    name: 'String',
                },
            },
        },
    },
};
exports.TaskModuleRequest = {
    serializedName: 'TaskModuleRequest',
    type: {
        name: 'Composite',
        className: 'TaskModuleRequest',
        modelProperties: {
            data: {
                serializedName: 'data',
                type: {
                    name: 'Object',
                },
            },
            context: {
                serializedName: 'context',
                type: {
                    name: 'Composite',
                    className: 'TaskModuleRequestContext',
                },
            },
        },
    },
};
exports.MessagingExtensionAction = {
    serializedName: 'MessagingExtensionAction',
    type: {
        name: 'Composite',
        className: 'MessagingExtensionAction',
        modelProperties: Object.assign(Object.assign({}, exports.TaskModuleRequest.type.modelProperties), { commandId: {
                serializedName: 'commandId',
                type: {
                    name: 'String',
                },
            }, commandContext: {
                serializedName: 'commandContext',
                type: {
                    name: 'String',
                },
            }, botMessagePreviewAction: {
                serializedName: 'botMessagePreviewAction',
                type: {
                    name: 'String',
                },
            }, botActivityPreview: {
                serializedName: 'botActivityPreview',
                type: {
                    name: 'Sequence',
                    element: {
                        type: {
                            name: 'Composite',
                            className: 'Activity',
                        },
                    },
                },
            }, messagePayload: {
                serializedName: 'messagePayload',
                type: {
                    name: 'Composite',
                    className: 'MessageActionsPayload',
                },
            } }),
    },
};
exports.TaskModuleResponseBase = {
    serializedName: 'TaskModuleResponseBase',
    type: {
        name: 'Composite',
        className: 'TaskModuleResponseBase',
        modelProperties: {
            type: {
                serializedName: 'type',
                type: {
                    name: 'String',
                },
            },
        },
    },
};
exports.Attachment = {
    serializedName: 'Attachment',
    type: {
        name: 'Composite',
        className: 'Attachment',
        modelProperties: {
            contentType: {
                serializedName: 'contentType',
                type: {
                    name: 'String',
                },
            },
            contentUrl: {
                serializedName: 'contentUrl',
                type: {
                    name: 'String',
                },
            },
            content: {
                serializedName: 'content',
                type: {
                    name: 'Object',
                },
            },
            name: {
                serializedName: 'name',
                type: {
                    name: 'String',
                },
            },
            thumbnailUrl: {
                serializedName: 'thumbnailUrl',
                type: {
                    name: 'String',
                },
            },
        },
    },
};
exports.MessagingExtensionAttachment = {
    serializedName: 'MessagingExtensionAttachment',
    type: {
        name: 'Composite',
        className: 'MessagingExtensionAttachment',
        modelProperties: Object.assign(Object.assign({}, exports.Attachment.type.modelProperties), { preview: {
                serializedName: 'preview',
                type: {
                    name: 'Composite',
                    className: 'Attachment',
                },
            } }),
    },
};
exports.MessagingExtensionSuggestedAction = {
    serializedName: 'MessagingExtensionSuggestedAction',
    type: {
        name: 'Composite',
        className: 'MessagingExtensionSuggestedAction',
        modelProperties: {
            actions: {
                serializedName: 'actions',
                type: {
                    name: 'Sequence',
                    element: {
                        type: {
                            name: 'Composite',
                            className: 'CardAction',
                        },
                    },
                },
            },
        },
    },
};
exports.MessagingExtensionResult = {
    serializedName: 'MessagingExtensionResult',
    type: {
        name: 'Composite',
        className: 'MessagingExtensionResult',
        modelProperties: {
            attachmentLayout: {
                serializedName: 'attachmentLayout',
                type: {
                    name: 'String',
                },
            },
            type: {
                serializedName: 'type',
                type: {
                    name: 'String',
                },
            },
            attachments: {
                serializedName: 'attachments',
                type: {
                    name: 'Sequence',
                    element: {
                        type: {
                            name: 'Composite',
                            className: 'MessagingExtensionAttachment',
                        },
                    },
                },
            },
            suggestedActions: {
                serializedName: 'suggestedActions',
                type: {
                    name: 'Composite',
                    className: 'MessagingExtensionSuggestedAction',
                },
            },
            text: {
                serializedName: 'text',
                type: {
                    name: 'String',
                },
            },
            activityPreview: {
                serializedName: 'activityPreview',
                type: {
                    name: 'Composite',
                    className: 'Activity',
                },
            },
        },
    },
};
exports.CacheInfo = {
    serializedName: 'cacheInfo',
    type: {
        name: 'Composite',
        className: 'cacheInfo',
        modelProperties: {
            cacheType: {
                serializedName: 'cacheType',
                type: {
                    name: 'String',
                },
            },
            cacheDuration: {
                serializedName: 'cacheDuration',
                type: {
                    name: 'Number',
                },
            },
        },
    },
};
exports.MessagingExtensionActionResponse = {
    serializedName: 'MessagingExtensionActionResponse',
    type: {
        name: 'Composite',
        className: 'MessagingExtensionActionResponse',
        modelProperties: {
            task: {
                serializedName: 'task',
                type: {
                    name: 'Composite',
                    className: 'TaskModuleResponseBase',
                },
            },
            composeExtension: {
                serializedName: 'composeExtension',
                type: {
                    name: 'Composite',
                    className: 'MessagingExtensionResult',
                },
            },
            cacheInfo: {
                serializedName: 'cacheInfo',
                type: {
                    name: 'Composite',
                    className: 'cacheInfo',
                },
            },
        },
    },
};
exports.MessagingExtensionResponse = {
    serializedName: 'MessagingExtensionResponse',
    type: {
        name: 'Composite',
        className: 'MessagingExtensionResponse',
        modelProperties: {
            composeExtension: {
                serializedName: 'composeExtension',
                type: {
                    name: 'Composite',
                    className: 'MessagingExtensionResult',
                },
            },
            cacheInfo: {
                serializedName: 'cacheInfo',
                type: {
                    name: 'Composite',
                    className: 'cacheInfo',
                },
            },
        },
    },
};
exports.FileConsentCard = {
    serializedName: 'FileConsentCard',
    type: {
        name: 'Composite',
        className: 'FileConsentCard',
        modelProperties: {
            description: {
                serializedName: 'description',
                type: {
                    name: 'String',
                },
            },
            sizeInBytes: {
                serializedName: 'sizeInBytes',
                type: {
                    name: 'Number',
                },
            },
            acceptContext: {
                serializedName: 'acceptContext',
                type: {
                    name: 'Object',
                },
            },
            declineContext: {
                serializedName: 'declineContext',
                type: {
                    name: 'Object',
                },
            },
        },
    },
};
exports.FileDownloadInfo = {
    serializedName: 'FileDownloadInfo',
    type: {
        name: 'Composite',
        className: 'FileDownloadInfo',
        modelProperties: {
            downloadUrl: {
                serializedName: 'downloadUrl',
                type: {
                    name: 'String',
                },
            },
            uniqueId: {
                serializedName: 'uniqueId',
                type: {
                    name: 'String',
                },
            },
            fileType: {
                serializedName: 'fileType',
                type: {
                    name: 'String',
                },
            },
            etag: {
                serializedName: 'etag',
                type: {
                    name: 'Object',
                },
            },
        },
    },
};
exports.FileInfoCard = {
    serializedName: 'FileInfoCard',
    type: {
        name: 'Composite',
        className: 'FileInfoCard',
        modelProperties: {
            uniqueId: {
                serializedName: 'uniqueId',
                type: {
                    name: 'String',
                },
            },
            fileType: {
                serializedName: 'fileType',
                type: {
                    name: 'String',
                },
            },
            etag: {
                serializedName: 'etag',
                type: {
                    name: 'Object',
                },
            },
        },
    },
};
exports.FileUploadInfo = {
    serializedName: 'FileUploadInfo',
    type: {
        name: 'Composite',
        className: 'FileUploadInfo',
        modelProperties: {
            name: {
                serializedName: 'name',
                type: {
                    name: 'String',
                },
            },
            uploadUrl: {
                serializedName: 'uploadUrl',
                type: {
                    name: 'String',
                },
            },
            contentUrl: {
                serializedName: 'contentUrl',
                type: {
                    name: 'String',
                },
            },
            uniqueId: {
                serializedName: 'uniqueId',
                type: {
                    name: 'String',
                },
            },
            fileType: {
                serializedName: 'fileType',
                type: {
                    name: 'String',
                },
            },
        },
    },
};
exports.FileConsentCardResponse = {
    serializedName: 'FileConsentCardResponse',
    type: {
        name: 'Composite',
        className: 'FileConsentCardResponse',
        modelProperties: {
            action: {
                serializedName: 'action',
                type: {
                    name: 'String',
                },
            },
            context: {
                serializedName: 'context',
                type: {
                    name: 'Object',
                },
            },
            uploadInfo: {
                serializedName: 'uploadInfo',
                type: {
                    name: 'Composite',
                    className: 'FileUploadInfo',
                },
            },
        },
    },
};
exports.TaskModuleTaskInfo = {
    serializedName: 'TaskModuleTaskInfo',
    type: {
        name: 'Composite',
        className: 'TaskModuleTaskInfo',
        modelProperties: {
            title: {
                serializedName: 'title',
                type: {
                    name: 'String',
                },
            },
            height: {
                serializedName: 'height',
                type: {
                    name: 'Object',
                },
            },
            width: {
                serializedName: 'width',
                type: {
                    name: 'Object',
                },
            },
            url: {
                serializedName: 'url',
                type: {
                    name: 'String',
                },
            },
            card: {
                serializedName: 'card',
                type: {
                    name: 'Composite',
                    className: 'Attachment',
                },
            },
            fallbackUrl: {
                serializedName: 'fallbackUrl',
                type: {
                    name: 'String',
                },
            },
            completionBotId: {
                serializedName: 'completionBotId',
                type: {
                    name: 'String',
                },
            },
        },
    },
};
exports.TaskModuleContinueResponse = {
    serializedName: 'TaskModuleContinueResponse',
    type: {
        name: 'Composite',
        className: 'TaskModuleContinueResponse',
        modelProperties: Object.assign(Object.assign({}, exports.TaskModuleResponseBase.type.modelProperties), { value: {
                serializedName: 'value',
                type: {
                    name: 'Composite',
                    className: 'TaskModuleTaskInfo',
                },
            } }),
    },
};
exports.TaskModuleMessageResponse = {
    serializedName: 'TaskModuleMessageResponse',
    type: {
        name: 'Composite',
        className: 'TaskModuleMessageResponse',
        modelProperties: Object.assign(Object.assign({}, exports.TaskModuleResponseBase.type.modelProperties), { value: {
                serializedName: 'value',
                type: {
                    name: 'String',
                },
            } }),
    },
};
exports.TaskModuleResponse = {
    serializedName: 'TaskModuleResponse',
    type: {
        name: 'Composite',
        className: 'TaskModuleResponse',
        modelProperties: {
            task: {
                serializedName: 'task',
                type: {
                    name: 'Composite',
                    className: 'TaskModuleResponseBase',
                },
            },
            cacheInfo: {
                serializedName: 'cacheInfo',
                type: {
                    name: 'Composite',
                    className: 'cacheInfo',
                },
            },
        },
    },
};
exports.TaskModuleRequestContext = {
    serializedName: 'TaskModuleRequest_context',
    type: {
        name: 'Composite',
        className: 'TaskModuleRequestContext',
        modelProperties: {
            theme: {
                serializedName: 'theme',
                type: {
                    name: 'String',
                },
            },
        },
    },
};
exports.AppBasedLinkQuery = {
    serializedName: 'AppBasedLinkQuery',
    type: {
        name: 'Composite',
        className: 'AppBasedLinkQuery',
        modelProperties: {
            url: {
                serializedName: 'url',
                type: {
                    name: 'String',
                },
            },
            state: {
                serializedName: 'state',
                type: {
                    name: 'String',
                },
            },
        },
    },
};
