// botframework-connector/lib/teams/models/parameters.js
"use strict";
/**
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.tenantId = exports.participantId = exports.meetingId = exports.teamId = void 0;
exports.teamId = {
    parameterPath: 'teamId',
    mapper: {
        required: true,
        serializedName: 'teamId',
        type: {
            name: 'String',
        },
    },
};
exports.meetingId = {
    parameterPath: 'meetingId',
    mapper: {
        required: true,
        serializedName: 'meetingId',
        type: {
            name: 'String',
        },
    },
};
exports.participantId = {
    parameterPath: 'participantId',
    mapper: {
        required: true,
        serializedName: 'participantId',
        type: {
            name: 'String',
        },
    },
};
exports.tenantId = {
    parameterPath: ['options', 'tenantId'],
    mapper: {
        serializedName: 'tenantId',
        type: {
            name: 'String',
        },
    },
};
