// botframework-connector/lib/teams/models/teamsMappers.js
"use strict";
/**
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TeamsMeetingParticipant = exports.TeamsMeetingInfo = exports.TeamsMeetingDetails = exports.TeamsChannelAccount = exports.TeamDetails = exports.Meeting = exports.MessageActionsPayloadConversation = exports.ChannelInfo = exports.ConversationList = void 0;
var mappers_1 = require("./mappers");
Object.defineProperty(exports, "ConversationList", { enumerable: true, get: function () { return mappers_1.ConversationList; } });
Object.defineProperty(exports, "ChannelInfo", { enumerable: true, get: function () { return mappers_1.ChannelInfo; } });
Object.defineProperty(exports, "MessageActionsPayloadConversation", { enumerable: true, get: function () { return mappers_1.MessageActionsPayloadConversation; } });
Object.defineProperty(exports, "Meeting", { enumerable: true, get: function () { return mappers_1.Meeting; } });
Object.defineProperty(exports, "TeamDetails", { enumerable: true, get: function () { return mappers_1.TeamDetails; } });
Object.defineProperty(exports, "TeamsChannelAccount", { enumerable: true, get: function () { return mappers_1.TeamsChannelAccount; } });
Object.defineProperty(exports, "TeamsMeetingDetails", { enumerable: true, get: function () { return mappers_1.TeamsMeetingDetails; } });
Object.defineProperty(exports, "TeamsMeetingInfo", { enumerable: true, get: function () { return mappers_1.TeamsMeetingInfo; } });
Object.defineProperty(exports, "TeamsMeetingParticipant", { enumerable: true, get: function () { return mappers_1.TeamsMeetingParticipant; } });
