// @fluentui/react-northstar/dist/dts/src/components/Header/Header.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Header = exports.headerClassName = void 0;
var customPropTypes = require("@fluentui/react-proptypes");
var React = require("react");
var utils_1 = require("../../utils");
var HeaderDescription_1 = require("./HeaderDescription");
var react_bindings_1 = require("@fluentui/react-bindings");
exports.headerClassName = 'ui-header';
/**
 * A Header organises the content by declaring a content's topic.
 *
 * @accessibility
 * Headings communicate the organization of the content on the page. Web browsers, plug-ins, and assistive technologies can use them to provide in-page navigation.
 * Nest headings by their rank (or level). The most important heading has the rank 1 (<h1>), the least important heading rank 6 (<h6>). Headings with an equal or higher rank start a new section, headings with a lower rank start new subsections that are part of the higher ranked section.
 *
 * Other considerations:
 *  - when the description property is used in header, readers will narrate both header content and description within the element. In addition to that, both will be displayed in the list of headings.
 */
exports.Header = React.forwardRef(function (props, ref) {
    var context = react_bindings_1.useFluentContext();
    var _a = react_bindings_1.useTelemetry(exports.Header.displayName, context.telemetry), setStart = _a.setStart, setEnd = _a.setEnd;
    setStart();
    var children = props.children, content = props.content, variables = props.variables, align = props.align, className = props.className, design = props.design, styles = props.styles, description = props.description, color = props.color;
    var ElementType = react_bindings_1.getElementType(props);
    var unhandledProps = react_bindings_1.useUnhandledProps(exports.Header.handledProps, props);
    var hasChildren = utils_1.childrenExist(children);
    var contentElement = hasChildren ? children : content;
    var getA11yProps = react_bindings_1.useAccessibility(props.accessibility, {
        debugName: exports.Header.displayName,
        rtl: context.rtl,
    });
    var classes = react_bindings_1.useStyles(exports.Header.displayName, {
        className: exports.headerClassName,
        mapPropsToStyles: function () { return ({
            align: align,
            hasDescription: !!description,
            color: color,
        }); },
        mapPropsToInlineStyles: function () { return ({
            className: className,
            design: design,
            styles: styles,
            variables: variables,
        }); },
        rtl: context.rtl,
    }).classes;
    var element = (React.createElement(ElementType, __assign({}, getA11yProps('root', __assign(__assign({ className: classes.root, ref: ref }, unhandledProps), utils_1.rtlTextContainer.getAttributes({
        forElements: [children, content],
        condition: !description,
    })))),
        utils_1.rtlTextContainer.createFor({ element: contentElement, condition: !!description }),
        !hasChildren &&
            HeaderDescription_1.HeaderDescription.create(description, {
                defaultProps: function () { return getA11yProps('description', {}); },
            })));
    setEnd();
    return element;
});
exports.Header.displayName = 'Header';
exports.Header.propTypes = __assign(__assign({}, utils_1.commonPropTypes.createCommon({ color: true })), { description: customPropTypes.itemShorthand, align: customPropTypes.align });
exports.Header.defaultProps = {
    as: 'h1',
};
exports.Header.handledProps = Object.keys(exports.Header.propTypes);
exports.Header.Description = HeaderDescription_1.HeaderDescription;
exports.Header.create = utils_1.createShorthandFactory({ Component: exports.Header, mappedProp: 'content' });
