// @fluentui/react-northstar/dist/dts/src/components/Header/HeaderDescription.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.HeaderDescription = exports.headerDescriptionClassName = void 0;
var React = require("react");
var utils_1 = require("../../utils");
var react_bindings_1 = require("@fluentui/react-bindings");
exports.headerDescriptionClassName = 'ui-header__description';
/**
 * A HeaderDescription provides more detailed information about the Header.
 */
exports.HeaderDescription = React.forwardRef(function (props, ref) {
    var context = react_bindings_1.useFluentContext();
    var _a = react_bindings_1.useTelemetry(exports.HeaderDescription.displayName, context.telemetry), setStart = _a.setStart, setEnd = _a.setEnd;
    setStart();
    var children = props.children, content = props.content, color = props.color, className = props.className, design = props.design, styles = props.styles, variables = props.variables;
    var ElementType = react_bindings_1.getElementType(props);
    var unhandledProps = react_bindings_1.useUnhandledProps(exports.HeaderDescription.handledProps, props);
    var getA11yProps = react_bindings_1.useAccessibility(props.accessibility, {
        debugName: exports.HeaderDescription.displayName,
        rtl: context.rtl,
    });
    var classes = react_bindings_1.useStyles(exports.HeaderDescription.displayName, {
        className: exports.headerDescriptionClassName,
        mapPropsToStyles: function () { return ({
            color: color,
        }); },
        mapPropsToInlineStyles: function () { return ({
            className: className,
            design: design,
            styles: styles,
            variables: variables,
        }); },
        rtl: context.rtl,
    }).classes;
    var element = (React.createElement(ElementType, __assign({}, getA11yProps('root', __assign(__assign({ className: classes.root, ref: ref }, unhandledProps), utils_1.rtlTextContainer.getAttributes({ forElements: [children, content] })))), utils_1.childrenExist(children) ? children : content));
    setEnd();
    return element;
});
exports.HeaderDescription.displayName = 'HeaderDescription';
exports.HeaderDescription.propTypes = __assign({}, utils_1.commonPropTypes.createCommon({ color: true }));
exports.HeaderDescription.defaultProps = {
    as: 'p',
};
exports.HeaderDescription.handledProps = Object.keys(exports.HeaderDescription.propTypes);
exports.HeaderDescription.create = utils_1.createShorthandFactory({
    Component: exports.HeaderDescription,
    mappedProp: 'content',
});
