// @fluentui/react-northstar/dist/commonjs/components/Pill/PillGroup.js
"use strict";

exports.__esModule = true;
exports.PillGroup = exports.pillGroupClassName = void 0;

var React = _interopRequireWildcard(require("react"));

var _accessibility = require("@fluentui/accessibility");

var _reactBindings = require("@fluentui/react-bindings");

var _utils = require("../../utils");

var _pillContext = require("./pillContext");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var pillGroupClassName = 'ui-pills';
/**
 * A PillGroup can be used as container for Pill.
 */

exports.pillGroupClassName = pillGroupClassName;
var PillGroup = /*#__PURE__*/React.forwardRef(function (props, ref) {
  var _getA11Props$unstable;

  var context = (0, _reactBindings.useFluentContext)();

  var _useTelemetry = (0, _reactBindings.useTelemetry)(PillGroup.displayName, context.telemetry),
      setStart = _useTelemetry.setStart,
      setEnd = _useTelemetry.setEnd;

  setStart();
  var accessibility = props.accessibility,
      children = props.children,
      className = props.className,
      design = props.design,
      styles = props.styles,
      variables = props.variables;
  var getA11Props = (0, _reactBindings.useAccessibility)(accessibility, {
    debugName: PillGroup.displayName,
    rtl: context.rtl
  });

  var _useStyles = (0, _reactBindings.useStyles)(PillGroup.displayName, {
    className: pillGroupClassName,
    mapPropsToInlineStyles: function mapPropsToInlineStyles() {
      return {
        className: className,
        design: design,
        styles: styles,
        variables: variables
      };
    },
    rtl: context.rtl
  }),
      classes = _useStyles.classes;

  var ElementType = (0, _reactBindings.getElementType)(props);
  var unhandledProps = (0, _reactBindings.useUnhandledProps)(PillGroup.handledProps, props);
  var pillBehavior = (_getA11Props$unstable = getA11Props.unstable_behaviorDefinition().childBehaviors) == null ? void 0 : _getA11Props$unstable.pill;
  var childProps = React.useMemo(function () {
    return {
      pillBehavior: pillBehavior
    };
  }, [pillBehavior]);
  var element = getA11Props.unstable_wrapWithFocusZone( /*#__PURE__*/React.createElement(ElementType, getA11Props('root', Object.assign({
    className: classes.root,
    ref: ref
  }, _utils.rtlTextContainer.getAttributes({
    forElements: [children]
  }), unhandledProps)), /*#__PURE__*/React.createElement(_pillContext.PillsContextProvider, {
    value: childProps
  }, children)));
  setEnd();
  return element;
});
exports.PillGroup = PillGroup;
PillGroup.displayName = 'PillGroup';
PillGroup.propTypes = _utils.commonPropTypes.createCommon();
PillGroup.defaultProps = {
  accessibility: _accessibility.PillGroupBehavior
};
PillGroup.handledProps = Object.keys(PillGroup.propTypes);
PillGroup.shorthandConfig = {
  mappedProp: 'content'
};
