// @fluentui/accessibility/dist/commonjs/behaviors/Radio/radioGroupBehavior.js
"use strict";

exports.__esModule = true;
exports.radioGroupBehavior = void 0;

var _keyboardKey = require("../../keyboard-key");

/**
 * @description
 * Implements ARIA Radio Group design pattern.
 * @specification
 *  Adds role='radiogroup'. This allows screen readers to handle the component as a radio group.
 *  Triggers 'nextItem' action with 'ArrowDown' or 'ArrowRight' on 'root'.
 *  Triggers 'prevItem' action with 'ArrowUp' or 'ArrowLeft' on 'root'.
 */
var radioGroupBehavior = function radioGroupBehavior() {
  return {
    attributes: {
      root: {
        role: 'radiogroup'
      }
    },
    keyActions: {
      root: {
        nextItem: {
          keyCombinations: [{
            keyCode: _keyboardKey.keyboardKey.ArrowDown
          }, {
            keyCode: _keyboardKey.keyboardKey.ArrowRight
          }]
        },
        prevItem: {
          keyCombinations: [{
            keyCode: _keyboardKey.keyboardKey.ArrowUp
          }, {
            keyCode: _keyboardKey.keyboardKey.ArrowLeft
          }]
        }
      }
    }
  };
};

exports.radioGroupBehavior = radioGroupBehavior;
