// @fluentui/react-northstar/dist/commonjs/themes/teams/components/SplitButton/splitButtonVariables.js
"use strict";

exports.__esModule = true;
exports.splitButtonVariables = void 0;

var _utils = require("../../../../utils");

var splitButtonVariables = function splitButtonVariables(siteVars) {
  return {
    borderRadius: siteVars.borderRadiusMedium,
    borderColor: siteVars.colorScheme.default.border,
    borderColorPrimary: siteVars.colors.brand[500],
    borderColorDisabled: siteVars.colorScheme.brand.foregroundDisabled,
    smallDimension: (0, _utils.pxToRem)(24),
    smallPadding: "0 " + (0, _utils.pxToRem)(8),
    smallMinWidth: '0',
    smallBoxShadow: 'none',
    padding: "0 " + (0, _utils.pxToRem)(12),
    minWidth: '0',
    boxShadow: siteVars.shadowLevel1,
    toggleButtonHeight: (0, _utils.pxToRem)(32),
    dividerColor: siteVars.colorScheme.default.border,
    dividerPrimaryColor: siteVars.colors.white,
    toggleButtonColor: siteVars.colorScheme.default.foreground,
    toggleButtonBackgroundColor: siteVars.colorScheme.default.background,
    toggleButtonBorderRadius: siteVars.borderRadiusMedium,
    toggleButtonBorderColor: siteVars.colorScheme.default.border,
    toggleButtonBoxShadow: siteVars.shadowLevel1,
    toggleButtonColorHover: siteVars.colorScheme.default.foregroundHover,
    toggleButtonBackgroundColorHover: siteVars.colorScheme.default.backgroundHover1,
    toggleButtonBorderColorHover: siteVars.colorScheme.default.borderHover,
    toggleButtonColorActive: siteVars.colorScheme.default.foregroundPressed,
    toggleButtonBackgroundColorActive: siteVars.colorScheme.default.backgroundPressed,
    toggleButtonBorderColorActive: siteVars.colorScheme.default.borderPressed,
    toggleButtonBackgroundColorFocus: undefined,
    toggleButtonBorderColorFocus: undefined,
    toggleButtonColorFocus: undefined,
    toggleButtonPrimaryHoverBackgroundColor: siteVars.colorScheme.brand.backgroundHover,
    toggleButtonPrimaryHoverColor: siteVars.colorScheme.brand.foregroundHover1,
    toggleButtonPrimaryColor: siteVars.colorScheme.brand.foreground4,
    toggleButtonPrimaryBackgroundColor: siteVars.colorScheme.brand.background,
    toggleButtonPrimaryBorderColor: 'transparent',
    toggleButtonPrimaryBoxShadow: siteVars.shadowLevel1Dark,
    toggleButtonPrimaryBackgroundColorActive: siteVars.colorScheme.brand.backgroundPressed,
    toggleButtonPrimaryBackgroundColorFocus: undefined,
    toggleButtonColorDisabled: siteVars.colorScheme.brand.foregroundDisabled,
    toggleButtonBackgroundColorDisabled: siteVars.colorScheme.default.backgroundDisabled,
    toggleButtonIndicatorSize: (0, _utils.pxToRem)(16),
    focusBorderRadius: siteVars.borderRadiusMedium,
    focusBorderWidth: siteVars.borderWidth,
    focusInnerBorderColor: siteVars.focusInnerBorderColor,
    focusOuterBorderColor: siteVars.focusOuterBorderColor,
    focusBorderZIndex: siteVars.zIndexes.foreground
  };
};

exports.splitButtonVariables = splitButtonVariables;
