// @fluentui/react-northstar/dist/dts/src/components/Breadcrumb/BreadcrumbLink.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BreadcrumbLink = exports.breadcrumbLinkClassName = void 0;
var React = require("react");
var react_bindings_1 = require("@fluentui/react-bindings");
var utils_1 = require("../../utils");
var breadcrumbContext_1 = require("./breadcrumbContext");
exports.breadcrumbLinkClassName = 'ui-breadcrumb__link';
/**
 * An BreadcrumbLink represents a anchor to be used inside the Breadcrumb
 */
exports.BreadcrumbLink = react_bindings_1.compose(function (props, ref, composeOptions) {
    var context = react_bindings_1.useFluentContext();
    var _a = react_bindings_1.useTelemetry(composeOptions.displayName, context.telemetry), setStart = _a.setStart, setEnd = _a.setEnd;
    setStart();
    var accessibility = props.accessibility, children = props.children, content = props.content, className = props.className, design = props.design, styles = props.styles, variables = props.variables;
    var size = breadcrumbContext_1.useBreadcrumbContext().size;
    var getA11yProps = react_bindings_1.useAccessibility(accessibility, {
        debugName: composeOptions.displayName,
        rtl: context.rtl,
    });
    var classes = react_bindings_1.useStyles(composeOptions.displayName, {
        className: composeOptions.className,
        composeOptions: composeOptions,
        mapPropsToStyles: function () { return ({
            size: size,
        }); },
        mapPropsToInlineStyles: function () { return ({
            className: className,
            design: design,
            styles: styles,
            variables: variables,
        }); },
        rtl: context.rtl,
        unstable_props: props,
    }).classes;
    var unhandledProps = react_bindings_1.useUnhandledProps(composeOptions.handledProps, props);
    var ElementType = react_bindings_1.getElementType(props);
    var result = (React.createElement(ElementType, __assign({}, getA11yProps('root', __assign({ className: classes.root, ref: ref }, unhandledProps))), react_bindings_1.childrenExist(children) ? children : content));
    setEnd();
    return result;
}, {
    className: exports.breadcrumbLinkClassName,
    displayName: 'BreadcrumbLink',
    handledProps: ['accessibility', 'as', 'children', 'className', 'content', 'design', 'styles', 'variables'],
});
exports.BreadcrumbLink.defaultProps = {
    as: 'a',
};
exports.BreadcrumbLink.propTypes = utils_1.commonPropTypes.createCommon();
