// @fluentui/state/dist/es/managers/carouselManager.js
import { createManager } from '../createManager';
export var createCarouselManager = function createCarouselManager(config) {
  if (config === void 0) {
    config = {};
  }

  return createManager(Object.assign({}, config, {
    state: Object.assign({
      activeIndex: 0,
      ariaLiveOn: false,
      shouldFocusContainer: false,
      isFromKeyboard: false
    }, config.state),
    actions: Object.assign({
      setIndexes: function setIndexes(activeIndex) {
        return function () {
          return {
            activeIndex: activeIndex
          };
        };
      },
      setAriaLiveOn: function setAriaLiveOn(ariaLiveOn) {
        return function () {
          return {
            ariaLiveOn: ariaLiveOn
          };
        };
      },
      setShouldFocusContainer: function setShouldFocusContainer(shouldFocusContainer) {
        return function () {
          return {
            shouldFocusContainer: shouldFocusContainer
          };
        };
      },
      setIsFromKeyboard: function setIsFromKeyboard(isFromKeyboard) {
        return function () {
          return {
            isFromKeyboard: isFromKeyboard
          };
        };
      }
    }, config.actions)
  }));
};
