// @fluentui/react-northstar/dist/dts/src/themes/teams/components/Tree/treeTitleVariables.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.treeTitleVariables = void 0;
var utils_1 = require("../../../../utils");
var treeTitleVariables = function (siteVars) {
    return {
        color: siteVars.colorScheme.default.foreground,
        padding: utils_1.pxToRem(1) + " 0",
        paddingRight: utils_1.pxToRem(20),
        paddingLeft: utils_1.pxToRem(20),
        borderColor: siteVars.colorScheme.default.foreground1,
        borderStyle: 'solid',
        borderRadius: siteVars.borderRadiusMedium,
        borderWidth: utils_1.pxToRem(1),
        indicatorColor: 'transparent',
        selectionIndicatorMargin: "0 0 0 0",
        background: 'transparent',
        hoverBackground: siteVars.colorScheme.default.background1,
        checkedBackground: siteVars.colorScheme.brand.backgroundActive1,
        checkedBorderColor: siteVars.colorScheme.brand.backgroundActive1,
        checkedIndicatorColor: siteVars.colorScheme.default.background,
        disabledBackground: siteVars.colorScheme.default.background,
        disabledBackgroundChecked: siteVars.colorScheme.default.backgroundDisabled,
        disabledBorderColor: siteVars.colorScheme.default.foregroundDisabled1,
        disabledCheckedIndicatorColor: siteVars.colorScheme.default.foregroundDisabled,
    };
};
exports.treeTitleVariables = treeTitleVariables;
