// @fluentui/react-northstar/dist/dts/src/themes/teams/components/Table/tableVariables.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.tableVariables = void 0;
var utils_1 = require("../../../../utils");
var tableVariables = function (siteVariables) {
    return {
        color: siteVariables.colorScheme.default.foreground,
        hoverColor: siteVariables.colorScheme.default.foregroundHover,
        backgroundColor: siteVariables.colorScheme.default.background,
        backgroundHoverColor: siteVariables.colorScheme.default.backgroundHover1,
        defaultRowHeight: utils_1.pxToRem(48),
        compactRowHeight: utils_1.pxToRem(36),
        minCellWidth: '0',
        cellPadding: "0 " + utils_1.pxToRem(8),
        rowPadding: '0',
        headerFontSize: utils_1.pxToRem(12),
        bodyFontSize: utils_1.pxToRem(13),
        borderWidth: utils_1.pxToRem(1),
        headerBorderColor: siteVariables.colorScheme.default.backgroundHover1,
        headerBorderHoverColor: 'transparent',
        headerBorderFocusColor: 'transparent',
        rowBorderColor: siteVariables.colorScheme.default.backgroundHover1,
        rowBorderHoverColor: siteVariables.colorScheme.default.backgroundHover1,
        rowBorderFocusColor: siteVariables.colorScheme.default.borderFocus,
        cellBorderColor: 'transparent',
        cellBorderHoverColor: 'transparent',
        cellBorderFocusColor: siteVariables.colorScheme.default.borderFocus,
    };
};
exports.tableVariables = tableVariables;
