// @fluentui/react-northstar/dist/dts/src/themes/teams/components/Table/tableCellStyles.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.tableCellStyles = void 0;
var getBorderFocusStyles_1 = require("../../getBorderFocusStyles");
exports.tableCellStyles = {
    root: function (_a) {
        var v = _a.variables, siteVariables = _a.theme.siteVariables;
        var borderFocusStyles = getBorderFocusStyles_1.getBorderFocusStyles({
            variables: siteVariables,
        });
        return __assign(__assign({ display: 'flex', flexFlow: 'row nowrap', flexGrow: 1, flexBasis: 0, minWidth: v.minCellWidth, outline: 0, borderWidth: v.borderWidth, borderStyle: 'solid', borderColor: 'transparent' }, borderFocusStyles), { padding: v.cellPadding, position: 'relative', height: '100%' });
    },
    content: function (_a) {
        var p = _a.props;
        return __assign({ alignSelf: 'center' }, (p.truncateContent && {
            display: 'block',
            overflow: 'hidden',
            textOverflow: 'ellipsis',
            whiteSpace: 'nowrap',
        }));
    },
};
