// @fluentui/accessibility/dist/dts/src/behaviors/Card/cardSelectableBehavior.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.cardSelectableBehavior = void 0;
var keyboard_key_1 = require("../../keyboard-key");
var attributes_1 = require("../../attributes");
/**
 * @description
 * Behavior for a selectable card component
 * @specification
 * Adds role='group'.
 * Adds attribute 'aria-disabled=true' based on the property 'disabled'.
 * Adds attribute 'tabIndex=0' to 'root' slot.
 * Adds attribute 'data-is-focusable=true' to 'root' slot.
 * Triggers 'performClick' action with 'Spacebar' on 'root'.
 */
var cardSelectableBehavior = function (props) {
    var _a;
    return ({
        attributes: {
            root: (_a = {
                    role: 'group',
                    tabIndex: 0
                },
                _a[attributes_1.IS_FOCUSABLE_ATTRIBUTE] = true,
                _a['aria-disabled'] = props.disabled,
                _a),
        },
        keyActions: {
            root: {
                performClick: {
                    keyCombinations: [{ keyCode: keyboard_key_1.SpacebarKey }],
                },
            },
        },
    });
};
exports.cardSelectableBehavior = cardSelectableBehavior;
