// @fluentui/react-northstar/dist/dts/test/specs/components/Animation/Animation-test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var commonTests_1 = require("test/specs/commonTests");
var Animation_1 = require("src/components/Animation/Animation");
var utils_1 = require("test/utils");
describe('Animation', function () {
    commonTests_1.isConformant(Animation_1.Animation, {
        testPath: __filename,
        constructorName: 'Animation',
        hasAccessibilityProp: false,
        skipAsPropTests: true,
        requiredProps: { children: React.createElement("div", null) },
        testOptions: { 'consistent-callback-names': { ignoreProps: ['onEntered', 'onExited'] } },
        disabledTests: ['component-has-root-ref', 'component-handles-ref'],
    });
    test('does not throw if children is not passed', function () {
        expect(function () { return utils_1.mountWithProvider(React.createElement(Animation_1.Animation, null)); }).not.toThrowError();
    });
    test('does not throw if children function returns undefined', function () {
        expect(function () { return utils_1.mountWithProvider(React.createElement(Animation_1.Animation, { children: function () { return undefined; } })); }).not.toThrowError();
    });
});
