// @fluentui/react-northstar/dist/es/components/SplitButton/SplitButtonDivider.js
import * as React from 'react';
import * as PropTypes from 'prop-types';
import { useTelemetry, useAccessibility, getElementType, useFluentContext, useUnhandledProps, useStyles } from '@fluentui/react-bindings';
import { commonPropTypes } from '../../utils';
export var SplitButtonDividerClassName = 'ui-splitButtonDivider';
/**
 * A SplitButtonDivider visually segments content.
 */

export var SplitButtonDivider = /*#__PURE__*/function () {
  var SplitButtonDivider = /*#__PURE__*/React.forwardRef(function (props, ref) {
    var context = useFluentContext();

    var _useTelemetry = useTelemetry(SplitButtonDivider.displayName, context.telemetry),
        setStart = _useTelemetry.setStart,
        setEnd = _useTelemetry.setEnd;

    setStart();
    var className = props.className,
        design = props.design,
        styles = props.styles,
        variables = props.variables,
        accessibility = props.accessibility;
    var ElementType = getElementType(props);
    var primary = props.primary;
    var unhandledProps = useUnhandledProps(SplitButtonDivider.handledProps, props);
    var getA11yProps = useAccessibility(accessibility, {
      debugName: SplitButtonDivider.displayName,
      rtl: context.rtl
    });

    var _useStyles = useStyles(SplitButtonDivider.displayName, {
      className: SplitButtonDividerClassName,
      mapPropsToStyles: function mapPropsToStyles() {
        return {
          primary: primary
        };
      },
      mapPropsToInlineStyles: function mapPropsToInlineStyles() {
        return {
          className: className,
          design: design,
          styles: styles,
          variables: variables
        };
      },
      rtl: context.rtl
    }),
        classes = _useStyles.classes;

    var element = /*#__PURE__*/React.createElement(ElementType, getA11yProps('root', Object.assign({
      className: classes.root,
      ref: ref
    }, unhandledProps)));
    setEnd();
    return element;
  });
  SplitButtonDivider.displayName = 'SplitButtonDivider';
  SplitButtonDivider.propTypes = Object.assign({}, commonPropTypes.createCommon(), {
    primary: PropTypes.bool
  });
  SplitButtonDivider.handledProps = Object.keys(SplitButtonDivider.propTypes);
  return SplitButtonDivider;
}();
