// @fluentui/react-northstar/dist/commonjs/themes/teams/components/Toolbar/toolbarVariables.js
"use strict";

exports.__esModule = true;
exports.toolbarVariables = exports.toolbarColorAreas = void 0;

var _utils = require("../../../../utils");

var _colorUtils = require("../../../colorUtils");

var toolbarColorAreas = (0, _utils.stringLiteralsArray)('foreground1', 'background', 'border', 'foregroundActive', 'foregroundHover', 'backgroundHover', 'foregroundDisabled1', // custom
'menuItemForegroundHover', 'menuItemBackgroundHover');
exports.toolbarColorAreas = toolbarColorAreas;

var toolbarVariables = function toolbarVariables(siteVars) {
  return {
    colorScheme: (0, _colorUtils.pickValuesFromColorScheme)((0, _colorUtils.extendColorScheme)(siteVars.colorScheme, {
      default: {
        foregroundHover: siteVars.colorScheme.brand.foregroundHover,
        backgroundHover: 'transparent',
        foregroundActive: siteVars.colorScheme.brand.foregroundActive,
        menuItemForegroundHover: siteVars.colorScheme.default.foregroundHover,
        menuItemBackgroundHover: siteVars.colorScheme.default.backgroundHover
      }
    }), toolbarColorAreas),
    foreground: undefined,
    background: 'transparent',
    dividerBorder: undefined,
    foregroundHover: undefined,
    backgroundHover: undefined,
    foregroundActive: undefined,
    backgroundActive: 'transparent',
    foregroundDisabled: undefined,
    backgroundDisabled: 'transparent',
    itemHeight: (0, _utils.pxToRem)(32),
    itemPadding: '0',
    borderWidth: '2px',
    borderRadius: siteVars.borderRadiusMedium,
    dividerMargin: (0, _utils.pxToRem)(10) + " " + (0, _utils.pxToRem)(4),
    menuMaxWidth: (0, _utils.pxToRem)(312),
    menuPadding: (0, _utils.pxToRem)(8) + " 0",
    menuBackground: undefined,
    menuBorder: undefined,
    menuBorderWidth: '1px',
    menuBorderRadius: siteVars.borderRadiusMedium,
    menuBoxShadow: siteVars.shadowLevel3,
    menuItemForeground: undefined,
    menuItemForegroundHover: undefined,
    menuItemBackgroundHover: undefined,
    menuItemForegroundDisabled: undefined,
    menuItemBackgroundDisabled: 'transparent',
    menuItemPadding: (0, _utils.pxToRem)(5) + " " + (0, _utils.pxToRem)(15),
    menuDividerBorder: undefined,
    menuDividerMargin: (0, _utils.pxToRem)(8) + " 0",
    customItemHorizontalPadding: (0, _utils.pxToRem)(16),
    customItemVerticalPadding: (0, _utils.pxToRem)(4),
    overlayZIndex: siteVars.zIndexes.overlay,
    lineHeightBase: siteVars.lineHeightMedium
  };
};

exports.toolbarVariables = toolbarVariables;
