// @fluentui/react-northstar/dist/commonjs/themes/teams/components/Toolbar/toolbarCustomItemStyles.js
"use strict";

exports.__esModule = true;
exports.toolbarCustomItemStyles = void 0;

var _colors = require("../../colors");

var _getBorderFocusStyles = require("../../getBorderFocusStyles");

var toolbarCustomItemStyles = {
  root: function root(_ref) {
    var p = _ref.props,
        v = _ref.variables,
        siteVariables = _ref.theme.siteVariables;
    var colors = (0, _colors.getColorScheme)(v.colorScheme);
    var borderWidth = siteVariables.borderWidth;
    var borderFocusStyles = (0, _getBorderFocusStyles.getBorderFocusStyles)({
      variables: siteVariables
    });
    return Object.assign({
      position: 'relative',
      backgroundColor: v.background,
      borderColor: 'transparent',
      borderWidth: borderWidth,
      borderStyle: 'solid',
      height: v.itemHeight,
      color: v.foreground || colors.foreground1,
      display: 'flex',
      alignItems: 'center',
      justifyContent: 'center',
      userSelect: 'none'
    }, p.fitted !== true && p.fitted !== 'horizontally' && {
      paddingLeft: v.customItemHorizontalPadding,
      paddingRight: v.customItemHorizontalPadding
    }, p.fitted !== true && p.fitted !== 'vertically' && {
      paddingTop: v.customItemVerticalPadding,
      paddingBottom: v.customItemVerticalPadding
    }, {
      ':focus': borderFocusStyles[':focus'],
      ':focus-visible': borderFocusStyles[':focus-visible']
    });
  }
};
exports.toolbarCustomItemStyles = toolbarCustomItemStyles;
