// @fluentui/react-northstar/dist/commonjs/themes/teams/components/Toolbar/toolbarItemStyles.js
"use strict";

exports.__esModule = true;
exports.toolbarItemStyles = void 0;

var _getIconFillOrOutlineStyles = require("../../getIconFillOrOutlineStyles");

var _colors = require("../../colors");

var _getBorderFocusStyles = require("../../getBorderFocusStyles");

var toolbarItemStyles = {
  root: function root(_ref) {
    var p = _ref.props,
        v = _ref.variables,
        siteVariables = _ref.theme.siteVariables;
    var colors = (0, _colors.getColorScheme)(v.colorScheme);
    var borderWidth = siteVariables.borderWidth;
    var borderFocusStyles = (0, _getBorderFocusStyles.getBorderFocusStyles)({
      variables: siteVariables
    });
    return Object.assign({
      position: 'relative',
      backgroundColor: v.background,
      borderWidth: borderWidth,
      borderStyle: 'solid',
      borderColor: 'transparent',
      borderRadius: v.borderRadius,
      height: v.itemHeight,
      minWidth: v.itemHeight,
      padding: v.itemPadding,
      color: v.foreground || colors.foreground1,
      cursor: 'pointer',
      userSelect: 'none',
      ':focus': borderFocusStyles[':focus']
    }, p.active && Object.assign({
      color: v.foregroundActive || colors.foregroundActive,
      backgroundColor: v.backgroundActive
    }, (0, _getIconFillOrOutlineStyles.getIconFillOrOutlineStyles)({
      outline: false
    })), {
      ':hover': Object.assign({
        color: v.foregroundHover || colors.foregroundHover,
        backgroundColor: v.backgroundHover || colors.backgroundHover
      }, (0, _getIconFillOrOutlineStyles.getIconFillOrOutlineStyles)({
        outline: false
      })),
      ':focus-visible': borderFocusStyles[':focus-visible']
    }, p.disabled && {
      color: v.foregroundDisabled || colors.foregroundDisabled1,
      backgroundColor: v.backgroundDisabled,
      cursor: 'default',
      pointerEvents: 'none',
      ':hover': {// empty to overwrite all existing hover styles
      }
    });
  }
};
exports.toolbarItemStyles = toolbarItemStyles;
