// @fluentui/react-northstar/dist/es/components/Form/FormTextArea.js
import { compose } from '@fluentui/react-bindings';
import { TextArea } from '../TextArea/TextArea';
import { _FormFieldBase } from './utils/formFieldBase';
import { commonPropTypes } from '../../utils';
export var formTextAreaClassName = 'ui-form__textarea';
/**
 * An FormTextArea renders a TextArea wrapped by FormField.
 */

export var FormTextArea = /*#__PURE__*/function () {
  var FormTextArea = compose(_FormFieldBase, {
    className: formTextAreaClassName,
    displayName: 'FormTextArea',
    overrideStyles: true,
    slots: {
      control: TextArea
    },
    slotProps: function slotProps(_ref) {
      var errorMessage = _ref.errorMessage,
          required = _ref.required;
      return {
        label: {
          required: required
        },
        control: {
          error: !!errorMessage
        },
        message: {
          error: !!errorMessage
        }
      };
    }
  });
  FormTextArea.propTypes = commonPropTypes.createCommon({
    children: false,
    content: false
  });
  return FormTextArea;
}();
