// @fluentui/react-northstar/dist/es/components/Form/FormDatepicker.js
import { compose } from '@fluentui/react-bindings';
import { commonPropTypes } from '../../utils';
import { Datepicker } from '../Datepicker/Datepicker';
import { _FormFieldBase } from './utils/formFieldBase';
export var formDatepickerClassName = 'ui-form__datepicker';
/**
 * An FormDatepicker renders a Datepicker wrapped by FormField.
 */

export var FormDatepicker = /*#__PURE__*/function () {
  var FormDatepicker = compose(_FormFieldBase, {
    className: formDatepickerClassName,
    displayName: 'FormDatepicker',
    overrideStyles: true,
    slots: {
      control: Datepicker
    },
    slotProps: function slotProps(_ref) {
      var errorMessage = _ref.errorMessage,
          required = _ref.required;
      return {
        control: {
          required: required
        },
        message: {
          error: !!errorMessage
        }
      };
    }
  });
  FormDatepicker.propTypes = commonPropTypes.createCommon({
    accessibility: false,
    children: false,
    content: false
  });
  return FormDatepicker;
}();
