// @fluentui/react-northstar/dist/es/components/Form/FormCheckbox.js
import { compose } from '@fluentui/react-bindings';
import { commonPropTypes } from '../../utils';
import { Checkbox } from '../Checkbox/Checkbox';
import { _FormFieldBase } from './utils/formFieldBase';
export var formCheckboxClassName = 'ui-form__checkbox';
/**
 * An FormCheckbox renders a Checkbox wrapped by FormField.
 */

export var FormCheckbox = /*#__PURE__*/function () {
  var FormCheckbox = compose(_FormFieldBase, {
    className: formCheckboxClassName,
    displayName: 'FormCheckbox',
    overrideStyles: true,
    handledProps: ['label'],
    slots: {
      label: function label() {
        return null;
      },
      control: Checkbox
    },
    slotProps: function slotProps(_ref) {
      var label = _ref.label,
          errorMessage = _ref.errorMessage;
      return {
        control: {
          label: label
        },
        message: {
          error: !!errorMessage
        }
      };
    }
  });
  FormCheckbox.propTypes = commonPropTypes.createCommon({
    content: false,
    children: false
  });
  return FormCheckbox;
}();
