// @fluentui/react-northstar/dist/es/components/Skeleton/Skeleton.js
import * as PropTypes from 'prop-types';
import { compose } from '@fluentui/react-bindings';
import { commonPropTypes } from '../../utils';
import { Box } from '../Box/Box';
import { SkeletonLine } from './SkeletonLine';
import { SkeletonShape } from './SkeletonShape';
import { SkeletonButton } from './SkeletonButton';
import { SkeletonText } from './SkeletonText';
import { SkeletonInput } from './SkeletonInput';
import { SkeletonAvatar } from './SkeletonAvatar';
export var skeletonClassName = 'ui-skeleton';
/**
 * A Skeleton is a component to be placed while the content is loading.
 */

export var Skeleton = /*#__PURE__*/function () {
  var Skeleton = compose(Box, {
    className: skeletonClassName,
    displayName: 'Skeleton',
    overrideStyles: true,
    shorthandConfig: {},
    handledProps: ['animation'],
    mapPropsToStylesProps: function mapPropsToStylesProps(_ref) {
      var animation = _ref.animation;
      return {
        animation: animation
      };
    }
  });
  Skeleton.Line = SkeletonLine;
  Skeleton.Shape = SkeletonShape;
  Skeleton.Button = SkeletonButton;
  Skeleton.Text = SkeletonText;
  Skeleton.Input = SkeletonInput;
  Skeleton.Avatar = SkeletonAvatar;
  Skeleton.propTypes = Object.assign({}, commonPropTypes.createCommon({
    accessibility: false
  }), {
    animation: PropTypes.oneOf(['wave', 'pulse'])
  });
  return Skeleton;
}();
