// @fluentui/accessibility/dist/commonjs/behaviors/Checkbox/checkboxBehavior.js
"use strict";

exports.__esModule = true;
exports.checkboxBehavior = void 0;

var _keyboardKey = require("../../keyboard-key");

var _attributes = require("../../attributes");

/**
 * @specification
 * Adds role='checkbox'. This allows screen readers to handle the component as a checkbox button.
 * Adds attribute 'aria-checked=true' based on the property 'checked'.
 * Adds attribute 'aria-disabled=true' based on the property 'disabled'.
 * Adds attribute 'tabIndex=0' to 'root' slot.
 * Adds attribute 'data-is-focusable=true' to 'root' slot.
 */
var checkboxBehavior = function checkboxBehavior(props) {
  var _root;

  return {
    attributes: {
      root: (_root = {
        'aria-checked': props.checked === 'mixed' ? 'mixed' : !!props.checked,
        'aria-disabled': props.disabled,
        role: 'checkbox',
        tabIndex: 0
      }, _root[_attributes.IS_FOCUSABLE_ATTRIBUTE] = true, _root)
    },
    keyActions: {
      root: {
        performClick: {
          keyCombinations: [{
            keyCode: _keyboardKey.SpacebarKey
          }]
        }
      }
    }
  };
};

exports.checkboxBehavior = checkboxBehavior;
