// @fluentui/react-northstar/dist/dts/src/themes/teams/components/Provider/providerStyles.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.providerStyles = void 0;
exports.providerStyles = {
    root: function (_a) {
        var v = _a.variables;
        return ({
            background: v.background,
            color: v.color,
            textAlign: 'left',
            '& ::-webkit-scrollbar': {
                height: v.scrollbarHeight,
                width: v.scrollbarWidth,
                ':disabled': {
                    display: 'none',
                },
            },
            '& ::-webkit-scrollbar-thumb': {
                borderRadius: v.scrollbarThumbBorderRadius,
                border: "solid " + v.scrollbarThumbBorderSize + " transparent",
                backgroundClip: 'content-box',
                backgroundColor: v.scrollbarThumbBackgroundColor,
                ':hover': {
                    backgroundColor: v.scrollbarThumbHoverBackgroundColor,
                    border: "solid " + v.scrollbarThumbHoverBorderSize + " transparent",
                },
            },
            '& ::-webkit-scrollbar-track': {
                background: 'transparent',
            },
        });
    },
};
