// @fluentui/react-northstar/dist/dts/src/themes/teams-high-contrast/colors.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.categoryColorScheme = exports.colorScheme = exports.red = exports.accessibleCyan = exports.accessibleGreen = exports.accessibleYellow = void 0;
var siteVariables_1 = require("../teams/siteVariables");
var colors_1 = require("../teams/colors");
exports.accessibleYellow = '#ffff01';
exports.accessibleGreen = '#3ff23f'; // always disabled color in high contrast
exports.accessibleCyan = '#1aebff';
exports.red = '#f00';
// COLOR SCHEME
var createColorScheme = function (customValues) {
    if (customValues === void 0) { customValues = {}; }
    return __assign({ foreground: siteVariables_1.colors.white, foreground1: siteVariables_1.colors.black, background: siteVariables_1.colors.black, background1: siteVariables_1.colors.white, border: siteVariables_1.colors.white, shadow: siteVariables_1.colors.white, foregroundHover: exports.accessibleYellow, backgroundHover: siteVariables_1.colors.black, borderHover: exports.accessibleCyan, shadowHover: exports.accessibleCyan, foregroundActive: exports.accessibleCyan, backgroundActive: siteVariables_1.colors.black, borderActive: exports.accessibleCyan, shadowActive: exports.accessibleCyan, foregroundFocus: siteVariables_1.colors.black, backgroundFocus: exports.accessibleCyan, borderFocus: siteVariables_1.colors.white, shadowFocus: siteVariables_1.colors.white, foregroundPressed: siteVariables_1.colors.black, backgroundPressed: exports.accessibleCyan, borderPressed: siteVariables_1.colors.white, shadowPressed: siteVariables_1.colors.white, foregroundDisabled: siteVariables_1.colors.black, backgroundDisabled: exports.accessibleGreen, borderDisabled: siteVariables_1.colors.black, shadowDisabled: siteVariables_1.colors.black }, customValues);
};
exports.colorScheme = {
    default: colors_1.createColorScheme({
        foreground: siteVariables_1.colors.white,
        foreground1: siteVariables_1.colors.white,
        foreground2: siteVariables_1.colors.white,
        foreground3: siteVariables_1.colors.white,
        foreground4: siteVariables_1.colors.black,
        foreground5: siteVariables_1.colors.grey[600],
        foreground6: siteVariables_1.colors.grey[750],
        foreground7: siteVariables_1.colors.black,
        foreground8: siteVariables_1.colors.black,
        foreground9: siteVariables_1.colors.white,
        background: siteVariables_1.colors.black,
        background1: siteVariables_1.colors.black,
        background2: siteVariables_1.colors.black,
        background3: siteVariables_1.colors.black,
        background4: siteVariables_1.colors.black,
        background5: exports.accessibleYellow,
        background6: siteVariables_1.colors.black,
        border: siteVariables_1.colors.white,
        border1: siteVariables_1.colors.white,
        border2: siteVariables_1.colors.white,
        border3: siteVariables_1.colors.white,
        borderTransparent: siteVariables_1.colors.white,
        borderTransparentActive: exports.accessibleCyan,
        shadow: siteVariables_1.colors.black,
        shadowHover: siteVariables_1.colors.black,
        foregroundHover: siteVariables_1.colors.black,
        foregroundHover1: siteVariables_1.colors.black,
        foregroundHover2: exports.accessibleCyan,
        backgroundHover: exports.accessibleCyan,
        backgroundHover1: exports.accessibleCyan,
        backgroundHover2: exports.accessibleCyan,
        backgroundHover3: siteVariables_1.colors.black,
        backgroundHover4: siteVariables_1.colors.black,
        borderHover: exports.accessibleCyan,
        foregroundPressed: siteVariables_1.colors.black,
        backgroundPressed: exports.accessibleCyan,
        backgroundPressed3: exports.accessibleCyan,
        borderPressed: exports.accessibleCyan,
        foregroundActive: exports.accessibleCyan,
        foregroundActive1: siteVariables_1.colors.black,
        backgroundActive: siteVariables_1.colors.black,
        backgroundActive1: exports.accessibleCyan,
        backgroundActive2: exports.accessibleCyan,
        borderActive: exports.accessibleCyan,
        borderActive1: exports.accessibleCyan,
        borderActive2: exports.accessibleCyan,
        borderActive3: exports.accessibleCyan,
        borderActive4: exports.accessibleCyan,
        foregroundFocus: siteVariables_1.colors.black,
        foregroundFocus1: siteVariables_1.colors.black,
        foregroundFocus2: siteVariables_1.colors.black,
        foregroundFocus3: siteVariables_1.colors.black,
        backgroundFocus: exports.accessibleCyan,
        backgroundFocus1: exports.accessibleCyan,
        backgroundFocus2: exports.accessibleCyan,
        backgroundFocus3: exports.accessibleCyan,
        borderFocusWithin: siteVariables_1.colors.black,
        borderFocus: exports.accessibleCyan,
        foregroundDisabled1: exports.accessibleGreen,
        foregroundDisabled: siteVariables_1.colors.black,
        backgroundDisabled: exports.accessibleGreen,
        backgroundDisabled1: siteVariables_1.colors.black,
        backgroundDisabled2: exports.accessibleGreen,
        backgroundDisabled3: exports.accessibleGreen,
        borderDisabled: exports.accessibleGreen,
    }),
    brand: colors_1.createColorScheme({
        foreground: siteVariables_1.colors.white,
        foreground1: exports.accessibleYellow,
        foreground2: exports.accessibleYellow,
        foreground3: exports.accessibleYellow,
        foreground4: siteVariables_1.colors.black,
        foreground5: siteVariables_1.colors.black,
        background: siteVariables_1.colors.white,
        background1: siteVariables_1.colors.black,
        background2: siteVariables_1.colors.black,
        background3: siteVariables_1.colors.black,
        background4: siteVariables_1.colors.black,
        background5: siteVariables_1.colors.white,
        background6: siteVariables_1.colors.white,
        border: siteVariables_1.colors.white,
        border1: siteVariables_1.colors.white,
        border2: siteVariables_1.colors.white,
        shadow: siteVariables_1.colors.black,
        shadowHover: siteVariables_1.colors.black,
        foregroundHover: siteVariables_1.colors.black,
        foregroundHover1: siteVariables_1.colors.black,
        foregroundHover2: siteVariables_1.colors.black,
        borderHover: exports.accessibleCyan,
        backgroundHover: exports.accessibleCyan,
        backgroundHover1: exports.accessibleCyan,
        backgroundHover2: exports.accessibleYellow,
        backgroundHover3: siteVariables_1.colors.black,
        foregroundPressed: siteVariables_1.colors.black,
        foregroundPressed1: siteVariables_1.colors.black,
        backgroundPressed: exports.accessibleCyan,
        backgroundPressed1: exports.accessibleCyan,
        backgroundPressed2: siteVariables_1.colors.black,
        borderPressed: exports.accessibleCyan,
        foregroundActive: exports.accessibleCyan,
        foregroundActive1: siteVariables_1.colors.black,
        foregroundActive2: exports.accessibleCyan,
        backgroundActive: siteVariables_1.colors.black,
        backgroundActive1: exports.accessibleCyan,
        borderActive: exports.accessibleCyan,
        borderActive1: exports.accessibleCyan,
        borderActive2: exports.accessibleCyan,
        foregroundFocus: siteVariables_1.colors.black,
        foregroundFocus1: siteVariables_1.colors.black,
        foregroundFocus2: siteVariables_1.colors.black,
        foregroundFocus3: siteVariables_1.colors.black,
        foregroundFocus4: siteVariables_1.colors.black,
        backgroundFocus: exports.accessibleCyan,
        backgroundFocus1: exports.accessibleCyan,
        backgroundFocus2: exports.accessibleCyan,
        backgroundFocus3: exports.accessibleCyan,
        borderFocus: exports.accessibleCyan,
        borderFocusWithin: siteVariables_1.colors.black,
        borderFocus1: exports.accessibleCyan,
        foregroundDisabled: siteVariables_1.colors.black,
        foregroundDisabled1: exports.accessibleGreen,
        backgroundDisabled: exports.accessibleGreen,
        backgroundDisabled1: siteVariables_1.colors.black,
        borderDisabled: exports.accessibleGreen,
    }),
    black: createColorScheme(),
    white: createColorScheme(),
    grey: createColorScheme(),
    green: colors_1.createColorScheme({
        foreground: siteVariables_1.colors.white,
        foreground1: siteVariables_1.colors.black,
        foreground2: siteVariables_1.colors.white,
        foreground3: siteVariables_1.colors.black,
        background: siteVariables_1.colors.white,
        background1: siteVariables_1.colors.white,
        background2: siteVariables_1.colors.black,
        border: siteVariables_1.colors.white,
        border1: siteVariables_1.colors.white,
        shadow: undefined,
        foregroundHover: undefined,
        backgroundHover: undefined,
        borderHover: undefined,
        shadowHover: undefined,
        foregroundActive: undefined,
        backgroundActive: undefined,
        borderActive: undefined,
        shadowActive: undefined,
        foregroundFocus: undefined,
        backgroundFocus: undefined,
        borderFocus: undefined,
        shadowFocus: undefined,
        foregroundPressed: undefined,
        backgroundPressed: undefined,
        borderPressed: undefined,
        shadowPressed: undefined,
        foregroundDisabled: undefined,
        backgroundDisabled: undefined,
        borderDisabled: undefined,
        shadowDisabled: undefined,
    }),
    orange: colors_1.createColorScheme({
        foreground: exports.accessibleYellow,
        foreground1: exports.accessibleYellow,
        foreground2: siteVariables_1.colors.black,
        background: exports.accessibleYellow,
        background1: siteVariables_1.colors.white,
        border: exports.accessibleCyan,
    }),
    pink: colors_1.createColorScheme({
        foreground: siteVariables_1.colors.white,
        foreground1: siteVariables_1.colors.white,
        background: siteVariables_1.colors.black,
        border: siteVariables_1.colors.white,
    }),
    red: colors_1.createColorScheme({
        foreground: siteVariables_1.colors.white,
        foreground1: siteVariables_1.colors.black,
        foreground2: siteVariables_1.colors.black,
        background: siteVariables_1.colors.white,
        background1: siteVariables_1.colors.black,
        background2: siteVariables_1.colors.black,
        background3: exports.accessibleYellow,
        border: siteVariables_1.colors.white,
        border1: siteVariables_1.colors.white,
        foregroundHover: siteVariables_1.colors.black,
        backgroundHover: exports.accessibleCyan,
        backgroundHover1: exports.accessibleCyan,
        foregroundPressed: siteVariables_1.colors.black,
        backgroundPressed: exports.accessibleCyan,
    }),
    yellow: colors_1.createColorScheme({
        foreground: siteVariables_1.colors.white,
        foreground1: siteVariables_1.colors.black,
        foreground2: siteVariables_1.colors.black,
        foreground3: siteVariables_1.colors.black,
        foreground4: siteVariables_1.colors.white,
        background: siteVariables_1.colors.white,
        background1: 'transparent',
        background2: siteVariables_1.colors.white,
        background3: siteVariables_1.colors.black,
        border: siteVariables_1.colors.white,
        border1: siteVariables_1.colors.white,
    }),
    silver: colors_1.createColorScheme({
        foreground: siteVariables_1.colors.white,
        foreground1: siteVariables_1.colors.white,
        foregroundHover: siteVariables_1.colors.black,
        foregroundPressed: siteVariables_1.colors.black,
        background: siteVariables_1.colors.black,
        border: siteVariables_1.colors.white,
        backgroundHover: exports.accessibleCyan,
        borderHover: siteVariables_1.colors.white,
        backgroundPressed: exports.accessibleCyan,
        borderPressed: siteVariables_1.colors.white,
        foregroundDisabled: exports.accessibleGreen,
        backgroundDisabled: siteVariables_1.colors.black,
        borderDisabled: exports.accessibleGreen,
    }),
    onyx: colors_1.createColorScheme({
        background: siteVariables_1.colors.black,
        background1: siteVariables_1.colors.black,
        background2: siteVariables_1.colors.white,
        background3: siteVariables_1.colors.white,
        backgroundHover: exports.accessibleCyan,
        backgroundPressed: exports.accessibleCyan,
        border: siteVariables_1.colors.white,
        border1: siteVariables_1.colors.white,
        border2: siteVariables_1.colors.white,
    }),
    amethyst: colors_1.createColorScheme({
        background: siteVariables_1.colors.silver[900],
        backgroundHover: exports.accessibleCyan,
        backgroundHover1: exports.accessibleCyan,
        backgroundActive: exports.accessibleCyan,
    }),
};
var createCategoryColorScheme = function (customValues) {
    if (customValues === void 0) { customValues = {}; }
    return __assign({ foreground: siteVariables_1.colors.black, foreground1: exports.accessibleYellow, background: exports.accessibleYellow, borderActive: exports.accessibleCyan }, customValues);
};
exports.categoryColorScheme = {
    redDark: createCategoryColorScheme(),
    red: createCategoryColorScheme(),
    orangeDark: createCategoryColorScheme(),
    orange: createCategoryColorScheme(),
    orangeLight: createCategoryColorScheme(),
    yellowDark: createCategoryColorScheme(),
    yellow: createCategoryColorScheme(),
    brown: createCategoryColorScheme(),
    oliveDark: createCategoryColorScheme(),
    olive: createCategoryColorScheme(),
    greenDark: createCategoryColorScheme(),
    green: createCategoryColorScheme(),
    tealDark: createCategoryColorScheme(),
    teal: createCategoryColorScheme(),
    tealLight: createCategoryColorScheme(),
    blueDark: createCategoryColorScheme(),
    blue: createCategoryColorScheme(),
    purpleDark: createCategoryColorScheme(),
    purple: createCategoryColorScheme(),
    maroon: createCategoryColorScheme(),
    pink: createCategoryColorScheme(),
    smokeDark: createCategoryColorScheme(),
    smokeLight: createCategoryColorScheme(),
    steelDark: createCategoryColorScheme(),
    steelLight: createCategoryColorScheme(),
    neon: createCategoryColorScheme(),
    formatting: {
        foreground1: undefined,
        background1: undefined,
        foreground2: undefined,
        background2: undefined,
        foreground3: undefined,
        background3: undefined,
        foreground4: undefined,
        background4: undefined,
        foreground5: undefined,
        background5: undefined,
        foreground6: undefined,
        background6: undefined,
        foreground7: undefined,
        background7: undefined,
        foreground8: undefined,
        background8: undefined,
    },
};
