// @fluentui/react-northstar/dist/dts/src/themes/teams-high-contrast/siteVariables.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.bodyColor = exports.bodyBackground = exports.shadow64 = exports.shadow28 = exports.shadow16 = exports.shadow8 = exports.shadow4 = exports.shadow2 = exports.focusOuterBorderColor = exports.focusInnerBorderColor = void 0;
var siteVariables_1 = require("../teams/siteVariables");
var colors_1 = require("./colors");
__exportStar(require("./colors"), exports);
//
// BORDER STYLES
//
exports.focusInnerBorderColor = siteVariables_1.colors.black;
exports.focusOuterBorderColor = colors_1.accessibleCyan;
//
// SHADOW LEVELS
//
exports.shadow2 = '0 0 2px rgba(0, 0, 0, .24), 0 1px 2px rgba(0, 0, 0, .28)';
exports.shadow4 = '0 0 2px rgba(0, 0, 0, .24), 0 2px 4px rgba(0, 0, 0, .28)';
exports.shadow8 = '0 0 2px rgba(0, 0, 0, .24), 0 4px 8px rgba(0, 0, 0, .28)';
exports.shadow16 = '0 0 2px rgba(0, 0, 0, .24), 0 8px 16px rgba(0, 0, 0, .28)';
exports.shadow28 = '0 0 8px rgba(0, 0, 0, .40), 0 14px 28px rgba(0, 0, 0, .48)';
exports.shadow64 = '0 0 8px rgba(0, 0, 0, .40), 0 32px 64px rgba(0, 0, 0, .48)';
//
// SEMANTIC ASSIGNMENTS
//
exports.bodyBackground = siteVariables_1.colors.black;
exports.bodyColor = siteVariables_1.colors.white;
