// @fluentui/accessibility/dist/es/behaviors/MenuButton/menuButtonBehavior.js
import _includes from "lodash/includes";
import _merge from "lodash/merge";
import { keyboardKey } from '../../keyboard-key';
import { popupBehavior } from '../Popup/popupBehavior';
/**
 * @description
 * Implements ARIA [MenuButton](https://www.w3.org/TR/wai-aria-practices/#menubutton) design pattern.
 */

export var menuButtonBehavior = function menuButtonBehavior(props) {
  var behavior = popupBehavior(Object.assign({}, props, {
    inline: !props.contextMenu
  }));
  return _merge(behavior, {
    attributes: {
      trigger: Object.assign({
        'aria-controls': props.open ? props.menuId : undefined,
        'aria-expanded': props.open && !props.contextMenu || undefined,
        'aria-haspopup': props.contextMenu ? undefined : 'true',
        id: props.triggerId
      }, !props.contextMenu && props.open && {
        tabIndex: -1
      }),
      menu: {
        'aria-labelledby': props.triggerId,
        id: props.menuId
      }
    },
    keyActions: {
      root: Object.assign({}, props.open ? {
        closeMenu: {
          keyCombinations: [{
            keyCode: keyboardKey.Tab,
            shiftKey: false
          }, {
            keyCode: keyboardKey.Tab,
            shiftKey: true
          }]
        }
      } : _includes(props.on, 'click') && {
        openAndFocusFirst: {
          keyCombinations: [{
            keyCode: keyboardKey.ArrowDown
          }]
        },
        openAndFocusLast: {
          keyCombinations: [{
            keyCode: keyboardKey.ArrowUp
          }]
        }
      })
    }
  });
};
