// @fluentui/react-northstar/dist/es/components/Card/CardPreview.js
import { getElementType, useUnhandledProps, useAccessibility, useStyles, useTelemetry, useFluentContext } from '@fluentui/react-bindings';
import * as PropTypes from 'prop-types';
import * as React from 'react';
import { commonPropTypes, createShorthandFactory } from '../../utils';
export var cardPreviewClassName = 'ui-card__preview';
/**
 * A CardPreview is used to display data Card preview.
 */

export var CardPreview = /*#__PURE__*/function () {
  var CardPreview = /*#__PURE__*/React.forwardRef(function (props, ref) {
    var context = useFluentContext();

    var _useTelemetry = useTelemetry(CardPreview.displayName, context.telemetry),
        setStart = _useTelemetry.setStart,
        setEnd = _useTelemetry.setEnd;

    setStart();
    var className = props.className,
        design = props.design,
        styles = props.styles,
        variables = props.variables,
        children = props.children,
        horizontal = props.horizontal,
        fitted = props.fitted;
    var ElementType = getElementType(props);
    var unhandledProps = useUnhandledProps(CardPreview.handledProps, props);
    var getA11yProps = useAccessibility(props.accessibility, {
      debugName: CardPreview.displayName,
      rtl: context.rtl
    });

    var _useStyles = useStyles(CardPreview.displayName, {
      className: cardPreviewClassName,
      mapPropsToStyles: function mapPropsToStyles() {
        return {
          horizontal: horizontal,
          fitted: fitted
        };
      },
      mapPropsToInlineStyles: function mapPropsToInlineStyles() {
        return {
          className: className,
          design: design,
          styles: styles,
          variables: variables
        };
      },
      rtl: context.rtl
    }),
        classes = _useStyles.classes;

    var element = /*#__PURE__*/React.createElement(ElementType, getA11yProps('root', Object.assign({
      className: classes.root,
      ref: ref
    }, unhandledProps)), children);
    setEnd();
    return element;
  });
  CardPreview.displayName = 'CardPreview';
  CardPreview.propTypes = Object.assign({}, commonPropTypes.createCommon(), {
    horizontal: PropTypes.bool,
    fitted: PropTypes.bool
  });
  CardPreview.handledProps = Object.keys(CardPreview.propTypes);
  CardPreview.create = createShorthandFactory({
    Component: CardPreview
  });
  return CardPreview;
}();
