// @fluentui/react-northstar/dist/es/components/Card/CardFooter.js
import { getElementType, useUnhandledProps, useAccessibility, useFluentContext, useStyles, useTelemetry } from '@fluentui/react-bindings';
import * as PropTypes from 'prop-types';
import * as React from 'react';
import { commonPropTypes, createShorthandFactory } from '../../utils';
export var cardFooterClassName = 'ui-card__footer';
/**
 * A CardFooter is used to display data in Card component footer
 */

export var CardFooter = /*#__PURE__*/function () {
  var CardFooter = /*#__PURE__*/React.forwardRef(function (props, ref) {
    var context = useFluentContext();

    var _useTelemetry = useTelemetry(CardFooter.displayName, context.telemetry),
        setStart = _useTelemetry.setStart,
        setEnd = _useTelemetry.setEnd;

    setStart();
    var className = props.className,
        design = props.design,
        styles = props.styles,
        variables = props.variables,
        children = props.children,
        fitted = props.fitted;
    var ElementType = getElementType(props);
    var unhandledProps = useUnhandledProps(CardFooter.handledProps, props);
    var getA11yProps = useAccessibility(props.accessibility, {
      debugName: CardFooter.displayName,
      rtl: context.rtl
    });

    var _useStyles = useStyles(CardFooter.displayName, {
      className: cardFooterClassName,
      mapPropsToStyles: function mapPropsToStyles() {
        return {
          fitted: fitted
        };
      },
      mapPropsToInlineStyles: function mapPropsToInlineStyles() {
        return {
          className: className,
          design: design,
          styles: styles,
          variables: variables
        };
      },
      rtl: context.rtl
    }),
        classes = _useStyles.classes;

    var element = /*#__PURE__*/React.createElement(ElementType, getA11yProps('root', Object.assign({
      className: classes.root,
      ref: ref
    }, unhandledProps)), children);
    setEnd();
    return element;
  });
  CardFooter.displayName = 'CardFooter';
  CardFooter.propTypes = Object.assign({}, commonPropTypes.createCommon(), {
    fitted: PropTypes.bool
  });
  CardFooter.handledProps = Object.keys(CardFooter.propTypes);
  CardFooter.create = createShorthandFactory({
    Component: CardFooter
  });
  return CardFooter;
}();
