// @fluentui/react-northstar/dist/dts/src/components/List/ListItem.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ListItem = exports.listItemSlotClassNames = exports.listItemClassName = void 0;
var accessibility_1 = require("@fluentui/accessibility");
var react_bindings_1 = require("@fluentui/react-bindings");
var classnames_1 = require("classnames");
var _ = require("lodash");
var PropTypes = require("prop-types");
var React = require("react");
var utils_1 = require("../../utils");
var listContext_1 = require("./listContext");
var ListItemContent_1 = require("./ListItemContent");
var ListItemContentMedia_1 = require("./ListItemContentMedia");
var ListItemEndMedia_1 = require("./ListItemEndMedia");
var ListItemHeader_1 = require("./ListItemHeader");
var ListItemHeaderMedia_1 = require("./ListItemHeaderMedia");
var ListItemMedia_1 = require("./ListItemMedia");
exports.listItemClassName = 'ui-list__item';
exports.listItemSlotClassNames = {
    headerWrapper: exports.listItemClassName + "__headerWrapper",
    main: exports.listItemClassName + "__main",
    contentWrapper: exports.listItemClassName + "__contentWrapper",
};
/**
 * A ListItem contains a single piece of content within a List.
 */
exports.ListItem = React.forwardRef(function (props, ref) {
    var context = react_bindings_1.useFluentContext();
    var _a = react_bindings_1.useTelemetry(exports.ListItem.displayName, context.telemetry), setStart = _a.setStart, setEnd = _a.setEnd;
    setStart();
    var accessibility = props.accessibility, className = props.className, content = props.content, contentMedia = props.contentMedia, design = props.design, endMedia = props.endMedia, header = props.header, important = props.important, headerMedia = props.headerMedia, media = props.media, styles = props.styles;
    var parentProps = react_bindings_1.useContextSelectors(listContext_1.ListContext, {
        debug: function (v) { return v.debug; },
        navigable: function (v) { return v.navigable; },
        selectable: function (v) { return v.selectable; },
        truncateContent: function (v) { return v.truncateContent; },
        truncateHeader: function (v) { return v.truncateHeader; },
        variables: function (v) { return v.variables; },
        onItemClick: function (v) { return v.onItemClick; },
        selected: function (v) { return v.selectedIndex === props.index; },
    });
    var _b = props.debug, debug = _b === void 0 ? parentProps.debug : _b, _c = props.navigable, navigable = _c === void 0 ? parentProps.navigable : _c, _d = props.selectable, selectable = _d === void 0 ? parentProps.selectable : _d, _e = props.selected, selected = _e === void 0 ? parentProps.selected : _e, _f = props.truncateContent, truncateContent = _f === void 0 ? parentProps.truncateContent : _f, _g = props.truncateHeader, truncateHeader = _g === void 0 ? parentProps.truncateHeader : _g, _h = props.variables, variables = _h === void 0 ? parentProps.variables : _h;
    var getA11Props = react_bindings_1.useAccessibility(accessibility, {
        debugName: exports.ListItem.displayName,
        actionHandlers: {
            performClick: function (e) {
                e.preventDefault();
                handleClick(e);
            },
        },
        mapPropsToBehavior: function () { return ({
            navigable: navigable,
            selectable: selectable,
            selected: selected,
        }); },
        rtl: context.rtl,
    });
    var classes = react_bindings_1.useStyles(exports.ListItem.displayName, {
        className: exports.listItemClassName,
        mapPropsToStyles: function () { return ({
            debug: debug,
            navigable: navigable,
            important: important,
            selectable: selectable,
            selected: selected,
        }); },
        mapPropsToInlineStyles: function () { return ({ className: className, design: design, styles: styles, variables: variables }); },
        rtl: context.rtl,
    }).classes;
    var ElementType = react_bindings_1.getElementType(props);
    var unhandledProps = react_bindings_1.useUnhandledProps(exports.ListItem.handledProps, props);
    var handleClick = function (e) {
        _.invoke(props, 'onClick', e, props);
        parentProps.onItemClick(e, props.index);
    };
    var contentElement = utils_1.createShorthand(ListItemContent_1.ListItemContent, content, {
        defaultProps: function () { return ({
            hasContentMedia: !!contentMedia,
            hasHeader: !!header,
            truncate: truncateContent,
        }); },
    });
    var contentMediaElement = utils_1.createShorthand(ListItemContentMedia_1.ListItemContentMedia, contentMedia);
    var headerElement = utils_1.createShorthand(ListItemHeader_1.ListItemHeader, header, {
        defaultProps: function () { return ({
            hasContent: !!content,
            hasHeaderMedia: !!headerMedia,
            truncate: truncateHeader,
        }); },
    });
    var headerMediaElement = utils_1.createShorthand(ListItemHeaderMedia_1.ListItemHeaderMedia, headerMedia);
    var endMediaElement = utils_1.createShorthand(ListItemEndMedia_1.ListItemEndMedia, endMedia, {
        defaultProps: function () { return ({
            navigable: navigable,
            selectable: selectable,
        }); },
    });
    var mediaElement = utils_1.createShorthand(ListItemMedia_1.ListItemMedia, media, {
        defaultProps: function () { return ({
            hasContent: !!content,
            hasHeader: !!header,
            important: important,
        }); },
    });
    var element = getA11Props.unstable_wrapWithFocusZone(React.createElement(ElementType, __assign({}, getA11Props('root', __assign(__assign({ className: classes.root, onClick: handleClick }, unhandledProps), { ref: ref }))),
        mediaElement,
        React.createElement("div", { className: classnames_1.default(exports.listItemSlotClassNames.main, classes.main) },
            (headerElement || headerMediaElement) && (React.createElement("div", { className: classnames_1.default(exports.listItemSlotClassNames.headerWrapper, classes.headerWrapper) },
                headerElement,
                headerMediaElement)),
            (contentElement || contentMediaElement) && (React.createElement("div", { className: classnames_1.default(exports.listItemSlotClassNames.contentWrapper, classes.contentWrapper) },
                contentElement,
                contentMediaElement))),
        endMediaElement));
    setEnd();
    return element;
});
exports.ListItem.displayName = 'ListItem';
exports.ListItem.defaultProps = {
    as: 'li',
    accessibility: accessibility_1.listItemBehavior,
};
exports.ListItem.propTypes = __assign(__assign({}, utils_1.commonPropTypes.createCommon({
    content: false,
})), { contentMedia: PropTypes.any, content: PropTypes.any, debug: PropTypes.bool, header: PropTypes.any, endMedia: PropTypes.any, headerMedia: PropTypes.any, important: PropTypes.bool, media: PropTypes.any, selectable: PropTypes.bool, navigable: PropTypes.bool, index: PropTypes.number, selected: PropTypes.bool, truncateContent: PropTypes.bool, truncateHeader: PropTypes.bool, onClick: PropTypes.func });
exports.ListItem.handledProps = Object.keys(exports.ListItem.propTypes);
exports.ListItem.create = utils_1.createShorthandFactory({ Component: exports.ListItem, mappedProp: 'content' });
