// @fluentui/react-northstar/dist/dts/src/components/List/ListItemContent.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ListItemContent = exports.listItemContentClassName = void 0;
var react_bindings_1 = require("@fluentui/react-bindings");
var utils_1 = require("../../utils");
var Box_1 = require("../Box/Box");
exports.listItemContentClassName = 'ui-list__itemcontent';
/**
 * Provides a content for the ListItem.
 */
exports.ListItemContent = react_bindings_1.compose(Box_1.Box, {
    className: exports.listItemContentClassName,
    displayName: 'ListItemContent',
    mapPropsToStylesProps: function (_a) {
        var hasContentMedia = _a.hasContentMedia, hasHeader = _a.hasHeader, truncate = _a.truncate;
        return ({ hasContentMedia: hasContentMedia, hasHeader: hasHeader, truncate: truncate });
    },
    handledProps: ['hasContentMedia', 'hasHeader', 'truncate'],
    overrideStyles: true,
    shorthandConfig: {
        mappedProp: 'content',
    },
});
exports.ListItemContent.propTypes = utils_1.commonPropTypes.createCommon();
