// @fluentui/react-northstar/dist/dts/src/components/List/ListItemHeader.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ListItemHeader = exports.listItemHeaderClassName = void 0;
var react_bindings_1 = require("@fluentui/react-bindings");
var utils_1 = require("../../utils");
var Box_1 = require("../Box/Box");
exports.listItemHeaderClassName = 'ui-list__itemheader';
/**
 * Provides a header for the ListItem.
 */
exports.ListItemHeader = react_bindings_1.compose(Box_1.Box, {
    className: exports.listItemHeaderClassName,
    displayName: 'ListItemHeader',
    mapPropsToStylesProps: function (_a) {
        var hasContent = _a.hasContent, hasHeaderMedia = _a.hasHeaderMedia, truncate = _a.truncate;
        return ({ hasContent: hasContent, hasHeaderMedia: hasHeaderMedia, truncate: truncate });
    },
    handledProps: ['hasContent', 'hasHeaderMedia', 'truncate'],
    overrideStyles: true,
    shorthandConfig: {
        mappedProp: 'content',
    },
});
exports.ListItemHeader.propTypes = utils_1.commonPropTypes.createCommon();
