// @fluentui/react-northstar-fela-renderer/dist/dts/src/felaExpandCssShorthandsPlugin.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.felaExpandCssShorthandsPlugin = void 0;
// @ts-ignore
var inline_style_expand_shorthand_1 = require("inline-style-expand-shorthand");
// https://jsperf.com/array-indexof-vs-object-key-lookup2/12
var handledCssProps = {
    // 'font', Oops, is not supported by inline-style-expand-shorthand
    padding: true,
    margin: true,
    border: true,
    borderWidth: true,
    borderStyle: true,
    borderColor: true,
    borderTop: true,
    borderRight: true,
    borderBottom: true,
    borderLeft: true,
    borderRadius: true,
    background: true,
    outline: true,
};
var felaExpandCssShorthandsPlugin = function (styles) {
    return Object.keys(styles).reduce(function (acc, cssPropertyName) {
        var _a, _b, _c, _d;
        var cssPropertyValue = styles[cssPropertyName];
        if (cssPropertyValue === null || typeof cssPropertyValue === 'undefined') {
            return __assign(__assign({}, acc), (_a = {}, _a[cssPropertyName] = cssPropertyValue, _a));
        }
        if (handledCssProps[cssPropertyName]) {
            var expandedProps = inline_style_expand_shorthand_1.expandProperty(cssPropertyName, cssPropertyValue);
            if (expandedProps) {
                return __assign(__assign({}, acc), expandedProps);
            }
        }
        if (Array.isArray(cssPropertyValue)) {
            return __assign(__assign({}, acc), (_b = {}, _b[cssPropertyName] = cssPropertyValue, _b));
        }
        if (typeof cssPropertyValue === 'object') {
            return __assign(__assign({}, acc), (_c = {}, _c[cssPropertyName] = exports.felaExpandCssShorthandsPlugin(cssPropertyValue), _c));
        }
        return __assign(__assign({}, acc), (_d = {}, _d[cssPropertyName] = cssPropertyValue, _d));
    }, {});
};
exports.felaExpandCssShorthandsPlugin = felaExpandCssShorthandsPlugin;
