// @fluentui/react-northstar-fela-renderer/dist/dts/src/felaStylisEnhancer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.felaStylisEnhancer = void 0;
var fela_utils_1 = require("fela-utils");
// @ts-ignore
var stylis_1 = require("stylis");
// `stylis@3` is a CJS library, there are known issues with them:
// https://github.com/rollup/rollup/issues/1267#issuecomment-446681320
var Stylis = stylis_1.default.default || stylis_1.default;
// We use Stylis only for vendor prefixing, all other capabilities are disabled
var stylis = new Stylis({
    cascade: false,
    compress: false,
    global: false,
    keyframe: false,
    preserve: false,
    semicolon: false,
});
var felaStylisEnhancer = function (renderer) {
    var existingEmitChange = renderer._emitChange.bind(renderer);
    renderer._emitChange = function (change) {
        if (change.type === fela_utils_1.RULE_TYPE) {
            var prefixed = stylis('', change.declaration);
            // Fela uses objects by references, it's safe to override properties
            change.declaration = prefixed.slice(1, -1);
        }
        existingEmitChange(change);
    };
    return renderer;
};
exports.felaStylisEnhancer = felaStylisEnhancer;
