// @fluentui/react-northstar/dist/dts/src/components/Provider/Provider.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Provider = exports.providerClassName = void 0;
var _ = require("lodash");
var react_bindings_1 = require("@fluentui/react-bindings");
var styles_1 = require("@fluentui/styles");
var PropTypes = require("prop-types");
var React = require("react");
var utils_1 = require("../../utils");
var mergeProviderContexts_1 = require("../../utils/mergeProviderContexts");
var ProviderConsumer_1 = require("./ProviderConsumer");
var usePortalBox_1 = require("./usePortalBox");
var renderFontFaces = function (renderer, theme) {
    if (!theme.fontFaces) {
        return;
    }
    var renderFontObject = function (font) {
        if (!_.isPlainObject(font)) {
            throw new Error("fontFaces must be objects, got: " + typeof font);
        }
        renderer.renderFont(font);
    };
    theme.fontFaces.forEach(function (font) {
        renderFontObject(font);
    });
};
var renderStaticStyles = function (renderer, theme, siteVariables) {
    if (!theme.staticStyles) {
        return;
    }
    var renderObject = function (object) {
        _.forEach(object, function (style, selector) {
            renderer.renderGlobal(style, selector);
        });
    };
    theme.staticStyles.forEach(function (staticStyle) {
        if (typeof staticStyle === 'string') {
            renderer.renderGlobal(staticStyle);
        }
        else if (_.isPlainObject(staticStyle)) {
            renderObject(staticStyle);
        }
        else if (_.isFunction(staticStyle)) {
            var preparedSiteVariables = styles_1.mergeSiteVariables(siteVariables);
            renderObject(staticStyle(preparedSiteVariables));
        }
        else {
            throw new Error("staticStyles array must contain CSS strings, style objects, or style functions, got: " + typeof staticStyle);
        }
    });
};
exports.providerClassName = 'ui-provider';
/**
 * The Provider passes the CSS-in-JS renderer, theme styles and other settings to Fluent UI components.
 */
var Provider = function (props) {
    var children = props.children, className = props.className, design = props.design, overwrite = props.overwrite, styles = props.styles, variables = props.variables, telemetryRef = props.telemetryRef;
    var ElementType = react_bindings_1.getElementType(props);
    var unhandledProps = react_bindings_1.useUnhandledProps(exports.Provider.handledProps, props);
    var telemetry = React.useMemo(function () {
        if (!telemetryRef) {
            return undefined;
        }
        if (!telemetryRef.current) {
            telemetryRef.current = new react_bindings_1.Telemetry();
        }
        return telemetryRef.current;
    }, [telemetryRef]);
    var consumedContext = react_bindings_1.useFluentContext();
    var incomingContext = overwrite
        ? react_bindings_1.defaultContextValue
        : consumedContext;
    var createRenderer = React.useContext(react_bindings_1.RendererContext);
    // Memoization of `inputContext` & `outgoingContext` is required to avoid useless notifications of components that
    // consume `useFluentContext()` on each render
    // @see https://reactjs.org/docs/context.html#caveats
    var inputContext = React.useMemo(function () { return ({
        disableAnimations: props.disableAnimations,
        performance: props.performance,
        rtl: props.rtl,
        target: props.target,
        telemetry: telemetry,
        theme: props.theme,
    }); }, [props.disableAnimations, props.performance, props.rtl, props.target, telemetry, props.theme]);
    var outgoingContext = React.useMemo(function () { return mergeProviderContexts_1.mergeProviderContexts(createRenderer, incomingContext, inputContext); }, [
        createRenderer,
        incomingContext,
        inputContext,
    ]);
    var rtlProps = {};
    // only add dir attribute for top level provider or when direction changes from parent to child
    if (consumedContext.rtl !== outgoingContext.rtl && _.isBoolean(outgoingContext.rtl)) {
        rtlProps.dir = outgoingContext.rtl ? 'rtl' : 'ltr';
    }
    var classes = react_bindings_1.unstable_getStyles({
        allDisplayNames: [exports.Provider.displayName],
        className: exports.providerClassName,
        primaryDisplayName: exports.Provider.displayName,
        componentProps: {},
        inlineStylesProps: {
            className: className,
            design: design,
            styles: styles,
            variables: variables,
        },
        disableAnimations: outgoingContext.disableAnimations,
        performance: outgoingContext.performance,
        renderer: outgoingContext.renderer,
        rtl: outgoingContext.rtl,
        theme: outgoingContext.theme,
        saveDebug: _.noop,
        telemetry: undefined,
    }).classes;
    var element = usePortalBox_1.usePortalBox({
        className: classes.root,
        target: outgoingContext.target,
        rtl: outgoingContext.rtl,
    });
    react_bindings_1.useIsomorphicLayoutEffect(function () {
        renderFontFaces(outgoingContext.renderer, props.theme);
        renderStaticStyles(outgoingContext.renderer, props.theme, outgoingContext.theme.siteVariables);
        if (props.target) {
            utils_1.setUpWhatInput(props.target);
        }
        outgoingContext.renderer.registerUsage();
        return function () {
            if (props.target) {
                utils_1.tryCleanupWhatInput(props.target);
            }
            outgoingContext.renderer.unregisterUsage();
        };
    }, []);
    // do not spread anything - React.Fragment can only have `key` and `children` props
    var elementProps = ElementType === React.Fragment
        ? {}
        : __assign(__assign({ className: classes.root }, rtlProps), unhandledProps);
    var RenderProvider = outgoingContext.renderer.Provider;
    return (React.createElement(RenderProvider, null,
        React.createElement(react_bindings_1.Unstable_FluentContextProvider, { value: outgoingContext },
            React.createElement(usePortalBox_1.PortalBoxContext.Provider, { value: element },
                React.createElement(ElementType, __assign({}, elementProps), children)))));
};
exports.Provider = Provider;
exports.Provider.displayName = 'Provider';
exports.Provider.defaultProps = {
    theme: {},
};
exports.Provider.propTypes = {
    as: PropTypes.elementType,
    design: PropTypes.object,
    variables: PropTypes.oneOfType([PropTypes.object, PropTypes.func]),
    styles: PropTypes.oneOfType([PropTypes.object, PropTypes.func]),
    theme: PropTypes.shape({
        siteVariables: PropTypes.object,
        componentVariables: PropTypes.object,
        componentStyles: PropTypes.objectOf(PropTypes.any),
        fontFaces: PropTypes.arrayOf(PropTypes.exact({
            name: PropTypes.string.isRequired,
            paths: PropTypes.arrayOf(PropTypes.string),
            props: PropTypes.shape({
                fontStretch: PropTypes.string,
                fontStyle: PropTypes.string,
                fontVariant: PropTypes.string,
                fontWeight: PropTypes.number,
                localAlias: PropTypes.oneOfType([PropTypes.string, PropTypes.arrayOf(PropTypes.string)]),
                unicodeRange: PropTypes.string,
            }),
        })),
        staticStyles: PropTypes.array,
        animations: PropTypes.objectOf(PropTypes.any),
    }),
    rtl: PropTypes.bool,
    disableAnimations: PropTypes.bool,
    // Heads Up!
    // Keep in sync with packages/react-bindings/src/styles/types.ts
    performance: PropTypes.shape({
        enableSanitizeCssPlugin: PropTypes.bool,
        enableStylesCaching: PropTypes.bool,
        enableVariablesCaching: PropTypes.bool,
    }),
    children: PropTypes.node.isRequired,
    overwrite: PropTypes.bool,
    target: PropTypes.object,
    telemetryRef: PropTypes.object,
};
exports.Provider.handledProps = Object.keys(exports.Provider.propTypes);
exports.Provider.Consumer = ProviderConsumer_1.ProviderConsumer;
