// @fluentui/react-northstar/dist/commonjs/themes/teams/components/Segment/segmentStyles.js
"use strict";

exports.__esModule = true;
exports.segmentStyles = void 0;

var _colors = require("../../colors");

var segmentStyles = {
  root: function root(_ref) {
    var p = _ref.props,
        v = _ref.variables;
    var colors = (0, _colors.getColorScheme)(v.colorScheme, p.color);
    return Object.assign({
      borderColor: 'transparent',
      borderRadius: v.borderRadius,
      borderStyle: v.borderStyle,
      borderWidth: v.borderWidth,
      boxShadow: v.boxShadow,
      padding: v.padding,
      color: v.color,
      backgroundColor: v.backgroundColor
    }, p.color && {
      borderColor: colors.foreground
    }, p.inverted && {
      color: v.backgroundColor,
      backgroundColor: p.color ? colors.foreground : v.color
    }, p.disabled && Object.assign({
      boxShadow: 'none',
      borderColor: v.disabledBorderColor,
      color: v.disabledColor,
      backgroundColor: v.disabledBackgroundColor
    }, p.inverted && {
      color: v.disabledBackgroundColor,
      backgroundColor: v.disabledColor
    }));
  }
};
exports.segmentStyles = segmentStyles;
