// @fluentui/react-northstar/dist/dts/src/components/Pill/PillImage.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PillImage = exports.pillImageClassName = void 0;
var accessibility_1 = require("@fluentui/accessibility");
var react_bindings_1 = require("@fluentui/react-bindings");
var React = require("react");
var utils_1 = require("../../utils");
exports.pillImageClassName = 'ui-pill__image';
/**
 * An PillImage is a graphic representation used by Pill.
 */
exports.PillImage = React.forwardRef(function (props, ref) {
    var context = react_bindings_1.useFluentContext();
    var _a = react_bindings_1.useTelemetry(exports.PillImage.displayName, context.telemetry), setStart = _a.setStart, setEnd = _a.setEnd;
    setStart();
    var accessibility = props.accessibility, className = props.className, design = props.design, styles = props.styles, variables = props.variables, size = props.size;
    var getA11Props = react_bindings_1.useAccessibility(accessibility, {
        debugName: exports.PillImage.displayName,
        rtl: context.rtl,
    });
    var classes = react_bindings_1.useStyles(exports.PillImage.displayName, {
        className: exports.pillImageClassName,
        mapPropsToStyles: function () { return ({
            size: size,
        }); },
        mapPropsToInlineStyles: function () { return ({
            className: className,
            design: design,
            styles: styles,
            variables: variables,
        }); },
        rtl: context.rtl,
    }).classes;
    var ElementType = react_bindings_1.getElementType(props);
    var unhandledProps = react_bindings_1.useUnhandledProps(exports.PillImage.handledProps, props);
    var result = React.createElement(ElementType, __assign({}, getA11Props('root', __assign({ className: classes.root, ref: ref }, unhandledProps))));
    setEnd();
    return result;
});
exports.PillImage.displayName = 'PillImage';
exports.PillImage.defaultProps = {
    accessibility: accessibility_1.pillImageBehavior,
    as: 'img',
};
exports.PillImage.propTypes = __assign({}, utils_1.commonPropTypes.createCommon({
    children: false,
    content: false,
}));
exports.PillImage.handledProps = Object.keys(exports.PillImage.propTypes);
exports.PillImage.shorthandConfig = {
    mappedProp: 'src',
    allowsJSX: false,
};
