// @fluentui/react-northstar/dist/dts/src/components/Pill/PillContent.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PillContent = exports.pillContentClassName = void 0;
var React = require("react");
var PropTypes = require("prop-types");
var react_bindings_1 = require("@fluentui/react-bindings");
var utils_1 = require("../../utils");
exports.pillContentClassName = 'ui-pillcontent';
/**
 * A PillContent allows user to classify content.
 */
exports.PillContent = React.forwardRef(function (props, ref) {
    var context = react_bindings_1.useFluentContext();
    var _a = react_bindings_1.useTelemetry(exports.PillContent.displayName, context.telemetry), setStart = _a.setStart, setEnd = _a.setEnd;
    setStart();
    var accessibility = props.accessibility, children = props.children, className = props.className, content = props.content, design = props.design, styles = props.styles, variables = props.variables, size = props.size, actionable = props.actionable;
    var getA11Props = react_bindings_1.useAccessibility(accessibility, {
        debugName: exports.PillContent.displayName,
        rtl: context.rtl,
    });
    var classes = react_bindings_1.useStyles(exports.PillContent.displayName, {
        className: exports.pillContentClassName,
        mapPropsToStyles: function () { return ({ size: size, actionable: actionable }); },
        mapPropsToInlineStyles: function () { return ({ className: className, design: design, styles: styles, variables: variables }); },
        rtl: context.rtl,
    }).classes;
    var ElementType = react_bindings_1.getElementType(props);
    var unhandledProps = react_bindings_1.useUnhandledProps(exports.PillContent.handledProps, props);
    var element = (React.createElement(ElementType, __assign({}, getA11Props('root', __assign(__assign({ className: classes.root, ref: ref }, utils_1.rtlTextContainer.getAttributes({ forElements: [children] })), unhandledProps))), utils_1.childrenExist(children) ? children : content));
    setEnd();
    return element;
});
exports.PillContent.displayName = 'PillContent';
exports.PillContent.propTypes = __assign(__assign({}, utils_1.commonPropTypes.createCommon()), { actionable: PropTypes.bool });
exports.PillContent.handledProps = Object.keys(exports.PillContent.propTypes);
exports.PillContent.defaultProps = {
    as: 'span',
};
exports.PillContent.shorthandConfig = {
    mappedProp: 'content',
};
