// @fluentui/react-northstar/dist/commonjs/components/Toolbar/ToolbarMenuItemSubmenuIndicator.js
"use strict";

exports.__esModule = true;
exports.ToolbarMenuItemSubmenuIndicator = exports.toolbarMenuItemSubmenuIndicatorClassName = void 0;

var _reactBindings = require("@fluentui/react-bindings");

var _utils = require("../../utils");

var _Box = require("../Box/Box");

var toolbarMenuItemSubmenuIndicatorClassName = 'ui-toolbar__menuitemsubmenuindicator';
/**
 * A ToolbarMenuItemSubmenuIndicator allows a user to have a dedicated component that can be targeted from the theme.
 */

exports.toolbarMenuItemSubmenuIndicatorClassName = toolbarMenuItemSubmenuIndicatorClassName;
var ToolbarMenuItemSubmenuIndicator = (0, _reactBindings.compose)(_Box.Box, {
  className: toolbarMenuItemSubmenuIndicatorClassName,
  displayName: 'ToolbarMenuItemSubmenuIndicator',
  mapPropsToStylesProps: function mapPropsToStylesProps(props) {
    return {
      hasContent: props.hasContent
    };
  },
  shorthandConfig: {
    mappedProp: 'content'
  },
  overrideStyles: true
});
exports.ToolbarMenuItemSubmenuIndicator = ToolbarMenuItemSubmenuIndicator;
ToolbarMenuItemSubmenuIndicator.defaultProps = {
  as: 'span'
};
ToolbarMenuItemSubmenuIndicator.propTypes = _utils.commonPropTypes.createCommon();
