// @fluentui/react-northstar/dist/commonjs/components/Toolbar/ToolbarMenuRadioGroupWrapper.js
"use strict";

exports.__esModule = true;
exports.ToolbarMenuRadioGroupWrapper = exports.toolbarMenuRadioGroupWrapperClassName = void 0;

var _reactBindings = require("@fluentui/react-bindings");

var _accessibility = require("@fluentui/accessibility");

var _utils = require("../../utils");

var _Box = require("../Box/Box");

var toolbarMenuRadioGroupWrapperClassName = 'ui-toolbars'; // FIXME: required by getComponentInfo/isConformant. But this is group inside a toolbar not a group of toolbars

/**
 * An ToolbarMenuRadioGroupWrapper provides a wrapping slot for in  ToolbarMenuRadioGroup.
 */

exports.toolbarMenuRadioGroupWrapperClassName = toolbarMenuRadioGroupWrapperClassName;
var ToolbarMenuRadioGroupWrapper = (0, _reactBindings.compose)(_Box.Box, {
  className: toolbarMenuRadioGroupWrapperClassName,
  displayName: 'ToolbarMenuRadioGroupWrapper',
  overrideStyles: true,
  shorthandConfig: {
    mappedProp: 'content'
  }
});
exports.ToolbarMenuRadioGroupWrapper = ToolbarMenuRadioGroupWrapper;
ToolbarMenuRadioGroupWrapper.defaultProps = {
  as: 'li',
  accessibility: _accessibility.toolbarMenuRadioGroupWrapperBehavior
};
ToolbarMenuRadioGroupWrapper.propTypes = _utils.commonPropTypes.createCommon();
