// @fluentui/react-northstar/dist/commonjs/components/Toolbar/ToolbarMenuItemActiveIndicator.js
"use strict";

exports.__esModule = true;
exports.ToolbarMenuItemActiveIndicator = exports.toolbarMenuItemActiveIndicatorClassName = void 0;

var _reactBindings = require("@fluentui/react-bindings");

var _utils = require("../../utils");

var _Box = require("../Box/Box");

var toolbarMenuItemActiveIndicatorClassName = 'ui-toolbar__menuitemactiveindicator';
/**
 * A ToolbarMenuItemActiveIndicator allows a user to have a dedicated component that can be targeted from the theme.
 */

exports.toolbarMenuItemActiveIndicatorClassName = toolbarMenuItemActiveIndicatorClassName;
var ToolbarMenuItemActiveIndicator = (0, _reactBindings.compose)(_Box.Box, {
  className: toolbarMenuItemActiveIndicatorClassName,
  displayName: 'ToolbarMenuItemActiveIndicator',
  shorthandConfig: {
    mappedProp: 'content'
  },
  overrideStyles: true
});
exports.ToolbarMenuItemActiveIndicator = ToolbarMenuItemActiveIndicator;
ToolbarMenuItemActiveIndicator.defaultProps = {
  as: 'span'
};
ToolbarMenuItemActiveIndicator.propTypes = _utils.commonPropTypes.createCommon();
