// @fluentui/react-northstar/dist/commonjs/themes/teams/components/Reaction/reactionStyles.js
"use strict";

exports.__esModule = true;
exports.reactionStyles = void 0;

var _Reaction = require("../../../../components/Reaction/Reaction");

var _utils = require("../../../../utils");

var contentClassNameSelector = "& ." + _Reaction.reactionSlotClassNames.content;
var reactionStyles = {
  root: function root(_ref) {
    var _hover;

    var p = _ref.props,
        v = _ref.variables,
        siteVariables = _ref.theme.siteVariables;
    return {
      cursor: 'pointer',
      background: 'transparent',
      border: (0, _utils.pxToRem)(0),
      padding: (0, _utils.pxToRem)(0),
      display: 'inline-flex',
      justifyContent: 'center',
      alignItems: 'center',
      color: v.meReacting ? v.meReactingColor : v.otherReactingColor,
      ':hover': (_hover = {
        color: v.meReacting ? v.meReactingColorHover : v.otherReactingColorHover
      }, _hover[contentClassNameSelector] = {
        fontWeight: v.fontWeightHover
      }, _hover),
      position: 'relative',
      ':focus': {
        outline: 'none'
      },
      ':focus-visible': {
        ':after': {
          content: '""',
          position: 'absolute',
          top: "-" + (0, _utils.pxToRem)(2),
          right: "-" + (0, _utils.pxToRem)(2),
          bottom: "-" + (0, _utils.pxToRem)(2),
          left: "-" + (0, _utils.pxToRem)(2),
          borderWidth: '1px',
          borderStyle: 'solid',
          borderColor: v.borderColorFocus,
          borderRadius: siteVariables.borderRadiusMedium,
          boxShadow: "0px 0px 0px 1px " + v.boxShadowColor + " inset"
        }
      }
    };
  },
  icon: function icon(_ref2) {
    var p = _ref2.props;
    return {
      display: 'inline-flex',
      alignItems: 'center',
      justifyContent: 'center',
      marginRight: p.hasContent ? (0, _utils.pxToRem)(4) : (0, _utils.pxToRem)(0)
    };
  },
  content: function content(_ref3) {
    var v = _ref3.variables;
    return {
      fontSize: v.contentFontSize
    };
  }
};
exports.reactionStyles = reactionStyles;
