// @fluentui/accessibility/dist/commonjs/behaviors/Table/gridHeaderRowBehavior.js
"use strict";

exports.__esModule = true;
exports.gridHeaderRowBehavior = void 0;

var _attributes = require("../../attributes");

var _types = require("../../focusZone/types");

var _keyboardKey = require("../../keyboard-key");

var _gridHeaderCellBehavior = require("./gridHeaderCellBehavior");

/**
 * @specification
 * Adds role='row'.
 * Adds attribute 'data-is-focusable=true' to 'root' slot.
 * Provides arrow key navigation in horizontal direction.
 * Focused active element of the component is reset when TAB from the component.
 * When component's container element receives focus, focus will be set to the default focusable child element of the component.
 * Triggers 'unsetRowTabbable' action using 'shiftKey' + 'Tab' key on 'root'.
 * Applies 'gridHeaderCellBehavior' for 'cell' child component.
 */
var gridHeaderRowBehavior = function gridHeaderRowBehavior(props) {
  var _root;

  return {
    attributes: {
      root: (_root = {}, _root[_attributes.IS_FOCUSABLE_ATTRIBUTE] = true, _root.role = 'row', _root)
    },
    focusZone: {
      props: {
        direction: _types.FocusZoneDirection.horizontal,
        shouldFocusInnerElementWhenReceivedFocus: true,
        shouldResetActiveElementWhenTabFromZone: true
      }
    },
    keyActions: {
      root: {
        unsetRowTabbable: {
          keyCombinations: [{
            keyCode: _keyboardKey.keyboardKey.Tab,
            shiftKey: true
          }]
        }
      }
    },
    childBehaviors: {
      cell: _gridHeaderCellBehavior.gridHeaderCellBehavior
    }
  };
};

exports.gridHeaderRowBehavior = gridHeaderRowBehavior;
