// @fluentui/accessibility/dist/commonjs/behaviors/Table/gridCellMultipleFocusableBehavior.js
"use strict";

exports.__esModule = true;
exports.gridCellMultipleFocusableBehavior = void 0;

var _attributes = require("../../attributes");

var _types = require("../../focusZone/types");

var _keyboardKey = require("../../keyboard-key");

/**
 * @specification
 * Adds role='gridcell'.
 * Adds attribute 'data-is-focusable=true' to 'root' slot.
 * Provides arrow key navigation in bidirectional direction.
 * Triggers 'focusCell' action with 'Escape' on 'root'.
 */
var gridCellMultipleFocusableBehavior = function gridCellMultipleFocusableBehavior() {
  var _root;

  return {
    attributes: {
      root: (_root = {
        role: 'gridcell'
      }, _root[_attributes.IS_FOCUSABLE_ATTRIBUTE] = true, _root)
    },
    focusZone: {
      props: {
        direction: _types.FocusZoneDirection.bidirectional
      }
    },
    keyActions: {
      root: {
        focusCell: {
          keyCombinations: [{
            keyCode: _keyboardKey.keyboardKey.Escape
          }]
        }
      }
    }
  };
};

exports.gridCellMultipleFocusableBehavior = gridCellMultipleFocusableBehavior;
