// @fluentui/react-northstar/dist/dts/src/components/Button/ButtonContent.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ButtonContent = exports.buttonContentClassName = void 0;
var react_bindings_1 = require("@fluentui/react-bindings");
var customPropTypes = require("@fluentui/react-proptypes");
var utils_1 = require("../../utils");
var Box_1 = require("../Box/Box");
exports.buttonContentClassName = 'ui-button__content';
/**
 * A ButtonContent allows a user to have a dedicated component that can be targeted from the theme.
 */
exports.ButtonContent = react_bindings_1.compose(Box_1.Box, {
    className: exports.buttonContentClassName,
    displayName: 'ButtonContent',
    mapPropsToStylesProps: function (props) { return ({ size: props.size }); },
    handledProps: ['size'],
    overrideStyles: true,
    shorthandConfig: {
        mappedProp: 'content',
    },
});
exports.ButtonContent.defaultProps = {
    as: 'span',
};
exports.ButtonContent.propTypes = __assign(__assign({}, utils_1.commonPropTypes.createCommon()), { size: customPropTypes.size });
