// @fluentui/react-northstar/dist/dts/src/components/Button/ButtonGroup.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ButtonGroup = exports.buttonGroupClassName = void 0;
var accessibility_1 = require("@fluentui/accessibility");
var customPropTypes = require("@fluentui/react-proptypes");
var PropTypes = require("prop-types");
var React = require("react");
var _ = require("lodash");
var utils_1 = require("../../utils");
var Button_1 = require("./Button");
var react_bindings_1 = require("@fluentui/react-bindings");
exports.buttonGroupClassName = 'ui-buttons';
/**
 * A ButtonGroup represents multiple related actions as a group.
 */
exports.ButtonGroup = React.forwardRef(function (props, ref) {
    var context = react_bindings_1.useFluentContext();
    var _a = react_bindings_1.useTelemetry(exports.ButtonGroup.displayName, context.telemetry), setStart = _a.setStart, setEnd = _a.setEnd;
    setStart();
    var children = props.children, buttons = props.buttons, circular = props.circular, content = props.content, className = props.className, design = props.design, styles = props.styles, variables = props.variables;
    var ElementType = react_bindings_1.getElementType(props);
    var unhandledProps = react_bindings_1.useUnhandledProps(exports.ButtonGroup.handledProps, props);
    var _b = react_bindings_1.useStyles(exports.ButtonGroup.displayName, {
        className: exports.buttonGroupClassName,
        mapPropsToStyles: function () { return ({
            circular: circular,
        }); },
        mapPropsToInlineStyles: function () { return ({
            className: className,
            design: design,
            styles: styles,
            variables: variables,
        }); },
        rtl: context.rtl,
    }), classes = _b.classes, resolvedStyles = _b.styles;
    var getA11yProps = react_bindings_1.useAccessibility(props.accessibility, {
        debugName: exports.ButtonGroup.displayName,
        rtl: context.rtl,
    });
    var getStyleForButtonIndex = function (styles, isFirst, isLast) {
        var resultStyles = {};
        if (isFirst) {
            resultStyles = styles.firstButton;
        }
        if (isLast) {
            resultStyles = __assign(__assign({}, resultStyles), styles.lastButton);
        }
        if (!isFirst && !isLast) {
            resultStyles = styles.middleButton;
        }
        return resultStyles;
    };
    var emptyButtons = _.isNil(buttons);
    var element = (React.createElement(ElementType, __assign({}, __assign(__assign({}, getA11yProps('root', __assign({ className: classes.root, ref: ref }, unhandledProps))), (emptyButtons && __assign({}, utils_1.rtlTextContainer.getAttributes({ forElements: [children, content] }))))), emptyButtons
        ? utils_1.childrenExist(children)
            ? children
            : content
        : _.map(buttons, function (button, idx) {
            return utils_1.createShorthand(Button_1.Button, button, {
                defaultProps: function () { return ({
                    circular: circular,
                    styles: getStyleForButtonIndex(resolvedStyles, idx === 0, idx === buttons.length - 1),
                }); },
            });
        })));
    setEnd();
    return element;
});
exports.ButtonGroup.displayName = 'ButtonGroup';
exports.ButtonGroup.propTypes = __assign(__assign({}, utils_1.commonPropTypes.createCommon()), { buttons: customPropTypes.collectionShorthand, circular: PropTypes.bool });
exports.ButtonGroup.defaultProps = {
    accessibility: accessibility_1.buttonGroupBehavior,
};
exports.ButtonGroup.handledProps = Object.keys(exports.ButtonGroup.propTypes);
exports.ButtonGroup.create = utils_1.createShorthandFactory({
    Component: exports.ButtonGroup,
    mappedProp: 'content',
    mappedArrayProp: 'buttons',
});
