// @fluentui/react-northstar/dist/dts/src/components/Button/Button.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Button = exports.buttonClassName = void 0;
var accessibility_1 = require("@fluentui/accessibility");
var react_bindings_1 = require("@fluentui/react-bindings");
var customPropTypes = require("@fluentui/react-proptypes");
var _ = require("lodash");
var PropTypes = require("prop-types");
var React = require("react");
var utils_1 = require("../../utils");
var Box_1 = require("../Box/Box");
var Loader_1 = require("../Loader/Loader");
var ButtonGroup_1 = require("./ButtonGroup");
var ButtonContent_1 = require("./ButtonContent");
exports.buttonClassName = 'ui-button';
/**
 * A Button enables users to take an action, such as submitting a form, opening a dialog, etc.
 *
 * @accessibility
 * Implements [ARIA Button](https://www.w3.org/TR/wai-aria-practices-1.1/#button) design pattern.
 */
exports.Button = react_bindings_1.compose(function (props, ref, composeOptions) {
    var context = react_bindings_1.useFluentContext();
    var _a = react_bindings_1.useTelemetry(composeOptions.displayName, context.telemetry), setStart = _a.setStart, setEnd = _a.setEnd;
    setStart();
    var accessibility = props.accessibility, 
    // @ts-ignore
    active = props.active, as = props.as, children = props.children, content = props.content, icon = props.icon, loader = props.loader, disabled = props.disabled, disabledFocusable = props.disabledFocusable, flat = props.flat, iconPosition = props.iconPosition, loading = props.loading, text = props.text, primary = props.primary, inverted = props.inverted, size = props.size, iconOnly = props.iconOnly, fluid = props.fluid, circular = props.circular, className = props.className, styles = props.styles, tinted = props.tinted, variables = props.variables, design = props.design;
    var hasChildren = utils_1.childrenExist(children);
    var getA11yProps = react_bindings_1.useAccessibility(accessibility, {
        debugName: composeOptions.displayName,
        mapPropsToBehavior: function () { return ({
            as: as,
            active: active,
            disabled: disabled,
            disabledFocusable: disabledFocusable,
        }); },
        actionHandlers: {
            performClick: function (event) {
                event.preventDefault();
                handleClick(event);
            },
        },
        rtl: context.rtl,
    });
    var _b = react_bindings_1.useStyles(composeOptions.displayName, {
        className: composeOptions.className,
        mapPropsToStyles: function () { return ({
            text: text,
            primary: primary,
            disabled: disabled,
            tinted: tinted,
            disabledFocusable: disabledFocusable,
            flat: flat,
            circular: circular,
            size: size,
            loading: loading,
            inverted: inverted,
            iconOnly: iconOnly,
            iconPosition: iconPosition,
            fluid: fluid,
            hasContent: !!content,
        }); },
        mapPropsToInlineStyles: function () { return ({
            className: className,
            design: design,
            styles: styles,
            variables: variables,
        }); },
        rtl: context.rtl,
        composeOptions: composeOptions,
        unstable_props: props,
    }), classes = _b.classes, resolvedStyles = _b.styles;
    var slotProps = composeOptions.resolveSlotProps(props);
    var unhandledProps = react_bindings_1.useUnhandledProps(composeOptions.handledProps, props);
    var ElementType = react_bindings_1.getElementType(props);
    var renderIcon = function () {
        return utils_1.createShorthand(composeOptions.slots.icon, icon, {
            defaultProps: function () {
                return getA11yProps('icon', __assign({ styles: resolvedStyles.icon }, slotProps.icon));
            },
        });
    };
    var renderLoader = function () {
        return utils_1.createShorthand(composeOptions.slots.loader, loader || {}, {
            defaultProps: function () {
                return getA11yProps('loader', __assign({ styles: resolvedStyles.loader }, slotProps.loader));
            },
        });
    };
    var renderContent = function () {
        return utils_1.createShorthand(composeOptions.slots.content, content, {
            defaultProps: function () { return getA11yProps('content', slotProps.content); },
        });
    };
    var handleClick = function (e) {
        if (disabled || disabledFocusable) {
            e.preventDefault();
            return;
        }
        _.invoke(props, 'onClick', e, props);
    };
    var handleFocus = function (e) {
        _.invoke(props, 'onFocus', e, props);
    };
    var result = (React.createElement(ElementType, __assign({}, utils_1.rtlTextContainer.getAttributes({ forElements: [children] }), getA11yProps('root', __assign({ onClick: handleClick, className: classes.root, onFocus: handleFocus, ref: ref }, unhandledProps))), hasChildren ? (children) : (React.createElement(React.Fragment, null,
        loading && renderLoader(),
        iconPosition !== 'after' && renderIcon(),
        renderContent(),
        iconPosition === 'after' && renderIcon()))));
    setEnd();
    return result;
}, {
    className: exports.buttonClassName,
    displayName: 'Button',
    slots: {
        content: ButtonContent_1.ButtonContent,
        icon: Box_1.Box,
        loader: Loader_1.Loader,
    },
    slotProps: function (props) { return ({
        content: {
            size: props.size,
        },
        loader: {
            role: undefined,
            secondary: props.primary,
        },
    }); },
    shorthandConfig: {
        mappedProp: 'content',
    },
    handledProps: [
        'accessibility',
        'as',
        'children',
        'circular',
        'className',
        'content',
        'design',
        'disabled',
        'tinted',
        'disabledFocusable',
        'flat',
        'fluid',
        'icon',
        'iconOnly',
        'iconPosition',
        'inverted',
        'loader',
        'loading',
        'onClick',
        'onFocus',
        'primary',
        'text',
        'secondary',
        'size',
        'styles',
        'variables',
    ],
});
exports.Button.defaultProps = {
    as: 'button',
    accessibility: accessibility_1.buttonBehavior,
    size: 'medium',
};
exports.Button.propTypes = __assign(__assign({}, utils_1.commonPropTypes.createCommon({
    content: 'shorthand',
})), { circular: PropTypes.bool, disabled: PropTypes.bool, disabledFocusable: PropTypes.bool, flat: PropTypes.bool, fluid: PropTypes.bool, icon: customPropTypes.shorthandAllowingChildren, iconOnly: PropTypes.bool, iconPosition: PropTypes.oneOf(['before', 'after']), inverted: PropTypes.bool, loader: customPropTypes.itemShorthandWithoutJSX, loading: PropTypes.bool, onClick: PropTypes.func, onFocus: PropTypes.func, tinted: customPropTypes.every([customPropTypes.disallow(['secondary']), PropTypes.bool]), primary: customPropTypes.every([customPropTypes.disallow(['secondary']), PropTypes.bool]), text: PropTypes.bool, secondary: customPropTypes.every([customPropTypes.disallow(['primary']), PropTypes.bool]), size: PropTypes.oneOf(['medium', 'small']) });
exports.Button.Group = ButtonGroup_1.ButtonGroup;
exports.Button.Content = ButtonContent_1.ButtonContent;
exports.Button.create = utils_1.createShorthandFactory({ Component: exports.Button, mappedProp: 'content' });
