// @fluentui/react-northstar/dist/es/components/Avatar/AvatarLabel.js
import { getElementType, useUnhandledProps, useAccessibility, useFluentContext, useStyles, useTelemetry } from '@fluentui/react-bindings';
import * as customPropTypes from '@fluentui/react-proptypes';
import * as PropTypes from 'prop-types';
import * as React from 'react';
import { childrenExist, commonPropTypes, rtlTextContainer } from '../../utils';
import { labelClassName } from '../Label/Label';
export var avatarlabelClassName = labelClassName;
/**
 * A AvatarLabel allows user to classify content.
 */

export var AvatarLabel = /*#__PURE__*/function () {
  var AvatarLabel = /*#__PURE__*/React.forwardRef(function (props, ref) {
    var context = useFluentContext();

    var _useTelemetry = useTelemetry(AvatarLabel.displayName, context.telemetry),
        setStart = _useTelemetry.setStart,
        setEnd = _useTelemetry.setEnd;

    setStart();
    var accessibility = props.accessibility,
        children = props.children,
        className = props.className,
        content = props.content,
        design = props.design,
        styles = props.styles,
        variables = props.variables,
        square = props.square,
        size = props.size,
        circular = props.circular;
    var getA11Props = useAccessibility(accessibility, {
      debugName: AvatarLabel.displayName,
      rtl: context.rtl
    });

    var _useStyles = useStyles(AvatarLabel.displayName, {
      className: avatarlabelClassName,
      mapPropsToStyles: function mapPropsToStyles() {
        return {
          square: square,
          size: size,
          circular: circular
        };
      },
      mapPropsToInlineStyles: function mapPropsToInlineStyles() {
        return {
          className: className,
          design: design,
          styles: styles,
          variables: variables
        };
      },
      rtl: context.rtl
    }),
        classes = _useStyles.classes;

    var ElementType = getElementType(props);
    var unhandledProps = useUnhandledProps(AvatarLabel.handledProps, props);
    var element = /*#__PURE__*/React.createElement(ElementType, getA11Props('root', Object.assign({
      className: classes.root,
      ref: ref
    }, rtlTextContainer.getAttributes({
      forElements: [children]
    }), unhandledProps)), childrenExist(children) ? children : content);
    setEnd();
    return element;
  });
  AvatarLabel.displayName = 'AvatarLabel';
  AvatarLabel.propTypes = Object.assign({}, commonPropTypes.createCommon(), {
    square: PropTypes.bool,
    size: customPropTypes.size,
    circular: PropTypes.bool
  });
  AvatarLabel.handledProps = Object.keys(AvatarLabel.propTypes);
  AvatarLabel.defaultProps = {
    as: 'span'
  };
  AvatarLabel.shorthandConfig = {
    mappedProp: 'content'
  };
  return AvatarLabel;
}();
