// @fluentui/react-northstar/dist/dts/src/components/Design/Design.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Design = void 0;
var react_bindings_1 = require("@fluentui/react-bindings");
var customPropTypes = require("@fluentui/react-proptypes");
var React = require("react");
var PropTypes = require("prop-types");
/**
 * The Design component provides a theme safe subset of CSS for designing layouts.
 */
// eslint-disable-next-line @typescript-eslint/no-unused-vars
function Design(_a) {
    var config = _a.config, children = _a.children;
    var context = react_bindings_1.useFluentContext();
    var getConfig = React.useCallback(function () { return config; }, [config]);
    // Heads Up! Keep in sync with renderComponent.tsx
    var styleParam = {
        displayName: Design.displayName,
        disableAnimations: context.disableAnimations,
        direction: context.rtl ? 'rtl' : 'ltr',
        sanitizeCss: context.performance.enableSanitizeCssPlugin,
    };
    var className = context.renderer.renderRule(getConfig, styleParam);
    return children({ className: className });
}
exports.Design = Design;
Design.displayName = 'Design';
Design.propTypes = {
    children: PropTypes.func.isRequired,
    config: customPropTypes.design.isRequired,
};
