// @fluentui/react-northstar/dist/commonjs/themes/teams/componentVariables.js
"use strict";

exports.__esModule = true;
exports.ToolbarMenuRadioGroup = exports.ToolbarMenuItemActiveIndicator = exports.ToolbarMenuItemSubmenuIndicator = exports.ToolbarMenuItemIcon = exports.ToolbarMenuItem = exports.ToolbarMenuDivider = exports.ToolbarMenu = exports.ToolbarDivider = exports.ToolbarItem = exports.ToolbarCustomItem = exports.Toolbar = exports.TextArea = exports.Text = exports.Status = exports.Slider = exports.SkeletonAvatar = exports.SkeletonInput = exports.SkeletonText = exports.SkeletonButton = exports.SkeletonShape = exports.SkeletonLine = exports.Skeleton = exports.Segment = exports.ReactionGroup = exports.Reaction = exports.RadioGroupItem = exports.Provider = exports.PopupContent = exports.PillIcon = exports.PillImage = exports.PillAction = exports.PillContent = exports.Pill = exports.MenuDivider = exports.MenuItemWrapper = exports.MenuItemIndicator = exports.MenuItemContent = exports.MenuItemIcon = exports.MenuItem = exports.Menu = exports.ListItemMedia = exports.ListItemHeaderMedia = exports.ListItemHeader = exports.ListItemContentMedia = exports.ListItemContent = exports.ListItem = exports.ItemLayout = exports.Loader = exports.Layout = exports.Label = exports.InputLabel = exports.Input = exports.Image = exports.SvgIcon = exports.HeaderDescription = exports.Header = exports.Grid = exports.FormLabel = exports.FormMessage = exports.FormFieldCustom = exports.FormField = exports.Form = exports.FlexItem = exports.Flex = exports.Embed = exports.DropdownSelectedItem = exports.DropdownSearchInput = exports.DropdownItem = exports.Dropdown = exports.Divider = exports.Dialog = exports.Checkbox = exports.ChatMessageReadStatus = exports.ChatMessageDetails = exports.ChatMessage = exports.ChatItem = exports.Chat = exports.BreadcrumbDivider = exports.BreadcrumbLink = exports.BreadcrumbItem = exports.Breadcrumb = exports.ButtonContent = exports.ButtonGroup = exports.Button = exports.AvatarLabel = exports.AvatarIcon = exports.AvatarImage = exports.AvatarStatusIcon = exports.AvatarStatus = exports.Avatar = exports.AlertDismissAction = exports.Alert = exports.AttachmentIcon = exports.AttachmentHeader = exports.AttachmentDescription = exports.AttachmentBody = exports.AttachmentAction = exports.Attachment = exports.AccordionTitle = exports.Accordion = void 0;
exports.DatepickerCalendarGridRow = exports.DatepickerCalendarGrid = exports.DatepickerCalendarHeaderCell = exports.DatepickerCalendarHeader = exports.DatepickerCalendar = exports.DatepickerCalendarCellButton = exports.DatepickerCalendarCell = exports.Datepicker = exports.CardExpandableBox = exports.CardTopControls = exports.CardPreview = exports.CardFooter = exports.CardBody = exports.CardHeader = exports.Card = exports.TableCell = exports.TableRow = exports.Table = exports.CarouselPaddle = exports.CarouselNavigationItem = exports.CarouselNavigation = exports.CarouselPaddlesContainer = exports.CarouselItem = exports.Carousel = exports.SplitButtonDivider = exports.SplitButtonToggle = exports.SplitButton = exports.TooltipContent = exports.Video = exports.TreeTitle = exports.ToolbarMenuRadioGroupWrapper = void 0;

var _accordionVariables = require("./components/Accordion/accordionVariables");

exports.Accordion = _accordionVariables.accordionVariables;

var _accordionTitleVariables = require("./components/Accordion/accordionTitleVariables");

exports.AccordionTitle = _accordionTitleVariables.accordionTitleVariables;

var _attachmentVariables = require("./components/Attachment/attachmentVariables");

exports.Attachment = _attachmentVariables.attachmentVariables;

var _attachmentActionVariables = require("./components/Attachment/attachmentActionVariables");

exports.AttachmentAction = _attachmentActionVariables.attachmentActionVariables;

var _attachmentBodyVariables = require("./components/Attachment/attachmentBodyVariables");

exports.AttachmentBody = _attachmentBodyVariables.attachmentBodyVariables;

var _attachmentDescriptionVariables = require("./components/Attachment/attachmentDescriptionVariables");

exports.AttachmentDescription = _attachmentDescriptionVariables.attachmentDescriptionVariables;

var _attachmentHeaderVariables = require("./components/Attachment/attachmentHeaderVariables");

exports.AttachmentHeader = _attachmentHeaderVariables.attachmentHeaderVariables;

var _attachmentIconVariables = require("./components/Attachment/attachmentIconVariables");

exports.AttachmentIcon = _attachmentIconVariables.attachmentIconVariables;

var _alertVariables = require("./components/Alert/alertVariables");

exports.Alert = _alertVariables.alertVariables;

var _alertDismissActionVariables = require("./components/Alert/alertDismissActionVariables");

exports.AlertDismissAction = _alertDismissActionVariables.alertDismissActionVariables;

var _avatarVariables = require("./components/Avatar/avatarVariables");

exports.Avatar = _avatarVariables.avatarVariables;

var _avatarStatusVariables = require("./components/Avatar/avatarStatusVariables");

exports.AvatarStatus = _avatarStatusVariables.avatarStatusVariables;

var _avatarStatusIconVariables = require("./components/Avatar/avatarStatusIconVariables");

exports.AvatarStatusIcon = _avatarStatusIconVariables.avatarStatusIconVariables;

var _avatarImageVariables = require("./components/Avatar/avatarImageVariables");

exports.AvatarImage = _avatarImageVariables.avatarImageVariables;

var _avatarIconVariables = require("./components/Avatar/avatarIconVariables");

exports.AvatarIcon = _avatarIconVariables.avatarIconVariables;

var _avatarLabelVariables = require("./components/Avatar/avatarLabelVariables");

exports.AvatarLabel = _avatarLabelVariables.avatarLabelVariables;

var _buttonVariables = require("./components/Button/buttonVariables");

exports.Button = _buttonVariables.buttonVariables;

var _buttonGroupVariables = require("./components/Button/buttonGroupVariables");

exports.ButtonGroup = _buttonGroupVariables.buttonGroupVariables;

var _buttonContentVariables = require("./components/Button/buttonContentVariables");

exports.ButtonContent = _buttonContentVariables.buttonContentVariables;

var _breadcrumbVariables = require("./components/Breadcrumb/breadcrumbVariables");

exports.Breadcrumb = _breadcrumbVariables.breadcrumbVariables;

var _breadcrumbItemVariables = require("./components/Breadcrumb/breadcrumbItemVariables");

exports.BreadcrumbItem = _breadcrumbItemVariables.breadcrumbItemVariables;

var _breadcrumbLinkVariables = require("./components/Breadcrumb/breadcrumbLinkVariables");

exports.BreadcrumbLink = _breadcrumbLinkVariables.breadcrumbLinkVariables;

var _breadcrumbDividerVariables = require("./components/Breadcrumb/breadcrumbDividerVariables");

exports.BreadcrumbDivider = _breadcrumbDividerVariables.breadcrumbDividerVariables;

var _chatVariables = require("./components/Chat/chatVariables");

exports.Chat = _chatVariables.chatVariables;

var _chatItemVariables = require("./components/Chat/chatItemVariables");

exports.ChatItem = _chatItemVariables.chatItemVariables;

var _chatMessageVariables = require("./components/Chat/chatMessageVariables");

exports.ChatMessage = _chatMessageVariables.chatMessageVariables;

var _chatMessageDetailsVariables = require("./components/Chat/chatMessageDetailsVariables");

exports.ChatMessageDetails = _chatMessageDetailsVariables.chatMessageDetailsVariables;

var _chatMessageReadStatusVariables = require("./components/Chat/chatMessageReadStatusVariables");

exports.ChatMessageReadStatus = _chatMessageReadStatusVariables.chatMessageReadStatusVariables;

var _checkboxVariables = require("./components/Checkbox/checkboxVariables");

exports.Checkbox = _checkboxVariables.checkboxVariables;

var _dialogVariables = require("./components/Dialog/dialogVariables");

exports.Dialog = _dialogVariables.dialogVariables;

var _dividerVariables = require("./components/Divider/dividerVariables");

exports.Divider = _dividerVariables.dividerVariables;

var _dropdownVariables = require("./components/Dropdown/dropdownVariables");

exports.Dropdown = _dropdownVariables.dropdownVariables;

var _dropdownItemVariables = require("./components/Dropdown/dropdownItemVariables");

exports.DropdownItem = _dropdownItemVariables.dropdownItemVariables;

var _dropdownSearchInputVariables = require("./components/Dropdown/dropdownSearchInputVariables");

exports.DropdownSearchInput = _dropdownSearchInputVariables.dropdownSearchInputVariables;

var _dropdownSelectedItemVariables = require("./components/Dropdown/dropdownSelectedItemVariables");

exports.DropdownSelectedItem = _dropdownSelectedItemVariables.dropdownSelectedItemVariables;

var _embedVariables = require("./components/Embed/embedVariables");

exports.Embed = _embedVariables.embedVariables;

var _flexVariables = require("./components/Flex/flexVariables");

exports.Flex = _flexVariables.flexVariables;

var _flexItemVariables = require("./components/Flex/flexItemVariables");

exports.FlexItem = _flexItemVariables.flexItemVariables;

var _formVariables = require("./components/Form/formVariables");

exports.Form = _formVariables.formVariables;

var _formFieldVariables = require("./components/Form/formFieldVariables");

exports.FormField = _formFieldVariables.formFieldVariables;
exports.FormFieldCustom = _formFieldVariables.formFieldVariables;

var _formMessageVariables = require("./components/Form/formMessageVariables");

exports.FormMessage = _formMessageVariables.formMessageVariables;

var _formLabelVariables = require("./components/Form/formLabelVariables");

exports.FormLabel = _formLabelVariables.formLabelVariables;

var _gridVariables = require("./components/Grid/gridVariables");

exports.Grid = _gridVariables.gridVariables;

var _headerVariables = require("./components/Header/headerVariables");

exports.Header = _headerVariables.headerVariables;

var _headerDescriptionVariables = require("./components/Header/headerDescriptionVariables");

exports.HeaderDescription = _headerDescriptionVariables.headerDescriptionVariables;

var _svgIconVariables = require("./components/SvgIcon/svgIconVariables");

exports.SvgIcon = _svgIconVariables.svgIconVariables;

var _imageVariables = require("./components/Image/imageVariables");

exports.Image = _imageVariables.imageVariables;

var _inputVariables = require("./components/Input/inputVariables");

exports.Input = _inputVariables.inputVariables;

var _inputLabelVariables = require("./components/Input/inputLabelVariables");

exports.InputLabel = _inputLabelVariables.inputLabelVariables;

var _labelVariables = require("./components/Label/labelVariables");

exports.Label = _labelVariables.labelVariables;

var _layoutVariables = require("./components/Layout/layoutVariables");

exports.Layout = _layoutVariables.layoutVariables;

var _loaderVariables = require("./components/Loader/loaderVariables");

exports.Loader = _loaderVariables.loaderVariables;

var _itemLayoutVariables = require("./components/ItemLayout/itemLayoutVariables");

exports.ItemLayout = _itemLayoutVariables.itemLayoutVariables;

var _listItemVariables = require("./components/List/listItemVariables");

exports.ListItem = _listItemVariables.listItemVariables;

var _listItemContentVariables = require("./components/List/listItemContentVariables");

exports.ListItemContent = _listItemContentVariables.listItemContentVariables;

var _listItemContentMediaVariables = require("./components/List/listItemContentMediaVariables");

exports.ListItemContentMedia = _listItemContentMediaVariables.listItemContentMediaVariables;

var _listItemHeaderVariables = require("./components/List/listItemHeaderVariables");

exports.ListItemHeader = _listItemHeaderVariables.listItemHeaderVariables;

var _listItemHeaderMediaVariables = require("./components/List/listItemHeaderMediaVariables");

exports.ListItemHeaderMedia = _listItemHeaderMediaVariables.listItemHeaderMediaVariables;

var _listItemMediaVariables = require("./components/List/listItemMediaVariables");

exports.ListItemMedia = _listItemMediaVariables.listItemMediaVariables;

var _menuVariables = require("./components/Menu/menuVariables");

exports.Menu = _menuVariables.menuVariables;

var _menuItemVariables = require("./components/Menu/menuItemVariables");

exports.MenuItem = _menuItemVariables.menuItemVariables;

var _menuItemIconVariables = require("./components/Menu/menuItemIconVariables");

exports.MenuItemIcon = _menuItemIconVariables.menuItemIconVariables;

var _menuItemContentVariables = require("./components/Menu/menuItemContentVariables");

exports.MenuItemContent = _menuItemContentVariables.menuItemContentVariables;
exports.MenuItemIndicator = _menuItemContentVariables.menuItemContentVariables;

var _menuItemWrapperVariables = require("./components/Menu/menuItemWrapperVariables");

exports.MenuItemWrapper = _menuItemWrapperVariables.menuItemWrapperVariables;

var _menuDividerVariables = require("./components/Menu/menuDividerVariables");

exports.MenuDivider = _menuDividerVariables.menuDividerVariables;

var _pillVariables = require("./components/Pill/pillVariables");

exports.Pill = _pillVariables.pillVariables;
exports.PillContent = _pillVariables.pillVariables;
exports.PillAction = _pillVariables.pillVariables;
exports.PillImage = _pillVariables.pillVariables;
exports.PillIcon = _pillVariables.pillVariables;

var _popupContentVariables = require("./components/Popup/popupContentVariables");

exports.PopupContent = _popupContentVariables.popupContentVariables;

var _providerVariables = require("./components/Provider/providerVariables");

exports.Provider = _providerVariables.providerVariables;

var _radioGroupItemVariables = require("./components/RadioGroup/radioGroupItemVariables");

exports.RadioGroupItem = _radioGroupItemVariables.radioGroupItemVariables;

var _reactionVariables = require("./components/Reaction/reactionVariables");

exports.Reaction = _reactionVariables.reactionVariables;

var _reactionGroupVariables = require("./components/Reaction/reactionGroupVariables");

exports.ReactionGroup = _reactionGroupVariables.reactionGroupVariables;

var _segmentVariables = require("./components/Segment/segmentVariables");

exports.Segment = _segmentVariables.segmentVariables;

var _skeletonVariables = require("./components/Skeleton/skeletonVariables");

exports.Skeleton = _skeletonVariables.skeletonVariables;
exports.SkeletonAvatar = _skeletonVariables.skeletonVariables;

var _skeletonLineVariables = require("./components/Skeleton/skeletonLineVariables");

exports.SkeletonLine = _skeletonLineVariables.skeletonVariables;

var _skeletonShapeVariables = require("./components/Skeleton/skeletonShapeVariables");

exports.SkeletonShape = _skeletonShapeVariables.skeletonVariables;

var _skeletonButtonVariables = require("./components/Skeleton/skeletonButtonVariables");

exports.SkeletonButton = _skeletonButtonVariables.skeletonVariables;

var _skeletonTextVariables = require("./components/Skeleton/skeletonTextVariables");

exports.SkeletonText = _skeletonTextVariables.skeletonVariables;

var _skeletonInputVariables = require("./components/Skeleton/skeletonInputVariables");

exports.SkeletonInput = _skeletonInputVariables.skeletonVariables;

var _sliderVariables = require("./components/Slider/sliderVariables");

exports.Slider = _sliderVariables.sliderVariables;

var _statusVariables = require("./components/Status/statusVariables");

exports.Status = _statusVariables.statusVariables;

var _textVariables = require("./components/Text/textVariables");

exports.Text = _textVariables.textVariables;

var _textAreaVariables = require("./components/TextArea/textAreaVariables");

exports.TextArea = _textAreaVariables.textAreaVariables;

var _toolbarVariables = require("./components/Toolbar/toolbarVariables");

exports.Toolbar = _toolbarVariables.toolbarVariables;

var _toolbarCustomItemVariables = require("./components/Toolbar/toolbarCustomItemVariables");

exports.ToolbarCustomItem = _toolbarCustomItemVariables.toolbarCustomItemVariables;

var _toolbarItemVariables = require("./components/Toolbar/toolbarItemVariables");

exports.ToolbarItem = _toolbarItemVariables.toolbarItemVariables;

var _toolbarDividerVariables = require("./components/Toolbar/toolbarDividerVariables");

exports.ToolbarDivider = _toolbarDividerVariables.toolbarDividerVariables;

var _toolbarMenuVariables = require("./components/Toolbar/toolbarMenuVariables");

exports.ToolbarMenu = _toolbarMenuVariables.toolbarMenuVariables;

var _toolbarMenuDividerVariables = require("./components/Toolbar/toolbarMenuDividerVariables");

exports.ToolbarMenuDivider = _toolbarMenuDividerVariables.toolbarMenuDividerVariables;

var _toolbarMenuItemVariables = require("./components/Toolbar/toolbarMenuItemVariables");

exports.ToolbarMenuItem = _toolbarMenuItemVariables.toolbarMenuItemVariables;

var _toolbarMenuItemIconVariables = require("./components/Toolbar/toolbarMenuItemIconVariables");

exports.ToolbarMenuItemIcon = _toolbarMenuItemIconVariables.toolbarMenuItemIconVariables;

var _toolbarMenuItemSubmenuIndicatorVariables = require("./components/Toolbar/toolbarMenuItemSubmenuIndicatorVariables");

exports.ToolbarMenuItemSubmenuIndicator = _toolbarMenuItemSubmenuIndicatorVariables.toolbarMenuItemSubmenuIndicatorVariables;

var _toolbarMenuItemActiveIndicatorVariables = require("./components/Toolbar/toolbarMenuItemActiveIndicatorVariables");

exports.ToolbarMenuItemActiveIndicator = _toolbarMenuItemActiveIndicatorVariables.toolbarMenuItemActiveIndicatorVariables;

var _toolbarMenuRadioGroupVariables = require("./components/Toolbar/toolbarMenuRadioGroupVariables");

exports.ToolbarMenuRadioGroup = _toolbarMenuRadioGroupVariables.toolbarMenuRadioGroupVariables;

var _toolbarMenuRadioGroupWrapperVariables = require("./components/Toolbar/toolbarMenuRadioGroupWrapperVariables");

exports.ToolbarMenuRadioGroupWrapper = _toolbarMenuRadioGroupWrapperVariables.toolbarMenuRadioGroupWrapperVariables;

var _treeTitleVariables = require("./components/Tree/treeTitleVariables");

exports.TreeTitle = _treeTitleVariables.treeTitleVariables;

var _videoVariables = require("./components/Video/videoVariables");

exports.Video = _videoVariables.videoVariables;

var _tooltipContentVariables = require("./components/Tooltip/tooltipContentVariables");

exports.TooltipContent = _tooltipContentVariables.tooltipContentVariables;

var _splitButtonVariables = require("./components/SplitButton/splitButtonVariables");

exports.SplitButton = _splitButtonVariables.splitButtonVariables;

var _splitButtonToggleVariables = require("./components/SplitButton/splitButtonToggleVariables");

exports.SplitButtonToggle = _splitButtonToggleVariables.splitButtonToggleVariables;

var _splitButtonDividerVariables = require("./components/SplitButton/splitButtonDividerVariables");

exports.SplitButtonDivider = _splitButtonDividerVariables.splitButtonDividerVariables;

var _carouselVariables = require("./components/Carousel/carouselVariables");

exports.Carousel = _carouselVariables.carouselVariables;

var _carouselItemVariables = require("./components/Carousel/carouselItemVariables");

exports.CarouselItem = _carouselItemVariables.carouselItemVariables;

var _carouselPaddlesContainerVariables = require("./components/Carousel/carouselPaddlesContainerVariables");

exports.CarouselPaddlesContainer = _carouselPaddlesContainerVariables.carouselPaddlesContainerVariables;

var _carouselNavigationVariables = require("./components/Carousel/carouselNavigationVariables");

exports.CarouselNavigation = _carouselNavigationVariables.carouselNavigationVariables;

var _carouselNavigationItemVariables = require("./components/Carousel/carouselNavigationItemVariables");

exports.CarouselNavigationItem = _carouselNavigationItemVariables.carouselNavigationItemVariables;

var _carouselPaddleVariables = require("./components/Carousel/carouselPaddleVariables");

exports.CarouselPaddle = _carouselPaddleVariables.carouselPaddleVariables;

var _tableVariables = require("./components/Table/tableVariables");

exports.Table = _tableVariables.tableVariables;

var _tableRowVariables = require("./components/Table/tableRowVariables");

exports.TableRow = _tableRowVariables.tableRowVariables;

var _tableCellVariables = require("./components/Table/tableCellVariables");

exports.TableCell = _tableCellVariables.tableCellVariables;

var _cardVariables = require("./components/Card/cardVariables");

exports.Card = _cardVariables.cardVariables;

var _cardHeaderVariables = require("./components/Card/cardHeaderVariables");

exports.CardHeader = _cardHeaderVariables.cardHeaderVariables;

var _cardBodyVariables = require("./components/Card/cardBodyVariables");

exports.CardBody = _cardBodyVariables.cardBodyVariables;

var _cardFooterVariables = require("./components/Card/cardFooterVariables");

exports.CardFooter = _cardFooterVariables.cardFooterVariables;

var _cardPreviewVariables = require("./components/Card/cardPreviewVariables");

exports.CardPreview = _cardPreviewVariables.cardPreviewVariables;

var _cardTopControlsVariables = require("./components/Card/cardTopControlsVariables");

exports.CardTopControls = _cardTopControlsVariables.cardTopControlsVariables;

var _cardExpandableBoxVariables = require("./components/Card/cardExpandableBoxVariables");

exports.CardExpandableBox = _cardExpandableBoxVariables.cardExpandableBoxVariables;

var _datepickerVariables = require("./components/Datepicker/datepickerVariables");

exports.Datepicker = _datepickerVariables.datepickerVariables;

var _datepickerCalendarCellVariables = require("./components/Datepicker/datepickerCalendarCellVariables");

exports.DatepickerCalendarCell = _datepickerCalendarCellVariables.datepickerCalendarCellVariables;

var _datepickerCalendarCellButtonVariables = require("./components/Datepicker/datepickerCalendarCellButtonVariables");

exports.DatepickerCalendarCellButton = _datepickerCalendarCellButtonVariables.datepickerCalendarCellButtonVariables;

var _datepickerCalendarVariables = require("./components/Datepicker/datepickerCalendarVariables");

exports.DatepickerCalendar = _datepickerCalendarVariables.datepickerCalendarVariables;

var _datepickerCalendarHeaderVariables = require("./components/Datepicker/datepickerCalendarHeaderVariables");

exports.DatepickerCalendarHeader = _datepickerCalendarHeaderVariables.datepickerCalendarHeaderVariables;

var _datepickerCalendarHeaderCellVariables = require("./components/Datepicker/datepickerCalendarHeaderCellVariables");

exports.DatepickerCalendarHeaderCell = _datepickerCalendarHeaderCellVariables.datepickerCalendarHeaderCellVariables;

var _datepickerCalendarGridVariables = require("./components/Datepicker/datepickerCalendarGridVariables");

exports.DatepickerCalendarGrid = _datepickerCalendarGridVariables.datepickerCalendarGridVariables;

var _datepickerCalendarGridRowVariables = require("./components/Datepicker/datepickerCalendarGridRowVariables");

exports.DatepickerCalendarGridRow = _datepickerCalendarGridRowVariables.datepickerCalendarGridRowVariables;
