// botbuilder/lib/index.js
"use strict";
/**
 * @module botbuilder
 */
/**
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TokenResolver = exports.StreamingHttpClient = exports.StatusCodeError = exports.HandoffEventNames = exports.EventFactory = exports.CloudAdapter = exports.ChannelServiceRoutes = exports.ChannelServiceHandlerBase = exports.ChannelServiceHandler = exports.BotFrameworkHttpClient = exports.BotFrameworkAdapter = void 0;
__exportStar(require("botbuilder-core"), exports);
__exportStar(require("./fileTranscriptStore"), exports);
__exportStar(require("./inspectionMiddleware"), exports);
__exportStar(require("./setSpeakMiddleware"), exports);
__exportStar(require("./skills"), exports);
__exportStar(require("./teams"), exports);
__exportStar(require("./teamsActivityHandler"), exports);
__exportStar(require("./teamsActivityHelpers"), exports);
__exportStar(require("./teamsInfo"), exports);
var botFrameworkAdapter_1 = require("./botFrameworkAdapter");
Object.defineProperty(exports, "BotFrameworkAdapter", { enumerable: true, get: function () { return botFrameworkAdapter_1.BotFrameworkAdapter; } });
var botFrameworkHttpClient_1 = require("./botFrameworkHttpClient");
Object.defineProperty(exports, "BotFrameworkHttpClient", { enumerable: true, get: function () { return botFrameworkHttpClient_1.BotFrameworkHttpClient; } });
var channelServiceHandler_1 = require("./channelServiceHandler");
Object.defineProperty(exports, "ChannelServiceHandler", { enumerable: true, get: function () { return channelServiceHandler_1.ChannelServiceHandler; } });
var channelServiceHandlerBase_1 = require("./channelServiceHandlerBase");
Object.defineProperty(exports, "ChannelServiceHandlerBase", { enumerable: true, get: function () { return channelServiceHandlerBase_1.ChannelServiceHandlerBase; } });
var channelServiceRoutes_1 = require("./channelServiceRoutes");
Object.defineProperty(exports, "ChannelServiceRoutes", { enumerable: true, get: function () { return channelServiceRoutes_1.ChannelServiceRoutes; } });
var cloudAdapter_1 = require("./cloudAdapter");
Object.defineProperty(exports, "CloudAdapter", { enumerable: true, get: function () { return cloudAdapter_1.CloudAdapter; } });
var eventFactory_1 = require("./eventFactory");
Object.defineProperty(exports, "EventFactory", { enumerable: true, get: function () { return eventFactory_1.EventFactory; } });
var handoffEventNames_1 = require("./handoffEventNames");
Object.defineProperty(exports, "HandoffEventNames", { enumerable: true, get: function () { return handoffEventNames_1.HandoffEventNames; } });
var statusCodeError_1 = require("./statusCodeError");
Object.defineProperty(exports, "StatusCodeError", { enumerable: true, get: function () { return statusCodeError_1.StatusCodeError; } });
var streaming_1 = require("./streaming");
Object.defineProperty(exports, "StreamingHttpClient", { enumerable: true, get: function () { return streaming_1.StreamingHttpClient; } });
Object.defineProperty(exports, "TokenResolver", { enumerable: true, get: function () { return streaming_1.TokenResolver; } });
