// @fluentui/react-northstar/dist/dts/src/themes/teams/components/Label/labelStyles.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.labelStyles = void 0;
var utils_1 = require("../../../../utils");
exports.labelStyles = {
    root: function (_a) {
        var p = _a.props, v = _a.variables, theme = _a.theme;
        var siteVariables = theme.siteVariables;
        return __assign(__assign(__assign(__assign(__assign(__assign(__assign(__assign(__assign(__assign({ display: 'inline-flex', alignItems: 'center', overflow: 'hidden', height: v.height, lineHeight: v.height, color: v.foreground, backgroundColor: v.background, fontSize: utils_1.pxToRem(14), borderRadius: siteVariables.borderRadiusMedium, padding: v.padding }, (p.hasImage &&
            (p.imagePosition === 'start' ? { paddingLeft: v.startPaddingLeft } : { paddingRight: v.endPaddingRight }))), (p.circular && {
            borderRadius: v.circularRadius,
        })), (p.color === 'black' && {
            color: v.blackForeground,
            backgroundColor: v.blackBackground,
        })), (p.color === 'white' && {
            color: v.whiteForeground,
            backgroundColor: v.whiteBackground,
        })), (p.color === 'brand' && {
            color: v.brandForeground,
            backgroundColor: v.brandBackground,
        })), (p.color === 'grey' && {
            color: v.greyForeground,
            backgroundColor: v.greyBackground,
        })), (p.color === 'orange' && {
            color: v.orangeForeground,
            backgroundColor: v.orangeBackground,
        })), (p.color === 'red' && {
            color: v.redForeground,
            backgroundColor: v.redBackground,
        })), (p.color === 'green' && {
            color: v.greenForeground,
            backgroundColor: v.greenBackground,
        })), (p.color === 'yellow' && {
            color: v.yellowForeground,
            backgroundColor: v.yellowBackground,
        }));
    },
    content: function (_a) {
        var p = _a.props, v = _a.variables;
        var hasStartElement = (p.hasImage && p.imagePosition === 'start') || (p.hasIcon && p.iconPosition === 'start');
        var hasEndElement = (p.hasImage && p.imagePosition === 'end') || (p.hasIcon && p.iconPosition === 'end');
        return __assign(__assign(__assign(__assign({}, (hasStartElement && {
            marginLeft: utils_1.pxToRem(3),
        })), (!hasStartElement &&
            p.circular && {
            marginLeft: utils_1.pxToRem(4),
        })), (hasEndElement && {
            marginRight: utils_1.pxToRem(3),
        })), (!hasEndElement &&
            p.circular && {
            marginRight: utils_1.pxToRem(4),
        }));
    },
    image: function (_a) {
        var v = _a.variables;
        return ({
            height: v.height,
            width: v.height,
        });
    },
    icon: function (_a) {
        var p = _a.props, v = _a.variables;
        return (__assign({ display: 'inline-flex', alignItems: 'center', justifyContent: 'center', width: v.iconSize, height: v.iconSize, '& > :first-child': {
                height: '100%',
                width: '100%',
                '& svg': {
                    height: '100%',
                    width: '100%',
                },
            } }, (p.hasActionableIcon && {
            cursor: 'pointer',
        })));
    },
};
