// @fluentui/react-northstar/dist/dts/src/themes/teams/components/Label/labelVariables.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.labelVariables = void 0;
var utils_1 = require("../../../../utils");
var labelVariables = function (siteVars) {
    return {
        circularRadius: utils_1.pxToRem(9999),
        iconSize: utils_1.pxToRem(16),
        padding: "0 " + utils_1.pxToRem(4) + " 0 " + utils_1.pxToRem(4),
        startPaddingLeft: '0px',
        endPaddingRight: '0px',
        height: utils_1.pxToRem(20),
        foreground: 'rgba(0, 0, 0, 0.6)',
        background: 'rgb(232, 232, 232)',
        blackForeground: siteVars.colorScheme.black.foreground1,
        blackBackground: siteVars.colorScheme.black.background1,
        whiteForeground: siteVars.colorScheme.white.foreground1,
        whiteBackground: siteVars.colorScheme.white.background1,
        brandForeground: siteVars.colorScheme.brand.foreground5,
        brandBackground: siteVars.colorScheme.brand.background6,
        greyForeground: siteVars.colorScheme.grey.foreground2,
        greyBackground: siteVars.colorScheme.grey.background3,
        orangeForeground: siteVars.colorScheme.orange.foreground2,
        orangeBackground: siteVars.colorScheme.orange.background1,
        redForeground: siteVars.colorScheme.red.foreground1,
        redBackground: siteVars.colorScheme.red.background,
        greenForeground: siteVars.colorScheme.green.foreground3,
        greenBackground: siteVars.colorScheme.green.background1,
        yellowForeground: siteVars.colorScheme.yellow.foreground3,
        yellowBackground: siteVars.colorScheme.yellow.background,
    };
};
exports.labelVariables = labelVariables;
