// @fluentui/react-northstar/dist/dts/src/components/Grid/Grid.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Grid = exports.gridClassName = void 0;
var react_bindings_1 = require("@fluentui/react-bindings");
var customPropTypes = require("@fluentui/react-proptypes");
var PropTypes = require("prop-types");
var React = require("react");
var utils_1 = require("../../utils");
exports.gridClassName = 'ui-grid';
/**
 * A Grid is a layout component that harmonizes negative space, by controlling both the row and column alignment.
 */
exports.Grid = React.forwardRef(function (props, ref) {
    var context = react_bindings_1.useFluentContext();
    var _a = react_bindings_1.useTelemetry(exports.Grid.displayName, context.telemetry), setStart = _a.setStart, setEnd = _a.setEnd;
    setStart();
    var accessibility = props.accessibility, children = props.children, className = props.className, columns = props.columns, content = props.content, design = props.design, rows = props.rows, styles = props.styles, variables = props.variables;
    var getA11yProps = react_bindings_1.useAccessibility(accessibility, {
        debugName: exports.Grid.displayName,
        rtl: context.rtl,
    });
    var classes = react_bindings_1.useStyles(exports.Grid.displayName, {
        className: exports.gridClassName,
        mapPropsToStyles: function () { return ({ columns: columns, rows: rows }); },
        mapPropsToInlineStyles: function () { return ({
            className: className,
            design: design,
            styles: styles,
            variables: variables,
        }); },
        rtl: context.rtl,
    }).classes;
    var ElementType = react_bindings_1.getElementType(props);
    var unhandledProps = react_bindings_1.useUnhandledProps(exports.Grid.handledProps, props);
    var element = getA11yProps.unstable_wrapWithFocusZone(React.createElement(ElementType, __assign({}, getA11yProps('root', __assign(__assign({ className: classes.root, ref: ref }, utils_1.rtlTextContainer.getAttributes({ forElements: [children, content] })), unhandledProps))), utils_1.childrenExist(children) ? children : content));
    setEnd();
    return element;
});
exports.Grid.displayName = 'Grid';
exports.Grid.propTypes = __assign(__assign({}, utils_1.commonPropTypes.createCommon({
    content: false,
})), { columns: PropTypes.oneOfType([PropTypes.string, PropTypes.number]), content: customPropTypes.every([
        customPropTypes.disallow(['children']),
        PropTypes.oneOfType([PropTypes.arrayOf(customPropTypes.nodeContent), customPropTypes.nodeContent]),
    ]), rows: PropTypes.oneOfType([PropTypes.string, PropTypes.number]) });
exports.Grid.handledProps = Object.keys(exports.Grid.propTypes);
