// @fluentui/accessibility/dist/commonjs/behaviors/SplitButton/splitButtonBehavior.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.splitButtonBehavior = void 0;

var _merge2 = _interopRequireDefault(require("lodash/merge"));

var _keyboardKey = require("../../keyboard-key");

var _attributes = require("../../attributes");

var _menuButtonBehavior = require("../MenuButton/menuButtonBehavior");

/**
 * @description
 * Adds attribute 'tabIndex=-1' based on the property 'open' to 'trigger' slot.
 * Adds attribute 'aria-controls=menu-id' based on the property 'open' to 'trigger' slot.
 * Adds attribute 'aria-expanded=true' based on the property 'open' to 'trigger' slot.
 * Adds attribute 'id=trigger-id' based on the property 'triggerId' to 'trigger' slot.
 * Adds attribute 'id=menu-id' based on the property 'menuId' to 'menu' slot.
 * Adds attribute 'aria-labelledby=trigger-id' based on the property 'triggerId' to 'menu' slot.
 * Triggers 'closeAndFocusTrigger' action with 'Escape' or 'altKey'+'ArrowUp'.
 * Triggers 'openAndFocusFirst' action with 'altKey'+'ArrowDown' on 'root' slot.
 * Triggers 'stopPropagation' action with 'ArrowLeft' or 'ArrowRight' on 'root' slot.
 *
 * @specification
 * Adds attribute 'tabIndex=-1' to 'toggleButton' slot.
 * Adds attribute 'aria-haspopup=true' to 'toggleButton' slot.
 * Adds attribute 'data-is-focusable=false' to 'toggleButton' slot.
 */
var splitButtonBehavior = function splitButtonBehavior(props) {
  var _toggleButton;

  var splitButtonMenuButtonBehavior = function splitButtonMenuButtonBehavior() {
    var menuButtonBehaviorData = (0, _menuButtonBehavior.menuButtonBehavior)(props);
    menuButtonBehaviorData.attributes.trigger['aria-haspopup'] = undefined;
    return (0, _merge2.default)(menuButtonBehaviorData, {
      keyActions: {
        popup: {
          closeAndFocusTrigger: {
            keyCombinations: [{
              keyCode: _keyboardKey.keyboardKey.Escape
            }, {
              keyCode: _keyboardKey.keyboardKey.ArrowUp,
              altKey: true
            }]
          },
          stopPropagation: {
            keyCombinations: [{
              keyCode: _keyboardKey.keyboardKey.ArrowLeft
            }, {
              keyCode: _keyboardKey.keyboardKey.ArrowRight
            }]
          }
        },
        root: Object.assign({}, !props.open && {
          openAndFocusFirst: {
            keyCombinations: [{
              keyCode: _keyboardKey.keyboardKey.ArrowDown,
              altKey: true
            }]
          }
        })
      }
    });
  };

  return {
    attributes: {
      root: {},
      toggleButton: (_toggleButton = {
        tabIndex: -1,
        'aria-haspopup': true
      }, _toggleButton[_attributes.IS_FOCUSABLE_ATTRIBUTE] = false, _toggleButton)
    },
    childBehaviors: {
      menuButton: splitButtonMenuButtonBehavior
    }
  };
};

exports.splitButtonBehavior = splitButtonBehavior;
