// @fluentui/react-northstar/dist/commonjs/themes/teams/components/Input/inputVariables.js
"use strict";

exports.__esModule = true;
exports.inputVariables = void 0;

var _utils = require("../../../../utils");

var inputVariables = function inputVariables(siteVars) {
  return {
    colorDisabled: siteVars.colorScheme.brand.foregroundDisabled,
    colorError: siteVars.colorScheme.red.foreground,
    borderColorError: siteVars.colorScheme.red.background,
    iconPosition: 'absolute',
    iconRight: (0, _utils.pxToRem)(10),
    iconLeft: (0, _utils.pxToRem)(9),
    inputPaddingWithIconAtStart: (0, _utils.pxToRem)(5) + " " + (0, _utils.pxToRem)(12) + " " + (0, _utils.pxToRem)(5) + " " + (0, _utils.pxToRem)(34),
    inputPaddingWithIconAtEnd: (0, _utils.pxToRem)(5) + " " + (0, _utils.pxToRem)(35) + " " + (0, _utils.pxToRem)(5) + " " + (0, _utils.pxToRem)(12),
    inputPadding: (0, _utils.pxToRem)(5) + " " + (0, _utils.pxToRem)(12),
    inputInsideLabelPaddingTop: (0, _utils.pxToRem)(14),
    borderColor: 'transparent',
    borderRadius: siteVars.borderRadiusMedium,
    borderWidth: "0 0 " + (0, _utils.pxToRem)(2) + " 0",
    backgroundColor: siteVars.colorScheme.default.background2,
    backgroundColorInverted: siteVars.colorScheme.default.background,
    fontColor: siteVars.colorScheme.default.foreground,
    fontSize: siteVars.fontSizes.medium,
    iconColor: siteVars.colorScheme.default.foreground,
    successfulColor: siteVars.colorScheme.green.foreground,
    inputFocusBorderColor: "transparent transparent " + siteVars.colorScheme.brand.borderFocus1 + " transparent",
    inputFocusBorderRadius: siteVars.borderRadiusMedium + " " + siteVars.borderRadiusMedium + " " + siteVars.borderRadiusSmall + " " + siteVars.borderRadiusSmall,
    placeholderColor: siteVars.colorScheme.default.foreground1
  };
};

exports.inputVariables = inputVariables;
