// @fluentui/react-northstar/dist/commonjs/themes/teams/components/Button/buttonGroupStyles.js
"use strict";

exports.__esModule = true;
exports.buttonGroupStyles = void 0;

var _utils = require("../../../../utils");

var commonButtonsStyles = function commonButtonsStyles(circular) {
  return circular ? {
    marginRight: (0, _utils.pxToRem)(8)
  } : {
    borderRadius: 0
  };
};

var buttonGroupStyles = {
  root: function root() {
    return {};
  },
  middleButton: function middleButton(_ref) {
    var p = _ref.props;
    return Object.assign({}, commonButtonsStyles(p.circular));
  },
  firstButton: function firstButton(_ref2) {
    var p = _ref2.props,
        v = _ref2.variables;
    return Object.assign({}, commonButtonsStyles(p.circular), !p.circular && {
      borderTopLeftRadius: v.borderRadius,
      borderBottomLeftRadius: v.borderRadius
    });
  },
  lastButton: function lastButton(_ref3) {
    var p = _ref3.props,
        v = _ref3.variables;
    return Object.assign({}, commonButtonsStyles(p.circular), !p.circular && {
      borderTopRightRadius: v.borderRadius,
      borderBottomRightRadius: v.borderRadius
    });
  }
};
exports.buttonGroupStyles = buttonGroupStyles;
