// @fluentui/react-northstar/dist/commonjs/themes/teams-dark/components/Menu/menuVariables.js
"use strict";

exports.__esModule = true;
exports.menuVariables = void 0;

var _menuVariables = require("../../../teams/components/Menu/menuVariables");

var _colorUtils = require("../../../colorUtils");

var menuVariables = function menuVariables(siteVars) {
  return {
    colorScheme: (0, _colorUtils.pickValuesFromColorScheme)((0, _colorUtils.extendColorScheme)(siteVars.colorScheme, {
      default: {
        borderActive: siteVars.colors.grey[600],
        backgroundFocus: siteVars.colors.grey[500],
        backgroundActive: siteVars.colorScheme.default.backgroundActive1,
        foregroundDisabled: siteVars.colorScheme.default.foregroundDisabled1
      },
      brand: {
        foregroundHover: siteVars.colors.white,
        backgroundHover: siteVars.colors.brand[300],
        borderActive: siteVars.colors.brand[400],
        foregroundActive: siteVars.colors.white,
        foregroundFocus: siteVars.colors.white,
        backgroundFocus: siteVars.colors.brand[300],
        foregroundDisabled: siteVars.colorScheme.brand.foregroundDisabled1
      }
    }), _menuVariables.menuColorAreas),
    color: siteVars.colors.grey[250],
    colorActive: siteVars.colors.white,
    borderColor: siteVars.colorScheme.default.border2,
    primaryBorderColor: siteVars.colors.grey[600],
    pointingIndicatorBackgroundColor: siteVars.colors.brand[400],
    activeUnderlinedBorderBottomColor: siteVars.colors.white,
    activeUnderlinedPrimaryColor: siteVars.colors.brand[400],
    verticalBackgroundColorFocus: siteVars.colors.grey[550],
    iconOnlyColorActive: siteVars.colors.brand[400],
    iconOnlyColorFocus: siteVars.colors.brand[400],
    iconOnlyColorHover: siteVars.colors.brand[400],
    wrapperColorActive: siteVars.colors.grey[250],
    wrapperColorFocus: siteVars.colors.white,
    wrapperColorHover: siteVars.colors.white,
    primaryIndicatorColorHover: siteVars.colors.white,
    wrapperBackgroundColorFocus: siteVars.colors.brand[400],
    underlinedWrapperColorHover: siteVars.colors.white,
    primaryWrapperColorFocus: siteVars.colors.white,
    primaryWrapperBackgroundColorFocus: siteVars.colors.brand[400],
    indicatorColor: siteVars.colors.grey[250],
    activeIndicatorColor: siteVars.colors.white,
    activePrimaryIndicatorColor: siteVars.colors.white,
    activePrimaryVerticalIndicatorColor: siteVars.colors.white,
    indicatorColorHover: siteVars.colors.white
  };
};

exports.menuVariables = menuVariables;
