// @fluentui/react-northstar/dist/dts/src/themes/teams/components/Checkbox/checkboxVariables.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.checkboxVariables = void 0;
var utils_1 = require("../../../../utils");
var _ = require("lodash");
var toggleMovementDistance = utils_1.pxToRem(20);
var padding = utils_1.pxToRem(2);
var defaultValue = 'red';
var checkboxVariables = function (siteVars) { return ({
    checkboxColor: 'transparent',
    toggleBackground: 'transparent',
    toggleBorderColor: siteVars.colors.grey[300],
    toggleBorderStyle: "solid",
    toggleBorderWidth: utils_1.pxToRem(1),
    toggleIndicatorColor: 'inherit',
    checkboxCheckedColor: siteVars.colors.grey[500],
    checkboxToggleCheckedBackground: 'transparent',
    checkboxToggleCheckedBorderColor: siteVars.colors.grey[500],
    checkboxToggleCheckedColor: 'inherit',
    disabledColor: siteVars.colors.grey[300],
    disabledCheckboxColor: siteVars.colors.grey[300],
    disabledToggleBackground: 'transparent',
    disabledToggleBorderColor: siteVars.colors.grey[200],
    textColor: _.get(siteVars, 'colorScheme.default.foreground1', defaultValue),
    background: 'transparent',
    borderColor: _.get(siteVars, 'colorScheme.default.foreground1', defaultValue),
    borderStyle: 'solid',
    borderRadius: siteVars.borderRadiusSmall,
    borderWidth: utils_1.pxToRem(1),
    indicatorColor: 'transparent',
    gap: utils_1.pxToRem(12),
    margin: utils_1.pxToRem(2.8) + " 0 0 0",
    padding: padding,
    rootPadding: '3px 5px',
    textColorHover: _.get(siteVars, 'colorScheme.default.foreground', defaultValue),
    borderColorHover: _.get(siteVars, 'colorScheme.default.foreground', defaultValue),
    checkedBackgroundHover: _.get(siteVars, 'colorScheme.brand.backgroundHover', defaultValue),
    toggleBorderRadius: utils_1.pxToRem(999),
    toggleIndicatorSize: utils_1.pxToRem(14),
    toggleMargin: '0',
    togglePadding: padding + " " + toggleMovementDistance + " " + padding + " " + padding,
    toggleWidth: utils_1.pxToRem(38),
    toggleHeight: utils_1.pxToRem(20),
    checkedTextColor: _.get(siteVars, 'colorScheme.default.foreground', defaultValue),
    checkedBackground: _.get(siteVars, 'colorScheme.brand.backgroundActive1', defaultValue),
    checkedBorderColor: _.get(siteVars, 'colorScheme.brand.backgroundActive1', defaultValue),
    checkedIndicatorColor: _.get(siteVars, 'colorScheme.default.background', defaultValue),
    toggleCheckedPadding: padding + " " + padding + " " + padding + " " + toggleMovementDistance,
    disabledBackground: _.get(siteVars, 'colorScheme.default.background', defaultValue),
    disabledBackgroundChecked: _.get(siteVars, 'colorScheme.default.backgroundDisabled', defaultValue),
    disabledBorderColor: _.get(siteVars, 'colorScheme.default.foregroundDisabled1', defaultValue),
    disabledCheckedIndicatorColor: _.get(siteVars, 'colorScheme.default.foregroundDisabled', defaultValue),
    disabledToggleIndicatorColor: _.get(siteVars, 'colorScheme.default.foregroundDisabled', defaultValue),
}); };
exports.checkboxVariables = checkboxVariables;
