// botframework-connector/lib/auth/credentialProvider.js
"use strict";
/**
 * @module botframework-connector
 */
/**
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.SimpleCredentialProvider = void 0;
/**
 * @deprecated Use `ConfigurationBotFrameworkAuthentication` instead to configure credentials.
 * A simple implementation of the [ICredentialProvider](xref:botframework-connector.ICredentialProvider) interface.
 */
class SimpleCredentialProvider {
    /**
     * Initializes a new instance of the [SimpleCredentialProvider](xref:botframework-connector.SimpleCredentialProvider) class with the provided credentials.
     *
     * @param {string} appId The app ID.
     * @param {string} appPassword The app password.
     */
    constructor(appId, appPassword) {
        this.appId = appId;
        this.appPassword = appPassword;
    }
    // Protects against JSON.stringify leaking secrets
    toJSON() {
        return { name: this.constructor.name, appId: this.appId };
    }
    /**
     * Validate AppId.
     *
     * This method is async to enable custom implementations
     * that may need to call out to service to validate the appId / password pair.
     *
     * @param {string} appId bot appid
     * @returns {Promise<boolean>} true if it is a valid AppId
     */
    isValidAppId(appId) {
        return Promise.resolve(this.appId === appId);
    }
    /**
     * Get the app password for a given bot appId, if it is not a valid appId, return Null
     *
     * This method is async to enable custom implementations
     * that may need to call out to serviced to validate the appId / password pair.
     *
     * @param {string} appId bot appid
     * @returns {Promise<string|null>} password or null for invalid appid
     */
    getAppPassword(appId) {
        return Promise.resolve(this.appId === appId ? this.appPassword : null);
    }
    /**
     * Checks if bot authentication is disabled.
     * Return true if bot authentication is disabled.
     *
     * This method is async to enable custom implementations
     * that may need to call out to serviced to validate the appId / password pair.
     *
     * @returns {Promise<boolean>} true if bot authentication is disabled.
     */
    isAuthenticationDisabled() {
        return Promise.resolve(!this.appId);
    }
}
exports.SimpleCredentialProvider = SimpleCredentialProvider;
