// @fluentui/react-northstar/dist/dts/test/specs/themes/colorUtils-test.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
var colorUtils_1 = require("src/themes/colorUtils");
var generateColorSchemeValues = function (color) { return ({
    foreground: color,
    background: color,
    border: color,
    shadow: color,
    foregroundHover: color,
    backgroundHover: color,
    borderHover: color,
    shadowHover: color,
    foregroundActive: color,
    backgroundActive: color,
    borderActive: color,
    shadowActive: color,
    foregroundFocus: color,
    backgroundFocus: color,
    borderFocus: color,
    shadowFocus: color,
    foregroundPressed: color,
    backgroundPressed: color,
    borderPressed: color,
    shadowPressed: color,
    foregroundDisabled: color,
    backgroundDisabled: color,
    borderDisabled: color,
    shadowDisabled: color,
}); };
describe('colorUtils', function () {
    describe('extendColorScheme', function () {
        test('replaced the defined color tokens with the overrides definition', function () {
            var baseColorScheme = {
                default: generateColorSchemeValues('grey'),
                primary: generateColorSchemeValues('purple'),
            };
            var overrides = {
                default: {
                    foreground: 'black',
                },
                primary: {
                    background: 'blue',
                },
            };
            expect(colorUtils_1.extendColorScheme(baseColorScheme, overrides)).toMatchObject({
                default: __assign(__assign({}, baseColorScheme.default), { foreground: 'black' }),
                primary: __assign(__assign({}, baseColorScheme.primary), { background: 'blue' }),
            });
        });
        test('extends the defined color tokens with new values if provided in the overrides', function () {
            var baseColorScheme = {
                default: generateColorSchemeValues('grey'),
                primary: generateColorSchemeValues('primary'),
            };
            var overrides = {
                primary: {
                    background1: 'red',
                },
            };
            expect(colorUtils_1.extendColorScheme(baseColorScheme, overrides)).toMatchObject({
                default: __assign({}, baseColorScheme.defualt),
                primary: __assign(__assign({}, baseColorScheme.primary), { background1: 'red' }),
            });
        });
    });
    describe('pickValuesFromColorScheme', function () {
        test('picks the provided values from the color scheme', function () {
            var baseColorScheme = {
                default: generateColorSchemeValues('grey'),
                primary: generateColorSchemeValues('purple'),
            };
            var componentAreas = ['foreground', 'background'];
            expect(colorUtils_1.pickValuesFromColorScheme(baseColorScheme, componentAreas)).toMatchObject({
                default: {
                    foreground: 'grey',
                    background: 'grey',
                },
                primary: {
                    foreground: 'purple',
                    background: 'purple',
                },
            });
        });
    });
});
