// @fluentui/react-northstar/dist/commonjs/components/List/ListItemHeader.js
"use strict";

exports.__esModule = true;
exports.ListItemHeader = exports.listItemHeaderClassName = void 0;

var _reactBindings = require("@fluentui/react-bindings");

var _utils = require("../../utils");

var _Box = require("../Box/Box");

var listItemHeaderClassName = 'ui-list__itemheader';
/**
 * Provides a header for the ListItem.
 */

exports.listItemHeaderClassName = listItemHeaderClassName;
var ListItemHeader = (0, _reactBindings.compose)(_Box.Box, {
  className: listItemHeaderClassName,
  displayName: 'ListItemHeader',
  mapPropsToStylesProps: function mapPropsToStylesProps(_ref) {
    var hasContent = _ref.hasContent,
        hasHeaderMedia = _ref.hasHeaderMedia,
        truncate = _ref.truncate;
    return {
      hasContent: hasContent,
      hasHeaderMedia: hasHeaderMedia,
      truncate: truncate
    };
  },
  handledProps: ['hasContent', 'hasHeaderMedia', 'truncate'],
  overrideStyles: true,
  shorthandConfig: {
    mappedProp: 'content'
  }
});
exports.ListItemHeader = ListItemHeader;
ListItemHeader.propTypes = _utils.commonPropTypes.createCommon();
