// @fluentui/react-northstar/dist/commonjs/components/List/ListItemMedia.js
"use strict";

exports.__esModule = true;
exports.ListItemMedia = exports.listItemMediaClassName = void 0;

var _reactBindings = require("@fluentui/react-bindings");

var _utils = require("../../utils");

var _Box = require("../Box/Box");

var listItemMediaClassName = 'ui-list__itemmedia';
/**
 * Provides a content for the ListItem.
 */

exports.listItemMediaClassName = listItemMediaClassName;
var ListItemMedia = (0, _reactBindings.compose)(_Box.Box, {
  className: listItemMediaClassName,
  displayName: 'ListItemMedia',
  mapPropsToStylesProps: function mapPropsToStylesProps(_ref) {
    var hasContent = _ref.hasContent,
        hasHeader = _ref.hasHeader,
        important = _ref.important;
    return {
      hasContent: hasContent,
      hasHeader: hasHeader,
      important: important
    };
  },
  handledProps: ['hasContent', 'hasHeader', 'important'],
  overrideStyles: true,
  shorthandConfig: {
    mappedProp: 'content'
  }
});
exports.ListItemMedia = ListItemMedia;
ListItemMedia.propTypes = _utils.commonPropTypes.createCommon();
