// @fluentui/react-northstar/dist/commonjs/components/List/ListItemEndMedia.js
"use strict";

exports.__esModule = true;
exports.ListItemEndMedia = exports.listItemEndMediaClassName = void 0;

var _reactBindings = require("@fluentui/react-bindings");

var _utils = require("../../utils");

var _Box = require("../Box/Box");

var listItemEndMediaClassName = 'ui-list__itemendmedia';
/**
 * Provides a header media for the ListItem.
 */

exports.listItemEndMediaClassName = listItemEndMediaClassName;
var ListItemEndMedia = (0, _reactBindings.compose)(_Box.Box, {
  className: listItemEndMediaClassName,
  displayName: 'ListItemEndMedia',
  mapPropsToStylesProps: function mapPropsToStylesProps(_ref) {
    var navigable = _ref.navigable,
        selectable = _ref.selectable;
    return {
      navigable: navigable,
      selectable: selectable
    };
  },
  handledProps: ['navigable', 'selectable'],
  overrideStyles: true,
  shorthandConfig: {
    mappedProp: 'content'
  }
});
exports.ListItemEndMedia = ListItemEndMedia;
ListItemEndMedia.propTypes = _utils.commonPropTypes.createCommon();
