// @fluentui/react-northstar/dist/commonjs/components/List/ListItemContent.js
"use strict";

exports.__esModule = true;
exports.ListItemContent = exports.listItemContentClassName = void 0;

var _reactBindings = require("@fluentui/react-bindings");

var _utils = require("../../utils");

var _Box = require("../Box/Box");

var listItemContentClassName = 'ui-list__itemcontent';
/**
 * Provides a content for the ListItem.
 */

exports.listItemContentClassName = listItemContentClassName;
var ListItemContent = (0, _reactBindings.compose)(_Box.Box, {
  className: listItemContentClassName,
  displayName: 'ListItemContent',
  mapPropsToStylesProps: function mapPropsToStylesProps(_ref) {
    var hasContentMedia = _ref.hasContentMedia,
        hasHeader = _ref.hasHeader,
        truncate = _ref.truncate;
    return {
      hasContentMedia: hasContentMedia,
      hasHeader: hasHeader,
      truncate: truncate
    };
  },
  handledProps: ['hasContentMedia', 'hasHeader', 'truncate'],
  overrideStyles: true,
  shorthandConfig: {
    mappedProp: 'content'
  }
});
exports.ListItemContent = ListItemContent;
ListItemContent.propTypes = _utils.commonPropTypes.createCommon();
