// @fluentui/react-northstar/dist/dts/src/components/Tooltip/TooltipContent.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TooltipContent = exports.tooltipContentClassName = void 0;
var react_bindings_1 = require("@fluentui/react-bindings");
var customPropTypes = require("@fluentui/react-proptypes");
var PropTypes = require("prop-types");
var React = require("react");
var utils_1 = require("../../utils");
var positioner_1 = require("../../utils/positioner");
exports.tooltipContentClassName = 'ui-tooltip__content';
/**
 * A TooltipContent contains the content of a Tooltip component.
 */
exports.TooltipContent = React.forwardRef(function (props, ref) {
    var context = react_bindings_1.useFluentContext();
    var _a = react_bindings_1.useTelemetry(exports.TooltipContent.displayName, context.telemetry), setStart = _a.setStart, setEnd = _a.setEnd;
    setStart();
    var accessibility = props.accessibility, children = props.children, className = props.className, content = props.content, design = props.design, open = props.open, placement = props.placement, pointing = props.pointing, pointerRef = props.pointerRef, styles = props.styles, variables = props.variables, subtle = props.subtle;
    var getA11Props = react_bindings_1.useAccessibility(accessibility, {
        debugName: exports.TooltipContent.displayName,
        rtl: context.rtl,
    });
    var classes = react_bindings_1.useStyles(exports.TooltipContent.displayName, {
        className: exports.tooltipContentClassName,
        mapPropsToStyles: function () { return ({
            basePlacement: positioner_1.getBasePlacement(placement, context.rtl),
            open: open,
            pointing: pointing,
            subtle: subtle,
        }); },
        mapPropsToInlineStyles: function () { return ({
            className: className,
            design: design,
            styles: styles,
            variables: variables,
        }); },
        rtl: context.rtl,
    }).classes;
    var ElementType = react_bindings_1.getElementType(props);
    var unhandledProps = react_bindings_1.useUnhandledProps(exports.TooltipContent.handledProps, props);
    var element = (React.createElement(ElementType, __assign({}, getA11Props('root', __assign(__assign({ className: classes.root, ref: ref }, utils_1.rtlTextContainer.getAttributes({ forElements: [children, content] })), unhandledProps))),
        open && pointing && React.createElement("div", { className: classes.pointer, ref: pointerRef }),
        React.createElement("div", __assign({}, getA11Props('content', { className: classes.content })), utils_1.childrenExist(children) ? children : content)));
    setEnd();
    return element;
});
exports.TooltipContent.displayName = 'TooltipContent';
exports.TooltipContent.propTypes = __assign(__assign({}, utils_1.commonPropTypes.createCommon()), { placement: PropTypes.oneOf([
        'auto-start',
        'auto',
        'auto-end',
        'top-start',
        'top',
        'top-end',
        'right-start',
        'right',
        'right-end',
        'bottom-end',
        'bottom',
        'bottom-start',
        'left-end',
        'left',
        'left-start',
    ]), pointing: PropTypes.bool, pointerRef: customPropTypes.ref, subtle: PropTypes.bool });
exports.TooltipContent.handledProps = Object.keys(exports.TooltipContent.propTypes);
exports.TooltipContent.create = utils_1.createShorthandFactory({ Component: exports.TooltipContent, mappedProp: 'content' });
