// @fluentui/accessibility/dist/dts/src/behaviors/Input/inputBehavior.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.inputBehavior = void 0;
var keyboard_key_1 = require("../../keyboard-key");
/**
 * @specification
 * Adds attribute 'aria-disabled=true' based on the property 'disabled' to 'input' slot.
 * Adds attribute 'aria-required=true' based on the property 'required' to 'input' slot.
 * Adds attribute 'aria-invalid=true' based on the property 'error' to 'input' slot.
 * Triggers 'clear' action with 'Escape' on 'input'.
 */
var inputBehavior = function (props) { return ({
    attributes: {
        input: __assign(__assign({ 'aria-disabled': props.disabled }, (props.required && { 'aria-required': true })), (props.error && { 'aria-invalid': true })),
    },
    keyActions: {
        input: {
            clear: {
                keyCombinations: [{ keyCode: keyboard_key_1.keyboardKey.Escape }],
            },
        },
    },
}); };
exports.inputBehavior = inputBehavior;
