// @fluentui/react-bindings/dist/dts/src/context.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Unstable_FluentContextProvider = exports.useFluentContext = exports.defaultContextValue = exports.defaultPerformanceFlags = void 0;
var react_northstar_styles_renderer_1 = require("@fluentui/react-northstar-styles-renderer");
var styles_1 = require("@fluentui/styles");
var React = require("react");
exports.defaultPerformanceFlags = {
    enableSanitizeCssPlugin: process.env.NODE_ENV !== 'production',
    enableStylesCaching: true,
    enableVariablesCaching: true,
    enableBooleanVariablesCaching: false,
};
exports.defaultContextValue = {
    // A default value for `rtl` is undefined to let compute `Provider` a proper one
    rtl: undefined,
    disableAnimations: false,
    performance: exports.defaultPerformanceFlags,
    renderer: react_northstar_styles_renderer_1.noopRenderer,
    theme: styles_1.emptyTheme,
    telemetry: undefined,
    target: undefined,
};
var FluentContext = React.createContext(exports.defaultContextValue);
function useFluentContext() {
    return React.useContext(FluentContext);
}
exports.useFluentContext = useFluentContext;
exports.Unstable_FluentContextProvider = FluentContext.Provider;
