// @fluentui/react-bindings/dist/dts/src/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.useContextSelectors = exports.useContextSelector = exports.createContext = exports.mergeVariablesOverrides = exports.getUnhandledProps = exports.getElementType = exports.childrenExist = exports.useTelemetry = exports.deprecated_getTelemetry = exports.unstable_getStyles = exports.unstable_calculateAnimationTimeout = exports.unstable_createAnimationStyles = exports.RendererContext = exports.useOnIFrameFocus = exports.usePrevious = exports.useUnhandledProps = exports.useTriggerElement = exports.useStateManager = exports.useMergedRefs = exports.useIsomorphicLayoutEffect = exports.useFirstMount = exports.useEventCallback = exports.useDeepMemo = exports.unstable_useDispatchEffect = exports.useCallbackRef = exports.useAutoControlled = exports.useAccessibility = exports.FocusZone = exports.FocusTrapZone = exports.AutoFocusZone = exports.unstable_getAccessibility = void 0;
var getAccessibility_1 = require("./accessibility/getAccessibility");
Object.defineProperty(exports, "unstable_getAccessibility", { enumerable: true, get: function () { return getAccessibility_1.getAccessibility; } });
__exportStar(require("./accessibility/types"), exports);
__exportStar(require("./compose"), exports);
var AutoFocusZone_1 = require("./FocusZone/AutoFocusZone");
Object.defineProperty(exports, "AutoFocusZone", { enumerable: true, get: function () { return AutoFocusZone_1.AutoFocusZone; } });
__exportStar(require("./FocusZone/AutoFocusZone.types"), exports);
var FocusTrapZone_1 = require("./FocusZone/FocusTrapZone");
Object.defineProperty(exports, "FocusTrapZone", { enumerable: true, get: function () { return FocusTrapZone_1.FocusTrapZone; } });
__exportStar(require("./FocusZone/FocusTrapZone.types"), exports);
var FocusZone_1 = require("./FocusZone/FocusZone");
Object.defineProperty(exports, "FocusZone", { enumerable: true, get: function () { return FocusZone_1.FocusZone; } });
__exportStar(require("./FocusZone/FocusZone.types"), exports);
__exportStar(require("./FocusZone/focusUtilities"), exports);
var useAccessibility_1 = require("./hooks/useAccessibility");
Object.defineProperty(exports, "useAccessibility", { enumerable: true, get: function () { return useAccessibility_1.useAccessibility; } });
var useAutoControlled_1 = require("./hooks/useAutoControlled");
Object.defineProperty(exports, "useAutoControlled", { enumerable: true, get: function () { return useAutoControlled_1.useAutoControlled; } });
var useCallbackRef_1 = require("./hooks/useCallbackRef");
Object.defineProperty(exports, "useCallbackRef", { enumerable: true, get: function () { return useCallbackRef_1.useCallbackRef; } });
var useDispatchEffect_1 = require("./hooks/useDispatchEffect");
Object.defineProperty(exports, "unstable_useDispatchEffect", { enumerable: true, get: function () { return useDispatchEffect_1.useDispatchEffect; } });
var useDeepMemo_1 = require("./hooks/useDeepMemo");
Object.defineProperty(exports, "useDeepMemo", { enumerable: true, get: function () { return useDeepMemo_1.useDeepMemo; } });
var useEventCallback_1 = require("./hooks/useEventCallback");
Object.defineProperty(exports, "useEventCallback", { enumerable: true, get: function () { return useEventCallback_1.useEventCallback; } });
var useFirstMount_1 = require("./hooks/useFirstMount");
Object.defineProperty(exports, "useFirstMount", { enumerable: true, get: function () { return useFirstMount_1.useFirstMount; } });
var useIsomorphicLayoutEffect_1 = require("./hooks/useIsomorphicLayoutEffect");
Object.defineProperty(exports, "useIsomorphicLayoutEffect", { enumerable: true, get: function () { return useIsomorphicLayoutEffect_1.useIsomorphicLayoutEffect; } });
var useMergedRefs_1 = require("./hooks/useMergedRefs");
Object.defineProperty(exports, "useMergedRefs", { enumerable: true, get: function () { return useMergedRefs_1.useMergedRefs; } });
var useStateManager_1 = require("./hooks/useStateManager");
Object.defineProperty(exports, "useStateManager", { enumerable: true, get: function () { return useStateManager_1.useStateManager; } });
__exportStar(require("./hooks/useStyles"), exports);
var useTriggerElement_1 = require("./hooks/useTriggerElement");
Object.defineProperty(exports, "useTriggerElement", { enumerable: true, get: function () { return useTriggerElement_1.useTriggerElement; } });
var useUnhandledProps_1 = require("./hooks/useUnhandledProps");
Object.defineProperty(exports, "useUnhandledProps", { enumerable: true, get: function () { return useUnhandledProps_1.useUnhandledProps; } });
var usePrevious_1 = require("./hooks/usePrevious");
Object.defineProperty(exports, "usePrevious", { enumerable: true, get: function () { return usePrevious_1.usePrevious; } });
var useOnIFrameFocus_1 = require("./hooks/useOnIFrameFocus");
Object.defineProperty(exports, "useOnIFrameFocus", { enumerable: true, get: function () { return useOnIFrameFocus_1.useOnIFrameFocus; } });
var RendererContext_1 = require("./renderer/RendererContext");
Object.defineProperty(exports, "RendererContext", { enumerable: true, get: function () { return RendererContext_1.RendererContext; } });
var createAnimationStyles_1 = require("./styles/createAnimationStyles");
Object.defineProperty(exports, "unstable_createAnimationStyles", { enumerable: true, get: function () { return createAnimationStyles_1.createAnimationStyles; } });
var calculateAnimationTimeout_1 = require("./styles/calculateAnimationTimeout");
Object.defineProperty(exports, "unstable_calculateAnimationTimeout", { enumerable: true, get: function () { return calculateAnimationTimeout_1.calculateAnimationTimeout; } });
var getStyles_1 = require("./styles/getStyles");
Object.defineProperty(exports, "unstable_getStyles", { enumerable: true, get: function () { return getStyles_1.getStyles; } });
__exportStar(require("./styles/types"), exports);
var useTelemetry_1 = require("./telemetry/useTelemetry");
Object.defineProperty(exports, "deprecated_getTelemetry", { enumerable: true, get: function () { return useTelemetry_1.getTelemetry; } });
Object.defineProperty(exports, "useTelemetry", { enumerable: true, get: function () { return useTelemetry_1.useTelemetry; } });
__exportStar(require("./telemetry/types"), exports);
var childrenExist_1 = require("./utils/childrenExist");
Object.defineProperty(exports, "childrenExist", { enumerable: true, get: function () { return childrenExist_1.childrenExist; } });
var getElementType_1 = require("./utils/getElementType");
Object.defineProperty(exports, "getElementType", { enumerable: true, get: function () { return getElementType_1.getElementType; } });
var getUnhandledProps_1 = require("./utils/getUnhandledProps");
Object.defineProperty(exports, "getUnhandledProps", { enumerable: true, get: function () { return getUnhandledProps_1.getUnhandledProps; } });
var mergeVariablesOverrides_1 = require("./utils/mergeVariablesOverrides");
Object.defineProperty(exports, "mergeVariablesOverrides", { enumerable: true, get: function () { return mergeVariablesOverrides_1.mergeVariablesOverrides; } });
__exportStar(require("./context"), exports);
var createContext_1 = require("./context-selector/createContext");
Object.defineProperty(exports, "createContext", { enumerable: true, get: function () { return createContext_1.createContext; } });
var useContextSelector_1 = require("./context-selector/useContextSelector");
Object.defineProperty(exports, "useContextSelector", { enumerable: true, get: function () { return useContextSelector_1.useContextSelector; } });
var useContextSelectors_1 = require("./context-selector/useContextSelectors");
Object.defineProperty(exports, "useContextSelectors", { enumerable: true, get: function () { return useContextSelectors_1.useContextSelectors; } });
__exportStar(require("./context-selector/types"), exports);
