// @fluentui/react-northstar/dist/dts/src/components/Avatar/AvatarStatus.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AvatarStatus = exports.avatarStatusClassName = void 0;
var React = require("react");
var react_bindings_1 = require("@fluentui/react-bindings");
var utils_1 = require("../../utils");
var customPropTypes = require("@fluentui/react-proptypes");
var PropTypes = require("prop-types");
var accessibility_1 = require("@fluentui/accessibility");
var AvatarStatusIcon_1 = require("./AvatarStatusIcon");
var Status_1 = require("../Status/Status");
// update in upcoming breaking change and enable a unit test
exports.avatarStatusClassName = Status_1.statusClassName;
/**
 * A AvatarStatus provides a status for the Avatar.
 */
exports.AvatarStatus = React.forwardRef(function (props, ref) {
    var context = react_bindings_1.useFluentContext();
    var _a = react_bindings_1.useTelemetry(exports.AvatarStatus.displayName, context.telemetry), setStart = _a.setStart, setEnd = _a.setEnd;
    setStart();
    var className = props.className, color = props.color, icon = props.icon, size = props.size, state = props.state, design = props.design, styles = props.styles, variables = props.variables;
    var classes = react_bindings_1.useStyles(exports.AvatarStatus.displayName, {
        className: exports.avatarStatusClassName,
        mapPropsToStyles: function () { return ({
            color: color,
            size: size,
            state: state,
        }); },
        mapPropsToInlineStyles: function () { return ({
            className: className,
            design: design,
            styles: styles,
            variables: variables,
        }); },
        rtl: context.rtl,
    }).classes;
    var getA11Props = react_bindings_1.useAccessibility(props.accessibility, {
        debugName: exports.AvatarStatus.displayName,
        rtl: context.rtl,
    });
    var ElementType = react_bindings_1.getElementType(props);
    var unhandledProps = react_bindings_1.useUnhandledProps(exports.AvatarStatus.handledProps, props);
    var iconElement = utils_1.createShorthand(AvatarStatusIcon_1.AvatarStatusIcon, icon, {
        defaultProps: function () { return ({
            state: state,
        }); },
    });
    var element = (React.createElement(ElementType, __assign({}, getA11Props('root', __assign({ className: classes.root, ref: ref }, unhandledProps))), iconElement));
    setEnd();
    return element;
});
exports.AvatarStatus.displayName = 'AvatarStatus';
exports.AvatarStatus.propTypes = __assign(__assign({}, utils_1.commonPropTypes.createCommon({
    children: false,
    content: false,
})), { color: PropTypes.string, icon: customPropTypes.shorthandAllowingChildren, size: customPropTypes.size, state: PropTypes.oneOf(['success', 'info', 'warning', 'error', 'unknown']) });
exports.AvatarStatus.handledProps = Object.keys(exports.AvatarStatus.propTypes);
exports.AvatarStatus.defaultProps = {
    accessibility: accessibility_1.statusBehavior,
    as: 'span',
    size: 'medium',
    state: 'unknown',
};
exports.AvatarStatus.create = utils_1.createShorthandFactory({ Component: exports.AvatarStatus, mappedProp: 'state' });
