// @fluentui/react-northstar/dist/dts/src/components/Avatar/AvatarImage.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AvatarImage = exports.avatarImageClassName = void 0;
var accessibility_1 = require("@fluentui/accessibility");
var react_bindings_1 = require("@fluentui/react-bindings");
var PropTypes = require("prop-types");
var React = require("react");
var utils_1 = require("../../utils");
var Image_1 = require("../Image/Image");
exports.avatarImageClassName = Image_1.imageClassName;
/**
 * An AvatarImage is a graphic representation used by Avatar.
 */
exports.AvatarImage = React.forwardRef(function (props, ref) {
    var context = react_bindings_1.useFluentContext();
    var _a = react_bindings_1.useTelemetry(exports.AvatarImage.displayName, context.telemetry), setStart = _a.setStart, setEnd = _a.setEnd;
    setStart();
    var accessibility = props.accessibility, alt = props.alt, ariaLabel = props["aria-label"], avatar = props.avatar, circular = props.circular, className = props.className, design = props.design, fluid = props.fluid, styles = props.styles, variables = props.variables, size = props.size;
    var getA11Props = react_bindings_1.useAccessibility(accessibility, {
        debugName: exports.AvatarImage.displayName,
        mapPropsToBehavior: function () { return ({
            alt: alt,
            'aria-label': ariaLabel,
        }); },
        rtl: context.rtl,
    });
    var classes = react_bindings_1.useStyles(exports.AvatarImage.displayName, {
        className: exports.avatarImageClassName,
        mapPropsToStyles: function () { return ({
            avatar: avatar,
            circular: circular,
            fluid: fluid,
            size: size,
        }); },
        mapPropsToInlineStyles: function () { return ({
            className: className,
            design: design,
            styles: styles,
            variables: variables,
        }); },
        rtl: context.rtl,
    }).classes;
    var ElementType = react_bindings_1.getElementType(props);
    var unhandledProps = react_bindings_1.useUnhandledProps(exports.AvatarImage.handledProps, props);
    var result = React.createElement(ElementType, __assign({}, getA11Props('root', __assign({ className: classes.root, ref: ref }, unhandledProps))));
    setEnd();
    return result;
});
exports.AvatarImage.displayName = 'AvatarImage';
exports.AvatarImage.defaultProps = {
    as: 'img',
    accessibility: accessibility_1.imageBehavior,
};
exports.AvatarImage.propTypes = __assign(__assign({}, utils_1.commonPropTypes.createCommon({
    children: false,
    content: false,
})), { avatar: PropTypes.bool, circular: PropTypes.bool, fluid: PropTypes.bool });
exports.AvatarImage.shorthandConfig = {
    mappedProp: 'src',
    allowsJSX: false,
};
exports.AvatarImage.handledProps = Object.keys(exports.AvatarImage.propTypes);
exports.AvatarImage.create = utils_1.createShorthandFactory({ Component: exports.AvatarImage, mappedProp: 'src', allowsJSX: false });
