// @fluentui/react-northstar/dist/dts/src/components/Avatar/Avatar.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Avatar = exports.avatarClassName = void 0;
var react_bindings_1 = require("@fluentui/react-bindings");
var customPropTypes = require("@fluentui/react-proptypes");
var PropTypes = require("prop-types");
var React = require("react");
var utils_1 = require("../../utils");
var AvatarStatus_1 = require("./AvatarStatus");
var AvatarImage_1 = require("./AvatarImage");
var AvatarIcon_1 = require("./AvatarIcon");
var AvatarLabel_1 = require("./AvatarLabel");
exports.avatarClassName = 'ui-avatar';
/**
 * An Avatar is a graphical representation of a user.
 */
exports.Avatar = React.forwardRef(function (props, ref) {
    var context = react_bindings_1.useFluentContext();
    var _a = react_bindings_1.useTelemetry(exports.Avatar.displayName, context.telemetry), setStart = _a.setStart, setEnd = _a.setEnd;
    setStart();
    var accessibility = props.accessibility, className = props.className, design = props.design, getInitials = props.getInitials, label = props.label, icon = props.icon, image = props.image, name = props.name, square = props.square, size = props.size, status = props.status, styles = props.styles, variables = props.variables;
    var getA11Props = react_bindings_1.useAccessibility(accessibility, {
        debugName: exports.Avatar.displayName,
        rtl: context.rtl,
    });
    var _b = react_bindings_1.useStyles(exports.Avatar.displayName, {
        className: exports.avatarClassName,
        mapPropsToStyles: function () { return ({ size: size, square: square }); },
        mapPropsToInlineStyles: function () { return ({
            className: className,
            design: design,
            styles: styles,
            variables: variables,
        }); },
    }), classes = _b.classes, resolvedStyles = _b.styles;
    var ElementType = react_bindings_1.getElementType(props);
    var unhandledProps = react_bindings_1.useUnhandledProps(exports.Avatar.handledProps, props);
    var imageElement = utils_1.createShorthand(AvatarImage_1.AvatarImage, image, {
        defaultProps: function () {
            return getA11Props('image', {
                fluid: true,
                avatar: !square,
                title: name,
                size: size,
                // remove in upcoming breaking change
                styles: resolvedStyles.image,
            });
        },
    });
    var iconElement = utils_1.createShorthand(AvatarIcon_1.AvatarIcon, icon, {
        defaultProps: function () {
            return getA11Props('icon', {
                title: name,
                styles: resolvedStyles.icon,
                size: size,
                square: square,
            });
        },
    });
    var labelElement = utils_1.createShorthand(AvatarLabel_1.AvatarLabel, label || {}, {
        defaultProps: function () {
            return getA11Props('label', {
                content: getInitials(name),
                circular: !square,
                title: name,
                size: size,
                square: square,
                styles: resolvedStyles.label,
            });
        },
    });
    var hasGlyph = !!image || !!icon;
    var result = (React.createElement(ElementType, __assign({}, getA11Props('root', __assign({ className: classes.root, ref: ref }, unhandledProps))),
        hasGlyph && (imageElement || iconElement),
        !hasGlyph && labelElement,
        utils_1.createShorthand(AvatarStatus_1.AvatarStatus, status, {
            defaultProps: function () {
                return getA11Props('status', {
                    size: size,
                    // remove in upcoming breaking change
                    styles: resolvedStyles.status,
                });
            },
            overrideProps: function (predefinedProps) { return ({
                variables: react_bindings_1.mergeVariablesOverrides(variables, predefinedProps.variables),
            }); },
        })));
    setEnd();
    return result;
});
exports.Avatar.displayName = 'Avatar';
exports.Avatar.defaultProps = {
    size: 'medium',
    getInitials: function (name) {
        if (!name) {
            return '';
        }
        var reducedName = name
            .replace(/\s+/g, ' ')
            .replace(/\s*\(.*?\)\s*/g, ' ')
            .replace(/\s*{.*?}\s*/g, ' ')
            .replace(/\s*\[.*?]\s*/g, ' ');
        var initials = reducedName
            .split(' ')
            .filter(function (item) { return item !== ''; })
            .map(function (item) { return item.charAt(0); })
            .reduce(function (accumulator, currentValue) { return accumulator + currentValue; }, '');
        if (initials.length > 2) {
            return initials.charAt(0) + initials.charAt(initials.length - 1);
        }
        return initials;
    },
};
exports.Avatar.propTypes = __assign(__assign({}, utils_1.commonPropTypes.createCommon({
    children: false,
    content: false,
})), { name: PropTypes.string, icon: customPropTypes.shorthandAllowingChildren, image: customPropTypes.itemShorthandWithoutJSX, label: customPropTypes.itemShorthand, square: PropTypes.bool, size: customPropTypes.size, status: customPropTypes.itemShorthand, getInitials: PropTypes.func });
exports.Avatar.handledProps = Object.keys(exports.Avatar.propTypes);
exports.Avatar.create = utils_1.createShorthandFactory({ Component: exports.Avatar, mappedProp: 'name' });
