// @fluentui/react-bindings/dist/commonjs/index.js
"use strict";

exports.__esModule = true;
var _exportNames = {
  unstable_getAccessibility: true,
  AutoFocusZone: true,
  FocusTrapZone: true,
  FocusZone: true,
  useAccessibility: true,
  useAutoControlled: true,
  useCallbackRef: true,
  unstable_useDispatchEffect: true,
  useDeepMemo: true,
  useEventCallback: true,
  useFirstMount: true,
  useIsomorphicLayoutEffect: true,
  useMergedRefs: true,
  useStateManager: true,
  useTriggerElement: true,
  useUnhandledProps: true,
  usePrevious: true,
  useOnIFrameFocus: true,
  RendererContext: true,
  unstable_createAnimationStyles: true,
  unstable_calculateAnimationTimeout: true,
  unstable_getStyles: true,
  deprecated_getTelemetry: true,
  useTelemetry: true,
  childrenExist: true,
  getElementType: true,
  getUnhandledProps: true,
  mergeVariablesOverrides: true,
  createContext: true,
  useContextSelector: true,
  useContextSelectors: true
};
exports.useContextSelectors = exports.useContextSelector = exports.createContext = exports.mergeVariablesOverrides = exports.getUnhandledProps = exports.getElementType = exports.childrenExist = exports.useTelemetry = exports.deprecated_getTelemetry = exports.unstable_getStyles = exports.unstable_calculateAnimationTimeout = exports.unstable_createAnimationStyles = exports.RendererContext = exports.useOnIFrameFocus = exports.usePrevious = exports.useUnhandledProps = exports.useTriggerElement = exports.useStateManager = exports.useMergedRefs = exports.useIsomorphicLayoutEffect = exports.useFirstMount = exports.useEventCallback = exports.useDeepMemo = exports.unstable_useDispatchEffect = exports.useCallbackRef = exports.useAutoControlled = exports.useAccessibility = exports.FocusZone = exports.FocusTrapZone = exports.AutoFocusZone = exports.unstable_getAccessibility = void 0;

var _getAccessibility = require("./accessibility/getAccessibility");

exports.unstable_getAccessibility = _getAccessibility.getAccessibility;

var _types = require("./accessibility/types");

Object.keys(_types).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _types[key]) return;
  exports[key] = _types[key];
});

var _compose = require("./compose");

Object.keys(_compose).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _compose[key]) return;
  exports[key] = _compose[key];
});

var _AutoFocusZone = require("./FocusZone/AutoFocusZone");

exports.AutoFocusZone = _AutoFocusZone.AutoFocusZone;

var _AutoFocusZone2 = require("./FocusZone/AutoFocusZone.types");

Object.keys(_AutoFocusZone2).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _AutoFocusZone2[key]) return;
  exports[key] = _AutoFocusZone2[key];
});

var _FocusTrapZone = require("./FocusZone/FocusTrapZone");

exports.FocusTrapZone = _FocusTrapZone.FocusTrapZone;

var _FocusTrapZone2 = require("./FocusZone/FocusTrapZone.types");

Object.keys(_FocusTrapZone2).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _FocusTrapZone2[key]) return;
  exports[key] = _FocusTrapZone2[key];
});

var _FocusZone = require("./FocusZone/FocusZone");

exports.FocusZone = _FocusZone.FocusZone;

var _FocusZone2 = require("./FocusZone/FocusZone.types");

Object.keys(_FocusZone2).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _FocusZone2[key]) return;
  exports[key] = _FocusZone2[key];
});

var _focusUtilities = require("./FocusZone/focusUtilities");

Object.keys(_focusUtilities).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _focusUtilities[key]) return;
  exports[key] = _focusUtilities[key];
});

var _useAccessibility = require("./hooks/useAccessibility");

exports.useAccessibility = _useAccessibility.useAccessibility;

var _useAutoControlled = require("./hooks/useAutoControlled");

exports.useAutoControlled = _useAutoControlled.useAutoControlled;

var _useCallbackRef = require("./hooks/useCallbackRef");

exports.useCallbackRef = _useCallbackRef.useCallbackRef;

var _useDispatchEffect = require("./hooks/useDispatchEffect");

exports.unstable_useDispatchEffect = _useDispatchEffect.useDispatchEffect;

var _useDeepMemo = require("./hooks/useDeepMemo");

exports.useDeepMemo = _useDeepMemo.useDeepMemo;

var _useEventCallback = require("./hooks/useEventCallback");

exports.useEventCallback = _useEventCallback.useEventCallback;

var _useFirstMount = require("./hooks/useFirstMount");

exports.useFirstMount = _useFirstMount.useFirstMount;

var _useIsomorphicLayoutEffect = require("./hooks/useIsomorphicLayoutEffect");

exports.useIsomorphicLayoutEffect = _useIsomorphicLayoutEffect.useIsomorphicLayoutEffect;

var _useMergedRefs = require("./hooks/useMergedRefs");

exports.useMergedRefs = _useMergedRefs.useMergedRefs;

var _useStateManager = require("./hooks/useStateManager");

exports.useStateManager = _useStateManager.useStateManager;

var _useStyles = require("./hooks/useStyles");

Object.keys(_useStyles).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _useStyles[key]) return;
  exports[key] = _useStyles[key];
});

var _useTriggerElement = require("./hooks/useTriggerElement");

exports.useTriggerElement = _useTriggerElement.useTriggerElement;

var _useUnhandledProps = require("./hooks/useUnhandledProps");

exports.useUnhandledProps = _useUnhandledProps.useUnhandledProps;

var _usePrevious = require("./hooks/usePrevious");

exports.usePrevious = _usePrevious.usePrevious;

var _useOnIFrameFocus = require("./hooks/useOnIFrameFocus");

exports.useOnIFrameFocus = _useOnIFrameFocus.useOnIFrameFocus;

var _RendererContext = require("./renderer/RendererContext");

exports.RendererContext = _RendererContext.RendererContext;

var _createAnimationStyles = require("./styles/createAnimationStyles");

exports.unstable_createAnimationStyles = _createAnimationStyles.createAnimationStyles;

var _calculateAnimationTimeout = require("./styles/calculateAnimationTimeout");

exports.unstable_calculateAnimationTimeout = _calculateAnimationTimeout.calculateAnimationTimeout;

var _getStyles = require("./styles/getStyles");

exports.unstable_getStyles = _getStyles.getStyles;

var _types2 = require("./styles/types");

Object.keys(_types2).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _types2[key]) return;
  exports[key] = _types2[key];
});

var _useTelemetry = require("./telemetry/useTelemetry");

exports.deprecated_getTelemetry = _useTelemetry.getTelemetry;
exports.useTelemetry = _useTelemetry.useTelemetry;

var _types3 = require("./telemetry/types");

Object.keys(_types3).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _types3[key]) return;
  exports[key] = _types3[key];
});

var _childrenExist = require("./utils/childrenExist");

exports.childrenExist = _childrenExist.childrenExist;

var _getElementType = require("./utils/getElementType");

exports.getElementType = _getElementType.getElementType;

var _getUnhandledProps = require("./utils/getUnhandledProps");

exports.getUnhandledProps = _getUnhandledProps.getUnhandledProps;

var _mergeVariablesOverrides = require("./utils/mergeVariablesOverrides");

exports.mergeVariablesOverrides = _mergeVariablesOverrides.mergeVariablesOverrides;

var _context = require("./context");

Object.keys(_context).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _context[key]) return;
  exports[key] = _context[key];
});

var _createContext = require("./context-selector/createContext");

exports.createContext = _createContext.createContext;

var _useContextSelector = require("./context-selector/useContextSelector");

exports.useContextSelector = _useContextSelector.useContextSelector;

var _useContextSelectors = require("./context-selector/useContextSelectors");

exports.useContextSelectors = _useContextSelectors.useContextSelectors;

var _types4 = require("./context-selector/types");

Object.keys(_types4).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _types4[key]) return;
  exports[key] = _types4[key];
});
