// @fluentui/accessibility/dist/dts/src/behaviors/Tab/tabListBehavior.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.tabListBehavior = void 0;
var types_1 = require("../../focusZone/types");
var tabBehavior_1 = require("./tabBehavior");
/**
 * @description
 * Implements ARIA Tabs design pattern.
 * Child item components need to have tabBehavior assigned.
 * @specification
 * Adds role 'tablist' to 'root' slot.
 * Provides arrow key navigation in bidirectionalDomOrder direction.
 * When component's container element receives focus, focus will be set to the default focusable child element of the component.
 */
var tabListBehavior = function () { return ({
    attributes: {
        root: {
            role: 'tablist',
        },
    },
    focusZone: {
        props: {
            shouldFocusInnerElementWhenReceivedFocus: true,
            direction: types_1.FocusZoneDirection.bidirectionalDomOrder,
        },
    },
    childBehaviors: {
        item: tabBehavior_1.tabBehavior,
    },
}); };
exports.tabListBehavior = tabListBehavior;
