// botbuilder-dialogs/lib/dialogsComponentRegistration.js
"use strict";
/**
 * @module botbuilder-dialogs
 */
/**
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.DialogsComponentRegistration = void 0;
const botbuilder_core_1 = require("botbuilder-core");
const botbuilder_dialogs_adaptive_runtime_core_1 = require("botbuilder-dialogs-adaptive-runtime-core");
const dialogsBotComponent_1 = require("./dialogsBotComponent");
/**
 * Makes dialogs component available to the system registering functionality.
 */
class DialogsComponentRegistration extends botbuilder_core_1.ComponentRegistration {
    constructor() {
        super();
        this.services = new botbuilder_dialogs_adaptive_runtime_core_1.ServiceCollection({
            memoryScopes: [],
            pathResolvers: [],
        });
        new dialogsBotComponent_1.DialogsBotComponent().configureServices(this.services, botbuilder_dialogs_adaptive_runtime_core_1.noOpConfiguration);
    }
    /**
     * Gets the dialogs memory scopes.
     *
     * @returns {MemoryScope[]} A list of [MemoryScope](xref:botbuilder-dialogs.MemoryScope).
     */
    getMemoryScopes() {
        return this.services.mustMakeInstance('memoryScopes');
    }
    /**
     * Gets the dialogs path resolvers.
     *
     * @returns {PathResolver[]} A list of [PathResolver](xref:botbuilder-dialogs.PathResolver).
     */
    getPathResolvers() {
        return this.services.mustMakeInstance('pathResolvers');
    }
}
exports.DialogsComponentRegistration = DialogsComponentRegistration;
