// botbuilder-dialogs/lib/i18n.js
"use strict";
/**
 * @module botbuilder-dialogs
 */
/**
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Spanish = exports.Portuguese = exports.NumberingSystem = exports.LikelySubtags = exports.Japanese = exports.Dutch = exports.German = exports.French = exports.English = exports.Chinese = void 0;
/* eslint-disable @typescript-eslint/no-var-requires */
const Chinese = require('../vendor/cldr-data/main/zh/numbers.json');
exports.Chinese = Chinese;
const English = require('../vendor/cldr-data/main/en/numbers.json');
exports.English = English;
const French = require('../vendor/cldr-data/main/fr/numbers.json');
exports.French = French;
const German = require('../vendor/cldr-data/main/de/numbers.json');
exports.German = German;
const Dutch = require('../vendor/cldr-data/main/nl/numbers.json');
exports.Dutch = Dutch;
const Japanese = require('../vendor/cldr-data/main/ja/numbers.json');
exports.Japanese = Japanese;
const LikelySubtags = require('../vendor/cldr-data/supplemental/likelySubtags.json');
exports.LikelySubtags = LikelySubtags;
const NumberingSystem = require('../vendor/cldr-data/supplemental/numberingSystems.json');
exports.NumberingSystem = NumberingSystem;
const Portuguese = require('../vendor/cldr-data/main/pt/numbers.json');
exports.Portuguese = Portuguese;
const Spanish = require('../vendor/cldr-data/main/es/numbers.json');
exports.Spanish = Spanish;
