// @fluentui/react-northstar/dist/commonjs/themes/teams/components/Tooltip/tooltipContentStyles.js
"use strict";

exports.__esModule = true;
exports.tooltipContentStyles = void 0;

var _getPointerStyles = require("../../getPointerStyles");

var _pointerSvgUrl = require("../../pointerSvgUrl");

var tooltipContentStyles = {
  root: function root(_ref) {
    var p = _ref.props,
        v = _ref.variables;
    return Object.assign({
      display: 'none',
      position: 'absolute',
      maxWidth: v.maxWidth,
      zIndex: v.zIndex
    }, p.pointing && Object.assign({
      pointerEvents: 'all'
    }, (0, _getPointerStyles.getContainerStyles)({
      placement: p.basePlacement,
      padding: v.pointerMargin
    })), p.open && {
      display: 'block'
    });
  },
  pointer: function pointer(_ref2) {
    var p = _ref2.props,
        v = _ref2.variables,
        rtl = _ref2.rtl;
    return Object.assign({
      display: 'block',
      position: 'absolute',
      width: v.pointerWidth,
      height: v.pointerHeight
    }, (0, _getPointerStyles.getPointerStyles)({
      backgroundColor: p.subtle ? v.subtleBackgroundColor : v.backgroundColor,
      borderSize: v.borderSize,
      borderColor: p.subtle ? v.subtleBorderColor : 'transparent',
      gap: v.pointerGap,
      padding: v.pointerMargin,
      height: v.pointerHeight,
      width: v.pointerWidth,
      placement: p.basePlacement,
      rtl: rtl,
      svg: (0, _pointerSvgUrl.pointerSvgUrl)(p.subtle ? v.subtleBackgroundColor : v.backgroundColor)
    }));
  },
  content: function content(_ref3) {
    var p = _ref3.props,
        v = _ref3.variables;
    return Object.assign({
      display: 'block',
      padding: v.padding,
      textAlign: 'left',
      color: v.color,
      background: v.backgroundColor,
      borderRadius: v.borderRadius,
      boxShadow: v.boxShadow
    }, p.subtle && {
      background: v.subtleBackgroundColor,
      color: v.subtleForegroundColor,
      borderStyle: 'solid',
      borderWidth: v.borderSize,
      borderColor: v.subtleBorderColor
    }, p.pointing && {
      pointerEvents: 'all'
    });
  }
};
exports.tooltipContentStyles = tooltipContentStyles;
