// @fluentui/react-northstar/dist/commonjs/themes/teams/components/Tooltip/tooltipContentVariables.js
"use strict";

exports.__esModule = true;
exports.tooltipContentVariables = void 0;

var _utils = require("../../../../utils");

var tooltipContentVariables = function tooltipContentVariables(siteVars) {
  return {
    boxShadow: siteVars.shadowLevel2,
    borderRadius: siteVars.borderRadiusMedium,
    borderSize: '1px',
    padding: (0, _utils.pxToRem)(5) + " " + (0, _utils.pxToRem)(12) + " " + (0, _utils.pxToRem)(7) + " " + (0, _utils.pxToRem)(12),
    maxWidth: (0, _utils.pxToRem)(246),
    pointerMargin: (0, _utils.pxToRem)(6),
    pointerGap: (0, _utils.pxToRem)(5),
    pointerWidth: (0, _utils.pxToRem)(16),
    pointerHeight: (0, _utils.pxToRem)(6),
    color: siteVars.colorScheme.default.foreground3,
    backgroundColor: siteVars.colors.grey[500],
    subtleBackgroundColor: siteVars.colorScheme.default.background,
    subtleForegroundColor: siteVars.colorScheme.default.foreground,
    subtleBorderColor: siteVars.colorScheme.onyx.border1,
    zIndex: siteVars.zIndexes.overlayPriority
  };
};

exports.tooltipContentVariables = tooltipContentVariables;
