// botframework-connector/lib/tokenApi/tokenApiClient.js
"use strict";
/**
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TokenApiMappers = exports.TokenApiModels = exports.TokenApiClientContext = exports.TokenApiClient = void 0;
const Models = require("./models");
exports.TokenApiModels = Models;
const Mappers = require("./models/mappers");
exports.TokenApiMappers = Mappers;
const operations = require("./operations");
const tokenApiClientContext_1 = require("./tokenApiClientContext");
Object.defineProperty(exports, "TokenApiClientContext", { enumerable: true, get: function () { return tokenApiClientContext_1.TokenApiClientContext; } });
class TokenApiClient extends tokenApiClientContext_1.TokenApiClientContext {
    /**
     * Initializes a new instance of the TokenApiClient class.
     * @param credentials Subscription credentials which uniquely identify client subscription.
     * @param [options] The parameter options
     */
    constructor(credentials, options) {
        super(credentials, options);
        this.botSignIn = new operations.BotSignIn(this);
        this.userToken = new operations.UserToken(this);
    }
}
exports.TokenApiClient = TokenApiClient;
__exportStar(require("./operations"), exports);
