// @fluentui/accessibility/dist/commonjs/behaviors/Accordion/accordionBehavior.js
"use strict";

exports.__esModule = true;
exports.accordionBehavior = void 0;

var _keyboardKey = require("../../keyboard-key");

/**
 * @specification
 * Adds attribute 'role=presentation' to 'root' slot.
 * Triggers 'moveNext' action with 'ArrowDown' on 'root'.
 * Triggers 'movePrevious' action with 'ArrowUp' on 'root'.
 * Triggers 'moveFirst' action with 'Home' on 'root'.
 * Triggers 'moveLast' action with 'End' on 'root'.
 */
var accordionBehavior = function accordionBehavior() {
  return {
    attributes: {
      root: {
        role: 'presentation'
      }
    },
    keyActions: {
      root: {
        moveNext: {
          keyCombinations: [{
            keyCode: _keyboardKey.keyboardKey.ArrowDown
          }]
        },
        movePrevious: {
          keyCombinations: [{
            keyCode: _keyboardKey.keyboardKey.ArrowUp
          }]
        },
        moveFirst: {
          keyCombinations: [{
            keyCode: _keyboardKey.keyboardKey.Home
          }]
        },
        moveLast: {
          keyCombinations: [{
            keyCode: _keyboardKey.keyboardKey.End
          }]
        }
      }
    }
  };
};

exports.accordionBehavior = accordionBehavior;
