// @fluentui/accessibility/dist/commonjs/behaviors/index.js
"use strict";

exports.__esModule = true;
var _exportNames = {
  alertWarningBehavior: true,
  alertBaseBehavior: true,
  gridHorizontalBehavior: true,
  tooltipAsLabelBehavior: true,
  treeAsListBehavior: true,
  treeItemAsListItemBehavior: true,
  treeTitleAsListItemTitleBehavior: true,
  treeAsListboxBehavior: true,
  treeItemAsOptionBehavior: true,
  treeTitleAsOptionBehavior: true,
  tableRowBehavior: true,
  tableHeaderCellBehavior: true,
  gridNestedBehavior: true,
  gridHeaderRowBehavior: true,
  gridHeaderCellBehavior: true,
  gridRowNestedBehavior: true,
  gridCellBehavior: true,
  gridCellMultipleFocusableBehavior: true,
  gridCellWithFocusableElementBehavior: true,
  cardFocusableBehavior: true,
  cardChildrenFocusableBehavior: true,
  cardSelectableBehavior: true,
  hiddenComponentBehavior: true
};
exports.hiddenComponentBehavior = exports.cardSelectableBehavior = exports.cardChildrenFocusableBehavior = exports.cardFocusableBehavior = exports.gridCellWithFocusableElementBehavior = exports.gridCellMultipleFocusableBehavior = exports.gridCellBehavior = exports.gridRowNestedBehavior = exports.gridHeaderCellBehavior = exports.gridHeaderRowBehavior = exports.gridNestedBehavior = exports.tableHeaderCellBehavior = exports.tableRowBehavior = exports.treeTitleAsOptionBehavior = exports.treeItemAsOptionBehavior = exports.treeAsListboxBehavior = exports.treeTitleAsListItemTitleBehavior = exports.treeItemAsListItemBehavior = exports.treeAsListBehavior = exports.tooltipAsLabelBehavior = exports.gridHorizontalBehavior = exports.alertBaseBehavior = exports.alertWarningBehavior = void 0;

var _alertBehavior = require("./Alert/alertBehavior");

Object.keys(_alertBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _alertBehavior[key]) return;
  exports[key] = _alertBehavior[key];
});

var _alertWarningBehavior = require("./Alert/alertWarningBehavior");

exports.alertWarningBehavior = _alertWarningBehavior.alertWarningBehavior;

var _alertBaseBehavior = require("./Alert/alertBaseBehavior");

exports.alertBaseBehavior = _alertBaseBehavior.alertBaseBehavior;

var _attachmentBehavior = require("./Attachment/attachmentBehavior");

Object.keys(_attachmentBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _attachmentBehavior[key]) return;
  exports[key] = _attachmentBehavior[key];
});

var _buttonBehavior = require("./Button/buttonBehavior");

Object.keys(_buttonBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _buttonBehavior[key]) return;
  exports[key] = _buttonBehavior[key];
});

var _buttonGroupBehavior = require("./Button/buttonGroupBehavior");

Object.keys(_buttonGroupBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _buttonGroupBehavior[key]) return;
  exports[key] = _buttonGroupBehavior[key];
});

var _toggleButtonBehavior = require("./Button/toggleButtonBehavior");

Object.keys(_toggleButtonBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _toggleButtonBehavior[key]) return;
  exports[key] = _toggleButtonBehavior[key];
});

var _breadcrumbBehavior = require("./Breadcrumb/breadcrumbBehavior");

Object.keys(_breadcrumbBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _breadcrumbBehavior[key]) return;
  exports[key] = _breadcrumbBehavior[key];
});

var _breadcrumbItemBehavior = require("./Breadcrumb/breadcrumbItemBehavior");

Object.keys(_breadcrumbItemBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _breadcrumbItemBehavior[key]) return;
  exports[key] = _breadcrumbItemBehavior[key];
});

var _breadcrumbDividerBehavior = require("./Breadcrumb/breadcrumbDividerBehavior");

Object.keys(_breadcrumbDividerBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _breadcrumbDividerBehavior[key]) return;
  exports[key] = _breadcrumbDividerBehavior[key];
});

var _dropdownSelectedItemBehavior = require("./Dropdown/dropdownSelectedItemBehavior");

Object.keys(_dropdownSelectedItemBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _dropdownSelectedItemBehavior[key]) return;
  exports[key] = _dropdownSelectedItemBehavior[key];
});

var _imageBehavior = require("./Image/imageBehavior");

Object.keys(_imageBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _imageBehavior[key]) return;
  exports[key] = _imageBehavior[key];
});

var _menuBehavior = require("./Menu/menuBehavior");

Object.keys(_menuBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _menuBehavior[key]) return;
  exports[key] = _menuBehavior[key];
});

var _menuItemBehavior = require("./Menu/menuItemBehavior");

Object.keys(_menuItemBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _menuItemBehavior[key]) return;
  exports[key] = _menuItemBehavior[key];
});

var _menuDividerBehavior = require("./Menu/menuDividerBehavior");

Object.keys(_menuDividerBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _menuDividerBehavior[key]) return;
  exports[key] = _menuDividerBehavior[key];
});

var _submenuBehavior = require("./Menu/submenuBehavior");

Object.keys(_submenuBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _submenuBehavior[key]) return;
  exports[key] = _submenuBehavior[key];
});

var _basicListBehavior = require("./List/basicListBehavior");

Object.keys(_basicListBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _basicListBehavior[key]) return;
  exports[key] = _basicListBehavior[key];
});

var _basicListItemBehavior = require("./List/basicListItemBehavior");

Object.keys(_basicListItemBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _basicListItemBehavior[key]) return;
  exports[key] = _basicListItemBehavior[key];
});

var _listBehavior = require("./List/listBehavior");

Object.keys(_listBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _listBehavior[key]) return;
  exports[key] = _listBehavior[key];
});

var _listItemBehavior = require("./List/listItemBehavior");

Object.keys(_listItemBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _listItemBehavior[key]) return;
  exports[key] = _listItemBehavior[key];
});

var _navigableListBehavior = require("./List/navigableListBehavior");

Object.keys(_navigableListBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _navigableListBehavior[key]) return;
  exports[key] = _navigableListBehavior[key];
});

var _navigableListItemBehavior = require("./List/navigableListItemBehavior");

Object.keys(_navigableListItemBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _navigableListItemBehavior[key]) return;
  exports[key] = _navigableListItemBehavior[key];
});

var _selectableListBehavior = require("./List/selectableListBehavior");

Object.keys(_selectableListBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _selectableListBehavior[key]) return;
  exports[key] = _selectableListBehavior[key];
});

var _selectableListItemBehavior = require("./List/selectableListItemBehavior");

Object.keys(_selectableListItemBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _selectableListItemBehavior[key]) return;
  exports[key] = _selectableListItemBehavior[key];
});

var _loaderBehavior = require("./Loader/loaderBehavior");

Object.keys(_loaderBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _loaderBehavior[key]) return;
  exports[key] = _loaderBehavior[key];
});

var _inputBehavior = require("./Input/inputBehavior");

Object.keys(_inputBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _inputBehavior[key]) return;
  exports[key] = _inputBehavior[key];
});

var _iconBehavior = require("./Icon/iconBehavior");

Object.keys(_iconBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _iconBehavior[key]) return;
  exports[key] = _iconBehavior[key];
});

var _indicatorBehavior = require("./Box/indicatorBehavior");

Object.keys(_indicatorBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _indicatorBehavior[key]) return;
  exports[key] = _indicatorBehavior[key];
});

var _tabBehavior = require("./Tab/tabBehavior");

Object.keys(_tabBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _tabBehavior[key]) return;
  exports[key] = _tabBehavior[key];
});

var _tabListBehavior = require("./Tab/tabListBehavior");

Object.keys(_tabListBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _tabListBehavior[key]) return;
  exports[key] = _tabListBehavior[key];
});

var _menuAsToolbarBehavior = require("./Toolbar/menuAsToolbarBehavior");

Object.keys(_menuAsToolbarBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _menuAsToolbarBehavior[key]) return;
  exports[key] = _menuAsToolbarBehavior[key];
});

var _toolbarMenuBehavior = require("./Toolbar/toolbarMenuBehavior");

Object.keys(_toolbarMenuBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _toolbarMenuBehavior[key]) return;
  exports[key] = _toolbarMenuBehavior[key];
});

var _toolbarMenuItemBehavior = require("./Toolbar/toolbarMenuItemBehavior");

Object.keys(_toolbarMenuItemBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _toolbarMenuItemBehavior[key]) return;
  exports[key] = _toolbarMenuItemBehavior[key];
});

var _toolbarMenuItemRadioBehavior = require("./Toolbar/toolbarMenuItemRadioBehavior");

Object.keys(_toolbarMenuItemRadioBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _toolbarMenuItemRadioBehavior[key]) return;
  exports[key] = _toolbarMenuItemRadioBehavior[key];
});

var _menuItemAsToolbarButtonBehavior = require("./Toolbar/menuItemAsToolbarButtonBehavior");

Object.keys(_menuItemAsToolbarButtonBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _menuItemAsToolbarButtonBehavior[key]) return;
  exports[key] = _menuItemAsToolbarButtonBehavior[key];
});

var _toolbarBehavior = require("./Toolbar/toolbarBehavior");

Object.keys(_toolbarBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _toolbarBehavior[key]) return;
  exports[key] = _toolbarBehavior[key];
});

var _toolbarItemBehavior = require("./Toolbar/toolbarItemBehavior");

Object.keys(_toolbarItemBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _toolbarItemBehavior[key]) return;
  exports[key] = _toolbarItemBehavior[key];
});

var _toolbarMenuItemCheckboxBehavior = require("./Toolbar/toolbarMenuItemCheckboxBehavior");

Object.keys(_toolbarMenuItemCheckboxBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _toolbarMenuItemCheckboxBehavior[key]) return;
  exports[key] = _toolbarMenuItemCheckboxBehavior[key];
});

var _toolbarMenuDividerBehavior = require("./Toolbar/toolbarMenuDividerBehavior");

Object.keys(_toolbarMenuDividerBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _toolbarMenuDividerBehavior[key]) return;
  exports[key] = _toolbarMenuDividerBehavior[key];
});

var _toolbarMenuRadioGroupBehavior = require("./Toolbar/toolbarMenuRadioGroupBehavior");

Object.keys(_toolbarMenuRadioGroupBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _toolbarMenuRadioGroupBehavior[key]) return;
  exports[key] = _toolbarMenuRadioGroupBehavior[key];
});

var _toolbarMenuRadioGroupWrapperBehavior = require("./Toolbar/toolbarMenuRadioGroupWrapperBehavior");

Object.keys(_toolbarMenuRadioGroupWrapperBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _toolbarMenuRadioGroupWrapperBehavior[key]) return;
  exports[key] = _toolbarMenuRadioGroupWrapperBehavior[key];
});

var _toolbarRadioGroupBehavior = require("./Toolbar/toolbarRadioGroupBehavior");

Object.keys(_toolbarRadioGroupBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _toolbarRadioGroupBehavior[key]) return;
  exports[key] = _toolbarRadioGroupBehavior[key];
});

var _toolbarRadioGroupItemBehavior = require("./Toolbar/toolbarRadioGroupItemBehavior");

Object.keys(_toolbarRadioGroupItemBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _toolbarRadioGroupItemBehavior[key]) return;
  exports[key] = _toolbarRadioGroupItemBehavior[key];
});

var _radioGroupBehavior = require("./Radio/radioGroupBehavior");

Object.keys(_radioGroupBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _radioGroupBehavior[key]) return;
  exports[key] = _radioGroupBehavior[key];
});

var _radioGroupItemBehavior = require("./Radio/radioGroupItemBehavior");

Object.keys(_radioGroupItemBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _radioGroupItemBehavior[key]) return;
  exports[key] = _radioGroupItemBehavior[key];
});

var _popupBehavior = require("./Popup/popupBehavior");

Object.keys(_popupBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _popupBehavior[key]) return;
  exports[key] = _popupBehavior[key];
});

var _chatBehavior = require("./Chat/chatBehavior");

Object.keys(_chatBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _chatBehavior[key]) return;
  exports[key] = _chatBehavior[key];
});

var _chatMessageBehavior = require("./Chat/chatMessageBehavior");

Object.keys(_chatMessageBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _chatMessageBehavior[key]) return;
  exports[key] = _chatMessageBehavior[key];
});

var _formFieldBehavior = require("./Form/formFieldBehavior");

Object.keys(_formFieldBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _formFieldBehavior[key]) return;
  exports[key] = _formFieldBehavior[key];
});

var _gridBehavior = require("./Grid/gridBehavior");

Object.keys(_gridBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _gridBehavior[key]) return;
  exports[key] = _gridBehavior[key];
});

var _gridHorizontalBehavior = require("./Grid/gridHorizontalBehavior");

exports.gridHorizontalBehavior = _gridHorizontalBehavior.gridHorizontalBehavior;

var _dialogBehavior = require("./Dialog/dialogBehavior");

Object.keys(_dialogBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _dialogBehavior[key]) return;
  exports[key] = _dialogBehavior[key];
});

var _statusBehavior = require("./Status/statusBehavior");

Object.keys(_statusBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _statusBehavior[key]) return;
  exports[key] = _statusBehavior[key];
});

var _embedBehavior = require("./Embed/embedBehavior");

Object.keys(_embedBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _embedBehavior[key]) return;
  exports[key] = _embedBehavior[key];
});

var _accordionBehavior = require("./Accordion/accordionBehavior");

Object.keys(_accordionBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _accordionBehavior[key]) return;
  exports[key] = _accordionBehavior[key];
});

var _accordionTitleBehavior = require("./Accordion/accordionTitleBehavior");

Object.keys(_accordionTitleBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _accordionTitleBehavior[key]) return;
  exports[key] = _accordionTitleBehavior[key];
});

var _accordionContentBehavior = require("./Accordion/accordionContentBehavior");

Object.keys(_accordionContentBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _accordionContentBehavior[key]) return;
  exports[key] = _accordionContentBehavior[key];
});

var _checkboxBehavior = require("./Checkbox/checkboxBehavior");

Object.keys(_checkboxBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _checkboxBehavior[key]) return;
  exports[key] = _checkboxBehavior[key];
});

var _tooltipAsDescriptionBehavior = require("./Tooltip/tooltipAsDescriptionBehavior");

Object.keys(_tooltipAsDescriptionBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _tooltipAsDescriptionBehavior[key]) return;
  exports[key] = _tooltipAsDescriptionBehavior[key];
});

var _tooltipAsLabelBehavior = require("./Tooltip/tooltipAsLabelBehavior");

exports.tooltipAsLabelBehavior = _tooltipAsLabelBehavior.tooltipAsLabelBehavior;

var _pillGroupBehavior = require("./Pill/pillGroupBehavior");

Object.keys(_pillGroupBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _pillGroupBehavior[key]) return;
  exports[key] = _pillGroupBehavior[key];
});

var _pillBehavior = require("./Pill/pillBehavior");

Object.keys(_pillBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _pillBehavior[key]) return;
  exports[key] = _pillBehavior[key];
});

var _pillIconBehavior = require("./Pill/pillIconBehavior");

Object.keys(_pillIconBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _pillIconBehavior[key]) return;
  exports[key] = _pillIconBehavior[key];
});

var _pillImageBehavior = require("./Pill/pillImageBehavior");

Object.keys(_pillImageBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _pillImageBehavior[key]) return;
  exports[key] = _pillImageBehavior[key];
});

var _pillActionBehavior = require("./Pill/pillActionBehavior");

Object.keys(_pillActionBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _pillActionBehavior[key]) return;
  exports[key] = _pillActionBehavior[key];
});

var _sliderBehavior = require("./Slider/sliderBehavior");

Object.keys(_sliderBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _sliderBehavior[key]) return;
  exports[key] = _sliderBehavior[key];
});

var _menuButtonBehavior = require("./MenuButton/menuButtonBehavior");

Object.keys(_menuButtonBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _menuButtonBehavior[key]) return;
  exports[key] = _menuButtonBehavior[key];
});

var _splitButtonBehavior = require("./SplitButton/splitButtonBehavior");

Object.keys(_splitButtonBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _splitButtonBehavior[key]) return;
  exports[key] = _splitButtonBehavior[key];
});

var _treeBehavior = require("./Tree/treeBehavior");

Object.keys(_treeBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _treeBehavior[key]) return;
  exports[key] = _treeBehavior[key];
});

var _treeItemBehavior = require("./Tree/treeItemBehavior");

Object.keys(_treeItemBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _treeItemBehavior[key]) return;
  exports[key] = _treeItemBehavior[key];
});

var _treeTitleBehavior = require("./Tree/treeTitleBehavior");

Object.keys(_treeTitleBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _treeTitleBehavior[key]) return;
  exports[key] = _treeTitleBehavior[key];
});

var _treeAsListBehavior = require("./Tree/treeAsListBehavior");

exports.treeAsListBehavior = _treeAsListBehavior.treeAsListBehavior;

var _treeItemAsListItemBehavior = require("./Tree/treeItemAsListItemBehavior");

exports.treeItemAsListItemBehavior = _treeItemAsListItemBehavior.treeItemAsListItemBehavior;

var _treeTitleAsListItemTitleBehavior = require("./Tree/treeTitleAsListItemTitleBehavior");

exports.treeTitleAsListItemTitleBehavior = _treeTitleAsListItemTitleBehavior.treeTitleAsListItemTitleBehavior;

var _treeAsListboxBehavior = require("./Tree/treeAsListboxBehavior");

exports.treeAsListboxBehavior = _treeAsListboxBehavior.treeAsListboxBehavior;

var _treeItemAsOptionBehavior = require("./Tree/treeItemAsOptionBehavior");

exports.treeItemAsOptionBehavior = _treeItemAsOptionBehavior.treeItemAsOptionBehavior;

var _treeTitleAsOptionBehavior = require("./Tree/treeTitleAsOptionBehavior");

exports.treeTitleAsOptionBehavior = _treeTitleAsOptionBehavior.treeTitleAsOptionBehavior;

var _textAreaBehavior = require("./TextArea/textAreaBehavior");

Object.keys(_textAreaBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _textAreaBehavior[key]) return;
  exports[key] = _textAreaBehavior[key];
});

var _carouselBehavior = require("./Carousel/carouselBehavior");

Object.keys(_carouselBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _carouselBehavior[key]) return;
  exports[key] = _carouselBehavior[key];
});

var _carouselItemBehavior = require("./Carousel/carouselItemBehavior");

Object.keys(_carouselItemBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _carouselItemBehavior[key]) return;
  exports[key] = _carouselItemBehavior[key];
});

var _tableBehavior = require("./Table/tableBehavior");

Object.keys(_tableBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _tableBehavior[key]) return;
  exports[key] = _tableBehavior[key];
});

var _tableRowBehavior = require("./Table/tableRowBehavior");

exports.tableRowBehavior = _tableRowBehavior.tableRowBehavior;

var _tableCellBehavior = require("./Table/tableCellBehavior");

Object.keys(_tableCellBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _tableCellBehavior[key]) return;
  exports[key] = _tableCellBehavior[key];
});

var _tableHeaderCellBehavior = require("./Table/tableHeaderCellBehavior");

exports.tableHeaderCellBehavior = _tableHeaderCellBehavior.tableHeaderCellBehavior;

var _gridNestedBehavior = require("./Table/gridNestedBehavior");

exports.gridNestedBehavior = _gridNestedBehavior.gridNestedBehavior;

var _gridHeaderRowBehavior = require("./Table/gridHeaderRowBehavior");

exports.gridHeaderRowBehavior = _gridHeaderRowBehavior.gridHeaderRowBehavior;

var _gridHeaderCellBehavior = require("./Table/gridHeaderCellBehavior");

exports.gridHeaderCellBehavior = _gridHeaderCellBehavior.gridHeaderCellBehavior;

var _gridRowBehavior = require("./Table/gridRowBehavior");

Object.keys(_gridRowBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _gridRowBehavior[key]) return;
  exports[key] = _gridRowBehavior[key];
});

var _gridRowNestedBehavior = require("./Table/gridRowNestedBehavior");

exports.gridRowNestedBehavior = _gridRowNestedBehavior.gridRowNestedBehavior;

var _gridCellBehavior = require("./Table/gridCellBehavior");

exports.gridCellBehavior = _gridCellBehavior.gridCellBehavior;

var _gridCellMultipleFocusableBehavior = require("./Table/gridCellMultipleFocusableBehavior");

exports.gridCellMultipleFocusableBehavior = _gridCellMultipleFocusableBehavior.gridCellMultipleFocusableBehavior;

var _gridCellWithFocusableElementBehavior = require("./Table/gridCellWithFocusableElementBehavior");

exports.gridCellWithFocusableElementBehavior = _gridCellWithFocusableElementBehavior.gridCellWithFocusableElementBehavior;

var _cardBehavior = require("./Card/cardBehavior");

Object.keys(_cardBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _cardBehavior[key]) return;
  exports[key] = _cardBehavior[key];
});

var _cardFocusableBehavior = require("./Card/cardFocusableBehavior");

exports.cardFocusableBehavior = _cardFocusableBehavior.cardFocusableBehavior;

var _cardChildrenFocusableBehavior = require("./Card/cardChildrenFocusableBehavior");

exports.cardChildrenFocusableBehavior = _cardChildrenFocusableBehavior.cardChildrenFocusableBehavior;

var _cardsContainerBehavior = require("./Card/cardsContainerBehavior");

Object.keys(_cardsContainerBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _cardsContainerBehavior[key]) return;
  exports[key] = _cardsContainerBehavior[key];
});

var _cardSelectableBehavior = require("./Card/cardSelectableBehavior");

exports.cardSelectableBehavior = _cardSelectableBehavior.cardSelectableBehavior;

var _videoBehavior = require("./Video/videoBehavior");

Object.keys(_videoBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _videoBehavior[key]) return;
  exports[key] = _videoBehavior[key];
});

var _hiddenComponentBehavior = require("./Common/hiddenComponentBehavior");

exports.hiddenComponentBehavior = _hiddenComponentBehavior.hiddenComponentBehavior;

var _datepickerBehavior = require("./Datepicker/datepickerBehavior");

Object.keys(_datepickerBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _datepickerBehavior[key]) return;
  exports[key] = _datepickerBehavior[key];
});

var _datepickerCalendarBehavior = require("./Datepicker/datepickerCalendarBehavior");

Object.keys(_datepickerCalendarBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _datepickerCalendarBehavior[key]) return;
  exports[key] = _datepickerCalendarBehavior[key];
});

var _datepickerCalendarHeaderBehavior = require("./Datepicker/datepickerCalendarHeaderBehavior");

Object.keys(_datepickerCalendarHeaderBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _datepickerCalendarHeaderBehavior[key]) return;
  exports[key] = _datepickerCalendarHeaderBehavior[key];
});

var _datepickerCalendarGridBehavior = require("./Datepicker/datepickerCalendarGridBehavior");

Object.keys(_datepickerCalendarGridBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _datepickerCalendarGridBehavior[key]) return;
  exports[key] = _datepickerCalendarGridBehavior[key];
});

var _datepickerCalendarGridRowBehavior = require("./Datepicker/datepickerCalendarGridRowBehavior");

Object.keys(_datepickerCalendarGridRowBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _datepickerCalendarGridRowBehavior[key]) return;
  exports[key] = _datepickerCalendarGridRowBehavior[key];
});

var _datepickerCalendarCellBehavior = require("./Datepicker/datepickerCalendarCellBehavior");

Object.keys(_datepickerCalendarCellBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _datepickerCalendarCellBehavior[key]) return;
  exports[key] = _datepickerCalendarCellBehavior[key];
});

var _datepickerCalendarCellButtonBehavior = require("./Datepicker/datepickerCalendarCellButtonBehavior");

Object.keys(_datepickerCalendarCellButtonBehavior).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _datepickerCalendarCellButtonBehavior[key]) return;
  exports[key] = _datepickerCalendarCellButtonBehavior[key];
});
