// @fluentui/react-northstar/dist/dts/src/components/Table/TableRow.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TableRow = exports.tableRowClassName = void 0;
var accessibility_1 = require("@fluentui/accessibility");
var react_bindings_1 = require("@fluentui/react-bindings");
var react_component_ref_1 = require("@fluentui/react-component-ref");
var customPropTypes = require("@fluentui/react-proptypes");
var _ = require("lodash");
var PropTypes = require("prop-types");
var React = require("react");
var utils_1 = require("../../utils");
var TableCell_1 = require("./TableCell");
exports.tableRowClassName = 'ui-table__row';
/**
 * Component represents a single row in a tabular structure
 */
exports.TableRow = React.forwardRef(function (props, ref) {
    var context = react_bindings_1.useFluentContext();
    var _a = react_bindings_1.useTelemetry(exports.TableRow.displayName, context.telemetry), setStart = _a.setStart, setEnd = _a.setEnd;
    setStart();
    var rowRef = React.useRef();
    var className = props.className, design = props.design, styles = props.styles, items = props.items, header = props.header, compact = props.compact, children = props.children, accessibility = props.accessibility, variables = props.variables, selected = props.selected;
    var hasChildren = utils_1.childrenExist(children);
    var ElementType = react_bindings_1.getElementType(props);
    var unhandledProps = react_bindings_1.useUnhandledProps(exports.TableRow.handledProps, props);
    var getA11yProps = react_bindings_1.useAccessibility(accessibility, {
        debugName: exports.TableRow.displayName,
        actionHandlers: {
            // https://github.com/microsoft/fluent-ui-react/issues/2150
            unsetRowTabbable: function (e) {
                rowRef.current.setAttribute('tabindex', '-1');
            },
            performClick: function (e) {
                if (e.currentTarget === e.target) {
                    _.invoke(props, 'onClick', e, props);
                    e.preventDefault();
                }
            },
        },
        mapPropsToBehavior: function () { return ({
            selected: selected,
            header: header,
        }); },
        rtl: context.rtl,
    });
    var classes = react_bindings_1.useStyles(exports.TableRow.displayName, {
        className: exports.tableRowClassName,
        mapPropsToStyles: function () { return ({
            header: header,
            compact: compact,
        }); },
        mapPropsToInlineStyles: function () { return ({
            className: className,
            design: design,
            styles: styles,
            variables: variables,
        }); },
        rtl: context.rtl,
    }).classes;
    var renderCells = function () {
        return _.map(items, function (item) {
            return TableCell_1.TableCell.create(item, {
                defaultProps: function () { return getA11yProps('cell', {}); },
                overrideProps: function (predefinedProps) { return ({
                    variables: react_bindings_1.mergeVariablesOverrides(variables, predefinedProps.variables),
                }); },
            });
        });
    };
    var element = (React.createElement(react_component_ref_1.Ref, { innerRef: rowRef }, getA11yProps.unstable_wrapWithFocusZone(React.createElement(ElementType, __assign({}, getA11yProps('root', __assign({ className: classes.root, ref: ref }, unhandledProps))),
        hasChildren && children,
        !hasChildren && renderCells()))));
    setEnd();
    return element;
});
exports.TableRow.displayName = 'TableRow';
exports.TableRow.propTypes = __assign(__assign({}, utils_1.commonPropTypes.createCommon({
    content: false,
})), { items: customPropTypes.collectionShorthand, header: PropTypes.bool, compact: PropTypes.bool, selected: PropTypes.bool });
exports.TableRow.handledProps = Object.keys(exports.TableRow.propTypes);
exports.TableRow.defaultProps = {
    accessibility: accessibility_1.tableRowBehavior,
};
exports.TableRow.create = utils_1.createShorthandFactory({ Component: exports.TableRow, mappedArrayProp: 'items' });
