// @fluentui/react-northstar/dist/es/components/Toolbar/ToolbarMenuRadioGroupWrapper.js
import { compose } from '@fluentui/react-bindings';
import { toolbarMenuRadioGroupWrapperBehavior } from '@fluentui/accessibility';
import { commonPropTypes } from '../../utils';
import { Box } from '../Box/Box';
export var toolbarMenuRadioGroupWrapperClassName = 'ui-toolbars'; // FIXME: required by getComponentInfo/isConformant. But this is group inside a toolbar not a group of toolbars

/**
 * An ToolbarMenuRadioGroupWrapper provides a wrapping slot for in  ToolbarMenuRadioGroup.
 */

export var ToolbarMenuRadioGroupWrapper = /*#__PURE__*/function () {
  var ToolbarMenuRadioGroupWrapper = compose(Box, {
    className: toolbarMenuRadioGroupWrapperClassName,
    displayName: 'ToolbarMenuRadioGroupWrapper',
    overrideStyles: true,
    shorthandConfig: {
      mappedProp: 'content'
    }
  });
  ToolbarMenuRadioGroupWrapper.defaultProps = {
    as: 'li',
    accessibility: toolbarMenuRadioGroupWrapperBehavior
  };
  ToolbarMenuRadioGroupWrapper.propTypes = commonPropTypes.createCommon();
  return ToolbarMenuRadioGroupWrapper;
}();
