// @fluentui/react-northstar/dist/es/components/Toolbar/ToolbarMenuItemSubmenuIndicator.js
import { compose } from '@fluentui/react-bindings';
import { commonPropTypes } from '../../utils';
import { Box } from '../Box/Box';
export var toolbarMenuItemSubmenuIndicatorClassName = 'ui-toolbar__menuitemsubmenuindicator';
/**
 * A ToolbarMenuItemSubmenuIndicator allows a user to have a dedicated component that can be targeted from the theme.
 */

export var ToolbarMenuItemSubmenuIndicator = /*#__PURE__*/function () {
  var ToolbarMenuItemSubmenuIndicator = compose(Box, {
    className: toolbarMenuItemSubmenuIndicatorClassName,
    displayName: 'ToolbarMenuItemSubmenuIndicator',
    mapPropsToStylesProps: function mapPropsToStylesProps(props) {
      return {
        hasContent: props.hasContent
      };
    },
    shorthandConfig: {
      mappedProp: 'content'
    },
    overrideStyles: true
  });
  ToolbarMenuItemSubmenuIndicator.defaultProps = {
    as: 'span'
  };
  ToolbarMenuItemSubmenuIndicator.propTypes = commonPropTypes.createCommon();
  return ToolbarMenuItemSubmenuIndicator;
}();
