// @uifabric/merge-styles/lib-commonjs/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var mergeStyles_1 = require("./mergeStyles");
exports.mergeStyles = mergeStyles_1.mergeStyles;
exports.mergeCss = mergeStyles_1.mergeCss;
var mergeStyleSets_1 = require("./mergeStyleSets");
exports.mergeStyleSets = mergeStyleSets_1.mergeStyleSets;
exports.mergeCssSets = mergeStyleSets_1.mergeCssSets;
var concatStyleSets_1 = require("./concatStyleSets");
exports.concatStyleSets = concatStyleSets_1.concatStyleSets;
var concatStyleSetsWithProps_1 = require("./concatStyleSetsWithProps");
exports.concatStyleSetsWithProps = concatStyleSetsWithProps_1.concatStyleSetsWithProps;
var fontFace_1 = require("./fontFace");
exports.fontFace = fontFace_1.fontFace;
var keyframes_1 = require("./keyframes");
exports.keyframes = keyframes_1.keyframes;
tslib_1.__exportStar(require("./Stylesheet"), exports);
var StyleOptionsState_1 = require("./StyleOptionsState");
exports.setRTL = StyleOptionsState_1.setRTL;
require("./version");
