// @fluentui/react-northstar/dist/commonjs/themes/teams/components/Slider/sliderVariables.js
"use strict";

exports.__esModule = true;
exports.sliderVariables = void 0;

var _utils = require("../../../../utils");

var sliderVariables = function sliderVariables(siteVars) {
  var colorScheme = siteVars.colorScheme;
  return {
    height: (0, _utils.pxToRem)(16),
    length: (0, _utils.pxToRem)(320),
    railColor: colorScheme.default.border,
    disabledRailColor: colorScheme.default.backgroundDisabled1,
    railHeight: (0, _utils.pxToRem)(2),
    thumbColor: colorScheme.default.foreground2,
    activeThumbColor: colorScheme.default.foreground1,
    disabledThumbColor: colorScheme.default.foregroundDisabled1,
    thumbHeight: (0, _utils.pxToRem)(10),
    activeThumbHeight: (0, _utils.pxToRem)(14),
    thumbBorderPadding: (0, _utils.pxToRem)(4),
    thumbWidth: (0, _utils.pxToRem)(10),
    activeThumbWidth: (0, _utils.pxToRem)(14),
    trackColor: colorScheme.brand.foregroundActive,
    disabledTrackColor: colorScheme.default.foregroundDisabled1
  };
};

exports.sliderVariables = sliderVariables;
