// @fluentui/react-northstar/dist/dts/src/components/Card/CardColumn.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CardColumn = exports.cardColumnClassName = void 0;
var React = require("react");
var utils_1 = require("../../utils");
var react_bindings_1 = require("@fluentui/react-bindings");
exports.cardColumnClassName = 'ui-card__column';
/**
 * A CardColumn is used to display content in card as column
 */
exports.CardColumn = React.forwardRef(function (props, ref) {
    var context = react_bindings_1.useFluentContext();
    var _a = react_bindings_1.useTelemetry(exports.CardColumn.displayName, context.telemetry), setStart = _a.setStart, setEnd = _a.setEnd;
    setStart();
    var className = props.className, design = props.design, styles = props.styles, variables = props.variables, children = props.children;
    var ElementType = react_bindings_1.getElementType(props);
    var unhandledProps = react_bindings_1.useUnhandledProps(exports.CardColumn.handledProps, props);
    var getA11yProps = react_bindings_1.useAccessibility(props.accessibility, {
        debugName: exports.CardColumn.displayName,
        rtl: context.rtl,
    });
    var classes = react_bindings_1.useStyles(exports.CardColumn.displayName, {
        className: exports.cardColumnClassName,
        mapPropsToInlineStyles: function () { return ({
            className: className,
            design: design,
            styles: styles,
            variables: variables,
        }); },
        rtl: context.rtl,
    }).classes;
    var element = (React.createElement(ElementType, __assign({}, getA11yProps('root', __assign({ className: classes.root, ref: ref }, unhandledProps))), children));
    setEnd();
    return element;
});
exports.CardColumn.displayName = 'CardColumn';
exports.CardColumn.propTypes = __assign({}, utils_1.commonPropTypes.createCommon());
exports.CardColumn.handledProps = Object.keys(exports.CardColumn.propTypes);
exports.CardColumn.create = utils_1.createShorthandFactory({ Component: exports.CardColumn });
