// @fluentui/react-northstar/dist/dts/src/components/Card/Card.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Card = exports.cardClassName = void 0;
var accessibility_1 = require("@fluentui/accessibility");
var react_bindings_1 = require("@fluentui/react-bindings");
var react_component_ref_1 = require("@fluentui/react-component-ref");
var CustomPropTypes = require("@fluentui/react-proptypes");
var _ = require("lodash");
var PropTypes = require("prop-types");
var React = require("react");
var utils_1 = require("../../utils");
var CardBody_1 = require("./CardBody");
var CardColumn_1 = require("./CardColumn");
var CardFooter_1 = require("./CardFooter");
var CardHeader_1 = require("./CardHeader");
var CardPreview_1 = require("./CardPreview");
var CardTopControls_1 = require("./CardTopControls");
var CardExpandableBox_1 = require("./CardExpandableBox");
exports.cardClassName = 'ui-card';
/**
 * A Card is used to display data in sematically grouped way.
 * * @accessibility
 * By default adds `group` role ([more information available in aria documentation](https://www.w3.org/TR/wai-aria-1.1/#group)), thus it's necessary to provide `aria-roledescription` for correct widget description. [More information available in aria documentation.](https://www.w3.org/TR/wai-aria-1.1/#aria-roledescription-property)
 * When card is actionable (i.e. has `onClick` property), use [cardFocusableBehavior](/components/card/accessibility#card-focusable). [More information available in aria documentation.](https://www.w3.org/TR/wai-aria-practices/#gridNav_focus)
 * When card contains actionable elements, use [cardChildrenFocusableBehavior](/components/card/accessibility#card-children-focusable).
 *
 */
exports.Card = React.forwardRef(function (props, ref) {
    var context = react_bindings_1.useFluentContext();
    var _a = react_bindings_1.useTelemetry(exports.Card.displayName, context.telemetry), setStart = _a.setStart, setEnd = _a.setEnd;
    setStart();
    var cardRef = React.useRef();
    var className = props.className, design = props.design, styles = props.styles, variables = props.variables, children = props.children, compact = props.compact, horizontal = props.horizontal, centered = props.centered, size = props.size, fluid = props.fluid, onClick = props.onClick, disabled = props.disabled, expandable = props.expandable, elevated = props.elevated, inverted = props.inverted, ghost = props.ghost, selected = props.selected;
    var ElementType = react_bindings_1.getElementType(props);
    var unhandledProps = react_bindings_1.useUnhandledProps(exports.Card.handledProps, props);
    var getA11yProps = react_bindings_1.useAccessibility(props.accessibility, {
        debugName: exports.Card.displayName,
        actionHandlers: {
            performClick: function (e) {
                // prevent Spacebar from scrolling
                e.preventDefault();
                handleClick(e);
            },
            focusCard: function (e) {
                cardRef.current.focus();
            },
        },
        mapPropsToBehavior: function () { return ({
            disabled: disabled,
        }); },
        rtl: context.rtl,
    });
    var classes = react_bindings_1.useStyles(exports.Card.displayName, {
        className: exports.cardClassName,
        mapPropsToStyles: function () { return ({
            centered: centered,
            horizontal: horizontal,
            compact: compact,
            size: size,
            fluid: fluid,
            actionable: !!onClick,
            disabled: disabled,
            expandable: expandable,
            elevated: elevated,
            inverted: inverted,
            ghost: ghost,
            selected: selected,
        }); },
        mapPropsToInlineStyles: function () { return ({
            className: className,
            design: design,
            styles: styles,
            variables: variables,
        }); },
        rtl: context.rtl,
    }).classes;
    var handleClick = function (e) {
        if (disabled) {
            e.preventDefault();
            return;
        }
        _.invoke(props, 'onClick', e, props);
    };
    var element = (React.createElement(react_component_ref_1.Ref, { innerRef: cardRef }, getA11yProps.unstable_wrapWithFocusZone(React.createElement(ElementType, __assign({}, getA11yProps('root', __assign({ className: classes.root, onClick: handleClick, ref: ref }, unhandledProps))), children))));
    setEnd();
    return element;
});
exports.Card.displayName = 'Card';
exports.Card.propTypes = __assign(__assign({}, utils_1.commonPropTypes.createCommon()), { onClick: PropTypes.func, compact: PropTypes.bool, horizontal: PropTypes.bool, centered: PropTypes.bool, size: CustomPropTypes.size, fluid: PropTypes.bool, expandable: PropTypes.bool, disabled: PropTypes.bool, elevated: PropTypes.bool, ghost: PropTypes.bool, inverted: PropTypes.bool, selected: PropTypes.bool });
exports.Card.defaultProps = {
    accessibility: accessibility_1.cardBehavior,
    size: 'medium',
};
exports.Card.handledProps = Object.keys(exports.Card.propTypes);
exports.Card.Header = CardHeader_1.CardHeader;
exports.Card.Body = CardBody_1.CardBody;
exports.Card.Footer = CardFooter_1.CardFooter;
exports.Card.Preview = CardPreview_1.CardPreview;
exports.Card.TopControls = CardTopControls_1.CardTopControls;
exports.Card.Column = CardColumn_1.CardColumn;
exports.Card.ExpandableBox = CardExpandableBox_1.CardExpandableBox;
exports.Card.create = utils_1.createShorthandFactory({ Component: exports.Card });
