// @fluentui/react-northstar/dist/dts/src/components/Dialog/DialogFooter.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DialogFooter = exports.dialogFooterClassName = void 0;
var React = require("react");
var utils_1 = require("../../utils");
var react_bindings_1 = require("@fluentui/react-bindings");
exports.dialogFooterClassName = 'ui-dialog__footer';
/**
 * A DialogFooter represents footer content in Dialog, usually shows dialog actions.
 */
exports.DialogFooter = React.forwardRef(function (props, ref) {
    var context = react_bindings_1.useFluentContext();
    var _a = react_bindings_1.useTelemetry(exports.DialogFooter.displayName, context.telemetry), setStart = _a.setStart, setEnd = _a.setEnd;
    setStart();
    var children = props.children, content = props.content, className = props.className, design = props.design, styles = props.styles, variables = props.variables, accessibility = props.accessibility;
    var ElementType = react_bindings_1.getElementType(props);
    var unhandledProps = react_bindings_1.useUnhandledProps(exports.DialogFooter.handledProps, props);
    var getA11yProps = react_bindings_1.useAccessibility(accessibility, {
        debugName: exports.DialogFooter.displayName,
        rtl: context.rtl,
    });
    var classes = react_bindings_1.useStyles(exports.DialogFooter.displayName, {
        className: exports.dialogFooterClassName,
        mapPropsToInlineStyles: function () { return ({
            className: className,
            design: design,
            styles: styles,
            variables: variables,
        }); },
        rtl: context.rtl,
    }).classes;
    var element = (React.createElement(ElementType, __assign({}, getA11yProps('root', __assign({ className: classes.root, ref: ref }, unhandledProps))), utils_1.childrenExist(children) ? children : content));
    setEnd();
    return element;
});
exports.DialogFooter.displayName = 'DialogFooter';
exports.DialogFooter.propTypes = __assign({}, utils_1.commonPropTypes.createCommon());
exports.DialogFooter.handledProps = Object.keys(exports.DialogFooter.propTypes);
exports.DialogFooter.create = utils_1.createShorthandFactory({ Component: exports.DialogFooter, mappedProp: 'content' });
