// @fluentui/react-northstar/dist/commonjs/themes/teams/components/Skeleton/utils/animations.js
"use strict";

exports.__esModule = true;
exports.getAnimations = void 0;

var _keyframes = require("./keyframes");

var _SkeletonLine = require("../../../../../components/Skeleton/SkeletonLine");

var _SkeletonShape = require("../../../../../components/Skeleton/SkeletonShape");

var _SkeletonButton = require("../../../../../components/Skeleton/SkeletonButton");

var _SkeletonInput = require("../../../../../components/Skeleton/SkeletonInput");

var _SkeletonAvatar = require("../../../../../components/Skeleton/SkeletonAvatar");

var _SkeletonText = require("../../../../../components/Skeleton/SkeletonText");

var waveStyles = function waveStyles(primaryColor, secondoryColor) {
  return {
    position: 'static',
    overflow: 'hidden',
    animationDuration: '4.5s',
    animationName: _keyframes.wave,
    animationIterationCount: 'infinite',
    backgroundColor: '#FFFFFF',
    animationTimingFunction: 'linear',
    backgroundImage: "linear-gradient(to right, " + secondoryColor + " 0%, " + primaryColor + " 20%, " + secondoryColor + " 40%)",
    backgroundSize: '1200px 100%'
  };
};

var pulseStyles = {
  animationName: _keyframes.pulse,
  animationDuration: '1.5s',
  animationIterationCount: 'infinite'
};

var getAnimations = function getAnimations(primaryColor, secondoryColor) {
  var _pulse, _wave;

  return {
    pulse: (_pulse = {}, _pulse["& ." + _SkeletonLine.skeletonLineClassName] = pulseStyles, _pulse["& ." + _SkeletonShape.skeletonShapeClassName] = pulseStyles, _pulse["& ." + _SkeletonButton.skeletonButtonClassName] = pulseStyles, _pulse["& ." + _SkeletonAvatar.skeletonAvatarClassName] = pulseStyles, _pulse["& ." + _SkeletonInput.skeletonInputClassName] = pulseStyles, _pulse["& ." + _SkeletonText.skeletonTextClassName] = pulseStyles, _pulse),
    wave: (_wave = {}, _wave["& ." + _SkeletonLine.skeletonLineClassName] = waveStyles(primaryColor, secondoryColor), _wave["& ." + _SkeletonShape.skeletonShapeClassName] = waveStyles(primaryColor, secondoryColor), _wave["& ." + _SkeletonButton.skeletonButtonClassName] = waveStyles(primaryColor, secondoryColor), _wave["& ." + _SkeletonAvatar.skeletonAvatarClassName] = waveStyles(primaryColor, secondoryColor), _wave["& ." + _SkeletonInput.skeletonInputClassName] = waveStyles(primaryColor, secondoryColor), _wave["& ." + _SkeletonText.skeletonTextClassName] = waveStyles(primaryColor, secondoryColor), _wave)
  };
};

exports.getAnimations = getAnimations;
