// @fluentui/react-northstar/dist/dts/src/components/Input/InputLabel.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.InputLabel = exports.inputLabelClassName = void 0;
var react_bindings_1 = require("@fluentui/react-bindings");
var utils_1 = require("../../utils");
var Box_1 = require("../Box/Box");
var PropTypes = require("prop-types");
exports.inputLabelClassName = 'ui-input__label';
/**
 * An InputLabel provides a slot for label in the Input.
 */
exports.InputLabel = react_bindings_1.compose(Box_1.Box, {
    className: exports.inputLabelClassName,
    displayName: 'InputLabel',
    overrideStyles: true,
    mapPropsToStylesProps: function (_a) {
        var labelPosition = _a.labelPosition, required = _a.required, hasValue = _a.hasValue;
        return ({
            labelPosition: labelPosition,
            required: required,
            hasValue: hasValue,
        });
    },
    handledProps: ['required', 'labelPosition', 'hasValue'],
    shorthandConfig: {
        mappedProp: 'content',
    },
});
exports.InputLabel.defaultProps = {
    as: 'label',
};
exports.InputLabel.propTypes = __assign(__assign({}, utils_1.commonPropTypes.createCommon()), { labelPosition: PropTypes.oneOf(['inline', 'above', 'inside']), required: PropTypes.bool, hasValue: PropTypes.bool });
