// @fluentui/accessibility/dist/dts/src/behaviors/Embed/embedBehavior.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.embedBehavior = void 0;
var keyboard_key_1 = require("../../keyboard-key");
/**
 * @description
 * GIFs are visual representation only so hidden unless alt or title applied.
 * Enter/space keys play and pause the gif respectively
 *
 * @specification
 * Adds role 'presentation' to 'root' slot.
 * Adds attribute 'aria-hidden=true', if there is no 'alt' property provided.
 * Adds attribute 'tabIndex=0' to 'root' slot.
 */
var embedBehavior = function (props) { return ({
    attributes: {
        root: {
            'aria-hidden': props.alt || props.title ? undefined : true,
            role: 'presentation',
            tabIndex: 0,
        },
    },
    keyActions: {
        root: {
            performClick: {
                keyCombinations: [{ keyCode: keyboard_key_1.keyboardKey.Enter }, { keyCode: keyboard_key_1.SpacebarKey }],
            },
        },
    },
}); };
exports.embedBehavior = embedBehavior;
