// @fluentui/react-northstar/dist/dts/src/themes/teams/components/Form/formFieldStyles.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.formFieldStyles = void 0;
var utils_1 = require("../../../../utils");
exports.formFieldStyles = {
    root: function (_a) {
        var props = _a.props, variables = _a.variables;
        return ({});
    },
    label: function (_a) {
        var props = _a.props;
        var type = props.type, inline = props.inline, required = props.required;
        return __assign(__assign(__assign({}, ((!type || (type !== 'radio' && type !== 'checkbox')) && {
            display: 'block',
            marginBottom: utils_1.pxToRem(4),
        })), (inline && { marginRight: utils_1.pxToRem(10), display: 'inline' })), (required && {
            '::after': {
                content: '"*"',
            },
        }));
    },
    control: function (_a) {
        var props = _a.props, v = _a.variables;
        var type = props.type;
        return __assign({}, (type &&
            (type === 'radio' || type === 'checkbox') && {
            marginRight: utils_1.pxToRem(10),
        }));
    },
    message: function (_a) {
        var p = _a.props, v = _a.variables;
        return (__assign(__assign({}, (p.hasErrorMessage && { color: v.colorScheme.red.foreground })), { display: 'block', paddingLeft: v.messagePaddingLeft }));
    },
    icon: function (_a) {
        var p = _a.props, v = _a.variables;
        return (__assign({}, (p.hasErrorMessage && { color: v.colorScheme.red.foreground })));
    },
};
