// @fluentui/react-northstar/dist/commonjs/themes/teams/components/Loader/loaderVariables.js
"use strict";

exports.__esModule = true;
exports.loaderVariables = void 0;

var _utils = require("../../../../utils");

var _loaderSvgDataUrl = require("./loaderSvgDataUrl");

var _loaderSecondarySvgDataUrl = require("./loaderSecondarySvgDataUrl");

var loaderVariables = function loaderVariables() {
  return {
    containerHeights: {
      smallest: (0, _utils.pxToRem)(24),
      smaller: (0, _utils.pxToRem)(24),
      small: (0, _utils.pxToRem)(24),
      medium: (0, _utils.pxToRem)(36),
      large: (0, _utils.pxToRem)(72),
      larger: (0, _utils.pxToRem)(72),
      largest: (0, _utils.pxToRem)(72)
    },
    containerWidths: {
      smallest: (0, _utils.pxToRem)(24),
      smaller: (0, _utils.pxToRem)(24),
      small: (0, _utils.pxToRem)(24),
      medium: (0, _utils.pxToRem)(36),
      large: (0, _utils.pxToRem)(72),
      larger: (0, _utils.pxToRem)(72),
      largest: (0, _utils.pxToRem)(72)
    },
    svgContent: _loaderSvgDataUrl.loaderSvgDataUrl,
    secondarySvgContent: _loaderSecondarySvgDataUrl.loaderSecondarySvgDataUrl,
    svgHeights: {
      smaller: (0, _utils.pxToRem)(1464),
      smallest: (0, _utils.pxToRem)(1464),
      small: (0, _utils.pxToRem)(1464),
      medium: (0, _utils.pxToRem)(2196),
      large: (0, _utils.pxToRem)(4392),
      larger: (0, _utils.pxToRem)(4392),
      largest: (0, _utils.pxToRem)(4392)
    },
    svgTranslatePosition: {
      smallest: (0, _utils.pxToRem)(-1440),
      smaller: (0, _utils.pxToRem)(-1440),
      small: (0, _utils.pxToRem)(-1440),
      medium: (0, _utils.pxToRem)(-2160),
      large: (0, _utils.pxToRem)(-4320),
      larger: (0, _utils.pxToRem)(-4320),
      largest: (0, _utils.pxToRem)(-4320)
    },
    svgWidths: {
      smallest: (0, _utils.pxToRem)(24),
      smaller: (0, _utils.pxToRem)(24),
      small: (0, _utils.pxToRem)(24),
      medium: (0, _utils.pxToRem)(36),
      large: (0, _utils.pxToRem)(72),
      larger: (0, _utils.pxToRem)(72),
      largest: (0, _utils.pxToRem)(72)
    }
  };
};

exports.loaderVariables = loaderVariables;
