// @fluentui/react-northstar/dist/dts/src/components/Status/Status.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Status = exports.statusClassName = void 0;
var accessibility_1 = require("@fluentui/accessibility");
var react_bindings_1 = require("@fluentui/react-bindings");
var customPropTypes = require("@fluentui/react-proptypes");
var PropTypes = require("prop-types");
var React = require("react");
var utils_1 = require("../../utils");
var Box_1 = require("../Box/Box");
exports.statusClassName = 'ui-status';
/**
 * A Status represents someone's or something's state.
 *
 * @accessibility
 * Implements [ARIA img](https://www.w3.org/TR/wai-aria-1.1/#img) role.
 */
exports.Status = React.forwardRef(function (props, ref) {
    var context = react_bindings_1.useFluentContext();
    var _a = react_bindings_1.useTelemetry(exports.Status.displayName, context.telemetry), setStart = _a.setStart, setEnd = _a.setEnd;
    setStart();
    var className = props.className, color = props.color, icon = props.icon, size = props.size, state = props.state, design = props.design, styles = props.styles, variables = props.variables;
    var _b = react_bindings_1.useStyles(exports.Status.displayName, {
        className: exports.statusClassName,
        mapPropsToStyles: function () { return ({
            color: color,
            size: size,
            state: state,
        }); },
        mapPropsToInlineStyles: function () { return ({
            className: className,
            design: design,
            styles: styles,
            variables: variables,
        }); },
        rtl: context.rtl,
    }), classes = _b.classes, resolvedStyles = _b.styles;
    var getA11Props = react_bindings_1.useAccessibility(props.accessibility, {
        debugName: exports.Status.displayName,
        rtl: context.rtl,
    });
    var ElementType = react_bindings_1.getElementType(props);
    var unhandledProps = react_bindings_1.useUnhandledProps(exports.Status.handledProps, props);
    var iconElement = Box_1.Box.create(icon, {
        defaultProps: function () {
            return getA11Props('icon', {
                styles: resolvedStyles.icon,
                as: 'span',
            });
        },
    });
    var element = (React.createElement(ElementType, __assign({}, getA11Props('root', __assign({ className: classes.root, ref: ref }, unhandledProps))), iconElement));
    setEnd();
    return element;
});
exports.Status.displayName = 'Status';
exports.Status.propTypes = __assign(__assign({}, utils_1.commonPropTypes.createCommon({
    children: false,
    content: false,
})), { color: PropTypes.string, icon: customPropTypes.shorthandAllowingChildren, size: customPropTypes.size, state: PropTypes.oneOf([
        'success',
        'info',
        'warning',
        'error',
        'unknown',
    ]) });
exports.Status.handledProps = Object.keys(exports.Status.propTypes);
exports.Status.defaultProps = {
    accessibility: accessibility_1.statusBehavior,
    as: 'span',
    size: 'medium',
    state: 'unknown',
};
exports.Status.create = utils_1.createShorthandFactory({ Component: exports.Status, mappedProp: 'state' });
