// @fluentui/accessibility/dist/dts/src/behaviors/Tree/treeItemAsListItemBehavior.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.treeItemAsListItemBehavior = void 0;
var _ = require("lodash");
var treeItemBehavior_1 = require("./treeItemBehavior");
var treeTitleAsListItemTitleBehavior_1 = require("./treeTitleAsListItemTitleBehavior");
/**
 * @description
 * Adds role 'listitem' to a non-leaf item and 'none' to a leaf item.
 */
var treeItemAsListItemBehavior = function (props) {
    var behavior = treeItemBehavior_1.treeItemBehavior(props);
    var definition = _.merge(behavior, {
        attributes: {
            root: __assign({}, (props.hasSubtree && {
                role: 'listitem',
            })),
        },
        childBehaviors: {
            title: treeTitleAsListItemTitleBehavior_1.treeTitleAsListItemTitleBehavior,
        },
    });
    if (process.env.NODE_ENV !== 'production' && props.hasSubtree) {
        // Override the default trigger's accessibility schema class.
        definition.attributes.root['data-aa-class'] = 'TreeItemList';
    }
    return definition;
};
exports.treeItemAsListItemBehavior = treeItemAsListItemBehavior;
