// @fluentui/accessibility/dist/dts/src/behaviors/Tree/treeAsListboxBehavior.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.treeAsListboxBehavior = void 0;
var treeBehavior_1 = require("./treeBehavior");
var treeItemAsOptionBehavior_1 = require("./treeItemAsOptionBehavior");
var treeTitleAsOptionBehavior_1 = require("./treeTitleAsOptionBehavior");
/**
 * @specification
 * Adds role='listbox'.
 * Adds attribute 'aria-multiselectable=true' to 'root' slot if 'selectable' property is true. Does not set the attribute otherwise.
 */
var treeAsListboxBehavior = function (props) {
    var behavior = treeBehavior_1.treeBehavior(props);
    var definition = {
        attributes: {
            root: __assign(__assign({}, behavior.attributes.root), { role: 'listbox' }),
        },
        keyActions: behavior.keyActions,
        focusZone: behavior.focusZone,
        childBehaviors: {
            item: treeItemAsOptionBehavior_1.treeItemAsOptionBehavior,
            title: treeTitleAsOptionBehavior_1.treeTitleAsOptionBehavior,
        },
    };
    if (process.env.NODE_ENV !== 'production') {
        definition.attributes.root['data-aa-class'] = 'TreeListbox';
    }
    return definition;
};
exports.treeAsListboxBehavior = treeAsListboxBehavior;
