// @fluentui/accessibility/dist/dts/src/behaviors/Tree/treeTitleBehavior.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.treeTitleBehavior = void 0;
var keyboard_key_1 = require("../../keyboard-key");
var attributes_1 = require("../../attributes");
/**
 * @description
 *  Adds attribute 'aria-checked=true' based on the properties 'selectable' & 'selected' if the component has 'hasSubtree' property false or undefined. Does not set anything if true.
 *  Triggers 'performClick' action with 'Spacebar' on 'root', when tree title is selectable.
 *  @specification
 * Adds attribute 'tabIndex=-1' to 'root' slot if 'hasSubtree' property is false or undefined. Does not set the attribute if true.
 * Adds attribute 'role=treeitem' to 'root' slot if 'hasSubtree' property is false or undefined. Does not set the attribute if true.
 * Adds attribute 'aria-setsize=3' based on the property 'treeSize' if the component has 'hasSubtree' property false or undefined. Does not set anything if true..
 * Adds attribute 'aria-posinset=2' based on the property 'index' if the component has 'hasSubtree' property false or undefined. Does not set anything if true..
 * Adds attribute 'aria-level=1' based on the property 'level' if the component has 'hasSubtree' property false or undefined. Does not set anything if true..
 * Triggers 'performClick' action with 'Spacebar' on 'root'.
 */
var treeTitleBehavior = function (props) {
    var _a;
    var definition = {
        attributes: {
            root: __assign({}, (!props.hasSubtree && __assign((_a = { tabIndex: -1 }, _a[attributes_1.IS_FOCUSABLE_ATTRIBUTE] = true, _a.role = 'treeitem', _a['aria-setsize'] = props.treeSize, _a['aria-posinset'] = props.index, _a['aria-level'] = props.level, _a), (props.selectable && { 'aria-checked': props.selected })))),
        },
        keyActions: {
            root: {
                performClick: {
                    keyCombinations: props.selectable
                        ? [{ keyCode: keyboard_key_1.SpacebarKey }]
                        : [{ keyCode: keyboard_key_1.SpacebarKey }, { keyCode: keyboard_key_1.EnterKey }],
                },
                focusParent: {
                    keyCombinations: [{ keyCode: keyboard_key_1.keyboardKey.ArrowLeft }],
                },
            },
        },
    };
    if (process.env.NODE_ENV !== 'production' && props.hasSubtree) {
        // Override the default trigger's accessibility schema class.
        definition.attributes.root['data-aa-class'] = 'ExpandableTreeTitle';
    }
    return definition;
};
exports.treeTitleBehavior = treeTitleBehavior;
