// @fluentui/accessibility/dist/dts/src/behaviors/Tree/treeAsListBehavior.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.treeAsListBehavior = void 0;
var treeBehavior_1 = require("./treeBehavior");
var treeItemAsListItemBehavior_1 = require("./treeItemAsListItemBehavior");
/**
 * @specification
 * Adds role 'list' to 'root' slot.
 */
var treeAsListBehavior = function (props) {
    var behavior = treeBehavior_1.treeBehavior(props);
    var definition = {
        attributes: {
            root: __assign(__assign({}, behavior.attributes.root), { role: 'list' }),
        },
        keyActions: behavior.keyActions,
        focusZone: behavior.focusZone,
        childBehaviors: {
            item: treeItemAsListItemBehavior_1.treeItemAsListItemBehavior,
        },
    };
    if (process.env.NODE_ENV !== 'production') {
        definition.attributes.root['data-aa-class'] = 'TreeList';
    }
    return definition;
};
exports.treeAsListBehavior = treeAsListBehavior;
