// @fluentui/react-northstar/dist/commonjs/components/Header/HeaderDescription.js
"use strict";

exports.__esModule = true;
exports.HeaderDescription = exports.headerDescriptionClassName = void 0;

var React = _interopRequireWildcard(require("react"));

var _utils = require("../../utils");

var _reactBindings = require("@fluentui/react-bindings");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var headerDescriptionClassName = 'ui-header__description';
exports.headerDescriptionClassName = headerDescriptionClassName;

/**
 * A HeaderDescription provides more detailed information about the Header.
 */
var HeaderDescription = /*#__PURE__*/React.forwardRef(function (props, ref) {
  var context = (0, _reactBindings.useFluentContext)();

  var _useTelemetry = (0, _reactBindings.useTelemetry)(HeaderDescription.displayName, context.telemetry),
      setStart = _useTelemetry.setStart,
      setEnd = _useTelemetry.setEnd;

  setStart();
  var children = props.children,
      content = props.content,
      color = props.color,
      className = props.className,
      design = props.design,
      styles = props.styles,
      variables = props.variables;
  var ElementType = (0, _reactBindings.getElementType)(props);
  var unhandledProps = (0, _reactBindings.useUnhandledProps)(HeaderDescription.handledProps, props);
  var getA11yProps = (0, _reactBindings.useAccessibility)(props.accessibility, {
    debugName: HeaderDescription.displayName,
    rtl: context.rtl
  });

  var _useStyles = (0, _reactBindings.useStyles)(HeaderDescription.displayName, {
    className: headerDescriptionClassName,
    mapPropsToStyles: function mapPropsToStyles() {
      return {
        color: color
      };
    },
    mapPropsToInlineStyles: function mapPropsToInlineStyles() {
      return {
        className: className,
        design: design,
        styles: styles,
        variables: variables
      };
    },
    rtl: context.rtl
  }),
      classes = _useStyles.classes;

  var element = /*#__PURE__*/React.createElement(ElementType, getA11yProps('root', Object.assign({
    className: classes.root,
    ref: ref
  }, unhandledProps, _utils.rtlTextContainer.getAttributes({
    forElements: [children, content]
  }))), (0, _utils.childrenExist)(children) ? children : content);
  setEnd();
  return element;
});
exports.HeaderDescription = HeaderDescription;
HeaderDescription.displayName = 'HeaderDescription';
HeaderDescription.propTypes = Object.assign({}, _utils.commonPropTypes.createCommon({
  color: true
}));
HeaderDescription.defaultProps = {
  as: 'p'
};
HeaderDescription.handledProps = Object.keys(HeaderDescription.propTypes);
HeaderDescription.create = (0, _utils.createShorthandFactory)({
  Component: HeaderDescription,
  mappedProp: 'content'
});
