// @fluentui/react-northstar/dist/dts/src/components/Datepicker/DatepickerCalendarGridRow.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DatepickerCalendarGridRow = exports.datepickerCalendarGridRowClassName = void 0;
var react_bindings_1 = require("@fluentui/react-bindings");
var utils_1 = require("../../utils");
var Box_1 = require("../Box/Box");
exports.datepickerCalendarGridRowClassName = 'ui-datepicker__calendargridrow';
/**
 * A DatepickerCalendarGridRow is used to display the calendar grid row.
 */
exports.DatepickerCalendarGridRow = react_bindings_1.compose(Box_1.Box, {
    className: exports.datepickerCalendarGridRowClassName,
    displayName: 'DatepickerCalendarGridRow',
    handledProps: ['isRowSelectionActive'],
    overrideStyles: true,
    mapPropsToStylesProps: function (_a) {
        var isRowSelectionActive = _a.isRowSelectionActive;
        return ({
            isRowSelectionActive: isRowSelectionActive,
        });
    },
    shorthandConfig: {
        mappedProp: 'content',
    },
});
exports.DatepickerCalendarGridRow.defaultProps = {
    as: 'tr',
};
exports.DatepickerCalendarGridRow.propTypes = utils_1.commonPropTypes.createCommon();
