// @fluentui/react-northstar/dist/dts/src/components/Datepicker/DatepickerCalendarHeaderAction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DatepickerCalendarHeaderAction = exports.datepickerCalendarHeaderActionClassName = void 0;
var React = require("react");
var accessibility_1 = require("@fluentui/accessibility");
var react_bindings_1 = require("@fluentui/react-bindings");
var react_icons_northstar_1 = require("@fluentui/react-icons-northstar");
var Button_1 = require("../Button/Button");
exports.datepickerCalendarHeaderActionClassName = 'ui-datepicker__calendarheaderaction';
/**
 * A DatepickerCalendarHeaderAction is used to display action button for DatepickerCalendarHeader.
 */
exports.DatepickerCalendarHeaderAction = react_bindings_1.compose(Button_1.Button, {
    className: exports.datepickerCalendarHeaderActionClassName,
    displayName: 'DatepickerCalendarHeaderAction',
    handledProps: ['direction', 'disabledNavigatableButton'],
    mapPropsToStylesProps: function (p) { return ({
        iconOnly: true,
        text: true,
        disabled: p.disabledNavigatableButton,
    }); },
    slotProps: function (props) { return ({
        icon: {
            content: props.direction === 'next' ? React.createElement(react_icons_northstar_1.ChevronEndIcon, null) : React.createElement(react_icons_northstar_1.ChevronStartIcon, null),
        },
    }); },
});
exports.DatepickerCalendarHeaderAction.defaultProps = {
    as: 'button',
    accessibility: accessibility_1.buttonBehavior,
    size: 'medium',
    icon: {},
};
