// @fluentui/react-northstar/dist/es/components/Breadcrumb/BreadcrumbItem.js
import * as React from 'react';
import { compose, useFluentContext, useTelemetry, useAccessibility, useStyles, useUnhandledProps, getElementType, childrenExist } from '@fluentui/react-bindings';
import { commonPropTypes } from '../../utils';
import { breadcrumbItemBehavior } from '@fluentui/accessibility';
import { useBreadcrumbContext } from './breadcrumbContext';
export var breadcrumbItemClassName = 'ui-breadcrumb__item';
/**
 * BreadcrumbItem an actionable item within a Breadcrumb
 */

export var BreadcrumbItem = /*#__PURE__*/function () {
  var BreadcrumbItem = compose(function (props, ref, composeOptions) {
    var context = useFluentContext();

    var _useTelemetry = useTelemetry(composeOptions.displayName, context.telemetry),
        setStart = _useTelemetry.setStart,
        setEnd = _useTelemetry.setEnd;

    setStart();
    var accessibility = props.accessibility,
        children = props.children,
        content = props.content,
        className = props.className,
        design = props.design,
        styles = props.styles,
        variables = props.variables,
        active = props.active,
        disabled = props.disabled;

    var _useBreadcrumbContext = useBreadcrumbContext(),
        size = _useBreadcrumbContext.size;

    var getA11yProps = useAccessibility(accessibility, {
      debugName: composeOptions.displayName,
      rtl: context.rtl
    });

    var _useStyles = useStyles(composeOptions.displayName, {
      className: composeOptions.className,
      composeOptions: composeOptions,
      mapPropsToStyles: function mapPropsToStyles() {
        return {
          size: size,
          active: active,
          disabled: disabled
        };
      },
      mapPropsToInlineStyles: function mapPropsToInlineStyles() {
        return {
          className: className,
          design: design,
          styles: styles,
          variables: variables
        };
      },
      rtl: context.rtl,
      unstable_props: props
    }),
        classes = _useStyles.classes;

    var unhandledProps = useUnhandledProps(composeOptions.handledProps, props);
    var ElementType = getElementType(props);
    var result = /*#__PURE__*/React.createElement(ElementType, getA11yProps('root', Object.assign({
      className: classes.root,
      ref: ref
    }, unhandledProps)), childrenExist(children) ? children : content);
    setEnd();
    return result;
  }, {
    className: breadcrumbItemClassName,
    displayName: 'BreadcrumbItem',
    handledProps: ['accessibility', 'as', 'children', 'className', 'content', 'design', 'styles', 'variables', 'active']
  });
  BreadcrumbItem.defaultProps = {
    accessibility: breadcrumbItemBehavior
  };
  BreadcrumbItem.propTypes = commonPropTypes.createCommon();
  return BreadcrumbItem;
}();
